# Phage-Host Interaction Prediction

## Description
**ABSTRACT**: With the growing interest in using phages to combat antimicrobial resistance, computational methods for predicting phage-host interactions have been explored to help shortlist candidate phages. While existing systems have been successful in integrating multiple features to improve performance, most consider entire proteomes and rely on manual feature engineering, which poses difficulty in selecting the most informative sequence properties to feed the model. In this paper, we focused on the phages' receptor-binding proteins, which are considered key machinery for host recognition, and explored different protein language models to automatically encode these protein sequences into meaningful dense embeddings without the need for additional alignment or structural information. Our experiments showed that, for phage-host interaction prediction, embeddings of receptor-binding proteins present improvements over handcrafted genomic and protein sequence features, with the highest performance obtained using the transformer-based protein language model ProtT5.

**INDEX TERMS**: Phage-host interaction, protein embedding, representation learning, transformers

## Project Structure

## Authors
- **Mark Edward M. Gonzales** <br>
  mark_gonzales@dlsu.edu.ph 
 
- **Ms. Jennifer C. Ureta** <br>
  jennifer.ureta@dlsu.edu.ph 
  
- **Dr. Anish M.S. Shrestha** <br>
  anish.shrestha@dlsu.edu.ph

This is a research project under the [Bioinformatics Laboratory](https://bioinfodlsu.com/), Advanced Research Institute for Informatics, Computing and Networking, De La Salle University, Philippines.
  
