{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4af558f4",
   "metadata": {},
   "source": [
    "# Protein embeddings improve phage-host interaction prediction\n",
    "\n",
    "**Mark Edward M. Gonzales<sup>1, 2</sup> & Anish M.S. Shrestha<sup>1, 2</sup>**\n",
    "\n",
    "<sup>1</sup> Department of Software Technology, College of Computer Studies, De La Salle University, Manila, Philippines <br>\n",
    "<sup>2</sup> Bioinformatics Laboratory, Advanced Research Institute for Informatics, Computing and Networking, De La Salle University, Manila, Philippines\n",
    "\n",
    "{mark_gonzales, anish.shrestha}@dlsu.edu.ph"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f20e790b",
   "metadata": {},
   "source": [
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6fecdbc",
   "metadata": {},
   "source": [
    "## üìÅ Output Files\n",
    "If you would like to skip running this notebook, you may download the \n",
    "\n",
    "Intermediate output files (i.e., those saved in `temp/preprocessing`) should have already been included when the repository was cloned."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98a2f6e3",
   "metadata": {},
   "source": [
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44d195eb",
   "metadata": {},
   "source": [
    "# Part I: Preliminaries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d9f66be",
   "metadata": {},
   "source": [
    "Import the necessary libraries and modules."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "69b9a927",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "import pickle\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from ConstantsUtil import ConstantsUtil\n",
    "from SequenceParsingUtil import SequenceParsingUtil\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b4e4ac5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_rows', None)\n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "pd.options.mode.chained_assignment = None"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6926243d",
   "metadata": {},
   "source": [
    "The parameter of `ConstantsUtil` corresponds to the download date of dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6e546b06",
   "metadata": {},
   "outputs": [],
   "source": [
    "constants = ConstantsUtil('16Sep2022')\n",
    "util = SequenceParsingUtil(constants.DISPLAY_PROGRESS, constants.MISSPELLING_THRESHOLD, constants.MIN_LEN_KEYWORD)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49e3e747",
   "metadata": {},
   "source": [
    "To load this month's version of the dataset via INPHARED:\n",
    "\n",
    "1. Clone the latest version of INPHARED from its official [repository](http://github.com/RyanCook94/inphared).\n",
    "2. Download the PHROG HMMs (`all_phrogs.hmm`) from [Millard Lab](https://millardlab.org/2021/11/21/phage-annotation-with-phrogs/).\n",
    "3. Run the following script to run the INPHARED pipeline (refer to the [documentation](https://github.com/RyanCook94/inphared#usage) for details): <br>\n",
    "    `perl inphared.pl -e exclusion_list.txt -P all_phrogs.hmm/all_phrogs.hmm -c 0`\n",
    "4. Move the following files to the `inphared` directory: <br>\n",
    "    4.1. `<download_date>_data_excluding_refseq.tsv` <br>\n",
    "    4.2. `<download_date>_phages_downloaded_from_genbank.gb` <br>\n",
    "   Note: `<download_date>` should be the same as the argument passed to `ConstantsUtil` in the previous code block."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fc6de8ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Description</th>\n",
       "      <th>Classification</th>\n",
       "      <th>Genome Length (bp)</th>\n",
       "      <th>Jumbophage</th>\n",
       "      <th>molGC (%)</th>\n",
       "      <th>Molecule</th>\n",
       "      <th>Modification Date</th>\n",
       "      <th>Number CDS</th>\n",
       "      <th>Positive Strand (%)</th>\n",
       "      <th>Negative Strand (%)</th>\n",
       "      <th>Coding Capacity(%)</th>\n",
       "      <th>Low Coding Capacity Warning</th>\n",
       "      <th>tRNAs</th>\n",
       "      <th>Host</th>\n",
       "      <th>Lowest Taxa</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Sub-family</th>\n",
       "      <th>Family</th>\n",
       "      <th>Order</th>\n",
       "      <th>Class</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Realm</th>\n",
       "      <th>Baltimore Group</th>\n",
       "      <th>Genbank Division</th>\n",
       "      <th>Isolation Host (beware inconsistent and nonsense values)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LR756511</td>\n",
       "      <td>uncultured phage</td>\n",
       "      <td>uncultured phage environmental samples Viruses</td>\n",
       "      <td>595163</td>\n",
       "      <td>True</td>\n",
       "      <td>42.110</td>\n",
       "      <td>DNA</td>\n",
       "      <td>26-MAR-2020</td>\n",
       "      <td>1080</td>\n",
       "      <td>20.648148</td>\n",
       "      <td>79.351852</td>\n",
       "      <td>93.248236</td>\n",
       "      <td>NaN</td>\n",
       "      <td>62</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>ENV</td>\n",
       "      <td>Unspecified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LR756508</td>\n",
       "      <td>uncultured phage</td>\n",
       "      <td>uncultured phage environmental samples Viruses</td>\n",
       "      <td>484177</td>\n",
       "      <td>True</td>\n",
       "      <td>39.553</td>\n",
       "      <td>DNA</td>\n",
       "      <td>26-MAR-2020</td>\n",
       "      <td>683</td>\n",
       "      <td>98.389458</td>\n",
       "      <td>1.610542</td>\n",
       "      <td>89.636022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>ENV</td>\n",
       "      <td>Unspecified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LR756504</td>\n",
       "      <td>uncultured phage</td>\n",
       "      <td>uncultured phage environmental samples Viruses</td>\n",
       "      <td>636363</td>\n",
       "      <td>True</td>\n",
       "      <td>26.393</td>\n",
       "      <td>DNA</td>\n",
       "      <td>26-MAR-2020</td>\n",
       "      <td>920</td>\n",
       "      <td>51.956522</td>\n",
       "      <td>48.043478</td>\n",
       "      <td>92.786193</td>\n",
       "      <td>NaN</td>\n",
       "      <td>34</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>ENV</td>\n",
       "      <td>Unspecified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LR756503</td>\n",
       "      <td>uncultured phage</td>\n",
       "      <td>uncultured phage environmental samples Viruses</td>\n",
       "      <td>735411</td>\n",
       "      <td>True</td>\n",
       "      <td>32.225</td>\n",
       "      <td>DNA</td>\n",
       "      <td>26-MAR-2020</td>\n",
       "      <td>1014</td>\n",
       "      <td>30.276134</td>\n",
       "      <td>69.723866</td>\n",
       "      <td>93.113239</td>\n",
       "      <td>NaN</td>\n",
       "      <td>56</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>ENV</td>\n",
       "      <td>Unspecified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LR756502</td>\n",
       "      <td>uncultured phage</td>\n",
       "      <td>uncultured phage environmental samples Viruses</td>\n",
       "      <td>642428</td>\n",
       "      <td>True</td>\n",
       "      <td>31.472</td>\n",
       "      <td>DNA</td>\n",
       "      <td>26-MAR-2020</td>\n",
       "      <td>971</td>\n",
       "      <td>54.582904</td>\n",
       "      <td>45.417096</td>\n",
       "      <td>94.525145</td>\n",
       "      <td>NaN</td>\n",
       "      <td>45</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>ENV</td>\n",
       "      <td>Unspecified</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Accession       Description                                  Classification  \\\n",
       "0  LR756511  uncultured phage  uncultured phage environmental samples Viruses   \n",
       "1  LR756508  uncultured phage  uncultured phage environmental samples Viruses   \n",
       "2  LR756504  uncultured phage  uncultured phage environmental samples Viruses   \n",
       "3  LR756503  uncultured phage  uncultured phage environmental samples Viruses   \n",
       "4  LR756502  uncultured phage  uncultured phage environmental samples Viruses   \n",
       "\n",
       "   Genome Length (bp)  Jumbophage  molGC (%) Molecule Modification Date  \\\n",
       "0              595163        True     42.110      DNA       26-MAR-2020   \n",
       "1              484177        True     39.553      DNA       26-MAR-2020   \n",
       "2              636363        True     26.393      DNA       26-MAR-2020   \n",
       "3              735411        True     32.225      DNA       26-MAR-2020   \n",
       "4              642428        True     31.472      DNA       26-MAR-2020   \n",
       "\n",
       "   Number CDS  Positive Strand (%)  Negative Strand (%)  Coding Capacity(%)  \\\n",
       "0        1080            20.648148            79.351852           93.248236   \n",
       "1         683            98.389458             1.610542           89.636022   \n",
       "2         920            51.956522            48.043478           92.786193   \n",
       "3        1014            30.276134            69.723866           93.113239   \n",
       "4         971            54.582904            45.417096           94.525145   \n",
       "\n",
       "  Low Coding Capacity Warning  tRNAs         Host   Lowest Taxa         Genus  \\\n",
       "0                         NaN     62  Unspecified  Unclassified  Unclassified   \n",
       "1                         NaN     31  Unspecified  Unclassified  Unclassified   \n",
       "2                         NaN     34  Unspecified  Unclassified  Unclassified   \n",
       "3                         NaN     56  Unspecified  Unclassified  Unclassified   \n",
       "4                         NaN     45  Unspecified  Unclassified  Unclassified   \n",
       "\n",
       "     Sub-family        Family         Order         Class        Phylum  \\\n",
       "0  Unclassified  Unclassified  Unclassified  Unclassified  Unclassified   \n",
       "1  Unclassified  Unclassified  Unclassified  Unclassified  Unclassified   \n",
       "2  Unclassified  Unclassified  Unclassified  Unclassified  Unclassified   \n",
       "3  Unclassified  Unclassified  Unclassified  Unclassified  Unclassified   \n",
       "4  Unclassified  Unclassified  Unclassified  Unclassified  Unclassified   \n",
       "\n",
       "        Kingdom         Realm Baltimore Group Genbank Division  \\\n",
       "0  Unclassified  Unclassified    Unclassified              ENV   \n",
       "1  Unclassified  Unclassified    Unclassified              ENV   \n",
       "2  Unclassified  Unclassified    Unclassified              ENV   \n",
       "3  Unclassified  Unclassified    Unclassified              ENV   \n",
       "4  Unclassified  Unclassified    Unclassified              ENV   \n",
       "\n",
       "  Isolation Host (beware inconsistent and nonsense values)  \n",
       "0                                        Unspecified        \n",
       "1                                        Unspecified        \n",
       "2                                        Unspecified        \n",
       "3                                        Unspecified        \n",
       "4                                        Unspecified        "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inphared_raw = pd.read_csv(f'{constants.INPHARED_TSV}', sep='\\t')\n",
    "inphared_raw = inphared_raw.drop_duplicates(ignore_index = True)\n",
    "inphared_raw.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "96759cef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(18389, 27)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inphared_raw.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b0c37f0",
   "metadata": {},
   "source": [
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29f8a02f",
   "metadata": {},
   "source": [
    "# Part II: Preprocessing of Host Information"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6079a0b0",
   "metadata": {},
   "source": [
    "Fo entries with `Unspecified` host in INPHARED, obtain the isolation host information from GenBank, if it makes sense."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dfd0b094",
   "metadata": {},
   "outputs": [],
   "source": [
    "util.set_inphared_gb(constants.INPHARED_GB)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "632d3d9e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Description</th>\n",
       "      <th>Classification</th>\n",
       "      <th>Genome Length (bp)</th>\n",
       "      <th>Jumbophage</th>\n",
       "      <th>molGC (%)</th>\n",
       "      <th>Molecule</th>\n",
       "      <th>Modification Date</th>\n",
       "      <th>Number CDS</th>\n",
       "      <th>Positive Strand (%)</th>\n",
       "      <th>Negative Strand (%)</th>\n",
       "      <th>Coding Capacity(%)</th>\n",
       "      <th>Low Coding Capacity Warning</th>\n",
       "      <th>tRNAs</th>\n",
       "      <th>Host</th>\n",
       "      <th>Lowest Taxa</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Sub-family</th>\n",
       "      <th>Family</th>\n",
       "      <th>Order</th>\n",
       "      <th>Class</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Realm</th>\n",
       "      <th>Baltimore Group</th>\n",
       "      <th>Genbank Division</th>\n",
       "      <th>Isolation Host (beware inconsistent and nonsense values)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LR756511</td>\n",
       "      <td>uncultured phage</td>\n",
       "      <td>uncultured phage environmental samples Viruses</td>\n",
       "      <td>595163</td>\n",
       "      <td>True</td>\n",
       "      <td>42.110</td>\n",
       "      <td>DNA</td>\n",
       "      <td>26-MAR-2020</td>\n",
       "      <td>1080</td>\n",
       "      <td>20.648148</td>\n",
       "      <td>79.351852</td>\n",
       "      <td>93.248236</td>\n",
       "      <td>NaN</td>\n",
       "      <td>62</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>ENV</td>\n",
       "      <td>Unspecified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LR756508</td>\n",
       "      <td>uncultured phage</td>\n",
       "      <td>uncultured phage environmental samples Viruses</td>\n",
       "      <td>484177</td>\n",
       "      <td>True</td>\n",
       "      <td>39.553</td>\n",
       "      <td>DNA</td>\n",
       "      <td>26-MAR-2020</td>\n",
       "      <td>683</td>\n",
       "      <td>98.389458</td>\n",
       "      <td>1.610542</td>\n",
       "      <td>89.636022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>ENV</td>\n",
       "      <td>Unspecified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LR756504</td>\n",
       "      <td>uncultured phage</td>\n",
       "      <td>uncultured phage environmental samples Viruses</td>\n",
       "      <td>636363</td>\n",
       "      <td>True</td>\n",
       "      <td>26.393</td>\n",
       "      <td>DNA</td>\n",
       "      <td>26-MAR-2020</td>\n",
       "      <td>920</td>\n",
       "      <td>51.956522</td>\n",
       "      <td>48.043478</td>\n",
       "      <td>92.786193</td>\n",
       "      <td>NaN</td>\n",
       "      <td>34</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>ENV</td>\n",
       "      <td>Unspecified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LR756503</td>\n",
       "      <td>uncultured phage</td>\n",
       "      <td>uncultured phage environmental samples Viruses</td>\n",
       "      <td>735411</td>\n",
       "      <td>True</td>\n",
       "      <td>32.225</td>\n",
       "      <td>DNA</td>\n",
       "      <td>26-MAR-2020</td>\n",
       "      <td>1014</td>\n",
       "      <td>30.276134</td>\n",
       "      <td>69.723866</td>\n",
       "      <td>93.113239</td>\n",
       "      <td>NaN</td>\n",
       "      <td>56</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>ENV</td>\n",
       "      <td>Unspecified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LR756502</td>\n",
       "      <td>uncultured phage</td>\n",
       "      <td>uncultured phage environmental samples Viruses</td>\n",
       "      <td>642428</td>\n",
       "      <td>True</td>\n",
       "      <td>31.472</td>\n",
       "      <td>DNA</td>\n",
       "      <td>26-MAR-2020</td>\n",
       "      <td>971</td>\n",
       "      <td>54.582904</td>\n",
       "      <td>45.417096</td>\n",
       "      <td>94.525145</td>\n",
       "      <td>NaN</td>\n",
       "      <td>45</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>ENV</td>\n",
       "      <td>Unspecified</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Accession       Description                                  Classification  \\\n",
       "0  LR756511  uncultured phage  uncultured phage environmental samples Viruses   \n",
       "1  LR756508  uncultured phage  uncultured phage environmental samples Viruses   \n",
       "2  LR756504  uncultured phage  uncultured phage environmental samples Viruses   \n",
       "3  LR756503  uncultured phage  uncultured phage environmental samples Viruses   \n",
       "4  LR756502  uncultured phage  uncultured phage environmental samples Viruses   \n",
       "\n",
       "   Genome Length (bp)  Jumbophage  molGC (%) Molecule Modification Date  \\\n",
       "0              595163        True     42.110      DNA       26-MAR-2020   \n",
       "1              484177        True     39.553      DNA       26-MAR-2020   \n",
       "2              636363        True     26.393      DNA       26-MAR-2020   \n",
       "3              735411        True     32.225      DNA       26-MAR-2020   \n",
       "4              642428        True     31.472      DNA       26-MAR-2020   \n",
       "\n",
       "   Number CDS  Positive Strand (%)  Negative Strand (%)  Coding Capacity(%)  \\\n",
       "0        1080            20.648148            79.351852           93.248236   \n",
       "1         683            98.389458             1.610542           89.636022   \n",
       "2         920            51.956522            48.043478           92.786193   \n",
       "3        1014            30.276134            69.723866           93.113239   \n",
       "4         971            54.582904            45.417096           94.525145   \n",
       "\n",
       "  Low Coding Capacity Warning  tRNAs         Host   Lowest Taxa         Genus  \\\n",
       "0                         NaN     62  Unspecified  Unclassified  Unclassified   \n",
       "1                         NaN     31  Unspecified  Unclassified  Unclassified   \n",
       "2                         NaN     34  Unspecified  Unclassified  Unclassified   \n",
       "3                         NaN     56  Unspecified  Unclassified  Unclassified   \n",
       "4                         NaN     45  Unspecified  Unclassified  Unclassified   \n",
       "\n",
       "     Sub-family        Family         Order         Class        Phylum  \\\n",
       "0  Unclassified  Unclassified  Unclassified  Unclassified  Unclassified   \n",
       "1  Unclassified  Unclassified  Unclassified  Unclassified  Unclassified   \n",
       "2  Unclassified  Unclassified  Unclassified  Unclassified  Unclassified   \n",
       "3  Unclassified  Unclassified  Unclassified  Unclassified  Unclassified   \n",
       "4  Unclassified  Unclassified  Unclassified  Unclassified  Unclassified   \n",
       "\n",
       "        Kingdom         Realm Baltimore Group Genbank Division  \\\n",
       "0  Unclassified  Unclassified    Unclassified              ENV   \n",
       "1  Unclassified  Unclassified    Unclassified              ENV   \n",
       "2  Unclassified  Unclassified    Unclassified              ENV   \n",
       "3  Unclassified  Unclassified    Unclassified              ENV   \n",
       "4  Unclassified  Unclassified    Unclassified              ENV   \n",
       "\n",
       "  Isolation Host (beware inconsistent and nonsense values)  \n",
       "0                                        Unspecified        \n",
       "1                                        Unspecified        \n",
       "2                                        Unspecified        \n",
       "3                                        Unspecified        \n",
       "4                                        Unspecified        "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inphared_unspec_host = inphared_raw.loc[inphared_raw['Host'] == 'Unspecified']\n",
    "inphared_unspec_host.reset_index(inplace = True, drop = True)\n",
    "inphared_unspec_host.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "98d5d77a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed 1000 records\n",
      "Processed 2000 records\n",
      "Processed 3000 records\n",
      "Processed 4000 records\n",
      "Processed 5000 records\n",
      "Processed 6000 records\n",
      "Processed 7000 records\n",
      "Processed 8000 records\n",
      "Processed 9000 records\n",
      "Processed 10000 records\n",
      "Processed 11000 records\n",
      "Processed 12000 records\n",
      "Processed 13000 records\n",
      "Processed 14000 records\n",
      "Processed 15000 records\n",
      "Processed 16000 records\n",
      "Processed 17000 records\n",
      "Processed 18000 records\n",
      "Processed 19000 records\n",
      "Processed 20000 records\n",
      "Processed 21000 records\n",
      "Processed 22000 records\n",
      "Processed 23000 records\n",
      "Processed 24000 records\n",
      "Processed 25000 records\n",
      "Processed 26000 records\n",
      "Processed 27000 records\n",
      "Processed 28000 records\n",
      "Processed 29000 records\n",
      "Processed 30000 records\n",
      "Processed 31000 records\n"
     ]
    }
   ],
   "source": [
    "phages_unspec_host = util.get_phages_unspec_host(inphared_unspec_host)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b2dee0ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "unfiltered_hosts = util.get_unfiltered_hosts()\n",
    "genus_typo = util.get_genus_typo(f'{constants.GENUS_TYPO}')\n",
    "unfiltered_suspected_genera = util.get_unfiltered_suspected_genera(constants.CANDIDATE_REGEX)\n",
    "\n",
    "excluded_hosts = util.get_excluded_hosts(constants.ARCHAEA, constants.BACTERIA_NOT_GENUS, constants.EXCLUDED_HOSTS)\n",
    "valid_hosts = util.get_valid_hosts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e229d780",
   "metadata": {},
   "source": [
    "Add another layer of checking to identify typos in GenBank annotations, as well as oversights during manual filtering:\n",
    "- Recheck genera in `valid_hosts` that are not in INPHARED's `Host` column.\n",
    "- Identify entries with minimum edit distance &leq; 2 (likely misspellings).\n",
    "<br>\n",
    "\n",
    "Adjust the exclusion and typo correction files in the `preprocessing` directory accordingly, and rerun the previous cell, as needed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "cb61b8fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'silvanigrella'}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inphared_hosts = set(inphared_raw['Host'].str.lower())\n",
    "valid_hosts - inphared_hosts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "dd7e2e75",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "microbacterium mycobacterium\n"
     ]
    }
   ],
   "source": [
    "valid_hosts_list = list(valid_hosts)\n",
    "for i in range(len(valid_hosts_list)):\n",
    "    for j in range(i + 1, len(valid_hosts_list)):\n",
    "        if util.is_possible_misspelling(valid_hosts_list[i], valid_hosts_list[j]):\n",
    "            print(valid_hosts_list[i], valid_hosts_list[j])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8e42ea6",
   "metadata": {},
   "source": [
    "Update `host` column based on the manually filtered info from GenBank."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b6b8e3a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index: 424 | Name: ON970599 | Host: gordonia\n",
      "Processed 1000 records\n",
      "Processed 2000 records\n",
      "Index: 2105 | Name: MN940411 | Host: staphylococcus\n",
      "Processed 3000 records\n",
      "Index: 3471 | Name: AF125163 | Host: vibrio\n",
      "Index: 3477 | Name: AF063097 | Host: escherichia\n",
      "Processed 4000 records\n",
      "Index: 3576 | Name: OK499975 | Host: proteus\n",
      "Index: 3705 | Name: OK649958 | Host: haloferax\n",
      "Index: 3761 | Name: MW822148 | Host: mycobacterium\n",
      "Processed 5000 records\n",
      "Index: 4552 | Name: MH590600 | Host: microbacterium\n",
      "Index: 4628 | Name: MH271321 | Host: microbacterium\n",
      "Index: 4643 | Name: MH155873 | Host: microbacterium\n",
      "Index: 5392 | Name: LC625742 | Host: enterococcus\n",
      "Processed 6000 records\n",
      "Processed 7000 records\n",
      "Processed 8000 records\n",
      "Processed 9000 records\n",
      "Index: 5662 | Name: MT588083 | Host: salmonella\n",
      "Processed 10000 records\n",
      "Processed 11000 records\n",
      "Processed 12000 records\n",
      "Processed 13000 records\n",
      "Processed 14000 records\n",
      "Processed 15000 records\n",
      "Processed 16000 records\n",
      "Processed 17000 records\n",
      "Processed 18000 records\n",
      "Index: 8414 | Name: MN857473 | Host: caulobacter\n",
      "Index: 8837 | Name: MN746332 | Host: caulobacter\n",
      "Processed 19000 records\n",
      "Index: 9294 | Name: MN161573 | Host: salmonella\n",
      "Index: 9451 | Name: MN882550 | Host: butyrivibrio\n",
      "Processed 20000 records\n",
      "Index: 10400 | Name: MN153797 | Host: escherichia\n",
      "Index: 10581 | Name: LC487997 | Host: escherichia\n",
      "Index: 10582 | Name: LC487996 | Host: escherichia\n",
      "Index: 10616 | Name: MG747435 | Host: ralstonia\n",
      "Processed 21000 records\n",
      "Index: 11284 | Name: MK310183 | Host: escherichia\n",
      "Processed 22000 records\n",
      "Index: 11806 | Name: MK310184 | Host: escherichia\n",
      "Processed 23000 records\n",
      "Processed 24000 records\n",
      "Index: 13732 | Name: MF155889 | Host: vibrio\n",
      "Index: 13982 | Name: CP017837 | Host: silvanigrella\n",
      "Index: 13984 | Name: CP017838 | Host: silvanigrella\n",
      "Processed 25000 records\n",
      "Index: 14453 | Name: KT232076 | Host: escherichia\n",
      "Index: 14569 | Name: JX412914 | Host: escherichia\n",
      "Index: 15013 | Name: AP012540 | Host: escherichia\n",
      "Index: 15014 | Name: AP012539 | Host: escherichia\n",
      "Index: 15015 | Name: AP012538 | Host: escherichia\n",
      "Index: 15016 | Name: AP012537 | Host: escherichia\n",
      "Index: 15017 | Name: AP012536 | Host: escherichia\n",
      "Index: 15018 | Name: AP012535 | Host: escherichia\n",
      "Index: 15019 | Name: AP012534 | Host: escherichia\n",
      "Index: 15020 | Name: AP012533 | Host: escherichia\n",
      "Index: 15021 | Name: AP012532 | Host: escherichia\n",
      "Index: 15022 | Name: AP012531 | Host: escherichia\n",
      "Index: 15023 | Name: AP012530 | Host: escherichia\n",
      "Index: 15024 | Name: AP012529 | Host: escherichia\n",
      "Index: 15025 | Name: AB008550 | Host: pseudomonas\n",
      "Processed 26000 records\n",
      "Index: 15577 | Name: GU942563 | Host: vibrio\n",
      "Processed 27000 records\n",
      "Index: 16396 | Name: KT254133 | Host: escherichia | chryseobacterium | pseudomonas\n",
      "Index: 16397 | Name: KT254132 | Host: escherichia | chryseobacterium | pseudomonas\n",
      "Index: 16398 | Name: KT254131 | Host: escherichia | chryseobacterium | pseudomonas\n",
      "Index: 16399 | Name: KT254130 | Host: escherichia | chryseobacterium | pseudomonas\n",
      "Index: 16621 | Name: KP296794 | Host: paenibacillus\n",
      "Index: 16623 | Name: KP296792 | Host: paenibacillus\n",
      "Processed 28000 records\n",
      "Index: 16741 | Name: KM236244 | Host: salmonella\n",
      "Index: 16806 | Name: KM458633 | Host: salmonella\n",
      "Index: 16878 | Name: FR719956 | Host: roseovarius\n",
      "Index: 17036 | Name: KJ619459 | Host: vibrio\n",
      "Index: 17123 | Name: AB720063 | Host: xanthomonas\n",
      "Index: 17158 | Name: KF598865 | Host: phormidium | plectonema\n",
      "Index: 17221 | Name: KC677662 | Host: salmonella\n",
      "Index: 17242 | Name: KC117378 | Host: haloarcula\n",
      "Index: 17286 | Name: KC292029 | Host: haloarcula\n",
      "Index: 17287 | Name: KC292028 | Host: haloarcula\n",
      "Index: 17288 | Name: KC292027 | Host: haloarcula\n",
      "Index: 17290 | Name: KC292025 | Host: haloarcula\n",
      "Index: 17291 | Name: KC292024 | Host: haloarcula\n",
      "Processed 29000 records\n",
      "Index: 17390 | Name: KC117377 | Host: haloarcula\n",
      "Index: 17391 | Name: KC117376 | Host: halorubrum\n",
      "Index: 17689 | Name: HQ711985 | Host: pseudomonas\n",
      "Index: 17855 | Name: HQ406778 | Host: salmonella\n",
      "Index: 17892 | Name: AM910650 | Host: pseudomonas\n",
      "Index: 17898 | Name: AM156909 | Host: escherichia\n",
      "Index: 17921 | Name: Y13918 | Host: pseudomonas\n",
      "Index: 17924 | Name: X99260 | Host: bacillus\n",
      "Index: 17948 | Name: AF424783 | Host: staphylococcus\n",
      "Index: 17959 | Name: FJ539135 | Host: escherichia\n",
      "Index: 17960 | Name: FJ539134 | Host: escherichia\n",
      "Index: 17961 | Name: FJ539133 | Host: escherichia\n",
      "Index: 17962 | Name: FJ539132 | Host: escherichia\n",
      "Processed 30000 records\n",
      "Index: 18007 | Name: BL2CG | Host: mycoplasma\n",
      "Index: 18033 | Name: AB370205 | Host: staphylococcus\n",
      "Index: 18262 | Name: AB255436 | Host: escherichia\n",
      "Index: 18264 | Name: AP004402 | Host: escherichia\n",
      "Index: 18269 | Name: AP001553 | Host: staphylococcus\n",
      "Index: 18278 | Name: AB012573 | Host: vibrio\n",
      "Index: 18280 | Name: AB370268 | Host: staphylococcus\n",
      "Index: 18281 | Name: AP008954 | Host: staphylococcus\n",
      "Index: 18282 | Name: AP008953 | Host: staphylococcus\n",
      "Index: 18288 | Name: EU794049 | Host: hamiltonella\n",
      "Index: 18359 | Name: AJ567472 | Host: acidianus\n",
      "Processed 31000 records\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "unspecified                                     2566\n",
       "mycobacterium                                   2189\n",
       "escherichia                                     1627\n",
       "vibrio                                           852\n",
       "pseudomonas                                      826\n",
       "salmonella                                       744\n",
       "streptococcus                                    730\n",
       "klebsiella                                       628\n",
       "gordonia                                         579\n",
       "staphylococcus                                   508\n",
       "enterobacteria                                   498\n",
       "microbacterium                                   459\n",
       "bacillus                                         377\n",
       "lactococcus                                      364\n",
       "synechococcus                                    362\n",
       "arthrobacter                                     340\n",
       "streptomyces                                     310\n",
       "flavobacterium                                   247\n",
       "acinetobacter                                    196\n",
       "enterococcus                                     185\n",
       "rhizobium                                        154\n",
       "aeromonas                                        140\n",
       "erwinia                                          137\n",
       "shigella                                         128\n",
       "propionibacterium                                125\n",
       "yersinia                                         120\n",
       "rheinheimera                                     110\n",
       "campylobacter                                     95\n",
       "pectobacterium                                    92\n",
       "xanthomonas                                       91\n",
       "lactobacillus                                     89\n",
       "sigmofec                                          88\n",
       "clostridium                                       83\n",
       "serratia                                          76\n",
       "ralstonia                                         76\n",
       "burkholderia                                      75\n",
       "sulfolobus                                        75\n",
       "paenibacillus                                     66\n",
       "listeria                                          62\n",
       "stenotrophomonas                                  61\n",
       "rhodococcus                                       60\n",
       "enterobacter                                      60\n",
       "peromfec                                          55\n",
       "rhodobacter                                       54\n",
       "cellulophaga                                      53\n",
       "prochlorococcus                                   51\n",
       "dipodfec                                          48\n",
       "proteus                                           48\n",
       "pelagibacter                                      47\n",
       "citrobacter                                       46\n",
       "halorubrum                                        45\n",
       "bacteroides                                       43\n",
       "dickeya                                           39\n",
       "cronobacter                                       39\n",
       "corynebacterium                                   38\n",
       "brucella                                          37\n",
       "helicobacter                                      37\n",
       "caulobacter                                       33\n",
       "pseudoalteromonas                                 33\n",
       "moraxella                                         32\n",
       "achromobacter                                     31\n",
       "haloarcula                                        30\n",
       "providencia                                       26\n",
       "pantoea                                           24\n",
       "leuconostoc                                       23\n",
       "roseobacter                                       21\n",
       "brevundimonas                                     20\n",
       "brevibacterium                                    19\n",
       "liberibacter                                      18\n",
       "polaribacter                                      18\n",
       "shewanella                                        18\n",
       "acidianus                                         18\n",
       "edwardsiella                                      17\n",
       "alteromonas                                       17\n",
       "cutibacterium                                     17\n",
       "thermus                                           17\n",
       "wolbachia                                         16\n",
       "clostridioides                                    15\n",
       "ruegeria                                          15\n",
       "prevotella                                        15\n",
       "mycoplasma                                        14\n",
       "microcystis                                       13\n",
       "sinorhizobium                                     13\n",
       "agrobacterium                                     13\n",
       "mannheimia                                        12\n",
       "bifidobacterium                                   11\n",
       "kosakonia                                         11\n",
       "bordetella                                        11\n",
       "hafnia                                            10\n",
       "brevibacillus                                     10\n",
       "salinibacter                                       9\n",
       "fusobacterium                                      9\n",
       "pasteurella                                        9\n",
       "xylella                                            9\n",
       "leptospira                                         9\n",
       "paracoccus                                         8\n",
       "faecalibacterium                                   8\n",
       "morganella                                         8\n",
       "spiroplasma                                        6\n",
       "nocardia                                           6\n",
       "maribacter                                         6\n",
       "oenococcus                                         6\n",
       "sphingomonas                                       6\n",
       "ruminococcus                                       6\n",
       "geobacillus                                        6\n",
       "nodularia                                          6\n",
       "dinoroseobacter                                    6\n",
       "sulfitobacter                                      6\n",
       "tenacibaculum                                      5\n",
       "tsukamurella                                       5\n",
       "marinomonas                                        5\n",
       "butyrivibrio                                       5\n",
       "curtobacterium                                     5\n",
       "chlamydia                                          5\n",
       "nitratiruptor                                      5\n",
       "nitrosopumilus                                     5\n",
       "haloferax                                          4\n",
       "riemerella                                         4\n",
       "escherichia | chryseobacterium | pseudomonas       4\n",
       "ochrobactrum                                       4\n",
       "elizabethkingia                                    4\n",
       "methanobacterium                                   4\n",
       "exiguobacterium                                    4\n",
       "brochothrix                                        4\n",
       "delftia                                            3\n",
       "nostoc                                             3\n",
       "psychrobacillus                                    3\n",
       "acidithiobacillus                                  3\n",
       "haemophilus                                        3\n",
       "natrinema                                          3\n",
       "raoultella                                         3\n",
       "lentibacter                                        3\n",
       "marinobacter                                       3\n",
       "weissella                                          3\n",
       "bdellovibrio                                       3\n",
       "psychrobacter                                      3\n",
       "glutamicibacter                                    3\n",
       "parabacteroides                                    3\n",
       "carnobacterium                                     3\n",
       "actinomyces                                        3\n",
       "acidovorax                                         3\n",
       "acetobacter                                        3\n",
       "komagataeibacter                                   3\n",
       "curvibacter                                        3\n",
       "halomonas                                          3\n",
       "hamiltonella                                       3\n",
       "aeropyrum                                          3\n",
       "alcaligenes                                        3\n",
       "metallosphaera                                     3\n",
       "sodalis                                            3\n",
       "leclercia                                          2\n",
       "mesorhizobium                                      2\n",
       "methanosarcina                                     2\n",
       "salinivibrio                                       2\n",
       "silvanigrella                                      2\n",
       "plectonema                                         2\n",
       "thermobifida                                       2\n",
       "rhodovulum                                         2\n",
       "pyrobaculum                                        2\n",
       "xenohaliotis                                       2\n",
       "marinitoga                                         2\n",
       "meiothermus                                        2\n",
       "natrialba                                          2\n",
       "croceibacter                                       2\n",
       "thermoproteus                                      2\n",
       "clavibacter                                        2\n",
       "erysipelothrix                                     2\n",
       "roseburia                                          2\n",
       "cylindrospermopsis                                 2\n",
       "halobacterium                                      2\n",
       "tetraselmis                                        2\n",
       "phormidium                                         2\n",
       "erythrobacter                                      2\n",
       "thermococcus                                       2\n",
       "citromicrobium                                     2\n",
       "oceanospirillum                                    2\n",
       "altiarchaeum                                       2\n",
       "myxococcus                                         2\n",
       "sphaerotilus                                       2\n",
       "anabaena                                           2\n",
       "octadecabacter                                     2\n",
       "paraglaciecola                                     2\n",
       "roseovarius                                        2\n",
       "nonlabens                                          2\n",
       "caldibacillus                                      1\n",
       "nitrososphaera                                     1\n",
       "plesiomonas                                        1\n",
       "rathayibacter                                      1\n",
       "planktothrix                                       1\n",
       "sphingobium                                        1\n",
       "dunaliella                                         1\n",
       "aphanizomenon                                      1\n",
       "tetrasphaera                                       1\n",
       "salisaeta                                          1\n",
       "lysinibacillus                                     1\n",
       "saccharomonospora                                  1\n",
       "azospirillum                                       1\n",
       "celeribacter                                       1\n",
       "peptoclostridium                                   1\n",
       "skermania                                          1\n",
       "methanothermobacter                                1\n",
       "aggregatibacter                                    1\n",
       "lelliottia                                         1\n",
       "aurantimonas                                       1\n",
       "rhodoferax                                         1\n",
       "acaryochloris                                      1\n",
       "thermoanaerobacterium                              1\n",
       "iodobacter                                         1\n",
       "stygiolobus                                        1\n",
       "nitrincola                                         1\n",
       "actinoplanes                                       1\n",
       "listonella                                         1\n",
       "thiobacimonas                                      1\n",
       "thalassomonas                                      1\n",
       "pelagibaca                                         1\n",
       "silicibacter                                       1\n",
       "phormidium | plectonema                            1\n",
       "colwellia                                          1\n",
       "puniceispirillum                                   1\n",
       "kluyvera                                           1\n",
       "halogranum                                         1\n",
       "loktanella                                         1\n",
       "hydrogenobaculum                                   1\n",
       "bradyrhizobium                                     1\n",
       "aquamicrobium                                      1\n",
       "akkermansia                                        1\n",
       "gluconobacter                                      1\n",
       "levilactobacillus                                  1\n",
       "desulfovibrio                                      1\n",
       "pseudoxanthomonas                                  1\n",
       "halogeometricum                                    1\n",
       "hydrogenophilus                                    1\n",
       "lentisphaerae                                      1\n",
       "rahnella                                           1\n",
       "parageobacillus                                    1\n",
       "pseudaeromonas                                     1\n",
       "pararheinheimera                                   1\n",
       "lokiarchaeota                                      1\n",
       "chryseobacterium                                   1\n",
       "lactiplantibacillus                                1\n",
       "kurthia                                            1\n",
       "stappia                                            1\n",
       "pseudarthrobacter                                  1\n",
       "luteibacter                                        1\n",
       "tetragenococcus                                    1\n",
       "cytobacillus                                       1\n",
       "eggerthella                                        1\n",
       "janthinobacterium                                  1\n",
       "mycolicibacterium                                  1\n",
       "cedecea                                            1\n",
       "muninn                                             1\n",
       "huginn                                             1\n",
       "verrucomicrobia                                    1\n",
       "stxxx                                              1\n",
       "solobacterium                                      1\n",
       "methanoculleus                                     1\n",
       "pontimonas                                         1\n",
       "saccharolobus                                      1\n",
       "anoxybacillus                                      1\n",
       "mastigocladus                                      1\n",
       "azobacteroides                                     1\n",
       "virgibacillus                                      1\n",
       "aerococcus                                         1\n",
       "rothia                                             1\n",
       "aeribacillus                                       1\n",
       "photobacterium                                     1\n",
       "pediococcus                                        1\n",
       "sporosarcina                                       1\n",
       "phaeobacter                                        1\n",
       "aminobacter                                        1\n",
       "pseudanabaena                                      1\n",
       "winogradskyella                                    1\n",
       "micromonospora                                     1\n",
       "buttiauxella                                       1\n",
       "haloterrigena                                      1\n",
       "methanocaldococcus                                 1\n",
       "arthronema                                         1\n",
       "protaetiibacter                                    1\n",
       "gardnerella                                        1\n",
       "arcanobacterium                                    1\n",
       "cellulosimicrobium                                 1\n",
       "thermocrinis                                       1\n",
       "sphingopyxis                                       1\n",
       "olleya                                             1\n",
       "pyrococcus                                         1\n",
       "Name: Host, dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inphared_augmented = inphared_raw.copy(deep = True)\n",
    "util.update_host_column(constants.CANDIDATE_REGEX, inphared_unspec_host, inphared_augmented)\n",
    "\n",
    "inphared_augmented['Host'] = inphared_augmented['Host'].str.lower()\n",
    "inphared_augmented['Host'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42f2d9cf",
   "metadata": {},
   "source": [
    "### Polyvalent Phages"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7493d1a0",
   "metadata": {},
   "source": [
    "There are only few polyvalent (multi-host) phages in our database. Hence, for simplicity, we map each polyvalent phage only to its host with the highest number of interacting phages in the dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8c7cc97d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "escherichia | chryseobacterium | pseudomonas : 4\n",
      "phormidium | plectonema : 1\n"
     ]
    }
   ],
   "source": [
    "multiple_hosts = set()\n",
    "\n",
    "for host, count in dict(inphared_augmented['Host'].value_counts()).items():\n",
    "    if '|' in host:\n",
    "        multiple_hosts.add(host)\n",
    "        print(host, ':', count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "20fea126",
   "metadata": {},
   "outputs": [],
   "source": [
    "for host in multiple_hosts:\n",
    "    # We have confirmed that the first host is the host with the highest number of interacting phages in the dataset.\n",
    "    inphared_augmented.loc[inphared_augmented['Host'] == host, 'Host'] = host.split(' | ')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a805a2d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Description</th>\n",
       "      <th>Classification</th>\n",
       "      <th>Genome Length (bp)</th>\n",
       "      <th>Jumbophage</th>\n",
       "      <th>molGC (%)</th>\n",
       "      <th>Molecule</th>\n",
       "      <th>Modification Date</th>\n",
       "      <th>Number CDS</th>\n",
       "      <th>Positive Strand (%)</th>\n",
       "      <th>Negative Strand (%)</th>\n",
       "      <th>Coding Capacity(%)</th>\n",
       "      <th>Low Coding Capacity Warning</th>\n",
       "      <th>tRNAs</th>\n",
       "      <th>Host</th>\n",
       "      <th>Lowest Taxa</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Sub-family</th>\n",
       "      <th>Family</th>\n",
       "      <th>Order</th>\n",
       "      <th>Class</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Realm</th>\n",
       "      <th>Baltimore Group</th>\n",
       "      <th>Genbank Division</th>\n",
       "      <th>Isolation Host (beware inconsistent and nonsense values)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MK250029</td>\n",
       "      <td>Prevotella phage Lak-C1</td>\n",
       "      <td>Prevotella phage Lak-C1 Myoviridae Caudovirice...</td>\n",
       "      <td>540217</td>\n",
       "      <td>True</td>\n",
       "      <td>25.796</td>\n",
       "      <td>DNA</td>\n",
       "      <td>13-JAN-2019</td>\n",
       "      <td>830</td>\n",
       "      <td>47.108434</td>\n",
       "      <td>52.891566</td>\n",
       "      <td>68.324951</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30</td>\n",
       "      <td>prevotella</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Caudoviricetes</td>\n",
       "      <td>Uroviricota</td>\n",
       "      <td>Heunggongvirae</td>\n",
       "      <td>Duplodnaviria</td>\n",
       "      <td>Group I</td>\n",
       "      <td>ENV</td>\n",
       "      <td>Prevotella sp.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MK250028</td>\n",
       "      <td>Prevotella phage Lak-B9</td>\n",
       "      <td>Prevotella phage Lak-B9 Myoviridae Caudovirice...</td>\n",
       "      <td>550053</td>\n",
       "      <td>True</td>\n",
       "      <td>26.012</td>\n",
       "      <td>DNA</td>\n",
       "      <td>13-JAN-2019</td>\n",
       "      <td>859</td>\n",
       "      <td>52.270081</td>\n",
       "      <td>47.729919</td>\n",
       "      <td>69.188424</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29</td>\n",
       "      <td>prevotella</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Caudoviricetes</td>\n",
       "      <td>Uroviricota</td>\n",
       "      <td>Heunggongvirae</td>\n",
       "      <td>Duplodnaviria</td>\n",
       "      <td>Group I</td>\n",
       "      <td>ENV</td>\n",
       "      <td>Prevotella sp.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MK250027</td>\n",
       "      <td>Prevotella phage Lak-B8</td>\n",
       "      <td>Prevotella phage Lak-B8 Myoviridae Caudovirice...</td>\n",
       "      <td>551627</td>\n",
       "      <td>True</td>\n",
       "      <td>26.022</td>\n",
       "      <td>DNA</td>\n",
       "      <td>13-JAN-2019</td>\n",
       "      <td>860</td>\n",
       "      <td>53.023256</td>\n",
       "      <td>46.976744</td>\n",
       "      <td>69.318761</td>\n",
       "      <td>NaN</td>\n",
       "      <td>33</td>\n",
       "      <td>prevotella</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Caudoviricetes</td>\n",
       "      <td>Uroviricota</td>\n",
       "      <td>Heunggongvirae</td>\n",
       "      <td>Duplodnaviria</td>\n",
       "      <td>Group I</td>\n",
       "      <td>ENV</td>\n",
       "      <td>Prevotella sp.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MK250026</td>\n",
       "      <td>Prevotella phage Lak-B7</td>\n",
       "      <td>Prevotella phage Lak-B7 Myoviridae Caudovirice...</td>\n",
       "      <td>550702</td>\n",
       "      <td>True</td>\n",
       "      <td>26.020</td>\n",
       "      <td>DNA</td>\n",
       "      <td>13-JAN-2019</td>\n",
       "      <td>859</td>\n",
       "      <td>53.201397</td>\n",
       "      <td>46.798603</td>\n",
       "      <td>69.363285</td>\n",
       "      <td>NaN</td>\n",
       "      <td>33</td>\n",
       "      <td>prevotella</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Caudoviricetes</td>\n",
       "      <td>Uroviricota</td>\n",
       "      <td>Heunggongvirae</td>\n",
       "      <td>Duplodnaviria</td>\n",
       "      <td>Group I</td>\n",
       "      <td>ENV</td>\n",
       "      <td>Prevotella sp.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MK250025</td>\n",
       "      <td>Prevotella phage Lak-B6</td>\n",
       "      <td>Prevotella phage Lak-B6 Myoviridae Caudovirice...</td>\n",
       "      <td>546689</td>\n",
       "      <td>True</td>\n",
       "      <td>26.029</td>\n",
       "      <td>DNA</td>\n",
       "      <td>13-JAN-2019</td>\n",
       "      <td>847</td>\n",
       "      <td>52.656434</td>\n",
       "      <td>47.343566</td>\n",
       "      <td>69.118274</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30</td>\n",
       "      <td>prevotella</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Caudoviricetes</td>\n",
       "      <td>Uroviricota</td>\n",
       "      <td>Heunggongvirae</td>\n",
       "      <td>Duplodnaviria</td>\n",
       "      <td>Group I</td>\n",
       "      <td>ENV</td>\n",
       "      <td>Prevotella sp.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Accession              Description  \\\n",
       "0  MK250029  Prevotella phage Lak-C1   \n",
       "1  MK250028  Prevotella phage Lak-B9   \n",
       "2  MK250027  Prevotella phage Lak-B8   \n",
       "3  MK250026  Prevotella phage Lak-B7   \n",
       "4  MK250025  Prevotella phage Lak-B6   \n",
       "\n",
       "                                      Classification  Genome Length (bp)  \\\n",
       "0  Prevotella phage Lak-C1 Myoviridae Caudovirice...              540217   \n",
       "1  Prevotella phage Lak-B9 Myoviridae Caudovirice...              550053   \n",
       "2  Prevotella phage Lak-B8 Myoviridae Caudovirice...              551627   \n",
       "3  Prevotella phage Lak-B7 Myoviridae Caudovirice...              550702   \n",
       "4  Prevotella phage Lak-B6 Myoviridae Caudovirice...              546689   \n",
       "\n",
       "   Jumbophage  molGC (%) Molecule Modification Date  Number CDS  \\\n",
       "0        True     25.796      DNA       13-JAN-2019         830   \n",
       "1        True     26.012      DNA       13-JAN-2019         859   \n",
       "2        True     26.022      DNA       13-JAN-2019         860   \n",
       "3        True     26.020      DNA       13-JAN-2019         859   \n",
       "4        True     26.029      DNA       13-JAN-2019         847   \n",
       "\n",
       "   Positive Strand (%)  Negative Strand (%)  Coding Capacity(%)  \\\n",
       "0            47.108434            52.891566           68.324951   \n",
       "1            52.270081            47.729919           69.188424   \n",
       "2            53.023256            46.976744           69.318761   \n",
       "3            53.201397            46.798603           69.363285   \n",
       "4            52.656434            47.343566           69.118274   \n",
       "\n",
       "  Low Coding Capacity Warning  tRNAs        Host Lowest Taxa         Genus  \\\n",
       "0                         NaN     30  prevotella  Myoviridae  Unclassified   \n",
       "1                         NaN     29  prevotella  Myoviridae  Unclassified   \n",
       "2                         NaN     33  prevotella  Myoviridae  Unclassified   \n",
       "3                         NaN     33  prevotella  Myoviridae  Unclassified   \n",
       "4                         NaN     30  prevotella  Myoviridae  Unclassified   \n",
       "\n",
       "     Sub-family      Family         Order           Class       Phylum  \\\n",
       "0  Unclassified  Myoviridae  Unclassified  Caudoviricetes  Uroviricota   \n",
       "1  Unclassified  Myoviridae  Unclassified  Caudoviricetes  Uroviricota   \n",
       "2  Unclassified  Myoviridae  Unclassified  Caudoviricetes  Uroviricota   \n",
       "3  Unclassified  Myoviridae  Unclassified  Caudoviricetes  Uroviricota   \n",
       "4  Unclassified  Myoviridae  Unclassified  Caudoviricetes  Uroviricota   \n",
       "\n",
       "          Kingdom          Realm Baltimore Group Genbank Division  \\\n",
       "0  Heunggongvirae  Duplodnaviria         Group I              ENV   \n",
       "1  Heunggongvirae  Duplodnaviria         Group I              ENV   \n",
       "2  Heunggongvirae  Duplodnaviria         Group I              ENV   \n",
       "3  Heunggongvirae  Duplodnaviria         Group I              ENV   \n",
       "4  Heunggongvirae  Duplodnaviria         Group I              ENV   \n",
       "\n",
       "  Isolation Host (beware inconsistent and nonsense values)  \n",
       "0                                     Prevotella sp.        \n",
       "1                                     Prevotella sp.        \n",
       "2                                     Prevotella sp.        \n",
       "3                                     Prevotella sp.        \n",
       "4                                     Prevotella sp.        "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inphared = inphared_augmented.loc[inphared_augmented['Host'] != 'unspecified']\n",
    "inphared.reset_index(inplace = True, drop = True)\n",
    "inphared.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f624efd",
   "metadata": {},
   "source": [
    "Replace the genera of some hosts with standard NCBI Taxonomy nomenclature."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "32235408",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "mycobacterium                2189\n",
       "escherichia                  1632\n",
       "vibrio                        852\n",
       "pseudomonas                   826\n",
       "salmonella                    744\n",
       "streptococcus                 730\n",
       "klebsiella                    628\n",
       "gordonia                      579\n",
       "enterobacter                  558\n",
       "staphylococcus                508\n",
       "microbacterium                459\n",
       "bacillus                      377\n",
       "lactococcus                   364\n",
       "synechococcus                 362\n",
       "arthrobacter                  340\n",
       "streptomyces                  310\n",
       "flavobacterium                247\n",
       "acinetobacter                 196\n",
       "enterococcus                  185\n",
       "rhizobium                     154\n",
       "aeromonas                     140\n",
       "erwinia                       137\n",
       "shigella                      128\n",
       "propionibacterium             125\n",
       "yersinia                      120\n",
       "rheinheimera                  110\n",
       "campylobacter                  95\n",
       "pectobacterium                 92\n",
       "xanthomonas                    91\n",
       "lactobacillus                  89\n",
       "sigmofec                       88\n",
       "clostridium                    83\n",
       "serratia                       76\n",
       "ralstonia                      76\n",
       "burkholderia                   75\n",
       "sulfolobus                     75\n",
       "paenibacillus                  66\n",
       "listeria                       62\n",
       "stenotrophomonas               61\n",
       "rhodococcus                    60\n",
       "peromfec                       55\n",
       "rhodobacter                    54\n",
       "cellulophaga                   53\n",
       "prochlorococcus                51\n",
       "dipodfec                       48\n",
       "proteus                        48\n",
       "pelagibacter                   47\n",
       "citrobacter                    46\n",
       "halorubrum                     45\n",
       "bacteroides                    43\n",
       "cronobacter                    39\n",
       "dickeya                        39\n",
       "corynebacterium                38\n",
       "helicobacter                   37\n",
       "brucella                       37\n",
       "pseudoalteromonas              33\n",
       "caulobacter                    33\n",
       "moraxella                      32\n",
       "achromobacter                  31\n",
       "haloarcula                     30\n",
       "providencia                    26\n",
       "pantoea                        24\n",
       "leuconostoc                    23\n",
       "roseobacter                    21\n",
       "brevundimonas                  20\n",
       "brevibacterium                 19\n",
       "acidianus                      18\n",
       "polaribacter                   18\n",
       "shewanella                     18\n",
       "liberibacter                   18\n",
       "edwardsiella                   17\n",
       "cutibacterium                  17\n",
       "alteromonas                    17\n",
       "thermus                        17\n",
       "wolbachia                      16\n",
       "prevotella                     15\n",
       "clostridioides                 15\n",
       "ruegeria                       15\n",
       "mycoplasma                     14\n",
       "agrobacterium                  13\n",
       "sinorhizobium                  13\n",
       "microcystis                    13\n",
       "mannheimia                     12\n",
       "bifidobacterium                11\n",
       "kosakonia                      11\n",
       "bordetella                     11\n",
       "brevibacillus                  10\n",
       "hafnia                         10\n",
       "leptospira                      9\n",
       "fusobacterium                   9\n",
       "pasteurella                     9\n",
       "salinibacter                    9\n",
       "xylella                         9\n",
       "faecalibacterium                8\n",
       "paracoccus                      8\n",
       "morganella                      8\n",
       "nodularia                       6\n",
       "dinoroseobacter                 6\n",
       "maribacter                      6\n",
       "spiroplasma                     6\n",
       "oenococcus                      6\n",
       "sphingomonas                    6\n",
       "nocardia                        6\n",
       "sulfitobacter                   6\n",
       "ruminococcus                    6\n",
       "geobacillus                     6\n",
       "butyrivibrio                    5\n",
       "chlamydia                       5\n",
       "curtobacterium                  5\n",
       "tenacibaculum                   5\n",
       "nitrosopumilus                  5\n",
       "tsukamurella                    5\n",
       "marinomonas                     5\n",
       "nitratiruptor                   5\n",
       "elizabethkingia                 4\n",
       "exiguobacterium                 4\n",
       "methanobacterium                4\n",
       "brochothrix                     4\n",
       "ochrobactrum                    4\n",
       "riemerella                      4\n",
       "haloferax                       4\n",
       "parabacteroides                 3\n",
       "glutamicibacter                 3\n",
       "metallosphaera                  3\n",
       "psychrobacter                   3\n",
       "delftia                         3\n",
       "halomonas                       3\n",
       "actinomyces                     3\n",
       "curvibacter                     3\n",
       "psychrobacillus                 3\n",
       "acidithiobacillus               3\n",
       "marinobacter                    3\n",
       "weissella                       3\n",
       "natrinema                       3\n",
       "aeropyrum                       3\n",
       "bdellovibrio                    3\n",
       "sodalis                         3\n",
       "carnobacterium                  3\n",
       "lentibacter                     3\n",
       "nostoc                          3\n",
       "hamiltonella                    3\n",
       "komagataeibacter                3\n",
       "acetobacter                     3\n",
       "haemophilus                     3\n",
       "acidovorax                      3\n",
       "raoultella                      3\n",
       "phormidium                      3\n",
       "alcaligenes                     3\n",
       "thermoproteus                   2\n",
       "marinitoga                      2\n",
       "mesorhizobium                   2\n",
       "halobacterium                   2\n",
       "sphaerotilus                    2\n",
       "plectonema                      2\n",
       "cylindrospermopsis              2\n",
       "methanosarcina                  2\n",
       "clavibacter                     2\n",
       "candidatus odinarchaeum         2\n",
       "paraglaciecola                  2\n",
       "pyrobaculum                     2\n",
       "croceibacter                    2\n",
       "rhodovulum                      2\n",
       "natrialba                       2\n",
       "meiothermus                     2\n",
       "erythrobacter                   2\n",
       "thermobifida                    2\n",
       "salinivibrio                    2\n",
       "xenohaliotis                    2\n",
       "silvanigrella                   2\n",
       "erysipelothrix                  2\n",
       "leclercia                       2\n",
       "oceanospirillum                 2\n",
       "roseovarius                     2\n",
       "thermococcus                    2\n",
       "tetraselmis                     2\n",
       "anabaena                        2\n",
       "roseburia                       2\n",
       "altiarchaeum                    2\n",
       "nonlabens                       2\n",
       "citromicrobium                  2\n",
       "octadecabacter                  2\n",
       "myxococcus                      2\n",
       "peptoclostridium                1\n",
       "sphingobium                     1\n",
       "nitrososphaera                  1\n",
       "azospirillum                    1\n",
       "methanothermobacter             1\n",
       "thalassomonas                   1\n",
       "parageobacillus                 1\n",
       "kluyvera                        1\n",
       "actinoplanes                    1\n",
       "thiobacimonas                   1\n",
       "pelagibaca                      1\n",
       "planktothrix                    1\n",
       "aquamicrobium                   1\n",
       "bradyrhizobium                  1\n",
       "listonella                      1\n",
       "rathayibacter                   1\n",
       "silicibacter                    1\n",
       "colwellia                       1\n",
       "rahnella                        1\n",
       "tetrasphaera                    1\n",
       "nitrincola                      1\n",
       "puniceispirillum                1\n",
       "halogranum                      1\n",
       "stygiolobus                     1\n",
       "loktanella                      1\n",
       "thermoanaerobacterium           1\n",
       "lysinibacillus                  1\n",
       "dunaliella                      1\n",
       "aurantimonas                    1\n",
       "saccharomonospora               1\n",
       "aggregatibacter                 1\n",
       "lelliottia                      1\n",
       "hydrogenobaculum                1\n",
       "pseudaeromonas                  1\n",
       "skermania                       1\n",
       "celeribacter                    1\n",
       "iodobacter                      1\n",
       "salisaeta                       1\n",
       "rhodoferax                      1\n",
       "acaryochloris                   1\n",
       "caldibacillus                   1\n",
       "pseudoxanthomonas               1\n",
       "aphanizomenon                   1\n",
       "gardnerella                     1\n",
       "micromonospora                  1\n",
       "cytobacillus                    1\n",
       "buttiauxella                    1\n",
       "tetragenococcus                 1\n",
       "haloterrigena                   1\n",
       "methanocaldococcus              1\n",
       "arthronema                      1\n",
       "protaetiibacter                 1\n",
       "luteibacter                     1\n",
       "pseudarthrobacter               1\n",
       "chryseobacterium                1\n",
       "arcanobacterium                 1\n",
       "pseudanabaena                   1\n",
       "cellulosimicrobium              1\n",
       "thermocrinis                    1\n",
       "sphingopyxis                    1\n",
       "lactiplantibacillus             1\n",
       "olleya                          1\n",
       "levilactobacillus               1\n",
       "winogradskyella                 1\n",
       "methanoculleus                  1\n",
       "stappia                         1\n",
       "solobacterium                   1\n",
       "kurthia                         1\n",
       "eggerthella                     1\n",
       "janthinobacterium               1\n",
       "plesiomonas                     1\n",
       "virgibacillus                   1\n",
       "lentisphaerae                   1\n",
       "gluconobacter                   1\n",
       "pontimonas                      1\n",
       "hydrogenophilus                 1\n",
       "anoxybacillus                   1\n",
       "halogeometricum                 1\n",
       "mastigocladus                   1\n",
       "candidatus azobacteroides       1\n",
       "pararheinheimera                1\n",
       "akkermansia                     1\n",
       "verrucomicrobia                 1\n",
       "aerococcus                      1\n",
       "saccharolobus                   1\n",
       "lokiarchaeota                   1\n",
       "rothia                          1\n",
       "aeribacillus                    1\n",
       "desulfovibrio                   1\n",
       "photobacterium                  1\n",
       "pediococcus                     1\n",
       "cedecea                         1\n",
       "sporosarcina                    1\n",
       "mycolicibacterium               1\n",
       "phaeobacter                     1\n",
       "aminobacter                     1\n",
       "pyrococcus                      1\n",
       "Name: Host, dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inphared['Host'].replace(util.get_ncbi_standard_nomenclature(constants.NCBI_STANDARD_NOMENCLATURE), inplace = True)\n",
    "inphared['Host'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d87b69a",
   "metadata": {},
   "source": [
    "Just verifying that all the entries are unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2e26de54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inphared.shape[0] == inphared['Accession'].nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b491fa6a",
   "metadata": {},
   "source": [
    "###  Some Statistics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d6e0865",
   "metadata": {},
   "source": [
    "Statistics on phages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2ed9a730",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original INPHARED:\t\t 18389\n",
      "With Specified Hosts:\t\t 15739\n",
      "Plus Manually Filtered Hosts:\t 15823\n"
     ]
    }
   ],
   "source": [
    "print(\"Original INPHARED:\\t\\t\", inphared_raw.shape[0])\n",
    "print(\"With Specified Hosts:\\t\\t\", inphared_raw.shape[0] - inphared_unspec_host.shape[0])\n",
    "print(\"Plus Manually Filtered Hosts:\\t\", inphared.shape[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46b8210f",
   "metadata": {},
   "source": [
    "Statistics on host genera"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "959eff7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num. of Unique Host Genera:\t 279\n"
     ]
    }
   ],
   "source": [
    "print(\"Num. of Unique Host Genera:\\t\", inphared['Host'].nunique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ab54a0c",
   "metadata": {},
   "source": [
    "Save the entries with host information to a CSV file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d8c61558",
   "metadata": {},
   "outputs": [],
   "source": [
    "inphared.to_csv(os.path.join(constants.TEMP_PREPROCESSING, constants.INPHARED_WITH_HOSTS), index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a235cca2",
   "metadata": {},
   "source": [
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "623f78dc",
   "metadata": {},
   "source": [
    "# Part III: Selection of Annotated Receptor-Binding Proteins (RBPs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f008255d",
   "metadata": {},
   "source": [
    "Load only the phage entries with host data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "c7c13f80",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(15823, 27)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inphared = pd.read_csv(f'{constants.TEMP_PREPROCESSING}/{constants.INPHARED_WITH_HOSTS}')\n",
    "util.set_inphared(inphared)\n",
    "\n",
    "inphared.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec56e781",
   "metadata": {},
   "source": [
    "Identify which phage entries have CDS information and which do not."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "aebf4985",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed 1000 records\n",
      "Processed 2000 records\n",
      "Processed 3000 records\n",
      "Processed 4000 records\n",
      "Processed 5000 records\n",
      "Processed 6000 records\n",
      "Processed 7000 records\n",
      "Processed 8000 records\n",
      "Processed 9000 records\n",
      "Processed 10000 records\n",
      "Processed 11000 records\n",
      "Processed 12000 records\n",
      "Processed 13000 records\n",
      "Processed 14000 records\n",
      "Processed 15000 records\n",
      "Processed 16000 records\n",
      "Processed 17000 records\n",
      "Processed 18000 records\n",
      "Processed 19000 records\n",
      "Processed 20000 records\n",
      "Processed 21000 records\n",
      "Processed 22000 records\n",
      "Processed 23000 records\n",
      "Processed 24000 records\n",
      "Processed 25000 records\n",
      "Processed 26000 records\n",
      "Processed 27000 records\n",
      "Processed 28000 records\n",
      "Processed 29000 records\n",
      "Processed 30000 records\n",
      "Processed 31000 records\n"
     ]
    }
   ],
   "source": [
    "no_cds_annot = util.get_no_cds_annot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "fe152e83",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(constants.TEMP_PREPROCESSING, constants.NO_CDS_ANNOT),'wb') as no_cds_annot_file:\n",
    "    pickle.dump(no_cds_annot, no_cds_annot_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d3a8d3e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f'{constants.TEMP_PREPROCESSING}/{constants.NO_CDS_ANNOT}','rb') as no_cds_annot_file:\n",
    "    no_cds_annot = pickle.load(no_cds_annot_file)\n",
    "    \n",
    "util.set_no_cds_annot(no_cds_annot)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04157f61",
   "metadata": {},
   "source": [
    "### Some statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "75111c9c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Without CDS info:\t 665\n",
      "With CDS info:\t\t 15158\n"
     ]
    }
   ],
   "source": [
    "print(\"Without CDS info:\\t\", len(no_cds_annot))\n",
    "print(\"With CDS info:\\t\\t\", inphared.shape[0] - len(no_cds_annot))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd16b5b1",
   "metadata": {},
   "source": [
    "## *A. Process entries with CDS information*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6bd4a85",
   "metadata": {},
   "source": [
    "- Use regex from this [RBP prediction study](https://www.mdpi.com/1999-4915/14/6/1329) by Boeckaerts <i>et al.</i> (2022) to select the annotated RBPs\n",
    "   - We modify this regex to:\n",
    "     - Make more accommodating to typos (e.g., look for `bind` instead of `binding`)\n",
    "     - Allow multiple spaces between tokens\n",
    "   - For comparison:\n",
    "     - Original regex: `tail.?(?:spike|fib(?:er|re))|^recept(?:o|e)r.?(?:binding|recognizing).*(?:protein)?|^RBP`\n",
    "     - Modified regex: `tail?(.?|\\s*)(?:spike?|fib(?:er|re))|recept(?:o|e)r(.?|\\s*)(?:bind|recogn).*(?:protein)?|(?<!\\w)RBP(?!a)`\n",
    "     - Annotations captured by modified regex but not by old regex:\n",
    "       - `receptor recognition protein`\n",
    "   - The same [study](https://www.mdpi.com/1999-4915/14/6/1329) has an exclusion list that covers proteins related to RBPs but are not RBPs themselves (e.g., assembly and portal proteins).\n",
    "       <br><br>\n",
    "\n",
    "- If product does not match regex for RBP, check if it is a hypothetical protein\n",
    "   - The same [study](https://www.mdpi.com/1999-4915/14/6/1329) has a list of keywords associated with hypothetical proteins\n",
    "      - We remove `putative`, `probable`, and `probably` since they are indicative of a product with a putative (rather than an unknown) function.\n",
    "   - Make more accommodating to typos by allowing keywords with minimum edit distance ‚â§ 2 (likely misspellings)\n",
    "   - Create a manual list to handle cases where the keyword `hypothetical` is present but the putative function is already given, as in the case of:\n",
    "       - `conserved hypothetical lipoprotein`\n",
    "       - `hypothetical host-like ribonucleoside diphosphate reductase`       <br><br>\n",
    "      \n",
    "- Predict whether these hypothetical proteins are RBPs using the XGBoost model from the [RBP prediction study](https://www.mdpi.com/1999-4915/14/6/1329) by Boeckaerts <i>et al.</i> (2022)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "698139cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "util.set_token_delimiter(constants.TOKEN_DELIMITER)\n",
    "util.construct_keyword_list(constants.HYPOTHETICAL_KEYWORDS, constants.RBP_RELATED_NOT_RBP, constants.PUTATIVE_FUNCTIONS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9168194f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed 1000 records\n",
      "Processed 2000 records\n",
      "Processed 3000 records\n",
      "Processed 4000 records\n",
      "Processed 5000 records\n",
      "Processed 6000 records\n",
      "Processed 7000 records\n",
      "Processed 8000 records\n",
      "Processed 9000 records\n",
      "Processed 10000 records\n",
      "Processed 11000 records\n",
      "Processed 12000 records\n",
      "Processed 13000 records\n",
      "Processed 14000 records\n",
      "Processed 15000 records\n",
      "Processed 16000 records\n",
      "Processed 17000 records\n",
      "Processed 18000 records\n",
      "Processed 19000 records\n",
      "Processed 20000 records\n",
      "Processed 21000 records\n",
      "Processed 22000 records\n",
      "Processed 23000 records\n"
     ]
    }
   ],
   "source": [
    "annot_products = util.get_annot_products()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ec1531f",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(constants.TEMP_PREPROCESSING, constants.ANNOT_PRODUCTS),'wb') as annot_products_file:\n",
    "    pickle.dump(annot_products, annot_products_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e4f85bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f'{constants.TEMP_PREPROCESSING}/{constants.ANNOT_PRODUCTS}','rb') as annot_products_file:\n",
    "    annot_products = pickle.load(annot_products_file)\n",
    "    \n",
    "util.set_annot_products(annot_products)\n",
    "for annot_product in sorted(list(annot_products)):\n",
    "    print(annot_product)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bd3f907",
   "metadata": {},
   "outputs": [],
   "source": [
    "rbp_products, hypothetical_proteins = util.get_rbp_hypothetical_proteins(constants.RBP_REGEX)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0fc6979",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(constants.TEMP_PREPROCESSING, constants.RBP_PRODUCTS),'wb') as rbp_products_file:\n",
    "    pickle.dump(rbp_products, rbp_products_file)\n",
    "    \n",
    "with open(os.path.join(constants.TEMP_PREPROCESSING, constants.HYPOTHETICAL_PRODUCTS),'wb') as hypothetical_proteins_file:\n",
    "    pickle.dump(hypothetical_proteins, hypothetical_proteins_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2b1a308",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f'{constants.TEMP_PREPROCESSING}/{constants.RBP_PRODUCTS}','rb') as rbp_products_file:\n",
    "    rbp_products = pickle.load(rbp_products_file)\n",
    "    \n",
    "util.set_rbp_products(rbp_products)\n",
    "rbp_products"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bf26916",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f'{constants.TEMP_PREPROCESSING}/{constants.HYPOTHETICAL_PRODUCTS}','rb') as hypothetical_proteins_file:\n",
    "    hypothetical_proteins = pickle.load(hypothetical_proteins_file)\n",
    "    \n",
    "util.set_hypothetical_proteins(hypothetical_proteins)\n",
    "for protein in hypothetical_proteins:\n",
    "    print(protein)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9983222e",
   "metadata": {},
   "source": [
    "### Get distribution of RBP lengths"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ee010fe",
   "metadata": {},
   "source": [
    "This preliminary analysis of the distribution of the RBP lengths covers both the sequences with CDS information and those where the CDS information features are predicted via Prokka. We carry it now in order since we need the lower and upper bounds for RBP lengths to generate the FASTA files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99b5f194",
   "metadata": {},
   "outputs": [],
   "source": [
    "len_distribution = util.generate_rbp_len_distribution()\n",
    "util.generate_rbp_len_distribution_prokka(len_distribution, constants.INPHARED_GENOME)\n",
    "\n",
    "lengths = []\n",
    "for length, freq in len_distribution.items():\n",
    "    for _ in range(freq):\n",
    "        lengths.append(length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eda06d32",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(constants.TEMP_PREPROCESSING, constants.RBP_LENGTHS),'wb') as rbp_lengths_file:\n",
    "    pickle.dump(lengths, rbp_lengths_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e287018",
   "metadata": {},
   "source": [
    "Statistically compute the lower and upper bounds for RBP lengths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c277da30",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f'{constants.TEMP_PREPROCESSING}/{constants.RBP_LENGTHS}','rb') as rbp_lengths_file:\n",
    "    lengths = pickle.load(rbp_lengths_file)\n",
    "\n",
    "Q1 = np.percentile(lengths, 25, interpolation = 'midpoint')\n",
    "Q3 = np.percentile(lengths, 75, interpolation = 'midpoint')\n",
    "IQR = Q3 - Q1\n",
    "\n",
    "lower_bound = Q1 - 1.5 * IQR\n",
    "upper_bound = Q3 + 1.5 * IQR\n",
    "\n",
    "print(\"Lower bound:\", lower_bound)\n",
    "print(\"Upper bound:\", upper_bound)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1173daf",
   "metadata": {},
   "source": [
    "### Generate FASTA for entries with CDS information"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "908b6c9b",
   "metadata": {},
   "source": [
    "Generate FASTA files containing protein sequences of hypothetical proteins and RBPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "206ad80f",
   "metadata": {},
   "outputs": [],
   "source": [
    "RBP_NEW_DIR = f'{constants.INPHARED}/{constants.FASTA}/{constants.RBP}/{constants.GENBANK}'\n",
    "HYPOTHETICAL_NEW_DIR = f'{constants.INPHARED}/{constants.FASTA}/{constants.HYPOTHETICAL}/{constants.GENBANK}'\n",
    "\n",
    "if not os.path.exists(RBP_NEW_DIR):\n",
    "    os.mkdir(RBP_NEW_DIR)\n",
    "    \n",
    "if not os.path.exists(HYPOTHETICAL_NEW_DIR):\n",
    "    os.mkdir(HYPOTHETICAL_NEW_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6722caa",
   "metadata": {},
   "outputs": [],
   "source": [
    "util.generate_rbp_hypothetical_fasta(RBP_NEW_DIR, HYPOTHETICAL_NEW_DIR, constants.LOWER_BOUND_RBP_LENGTH, \n",
    "                                     constants.UPPER_BOUND_RBP_LENGTH)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b12fedc0",
   "metadata": {},
   "source": [
    "The FASTA files are stored in the `inphared/fasta` directory, specifically in the `genbank` subdirectories."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96e10daa",
   "metadata": {},
   "source": [
    "## *B. Process entries without CDS information*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "909953a5",
   "metadata": {},
   "source": [
    "- Annotate genes using [Prokka](https://academic.oup.com/bioinformatics/article/30/14/2068/2390517), which calls [Prodigal](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-11-119) to predict coordinates of ORFs.\n",
    "- Configure Prokka to perform functional annotation using [PHROGs](https://academic.oup.com/nargab/article/3/3/lqab067/6342220).\n",
    "- Process in the same manner as the entries with CDS information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a20cf782",
   "metadata": {},
   "outputs": [],
   "source": [
    "RBP_NEW_DIR = f'{constants.INPHARED}/{constants.FASTA}/{constants.RBP}/{constants.PROKKA}'\n",
    "HYPOTHETICAL_NEW_DIR = f'{constants.INPHARED}/{constants.FASTA}/{constants.HYPOTHETICAL}/{constants.PROKKA}'\n",
    "\n",
    "if not os.path.exists(RBP_NEW_DIR):\n",
    "    os.mkdir(RBP_NEW_DIR)\n",
    "    \n",
    "if not os.path.exists(HYPOTHETICAL_NEW_DIR):\n",
    "    os.mkdir(HYPOTHETICAL_NEW_DIR)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a93dc14",
   "metadata": {},
   "source": [
    "Analyze the gene product annotations returned by PHROG."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6b737c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "annot_products_prokka = util.get_annot_products_prokka(constants.INPHARED_GENOME)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45276189",
   "metadata": {},
   "outputs": [],
   "source": [
    "x, y = util.get_rbp_hypothetical_proteins_prokka(constants.RBP_REGEX, annot_products_prokka)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "549ac5bd",
   "metadata": {},
   "source": [
    "Combine the keywords for RBPs and hypothetical proteins from GenBank (i.e., entries with CDS information) and PHROG annotations (i.e., entries without CDS information)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afbeee25",
   "metadata": {},
   "outputs": [],
   "source": [
    "util.set_rbp_products(rbp_products.union(x))\n",
    "util.set_hypothetical_proteins(hypothetical_proteins.union(y))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9cb71e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "rbp_products = rbp_products.union(x)\n",
    "hypothetical_proteins = hypothetical_proteins.union(y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1748d5e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(constants.TEMP_PREPROCESSING, constants.RBP_PRODUCTS),'wb') as rbp_products_file:\n",
    "    pickle.dump(rbp_products, rbp_products_file)\n",
    "    \n",
    "with open(os.path.join(constants.TEMP_PREPROCESSING, constants.HYPOTHETICAL_PRODUCTS),'wb') as hypothetical_proteins_file:\n",
    "    pickle.dump(hypothetical_proteins, hypothetical_proteins_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bee63c0",
   "metadata": {},
   "source": [
    "### Generate FASTA for entries without CDS information"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4844b2b0",
   "metadata": {},
   "source": [
    "Generate FASTA files containing protein sequences of hypothetical proteins and RBPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc3f8c6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "util.generate_rbp_hypothetical_fasta_prokka(constants.INPHARED_GENOME, RBP_NEW_DIR, HYPOTHETICAL_NEW_DIR, \n",
    "                                            lower_bound, upper_bound)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31c45553",
   "metadata": {},
   "source": [
    "The FASTA files are stored in the `inphared/fasta` directory, specifically in the `prokka` subdirectories."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "051d9040",
   "metadata": {},
   "source": [
    "### C. *Generate FFN containing the genomes of RBPs and hypothetical proteins*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a259f45",
   "metadata": {},
   "outputs": [],
   "source": [
    "util.generate_rbp_hypothetical_nucleotide(f'{constants.INPHARED}/{constants.FASTA}/{constants.NUCLEOTIDE}/{constants.RBP}/{constants.GENBANK}',\n",
    "                                          f'{constants.INPHARED}/{constants.FASTA}/{constants.NUCLEOTIDE}/{constants.HYPOTHETICAL}/{constants.GENBANK}', \n",
    "                                          lower_bound, upper_bound)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71029763",
   "metadata": {},
   "source": [
    "The FFN files are stored in the `inphared/fasta/nucleotide` directory, specifically in the `genbank` subdirectories."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "103cebbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "util.generate_rbp_hypothetical_nucleotide_prokka(constants.INPHARED_GENOME,\n",
    "                                                 f'{constants.INPHARED}/{constants.FASTA}/{constants.NUCLEOTIDE}/{constants.RBP}/{constants.PROKKA}',\n",
    "                                                 f'{constants.INPHARED}/{constants.FASTA}/{constants.NUCLEOTIDE}/{constants.HYPOTHETICAL}/{constants.PROKKA}', \n",
    "                                                 lower_bound, upper_bound)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9541defa",
   "metadata": {},
   "source": [
    "The FFN files are stored in the `inphared/fasta/nucleotide` directory, specifically in the `prokka` subdirectories."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
