{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c6e73016",
   "metadata": {},
   "source": [
    "# Protein embeddings improve phage-host interaction prediction\n",
    "\n",
    "**Mark Edward M. Gonzales<sup>1, 2</sup>, Jennifer C. Ureta<sup>1, 2</sup> & Anish M.S. Shrestha<sup>1, 2</sup>**\n",
    "\n",
    "<sup>1</sup> Bioinformatics Laboratory, Advanced Research Institute for Informatics, Computing and Networking, De La Salle University, Manila, Philippines <br>\n",
    "<sup>2</sup> Department of Software Technology, College of Computer Studies, De La Salle University, Manila, Philippines \n",
    "\n",
    "{mark_gonzales, jennifer.ureta, anish.shrestha}@dlsu.edu.ph"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0aa9a66",
   "metadata": {},
   "source": [
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "008ba397",
   "metadata": {},
   "source": [
    "## üí° Phage-Host-Features CSV Files\n",
    "\n",
    "This notebook assumes that you already have the phage-host-features CSV files (from running [`5. Data Consolidation.ipynb`](https://github.com/bioinfodlsu/phage-host-prediction/blob/main/experiments/5.%20Data%20Consolidation.ipynb)).\n",
    "\n",
    "Alternatively, you may download the CSV files from [Google Drive](https://drive.google.com/drive/folders/1xNoA6dxkN4jzVNCg_7YNjdPZzl51Jo9M?usp=sharing) and save the downloaded `data` folder inside the `inphared` directory located in the same folder as this notebook. The folder structure should look like this:\n",
    "\n",
    "`experiments` (parent folder of this notebook) <br> \n",
    "‚Ü≥ `inphared` <br>\n",
    "&nbsp; &nbsp;‚Ü≥ `data` <br>\n",
    "&nbsp; &nbsp;&nbsp; &nbsp; ‚Ü≥ `rbp.csv` <br>\n",
    "&nbsp; &nbsp;&nbsp; &nbsp; ‚Ü≥ `rbp_embeddings_esm.csv` <br>\n",
    "&nbsp; &nbsp;&nbsp; &nbsp; ‚Ü≥ ... <br>\n",
    "‚Ü≥ `6. Classifier Building & Evaluation.ipynb` (this notebook) <br>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19091a16",
   "metadata": {},
   "source": [
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c032873",
   "metadata": {},
   "source": [
    "## üìÅ Output Files\n",
    "\n",
    "If you would like to skip running this notebook, you may download the trained models from [Google Drive](https://drive.google.com/drive/folders/1U5ugmkhD4LHElYnLj3B8Xt2TcPx-TOjB?usp=sharing). Save the downloaded `models` folder in the same folder as this notebook. The folder structure should look like this:\n",
    "\n",
    "`experiments` (parent folder of this notebook) <br> \n",
    "‚Ü≥ `inphared` <br>\n",
    "‚Ü≥ `models` <br>\n",
    "&nbsp; &nbsp;‚Ü≥ `boeckaerts.joblib` <br>\n",
    "&nbsp; &nbsp;‚Ü≥ `esm.joblib` <br>\n",
    "&nbsp; &nbsp;‚Ü≥ ... <br>\n",
    "‚Ü≥ `6. Classifier Building & Evaluation.ipynb` (this notebook) <br>\n",
    "\n",
    "Intermediate output files (i.e., `temp/feature_importance.pickle` and those saved in `temp/results`) should have already been included when the repository was cloned.\n",
    "\n",
    "This notebook also generates `rbp_embeddings_boeckaerts.csv`, the phage-host-features CSV file for the state-of-the-art tool by Boeckaerts <i>et al.</i> (2021) with which we compared model performance. It should have already been downloaded as part of the [`data`](https://drive.google.com/drive/folders/1xNoA6dxkN4jzVNCg_7YNjdPZzl51Jo9M?usp=sharing) folder."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7be81b08",
   "metadata": {},
   "source": [
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad4d04c3",
   "metadata": {},
   "source": [
    "## Part I: Preliminaries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad0d4208",
   "metadata": {},
   "source": [
    "Import the necessary libraries and modules."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "367c759b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Edward\\anaconda3\\lib\\site-packages\\Bio\\SubsMat\\__init__.py:131: BiopythonDeprecationWarning: Bio.SubsMat has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.substitution_matrices as a replacement, and contact the Biopython developers if you still need the Bio.SubsMat module.\n",
      "  BiopythonDeprecationWarning,\n"
     ]
    }
   ],
   "source": [
    "import math\n",
    "import pickle\n",
    "import os\n",
    "import warnings\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sklearn\n",
    "\n",
    "from ConstantsUtil import ConstantsUtil\n",
    "from ClassificationUtil import ClassificationUtil\n",
    "import boeckaerts as RBP_f\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c1293e00",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_rows', None)\n",
    "pd.set_option('display.max_columns', 50)\n",
    "\n",
    "pd.options.mode.chained_assignment = None\n",
    "\n",
    "with warnings.catch_warnings(): \n",
    "    warnings.filterwarnings(\"ignore\", category=sklearn.exceptions.UndefinedMetricWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "34109ade",
   "metadata": {},
   "outputs": [],
   "source": [
    "constants = ConstantsUtil()\n",
    "util = ClassificationUtil()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c532db73",
   "metadata": {},
   "source": [
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc372796",
   "metadata": {},
   "source": [
    "## Part II: Classifier Building and Evaluation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e34d2522",
   "metadata": {},
   "source": [
    "Train a random forest model with the embeddings of the RBPs as the input and the host as the output. Display the results as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "83173f1f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Confidence threshold k: 0.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.3529    0.5217        17\n",
      "    acinetobacter     0.9385    0.6224    0.7485        98\n",
      "        aeromonas     0.8485    0.5895    0.6957        95\n",
      "    agrobacterium     1.0000    0.4167    0.5882        12\n",
      "     arthrobacter     0.9500    0.7308    0.8261        26\n",
      "         bacillus     0.7215    0.8382    0.7755       204\n",
      "      bacteroides     0.9032    0.9032    0.9032        31\n",
      "    brevundimonas     1.0000    0.6000    0.7500        10\n",
      "     burkholderia     0.6333    0.4872    0.5507        39\n",
      "    campylobacter     0.8559    0.8145    0.8347       124\n",
      "      caulobacter     1.0000    0.8400    0.9130        25\n",
      "      citrobacter     0.3077    0.2353    0.2667        34\n",
      "      clostridium     0.5000    0.6667    0.5714        24\n",
      "      cronobacter     0.5000    0.3333    0.4000        36\n",
      "          dickeya     0.8780    0.6923    0.7742        52\n",
      "     edwardsiella     1.0000    0.4000    0.5714        10\n",
      "     enterobacter     0.4874    0.4085    0.4444       142\n",
      "     enterococcus     0.9167    0.8980    0.9072        49\n",
      "          erwinia     0.6980    0.8566    0.7692       286\n",
      "      escherichia     0.5173    0.8301    0.6374      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     0.5513    0.8113    0.6565        53\n",
      "       halorubrum     0.3333    0.6154    0.4324        13\n",
      "     helicobacter     0.9412    1.0000    0.9697        16\n",
      "       klebsiella     0.7201    0.8035    0.7595       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.7692    0.4348    0.5556        23\n",
      "      lactococcus     0.9326    0.8300    0.8783       100\n",
      "      leuconostoc     1.0000    0.9000    0.9474        10\n",
      "         listeria     0.8889    0.8571    0.8727        28\n",
      "   microbacterium     0.8621    0.7576    0.8065        33\n",
      "    mycobacterium     0.9585    0.9315    0.9448       248\n",
      "           others     0.0000    0.0000    0.0000       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.8571    0.6761    0.7559        71\n",
      "     pelagibacter     0.6316    0.7273    0.6761        33\n",
      "       prevotella     0.8947    0.9444    0.9189        18\n",
      "  prochlorococcus     0.9394    0.5962    0.7294        52\n",
      "          proteus     0.8077    0.5250    0.6364        40\n",
      "      providencia     0.6667    0.2500    0.3636        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.7435    0.8928    0.8113       513\n",
      "        ralstonia     0.7667    0.5476    0.6389        42\n",
      "     rheinheimera     1.0000    1.0000    1.0000       176\n",
      "        rhizobium     0.9242    0.8243    0.8714        74\n",
      "      rhodobacter     0.8333    0.8824    0.8571        34\n",
      "      rhodococcus     1.0000    0.0769    0.1429        13\n",
      "       salmonella     0.8333    0.7991    0.8158       632\n",
      "         serratia     0.7857    0.5670    0.6587        97\n",
      "       shewanella     1.0000    0.2727    0.4286        11\n",
      "         shigella     0.4458    0.2587    0.3274       143\n",
      "   staphylococcus     0.7295    0.9836    0.8377       244\n",
      " stenotrophomonas     0.7917    0.4634    0.5846        41\n",
      "    streptococcus     0.8289    0.7778    0.8025        81\n",
      "     streptomyces     0.9697    0.7442    0.8421        43\n",
      "    synechococcus     0.7578    0.9731    0.8521       521\n",
      "           vibrio     0.6612    0.8769    0.7539       463\n",
      "      xanthomonas     0.6739    0.5345    0.5962        58\n",
      "         yersinia     0.5315    0.6344    0.5784        93\n",
      "\n",
      "         accuracy                         0.6970      8116\n",
      "        macro avg     0.7574    0.6329    0.6598      8116\n",
      "     weighted avg     0.6326    0.6970    0.6500      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 10.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2941    0.4545        17\n",
      "    acinetobacter     1.0000    0.5306    0.6933        98\n",
      "        aeromonas     0.9623    0.5368    0.6892        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.6923    0.8182        26\n",
      "         bacillus     0.8778    0.7745    0.8229       204\n",
      "      bacteroides     1.0000    0.7742    0.8727        31\n",
      "    brevundimonas     1.0000    0.4000    0.5714        10\n",
      "     burkholderia     0.9375    0.3846    0.5455        39\n",
      "    campylobacter     0.9167    0.7984    0.8534       124\n",
      "      caulobacter     1.0000    0.8400    0.9130        25\n",
      "      citrobacter     0.4167    0.1471    0.2174        34\n",
      "      clostridium     0.5417    0.5417    0.5417        24\n",
      "      cronobacter     0.6000    0.2500    0.3529        36\n",
      "          dickeya     0.9231    0.6923    0.7912        52\n",
      "     edwardsiella     1.0000    0.2000    0.3333        10\n",
      "     enterobacter     0.5106    0.3380    0.4068       142\n",
      "     enterococcus     0.9773    0.8776    0.9247        49\n",
      "          erwinia     0.9004    0.7902    0.8417       286\n",
      "      escherichia     0.7367    0.7606    0.7485      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     0.9750    0.7358    0.8387        53\n",
      "       halorubrum     0.3529    0.4615    0.4000        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.9148    0.7028    0.7949       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.7500    0.2609    0.3871        23\n",
      "      lactococcus     1.0000    0.8100    0.8950       100\n",
      "      leuconostoc     1.0000    0.8000    0.8889        10\n",
      "         listeria     0.8889    0.8571    0.8727        28\n",
      "   microbacterium     1.0000    0.7576    0.8621        33\n",
      "    mycobacterium     0.9870    0.9153    0.9498       248\n",
      "           others     0.3668    0.7373    0.4899       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.8936    0.5915    0.7119        71\n",
      "     pelagibacter     0.7500    0.7273    0.7385        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9333    0.5385    0.6829        52\n",
      "          proteus     0.8636    0.4750    0.6129        40\n",
      "      providencia     0.6667    0.2500    0.3636        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9501    0.8538    0.8994       513\n",
      "        ralstonia     0.9444    0.4048    0.5667        42\n",
      "     rheinheimera     1.0000    1.0000    1.0000       176\n",
      "        rhizobium     0.9818    0.7297    0.8372        74\n",
      "      rhodobacter     0.9355    0.8529    0.8923        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.8984    0.7278    0.8042       632\n",
      "         serratia     0.8548    0.5464    0.6667        97\n",
      "       shewanella     1.0000    0.1818    0.3077        11\n",
      "         shigella     0.5410    0.2308    0.3235       143\n",
      "   staphylococcus     0.8741    0.9672    0.9183       244\n",
      " stenotrophomonas     0.9375    0.3659    0.5263        41\n",
      "    streptococcus     0.9231    0.7407    0.8219        81\n",
      "     streptomyces     1.0000    0.6512    0.7887        43\n",
      "    synechococcus     0.8817    0.9443    0.9120       521\n",
      "           vibrio     0.8443    0.7732    0.8072       463\n",
      "      xanthomonas     0.8182    0.4655    0.5934        58\n",
      "         yersinia     0.5657    0.6022    0.5833        93\n",
      "\n",
      "         accuracy                         0.7357      8116\n",
      "        macro avg     0.8236    0.5867    0.6573      8116\n",
      "     weighted avg     0.7931    0.7357    0.7439      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 20.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2941    0.4545        17\n",
      "    acinetobacter     1.0000    0.4694    0.6389        98\n",
      "        aeromonas     0.9535    0.4316    0.5942        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.6154    0.7619        26\n",
      "         bacillus     0.9437    0.7402    0.8297       204\n",
      "      bacteroides     1.0000    0.7742    0.8727        31\n",
      "    brevundimonas     1.0000    0.3000    0.4615        10\n",
      "     burkholderia     1.0000    0.3590    0.5283        39\n",
      "    campylobacter     0.9417    0.7823    0.8546       124\n",
      "      caulobacter     1.0000    0.7600    0.8636        25\n",
      "      citrobacter     0.6250    0.1471    0.2381        34\n",
      "      clostridium     0.6316    0.5000    0.5581        24\n",
      "      cronobacter     0.7500    0.2500    0.3750        36\n",
      "          dickeya     0.9722    0.6731    0.7955        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.5844    0.3169    0.4110       142\n",
      "     enterococcus     1.0000    0.7551    0.8605        49\n",
      "          erwinia     0.9505    0.7378    0.8307       286\n",
      "      escherichia     0.8258    0.6880    0.7506      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.6604    0.7955        53\n",
      "       halorubrum     0.3529    0.4615    0.4000        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.9430    0.6247    0.7515       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.8333    0.2174    0.3448        23\n",
      "      lactococcus     1.0000    0.7900    0.8827       100\n",
      "      leuconostoc     1.0000    0.8000    0.8889        10\n",
      "         listeria     0.9231    0.8571    0.8889        28\n",
      "   microbacterium     1.0000    0.6970    0.8214        33\n",
      "    mycobacterium     0.9869    0.9113    0.9476       248\n",
      "           others     0.3273    0.8966    0.4795       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.8936    0.5915    0.7119        71\n",
      "     pelagibacter     0.8148    0.6667    0.7333        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9310    0.5192    0.6667        52\n",
      "          proteus     0.8571    0.4500    0.5902        40\n",
      "      providencia     0.6000    0.1875    0.2857        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9885    0.8363    0.9060       513\n",
      "        ralstonia     0.9444    0.4048    0.5667        42\n",
      "     rheinheimera     1.0000    1.0000    1.0000       176\n",
      "        rhizobium     1.0000    0.6622    0.7967        74\n",
      "      rhodobacter     0.9355    0.8529    0.8923        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9252    0.7041    0.7996       632\n",
      "         serratia     0.8475    0.5155    0.6410        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5714    0.1958    0.2917       143\n",
      "   staphylococcus     0.9707    0.9508    0.9607       244\n",
      " stenotrophomonas     0.9167    0.2683    0.4151        41\n",
      "    streptococcus     1.0000    0.7160    0.8345        81\n",
      "     streptomyces     1.0000    0.5814    0.7353        43\n",
      "    synechococcus     0.9498    0.9079    0.9284       521\n",
      "           vibrio     0.9329    0.6911    0.7940       463\n",
      "      xanthomonas     0.8710    0.4655    0.6067        58\n",
      "         yersinia     0.5730    0.5484    0.5604        93\n",
      "\n",
      "         accuracy                         0.7166      8116\n",
      "        macro avg     0.8317    0.5535    0.6344      8116\n",
      "     weighted avg     0.8293    0.7166    0.7361      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 30.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2941    0.4545        17\n",
      "    acinetobacter     1.0000    0.4184    0.5899        98\n",
      "        aeromonas     0.9714    0.3579    0.5231        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.5000    0.6667        26\n",
      "         bacillus     0.9467    0.6961    0.8023       204\n",
      "      bacteroides     1.0000    0.7419    0.8519        31\n",
      "    brevundimonas     1.0000    0.2000    0.3333        10\n",
      "     burkholderia     1.0000    0.2821    0.4400        39\n",
      "    campylobacter     0.9394    0.7500    0.8341       124\n",
      "      caulobacter     1.0000    0.6800    0.8095        25\n",
      "      citrobacter     0.6000    0.0882    0.1538        34\n",
      "      clostridium     0.6875    0.4583    0.5500        24\n",
      "      cronobacter     0.6667    0.1111    0.1905        36\n",
      "          dickeya     0.9706    0.6346    0.7674        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.6154    0.2817    0.3865       142\n",
      "     enterococcus     1.0000    0.6939    0.8193        49\n",
      "          erwinia     0.9608    0.6853    0.8000       286\n",
      "      escherichia     0.8663    0.6255    0.7265      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.6415    0.7816        53\n",
      "       halorubrum     0.4000    0.4615    0.4286        13\n",
      "     helicobacter     1.0000    0.9375    0.9677        16\n",
      "       klebsiella     0.9442    0.5970    0.7315       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.8333    0.2174    0.3448        23\n",
      "      lactococcus     1.0000    0.7700    0.8701       100\n",
      "      leuconostoc     1.0000    0.6000    0.7500        10\n",
      "         listeria     0.9200    0.8214    0.8679        28\n",
      "   microbacterium     1.0000    0.6970    0.8214        33\n",
      "    mycobacterium     0.9868    0.9032    0.9432       248\n",
      "           others     0.2940    0.9402    0.4479       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9268    0.5352    0.6786        71\n",
      "     pelagibacter     0.8000    0.4848    0.6038        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9310    0.5192    0.6667        52\n",
      "          proteus     0.8824    0.3750    0.5263        40\n",
      "      providencia     0.5000    0.1250    0.2000        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9952    0.8070    0.8913       513\n",
      "        ralstonia     0.9375    0.3571    0.5172        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     1.0000    0.6351    0.7769        74\n",
      "      rhodobacter     0.9286    0.7647    0.8387        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9256    0.6693    0.7769       632\n",
      "         serratia     0.8519    0.4742    0.6093        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5778    0.1818    0.2766       143\n",
      "   staphylococcus     0.9956    0.9262    0.9597       244\n",
      " stenotrophomonas     0.8889    0.1951    0.3200        41\n",
      "    streptococcus     1.0000    0.6667    0.8000        81\n",
      "     streptomyces     1.0000    0.5116    0.6769        43\n",
      "    synechococcus     0.9781    0.8560    0.9130       521\n",
      "           vibrio     0.9767    0.6328    0.7680       463\n",
      "      xanthomonas     0.8929    0.4310    0.5814        58\n",
      "         yersinia     0.6133    0.4946    0.5476        93\n",
      "\n",
      "         accuracy                         0.6836      8116\n",
      "        macro avg     0.8340    0.5111    0.6002      8116\n",
      "     weighted avg     0.8389    0.6836    0.7107      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 40.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2941    0.4545        17\n",
      "    acinetobacter     1.0000    0.3469    0.5152        98\n",
      "        aeromonas     0.9615    0.2632    0.4132        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.4615    0.6316        26\n",
      "         bacillus     0.9565    0.6471    0.7719       204\n",
      "      bacteroides     1.0000    0.7419    0.8519        31\n",
      "    brevundimonas     1.0000    0.2000    0.3333        10\n",
      "     burkholderia     1.0000    0.2308    0.3750        39\n",
      "    campylobacter     0.9362    0.7097    0.8073       124\n",
      "      caulobacter     1.0000    0.6800    0.8095        25\n",
      "      citrobacter     0.6000    0.0882    0.1538        34\n",
      "      clostridium     0.6923    0.3750    0.4865        24\n",
      "      cronobacter     0.6000    0.0833    0.1463        36\n",
      "          dickeya     1.0000    0.5962    0.7470        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.6458    0.2183    0.3263       142\n",
      "     enterococcus     1.0000    0.5918    0.7436        49\n",
      "          erwinia     0.9677    0.6294    0.7627       286\n",
      "      escherichia     0.8858    0.5629    0.6884      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.5472    0.7073        53\n",
      "       halorubrum     0.3571    0.3846    0.3704        13\n",
      "     helicobacter     1.0000    0.9375    0.9677        16\n",
      "       klebsiella     0.9444    0.5567    0.7005       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.5000    0.0435    0.0800        23\n",
      "      lactococcus     1.0000    0.7400    0.8506       100\n",
      "      leuconostoc     1.0000    0.6000    0.7500        10\n",
      "         listeria     0.9048    0.6786    0.7755        28\n",
      "   microbacterium     1.0000    0.6364    0.7778        33\n",
      "    mycobacterium     0.9868    0.9032    0.9432       248\n",
      "           others     0.2667    0.9533    0.4168       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9211    0.4930    0.6422        71\n",
      "     pelagibacter     0.8235    0.4242    0.5600        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9286    0.5000    0.6500        52\n",
      "          proteus     0.9333    0.3500    0.5091        40\n",
      "      providencia     0.6667    0.1250    0.2105        16\n",
      "pseudoalteromonas     1.0000    0.0588    0.1111        17\n",
      "      pseudomonas     0.9951    0.7895    0.8804       513\n",
      "        ralstonia     1.0000    0.3571    0.5263        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     1.0000    0.5946    0.7458        74\n",
      "      rhodobacter     0.9259    0.7353    0.8197        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9324    0.6329    0.7540       632\n",
      "         serratia     0.8750    0.4330    0.5793        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5429    0.1329    0.2135       143\n",
      "   staphylococcus     1.0000    0.9057    0.9505       244\n",
      " stenotrophomonas     0.8889    0.1951    0.3200        41\n",
      "    streptococcus     1.0000    0.6049    0.7538        81\n",
      "     streptomyces     1.0000    0.4651    0.6349        43\n",
      "    synechococcus     0.9861    0.8196    0.8952       521\n",
      "           vibrio     0.9963    0.5788    0.7322       463\n",
      "      xanthomonas     0.8800    0.3793    0.5301        58\n",
      "         yersinia     0.6164    0.4839    0.5422        93\n",
      "\n",
      "         accuracy                         0.6482      8116\n",
      "        macro avg     0.8325    0.4758    0.5684      8116\n",
      "     weighted avg     0.8412    0.6482    0.6805      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 50.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.3265    0.4923        98\n",
      "        aeromonas     1.0000    0.2526    0.4034        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.3462    0.5143        26\n",
      "         bacillus     0.9618    0.6176    0.7522       204\n",
      "      bacteroides     1.0000    0.7097    0.8302        31\n",
      "    brevundimonas     1.0000    0.2000    0.3333        10\n",
      "     burkholderia     1.0000    0.1795    0.3043        39\n",
      "    campylobacter     0.9438    0.6774    0.7887       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     1.0000    0.0588    0.1111        34\n",
      "      clostridium     0.7500    0.2500    0.3750        24\n",
      "      cronobacter     0.5000    0.0278    0.0526        36\n",
      "          dickeya     1.0000    0.5577    0.7160        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.7027    0.1831    0.2905       142\n",
      "     enterococcus     1.0000    0.5306    0.6933        49\n",
      "          erwinia     0.9760    0.5699    0.7196       286\n",
      "      escherichia     0.8975    0.4934    0.6368      1295\n",
      "   flavobacterium     1.0000    0.8958    0.9451        48\n",
      "         gordonia     1.0000    0.4528    0.6234        53\n",
      "       halorubrum     0.3077    0.3077    0.3077        13\n",
      "     helicobacter     1.0000    0.8750    0.9333        16\n",
      "       klebsiella     0.9507    0.4861    0.6433       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.5000    0.0435    0.0800        23\n",
      "      lactococcus     1.0000    0.6900    0.8166       100\n",
      "      leuconostoc     1.0000    0.4000    0.5714        10\n",
      "         listeria     0.9000    0.6429    0.7500        28\n",
      "   microbacterium     1.0000    0.5758    0.7308        33\n",
      "    mycobacterium     0.9866    0.8911    0.9364       248\n",
      "           others     0.2446    0.9635    0.3901       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9459    0.4930    0.6481        71\n",
      "     pelagibacter     0.9167    0.3333    0.4889        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9231    0.4615    0.6154        52\n",
      "          proteus     1.0000    0.3500    0.5185        40\n",
      "      providencia     1.0000    0.0625    0.1176        16\n",
      "pseudoalteromonas     1.0000    0.0588    0.1111        17\n",
      "      pseudomonas     0.9949    0.7602    0.8619       513\n",
      "        ralstonia     1.0000    0.3333    0.5000        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     1.0000    0.4865    0.6545        74\n",
      "      rhodobacter     0.9200    0.6765    0.7797        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9358    0.5997    0.7310       632\n",
      "         serratia     0.8571    0.3711    0.5180        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5862    0.1189    0.1977       143\n",
      "   staphylococcus     1.0000    0.8975    0.9460       244\n",
      " stenotrophomonas     0.8750    0.1707    0.2857        41\n",
      "    streptococcus     1.0000    0.5556    0.7143        81\n",
      "     streptomyces     1.0000    0.4419    0.6129        43\n",
      "    synechococcus     0.9928    0.7927    0.8815       521\n",
      "           vibrio     1.0000    0.5486    0.7085       463\n",
      "      xanthomonas     0.8636    0.3276    0.4750        58\n",
      "         yersinia     0.6618    0.4839    0.5590        93\n",
      "\n",
      "         accuracy                         0.6127      8116\n",
      "        macro avg     0.8321    0.4329    0.5271      8116\n",
      "     weighted avg     0.8452    0.6127    0.6487      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 60.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.2449    0.3934        98\n",
      "        aeromonas     1.0000    0.2000    0.3333        95\n",
      "    agrobacterium     1.0000    0.2500    0.4000        12\n",
      "     arthrobacter     1.0000    0.2692    0.4242        26\n",
      "         bacillus     0.9658    0.5539    0.7040       204\n",
      "      bacteroides     1.0000    0.6452    0.7843        31\n",
      "    brevundimonas     1.0000    0.2000    0.3333        10\n",
      "     burkholderia     1.0000    0.1026    0.1860        39\n",
      "    campylobacter     0.9737    0.5968    0.7400       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.8000    0.1667    0.2759        24\n",
      "      cronobacter     1.0000    0.0278    0.0541        36\n",
      "          dickeya     1.0000    0.5000    0.6667        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.7692    0.1408    0.2381       142\n",
      "     enterococcus     1.0000    0.4898    0.6575        49\n",
      "          erwinia     0.9865    0.5105    0.6728       286\n",
      "      escherichia     0.9161    0.4301    0.5854      1295\n",
      "   flavobacterium     1.0000    0.8542    0.9213        48\n",
      "         gordonia     1.0000    0.3208    0.4857        53\n",
      "       halorubrum     0.1818    0.1538    0.1667        13\n",
      "     helicobacter     1.0000    0.8750    0.9333        16\n",
      "       klebsiella     0.9649    0.4156    0.5810       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.6400    0.7805       100\n",
      "      leuconostoc     1.0000    0.2000    0.3333        10\n",
      "         listeria     0.9375    0.5357    0.6818        28\n",
      "   microbacterium     1.0000    0.4545    0.6250        33\n",
      "    mycobacterium     0.9863    0.8710    0.9251       248\n",
      "           others     0.2217    0.9767    0.3614       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9714    0.4789    0.6415        71\n",
      "     pelagibacter     1.0000    0.2424    0.3902        33\n",
      "       prevotella     1.0000    0.8889    0.9412        18\n",
      "  prochlorococcus     0.9500    0.3654    0.5278        52\n",
      "          proteus     1.0000    0.2250    0.3673        40\n",
      "      providencia     1.0000    0.0625    0.1176        16\n",
      "pseudoalteromonas     1.0000    0.0588    0.1111        17\n",
      "      pseudomonas     0.9972    0.6940    0.8184       513\n",
      "        ralstonia     1.0000    0.3095    0.4727        42\n",
      "     rheinheimera     1.0000    0.9886    0.9943       176\n",
      "        rhizobium     1.0000    0.4189    0.5905        74\n",
      "      rhodobacter     1.0000    0.6471    0.7857        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9448    0.5411    0.6881       632\n",
      "         serratia     0.8611    0.3196    0.4662        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.6154    0.1119    0.1893       143\n",
      "   staphylococcus     1.0000    0.8730    0.9322       244\n",
      " stenotrophomonas     1.0000    0.1463    0.2553        41\n",
      "    streptococcus     1.0000    0.4938    0.6612        81\n",
      "     streptomyces     1.0000    0.4186    0.5902        43\n",
      "    synechococcus     0.9950    0.7601    0.8618       521\n",
      "           vibrio     1.0000    0.4903    0.6580       463\n",
      "      xanthomonas     0.8824    0.2586    0.4000        58\n",
      "         yersinia     0.7778    0.3763    0.5072        93\n",
      "\n",
      "         accuracy                         0.5665      8116\n",
      "        macro avg     0.8254    0.3820    0.4775      8116\n",
      "     weighted avg     0.8496    0.5665    0.6058      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 70.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.1633    0.2807        98\n",
      "        aeromonas     1.0000    0.1368    0.2407        95\n",
      "    agrobacterium     1.0000    0.2500    0.4000        12\n",
      "     arthrobacter     1.0000    0.2308    0.3750        26\n",
      "         bacillus     0.9700    0.4755    0.6382       204\n",
      "      bacteroides     1.0000    0.5484    0.7083        31\n",
      "    brevundimonas     1.0000    0.1000    0.1818        10\n",
      "     burkholderia     1.0000    0.0769    0.1429        39\n",
      "    campylobacter     1.0000    0.5323    0.6947       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.7500    0.1250    0.2143        24\n",
      "      cronobacter     1.0000    0.0278    0.0541        36\n",
      "          dickeya     1.0000    0.4423    0.6133        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.7500    0.0634    0.1169       142\n",
      "     enterococcus     1.0000    0.4694    0.6389        49\n",
      "          erwinia     0.9832    0.4091    0.5778       286\n",
      "      escherichia     0.9201    0.3467    0.5036      1295\n",
      "   flavobacterium     1.0000    0.8333    0.9091        48\n",
      "         gordonia     1.0000    0.2642    0.4179        53\n",
      "       halorubrum     0.1818    0.1538    0.1667        13\n",
      "     helicobacter     1.0000    0.6250    0.7692        16\n",
      "       klebsiella     0.9630    0.3275    0.4887       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.6000    0.7500       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     0.9167    0.3929    0.5500        28\n",
      "   microbacterium     1.0000    0.3636    0.5333        33\n",
      "    mycobacterium     0.9861    0.8589    0.9181       248\n",
      "           others     0.2016    0.9797    0.3344       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.4507    0.6214        71\n",
      "     pelagibacter     1.0000    0.1212    0.2162        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     1.0000    0.2500    0.4000        52\n",
      "          proteus     1.0000    0.1250    0.2222        40\n",
      "      providencia     1.0000    0.0625    0.1176        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     0.9971    0.6667    0.7991       513\n",
      "        ralstonia     1.0000    0.2381    0.3846        42\n",
      "     rheinheimera     1.0000    0.9830    0.9914       176\n",
      "        rhizobium     1.0000    0.3514    0.5200        74\n",
      "      rhodobacter     1.0000    0.6471    0.7857        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9470    0.4810    0.6380       632\n",
      "         serratia     0.8462    0.2268    0.3577        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.6190    0.0909    0.1585       143\n",
      "   staphylococcus     1.0000    0.8566    0.9227       244\n",
      " stenotrophomonas     1.0000    0.1220    0.2174        41\n",
      "    streptococcus     1.0000    0.4198    0.5913        81\n",
      "     streptomyces     1.0000    0.3488    0.5172        43\n",
      "    synechococcus     0.9948    0.7274    0.8404       521\n",
      "           vibrio     1.0000    0.4384    0.6096       463\n",
      "      xanthomonas     0.8571    0.2069    0.3333        58\n",
      "         yersinia     0.9655    0.3011    0.4590        93\n",
      "\n",
      "         accuracy                         0.5160      8116\n",
      "        macro avg     0.7771    0.3286    0.4171      8116\n",
      "     weighted avg     0.8455    0.5160    0.5527      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 80.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.1224    0.2182        98\n",
      "        aeromonas     1.0000    0.0947    0.1731        95\n",
      "    agrobacterium     1.0000    0.2500    0.4000        12\n",
      "     arthrobacter     1.0000    0.2308    0.3750        26\n",
      "         bacillus     0.9859    0.3431    0.5091       204\n",
      "      bacteroides     1.0000    0.3871    0.5581        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     1.0000    0.0513    0.0976        39\n",
      "    campylobacter     1.0000    0.4113    0.5829       124\n",
      "      caulobacter     1.0000    0.6000    0.7500        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.6667    0.0833    0.1481        24\n",
      "      cronobacter     1.0000    0.0278    0.0541        36\n",
      "          dickeya     1.0000    0.4423    0.6133        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.6667    0.0423    0.0795       142\n",
      "     enterococcus     1.0000    0.4694    0.6389        49\n",
      "          erwinia     0.9890    0.3147    0.4775       286\n",
      "      escherichia     0.9367    0.2741    0.4241      1295\n",
      "   flavobacterium     1.0000    0.8333    0.9091        48\n",
      "         gordonia     1.0000    0.2075    0.3438        53\n",
      "       halorubrum     0.2000    0.0769    0.1111        13\n",
      "     helicobacter     1.0000    0.5625    0.7200        16\n",
      "       klebsiella     0.9800    0.2469    0.3944       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.5100    0.6755       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.3571    0.5263        28\n",
      "   microbacterium     1.0000    0.3030    0.4651        33\n",
      "    mycobacterium     0.9859    0.8468    0.9111       248\n",
      "           others     0.1850    0.9899    0.3117       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.3521    0.5208        71\n",
      "     pelagibacter     1.0000    0.0606    0.1143        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     1.0000    0.2500    0.4000        52\n",
      "          proteus     1.0000    0.1250    0.2222        40\n",
      "      providencia     1.0000    0.0625    0.1176        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     0.9967    0.5926    0.7433       513\n",
      "        ralstonia     1.0000    0.2143    0.3529        42\n",
      "     rheinheimera     1.0000    0.9716    0.9856       176\n",
      "        rhizobium     1.0000    0.2162    0.3556        74\n",
      "      rhodobacter     1.0000    0.5882    0.7407        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9455    0.4114    0.5733       632\n",
      "         serratia     0.8333    0.1546    0.2609        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.4615    0.0420    0.0769       143\n",
      "   staphylococcus     1.0000    0.7582    0.8625       244\n",
      " stenotrophomonas     1.0000    0.0732    0.1364        41\n",
      "    streptococcus     1.0000    0.3086    0.4717        81\n",
      "     streptomyces     1.0000    0.2093    0.3462        43\n",
      "    synechococcus     0.9946    0.7044    0.8247       521\n",
      "           vibrio     1.0000    0.3629    0.5325       463\n",
      "      xanthomonas     0.8889    0.1379    0.2388        58\n",
      "         yersinia     0.9583    0.2473    0.3932        93\n",
      "\n",
      "         accuracy                         0.4616      8116\n",
      "        macro avg     0.7572    0.2840    0.3669      8116\n",
      "     weighted avg     0.8420    0.4616    0.4929      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 90.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.0510    0.0971        98\n",
      "        aeromonas     1.0000    0.0211    0.0412        95\n",
      "    agrobacterium     0.0000    0.0000    0.0000        12\n",
      "     arthrobacter     1.0000    0.1154    0.2069        26\n",
      "         bacillus     1.0000    0.2206    0.3614       204\n",
      "      bacteroides     1.0000    0.0645    0.1212        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     1.0000    0.0256    0.0500        39\n",
      "    campylobacter     1.0000    0.2661    0.4204       124\n",
      "      caulobacter     1.0000    0.2800    0.4375        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     1.0000    0.0417    0.0800        24\n",
      "      cronobacter     0.0000    0.0000    0.0000        36\n",
      "          dickeya     1.0000    0.3654    0.5352        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.6000    0.0211    0.0408       142\n",
      "     enterococcus     1.0000    0.3878    0.5588        49\n",
      "          erwinia     0.9848    0.2273    0.3693       286\n",
      "      escherichia     0.9465    0.1776    0.2991      1295\n",
      "   flavobacterium     1.0000    0.6667    0.8000        48\n",
      "         gordonia     1.0000    0.1321    0.2333        53\n",
      "       halorubrum     1.0000    0.0769    0.1429        13\n",
      "     helicobacter     1.0000    0.1875    0.3158        16\n",
      "       klebsiella     0.9853    0.1688    0.2882       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.4600    0.6301       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.2857    0.4444        28\n",
      "   microbacterium     1.0000    0.1212    0.2162        33\n",
      "    mycobacterium     0.9854    0.8185    0.8943       248\n",
      "           others     0.1676    0.9959    0.2869       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.2958    0.4565        71\n",
      "     pelagibacter     1.0000    0.0303    0.0588        33\n",
      "       prevotella     1.0000    0.7222    0.8387        18\n",
      "  prochlorococcus     1.0000    0.2308    0.3750        52\n",
      "          proteus     1.0000    0.1250    0.2222        40\n",
      "      providencia     0.0000    0.0000    0.0000        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     1.0000    0.5088    0.6744       513\n",
      "        ralstonia     1.0000    0.1190    0.2128        42\n",
      "     rheinheimera     1.0000    0.9489    0.9738       176\n",
      "        rhizobium     1.0000    0.1351    0.2381        74\n",
      "      rhodobacter     1.0000    0.5000    0.6667        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9541    0.3291    0.4894       632\n",
      "         serratia     0.8333    0.0515    0.0971        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.8000    0.0280    0.0541       143\n",
      "   staphylococcus     1.0000    0.6639    0.7980       244\n",
      " stenotrophomonas     0.0000    0.0000    0.0000        41\n",
      "    streptococcus     1.0000    0.1111    0.2000        81\n",
      "     streptomyces     1.0000    0.0930    0.1702        43\n",
      "    synechococcus     0.9972    0.6833    0.8109       521\n",
      "           vibrio     1.0000    0.2937    0.4541       463\n",
      "      xanthomonas     0.8000    0.0690    0.1270        58\n",
      "         yersinia     1.0000    0.1720    0.2936        93\n",
      "\n",
      "         accuracy                         0.3948      8116\n",
      "        macro avg     0.7128    0.2083    0.2726      8116\n",
      "     weighted avg     0.8369    0.3948    0.4107      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 100.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.0102    0.0202        98\n",
      "        aeromonas     0.0000    0.0000    0.0000        95\n",
      "    agrobacterium     0.0000    0.0000    0.0000        12\n",
      "     arthrobacter     0.0000    0.0000    0.0000        26\n",
      "         bacillus     0.0000    0.0000    0.0000       204\n",
      "      bacteroides     0.0000    0.0000    0.0000        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     0.0000    0.0000    0.0000        39\n",
      "    campylobacter     1.0000    0.0565    0.1069       124\n",
      "      caulobacter     0.0000    0.0000    0.0000        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.0000    0.0000    0.0000        24\n",
      "      cronobacter     0.0000    0.0000    0.0000        36\n",
      "          dickeya     1.0000    0.2885    0.4478        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.5000    0.0070    0.0139       142\n",
      "     enterococcus     1.0000    0.0612    0.1154        49\n",
      "          erwinia     1.0000    0.0280    0.0544       286\n",
      "      escherichia     0.9558    0.0834    0.1534      1295\n",
      "   flavobacterium     1.0000    0.6250    0.7692        48\n",
      "         gordonia     0.0000    0.0000    0.0000        53\n",
      "       halorubrum     0.0000    0.0000    0.0000        13\n",
      "     helicobacter     0.0000    0.0000    0.0000        16\n",
      "       klebsiella     1.0000    0.0353    0.0681       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.1700    0.2906       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.0357    0.0690        28\n",
      "   microbacterium     0.0000    0.0000    0.0000        33\n",
      "    mycobacterium     0.9818    0.6532    0.7845       248\n",
      "           others     0.1425    0.9959    0.2493       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.1268    0.2250        71\n",
      "     pelagibacter     0.0000    0.0000    0.0000        33\n",
      "       prevotella     1.0000    0.0556    0.1053        18\n",
      "  prochlorococcus     1.0000    0.2115    0.3492        52\n",
      "          proteus     1.0000    0.0500    0.0952        40\n",
      "      providencia     0.0000    0.0000    0.0000        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     1.0000    0.3021    0.4641       513\n",
      "        ralstonia     0.0000    0.0000    0.0000        42\n",
      "     rheinheimera     1.0000    0.9318    0.9647       176\n",
      "        rhizobium     0.0000    0.0000    0.0000        74\n",
      "      rhodobacter     1.0000    0.1765    0.3000        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9857    0.1092    0.1966       632\n",
      "         serratia     1.0000    0.0206    0.0404        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.6667    0.0140    0.0274       143\n",
      "   staphylococcus     1.0000    0.2828    0.4409       244\n",
      " stenotrophomonas     0.0000    0.0000    0.0000        41\n",
      "    streptococcus     0.0000    0.0000    0.0000        81\n",
      "     streptomyces     1.0000    0.0465    0.0889        43\n",
      "    synechococcus     0.9963    0.5182    0.6818       521\n",
      "           vibrio     1.0000    0.1512    0.2627       463\n",
      "      xanthomonas     0.5000    0.0172    0.0333        58\n",
      "         yersinia     1.0000    0.1183    0.2115        93\n",
      "\n",
      "         accuracy                         0.2702      8116\n",
      "        macro avg     0.4530    0.1048    0.1293      8116\n",
      "     weighted avg     0.7374    0.2702    0.2459      8116\n",
      "\n",
      "===================\n",
      "Finished\n",
      "==============\n",
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Confidence threshold k: 0.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.3529    0.5217        17\n",
      "    acinetobacter     0.9242    0.6224    0.7439        98\n",
      "        aeromonas     0.9219    0.6211    0.7421        95\n",
      "    agrobacterium     0.8333    0.4167    0.5556        12\n",
      "     arthrobacter     0.9444    0.6538    0.7727        26\n",
      "         bacillus     0.6811    0.8480    0.7555       204\n",
      "      bacteroides     1.0000    0.9355    0.9667        31\n",
      "    brevundimonas     1.0000    0.5000    0.6667        10\n",
      "     burkholderia     0.5312    0.4359    0.4789        39\n",
      "    campylobacter     0.8487    0.8145    0.8313       124\n",
      "      caulobacter     0.8077    0.8400    0.8235        25\n",
      "      citrobacter     0.3793    0.3235    0.3492        34\n",
      "      clostridium     0.4324    0.6667    0.5246        24\n",
      "      cronobacter     0.5714    0.4444    0.5000        36\n",
      "          dickeya     0.9091    0.7692    0.8333        52\n",
      "     edwardsiella     1.0000    0.3000    0.4615        10\n",
      "     enterobacter     0.4615    0.3803    0.4170       142\n",
      "     enterococcus     0.9556    0.8776    0.9149        49\n",
      "          erwinia     0.6676    0.8287    0.7395       286\n",
      "      escherichia     0.5212    0.8263    0.6392      1295\n",
      "   flavobacterium     0.9574    0.9375    0.9474        48\n",
      "         gordonia     0.9024    0.6981    0.7872        53\n",
      "       halorubrum     0.3810    0.6154    0.4706        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.7082    0.7582    0.7324       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.7500    0.3913    0.5143        23\n",
      "      lactococcus     0.9011    0.8200    0.8586       100\n",
      "      leuconostoc     1.0000    0.9000    0.9474        10\n",
      "         listeria     0.8889    0.8571    0.8727        28\n",
      "   microbacterium     0.8800    0.6667    0.7586        33\n",
      "    mycobacterium     0.9746    0.9274    0.9504       248\n",
      "           others     0.0000    0.0000    0.0000       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.8571    0.6761    0.7559        71\n",
      "     pelagibacter     0.5000    0.5455    0.5217        33\n",
      "       prevotella     0.9444    0.9444    0.9444        18\n",
      "  prochlorococcus     0.9412    0.6154    0.7442        52\n",
      "          proteus     0.8214    0.5750    0.6765        40\n",
      "      providencia     0.5714    0.2500    0.3478        16\n",
      "pseudoalteromonas     1.0000    0.1765    0.3000        17\n",
      "      pseudomonas     0.6657    0.8889    0.7613       513\n",
      "        ralstonia     0.8400    0.5000    0.6269        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     0.9483    0.7432    0.8333        74\n",
      "      rhodobacter     0.8108    0.8824    0.8451        34\n",
      "      rhodococcus     1.0000    0.2308    0.3750        13\n",
      "       salmonella     0.8133    0.7722    0.7922       632\n",
      "         serratia     0.7722    0.6289    0.6932        97\n",
      "       shewanella     1.0000    0.2727    0.4286        11\n",
      "         shigella     0.4375    0.2448    0.3139       143\n",
      "   staphylococcus     0.7143    0.9836    0.8276       244\n",
      " stenotrophomonas     0.9412    0.3902    0.5517        41\n",
      "    streptococcus     0.8514    0.7778    0.8129        81\n",
      "     streptomyces     0.7619    0.7442    0.7529        43\n",
      "    synechococcus     0.6944    0.9463    0.8010       521\n",
      "           vibrio     0.6625    0.8099    0.7289       463\n",
      "      xanthomonas     0.7714    0.4655    0.5806        58\n",
      "         yersinia     0.5000    0.6344    0.5592        93\n",
      "\n",
      "         accuracy                         0.6831      8116\n",
      "        macro avg     0.7552    0.6224    0.6551      8116\n",
      "     weighted avg     0.6218    0.6831    0.6371      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 10.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2941    0.4545        17\n",
      "    acinetobacter     1.0000    0.5714    0.7273        98\n",
      "        aeromonas     0.9600    0.5053    0.6621        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.6154    0.7619        26\n",
      "         bacillus     0.8901    0.7941    0.8394       204\n",
      "      bacteroides     1.0000    0.9032    0.9492        31\n",
      "    brevundimonas     1.0000    0.4000    0.5714        10\n",
      "     burkholderia     0.8824    0.3846    0.5357        39\n",
      "    campylobacter     0.9706    0.7984    0.8761       124\n",
      "      caulobacter     1.0000    0.7600    0.8636        25\n",
      "      citrobacter     0.5000    0.2059    0.2917        34\n",
      "      clostridium     0.4643    0.5417    0.5000        24\n",
      "      cronobacter     0.6842    0.3611    0.4727        36\n",
      "          dickeya     0.9744    0.7308    0.8352        52\n",
      "     edwardsiella     1.0000    0.2000    0.3333        10\n",
      "     enterobacter     0.4719    0.2958    0.3636       142\n",
      "     enterococcus     0.9773    0.8776    0.9247        49\n",
      "          erwinia     0.8943    0.7692    0.8271       286\n",
      "      escherichia     0.7601    0.7436    0.7518      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.6226    0.7674        53\n",
      "       halorubrum     0.3889    0.5385    0.4516        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.8484    0.6625    0.7440       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.8000    0.3478    0.4848        23\n",
      "      lactococcus     0.9405    0.7900    0.8587       100\n",
      "      leuconostoc     1.0000    0.9000    0.9474        10\n",
      "         listeria     0.9231    0.8571    0.8889        28\n",
      "   microbacterium     1.0000    0.6364    0.7778        33\n",
      "    mycobacterium     0.9871    0.9274    0.9563       248\n",
      "           others     0.3473    0.7657    0.4778       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.8889    0.5634    0.6897        71\n",
      "     pelagibacter     0.7222    0.3939    0.5098        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9310    0.5192    0.6667        52\n",
      "          proteus     0.8800    0.5500    0.6769        40\n",
      "      providencia     0.6667    0.2500    0.3636        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9129    0.8577    0.8844       513\n",
      "        ralstonia     0.9444    0.4048    0.5667        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     0.9608    0.6622    0.7840        74\n",
      "      rhodobacter     0.9032    0.8235    0.8615        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9024    0.7168    0.7989       632\n",
      "         serratia     0.8387    0.5361    0.6541        97\n",
      "       shewanella     1.0000    0.1818    0.3077        11\n",
      "         shigella     0.4833    0.2028    0.2857       143\n",
      "   staphylococcus     0.8681    0.9713    0.9168       244\n",
      " stenotrophomonas     0.9286    0.3171    0.4727        41\n",
      "    streptococcus     0.9483    0.6790    0.7914        81\n",
      "     streptomyces     0.9643    0.6279    0.7606        43\n",
      "    synechococcus     0.8834    0.8868    0.8851       521\n",
      "           vibrio     0.9329    0.6911    0.7940       463\n",
      "      xanthomonas     0.8667    0.4483    0.5909        58\n",
      "         yersinia     0.5138    0.6022    0.5545        93\n",
      "\n",
      "         accuracy                         0.7212      8116\n",
      "        macro avg     0.8238    0.5765    0.6501      8116\n",
      "     weighted avg     0.7928    0.7212    0.7331      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 20.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2941    0.4545        17\n",
      "    acinetobacter     1.0000    0.5102    0.6757        98\n",
      "        aeromonas     0.9487    0.3895    0.5522        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.5385    0.7000        26\n",
      "         bacillus     0.9259    0.7353    0.8197       204\n",
      "      bacteroides     1.0000    0.8710    0.9310        31\n",
      "    brevundimonas     1.0000    0.4000    0.5714        10\n",
      "     burkholderia     0.9231    0.3077    0.4615        39\n",
      "    campylobacter     0.9796    0.7742    0.8649       124\n",
      "      caulobacter     1.0000    0.7600    0.8636        25\n",
      "      citrobacter     0.6250    0.1471    0.2381        34\n",
      "      clostridium     0.5417    0.5417    0.5417        24\n",
      "      cronobacter     0.6667    0.2222    0.3333        36\n",
      "          dickeya     1.0000    0.6731    0.8046        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.5588    0.2676    0.3619       142\n",
      "     enterococcus     1.0000    0.8163    0.8989        49\n",
      "          erwinia     0.9469    0.6853    0.7951       286\n",
      "      escherichia     0.8347    0.6865    0.7534      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.5472    0.7073        53\n",
      "       halorubrum     0.3529    0.4615    0.4000        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.9377    0.6071    0.7370       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.8571    0.2609    0.4000        23\n",
      "      lactococcus     0.9747    0.7700    0.8603       100\n",
      "      leuconostoc     1.0000    0.6000    0.7500        10\n",
      "         listeria     0.9231    0.8571    0.8889        28\n",
      "   microbacterium     1.0000    0.6364    0.7778        33\n",
      "    mycobacterium     0.9870    0.9194    0.9520       248\n",
      "           others     0.3130    0.9097    0.4657       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9070    0.5493    0.6842        71\n",
      "     pelagibacter     0.9000    0.2727    0.4186        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9286    0.5000    0.6500        52\n",
      "          proteus     0.8571    0.4500    0.5902        40\n",
      "      providencia     0.6000    0.1875    0.2857        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9860    0.8226    0.8969       513\n",
      "        ralstonia     0.9412    0.3810    0.5424        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     1.0000    0.6486    0.7869        74\n",
      "      rhodobacter     0.9259    0.7353    0.8197        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9127    0.6946    0.7889       632\n",
      "         serratia     0.8448    0.5052    0.6323        97\n",
      "       shewanella     1.0000    0.0909    0.1667        11\n",
      "         shigella     0.5102    0.1748    0.2604       143\n",
      "   staphylococcus     0.9467    0.9467    0.9467       244\n",
      " stenotrophomonas     0.8889    0.1951    0.3200        41\n",
      "    streptococcus     1.0000    0.6790    0.8088        81\n",
      "     streptomyces     1.0000    0.5349    0.6970        43\n",
      "    synechococcus     0.9632    0.8541    0.9054       521\n",
      "           vibrio     0.9864    0.6263    0.7662       463\n",
      "      xanthomonas     0.8519    0.3966    0.5412        58\n",
      "         yersinia     0.5417    0.5591    0.5503        93\n",
      "\n",
      "         accuracy                         0.6999      8116\n",
      "        macro avg     0.8439    0.5326    0.6178      8116\n",
      "     weighted avg     0.8292    0.6999    0.7218      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 30.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2353    0.3810        17\n",
      "    acinetobacter     1.0000    0.4184    0.5899        98\n",
      "        aeromonas     0.9667    0.3053    0.4640        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.5385    0.7000        26\n",
      "         bacillus     0.9533    0.7010    0.8079       204\n",
      "      bacteroides     1.0000    0.7742    0.8727        31\n",
      "    brevundimonas     1.0000    0.3000    0.4615        10\n",
      "     burkholderia     0.9000    0.2308    0.3673        39\n",
      "    campylobacter     0.9792    0.7581    0.8545       124\n",
      "      caulobacter     1.0000    0.6800    0.8095        25\n",
      "      citrobacter     0.6000    0.0882    0.1538        34\n",
      "      clostridium     0.6000    0.5000    0.5455        24\n",
      "      cronobacter     0.8889    0.2222    0.3556        36\n",
      "          dickeya     1.0000    0.6731    0.8046        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.6154    0.2254    0.3299       142\n",
      "     enterococcus     1.0000    0.7347    0.8471        49\n",
      "          erwinia     0.9686    0.6469    0.7757       286\n",
      "      escherichia     0.8623    0.6239    0.7240      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.4717    0.6410        53\n",
      "       halorubrum     0.3125    0.3846    0.3448        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.9461    0.5743    0.7147       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.8333    0.2174    0.3448        23\n",
      "      lactococcus     1.0000    0.7300    0.8439       100\n",
      "      leuconostoc     1.0000    0.6000    0.7500        10\n",
      "         listeria     0.9167    0.7857    0.8462        28\n",
      "   microbacterium     1.0000    0.5758    0.7308        33\n",
      "    mycobacterium     0.9868    0.9073    0.9454       248\n",
      "           others     0.2855    0.9473    0.4388       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9024    0.5211    0.6607        71\n",
      "     pelagibacter     1.0000    0.2727    0.4286        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9259    0.4808    0.6329        52\n",
      "          proteus     0.8421    0.4000    0.5424        40\n",
      "      providencia     0.5000    0.0625    0.1111        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9880    0.7992    0.8836       513\n",
      "        ralstonia     0.9412    0.3810    0.5424        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     1.0000    0.6081    0.7563        74\n",
      "      rhodobacter     0.9259    0.7353    0.8197        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9329    0.6598    0.7729       632\n",
      "         serratia     0.8519    0.4742    0.6093        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5238    0.1538    0.2378       143\n",
      "   staphylococcus     0.9785    0.9344    0.9560       244\n",
      " stenotrophomonas     0.8889    0.1951    0.3200        41\n",
      "    streptococcus     1.0000    0.6543    0.7910        81\n",
      "     streptomyces     1.0000    0.4419    0.6129        43\n",
      "    synechococcus     0.9862    0.8253    0.8986       521\n",
      "           vibrio     1.0000    0.5875    0.7401       463\n",
      "      xanthomonas     0.9091    0.3448    0.5000        58\n",
      "         yersinia     0.5904    0.5269    0.5568        93\n",
      "\n",
      "         accuracy                         0.6711      8116\n",
      "        macro avg     0.8356    0.4972    0.5872      8116\n",
      "     weighted avg     0.8393    0.6711    0.6993      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 40.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2353    0.3810        17\n",
      "    acinetobacter     1.0000    0.4082    0.5797        98\n",
      "        aeromonas     0.9630    0.2737    0.4262        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.4231    0.5946        26\n",
      "         bacillus     0.9504    0.6569    0.7768       204\n",
      "      bacteroides     1.0000    0.7419    0.8519        31\n",
      "    brevundimonas     1.0000    0.3000    0.4615        10\n",
      "     burkholderia     0.9000    0.2308    0.3673        39\n",
      "    campylobacter     0.9787    0.7419    0.8440       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     0.6000    0.0882    0.1538        34\n",
      "      clostridium     0.6875    0.4583    0.5500        24\n",
      "      cronobacter     0.7500    0.0833    0.1500        36\n",
      "          dickeya     1.0000    0.5769    0.7317        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.6047    0.1831    0.2811       142\n",
      "     enterococcus     1.0000    0.6735    0.8049        49\n",
      "          erwinia     0.9714    0.5944    0.7375       286\n",
      "      escherichia     0.8822    0.5552    0.6815      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.4151    0.5867        53\n",
      "       halorubrum     0.2667    0.3077    0.2857        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.9467    0.5365    0.6849       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.7500    0.1304    0.2222        23\n",
      "      lactococcus     1.0000    0.7200    0.8372       100\n",
      "      leuconostoc     1.0000    0.6000    0.7500        10\n",
      "         listeria     0.9091    0.7143    0.8000        28\n",
      "   microbacterium     1.0000    0.5455    0.7059        33\n",
      "    mycobacterium     0.9868    0.9073    0.9454       248\n",
      "           others     0.2625    0.9554    0.4118       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9444    0.4789    0.6355        71\n",
      "     pelagibacter     1.0000    0.1818    0.3077        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9259    0.4808    0.6329        52\n",
      "          proteus     0.9412    0.4000    0.5614        40\n",
      "      providencia     0.5000    0.0625    0.1111        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9951    0.7875    0.8792       513\n",
      "        ralstonia     1.0000    0.3810    0.5517        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     1.0000    0.5270    0.6903        74\n",
      "      rhodobacter     0.9259    0.7353    0.8197        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9326    0.6345    0.7552       632\n",
      "         serratia     0.8571    0.4330    0.5753        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5588    0.1329    0.2147       143\n",
      "   staphylococcus     0.9956    0.9262    0.9597       244\n",
      " stenotrophomonas     0.8750    0.1707    0.2857        41\n",
      "    streptococcus     1.0000    0.6173    0.7634        81\n",
      "     streptomyces     1.0000    0.4186    0.5902        43\n",
      "    synechococcus     0.9929    0.8004    0.8863       521\n",
      "           vibrio     1.0000    0.5637    0.7210       463\n",
      "      xanthomonas     0.9000    0.3103    0.4615        58\n",
      "         yersinia     0.6250    0.4839    0.5455        93\n",
      "\n",
      "         accuracy                         0.6417      8116\n",
      "        macro avg     0.8369    0.4687    0.5620      8116\n",
      "     weighted avg     0.8423    0.6417    0.6744      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 50.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.1176    0.2105        17\n",
      "    acinetobacter     1.0000    0.3265    0.4923        98\n",
      "        aeromonas     0.9583    0.2421    0.3866        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.4231    0.5946        26\n",
      "         bacillus     0.9624    0.6275    0.7596       204\n",
      "      bacteroides     1.0000    0.6452    0.7843        31\n",
      "    brevundimonas     1.0000    0.3000    0.4615        10\n",
      "     burkholderia     1.0000    0.1538    0.2667        39\n",
      "    campylobacter     0.9783    0.7258    0.8333       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     0.6667    0.0588    0.1081        34\n",
      "      clostridium     0.6667    0.4167    0.5128        24\n",
      "      cronobacter     0.6667    0.0556    0.1026        36\n",
      "          dickeya     1.0000    0.5385    0.7000        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.6452    0.1408    0.2312       142\n",
      "     enterococcus     1.0000    0.6327    0.7750        49\n",
      "          erwinia     0.9810    0.5420    0.6982       286\n",
      "      escherichia     0.9006    0.4826    0.6285      1295\n",
      "   flavobacterium     1.0000    0.8958    0.9451        48\n",
      "         gordonia     1.0000    0.3208    0.4857        53\n",
      "       halorubrum     0.3077    0.3077    0.3077        13\n",
      "     helicobacter     1.0000    0.9375    0.9677        16\n",
      "       klebsiella     0.9492    0.4710    0.6296       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.5000    0.0435    0.0800        23\n",
      "      lactococcus     1.0000    0.6500    0.7879       100\n",
      "      leuconostoc     1.0000    0.5000    0.6667        10\n",
      "         listeria     0.9048    0.6786    0.7755        28\n",
      "   microbacterium     1.0000    0.4848    0.6531        33\n",
      "    mycobacterium     0.9866    0.8911    0.9364       248\n",
      "           others     0.2401    0.9625    0.3843       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9444    0.4789    0.6355        71\n",
      "     pelagibacter     1.0000    0.0909    0.1667        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     0.9200    0.4423    0.5974        52\n",
      "          proteus     1.0000    0.3000    0.4615        40\n",
      "      providencia     1.0000    0.0625    0.1176        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9949    0.7563    0.8594       513\n",
      "        ralstonia     1.0000    0.3810    0.5517        42\n",
      "     rheinheimera     1.0000    0.9886    0.9943       176\n",
      "        rhizobium     1.0000    0.4865    0.6545        74\n",
      "      rhodobacter     0.9167    0.6471    0.7586        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9401    0.5965    0.7299       632\n",
      "         serratia     0.8810    0.3814    0.5324        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5758    0.1329    0.2159       143\n",
      "   staphylococcus     1.0000    0.8975    0.9460       244\n",
      " stenotrophomonas     0.8571    0.1463    0.2500        41\n",
      "    streptococcus     1.0000    0.5309    0.6935        81\n",
      "     streptomyces     1.0000    0.3953    0.5667        43\n",
      "    synechococcus     0.9951    0.7850    0.8777       521\n",
      "           vibrio     1.0000    0.5227    0.6865       463\n",
      "      xanthomonas     0.8889    0.2759    0.4211        58\n",
      "         yersinia     0.6338    0.4839    0.5488        93\n",
      "\n",
      "         accuracy                         0.6046      8116\n",
      "        macro avg     0.8451    0.4302    0.5256      8116\n",
      "     weighted avg     0.8461    0.6046    0.6411      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 60.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.0588    0.1111        17\n",
      "    acinetobacter     1.0000    0.2245    0.3667        98\n",
      "        aeromonas     1.0000    0.2000    0.3333        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.3077    0.4706        26\n",
      "         bacillus     0.9643    0.5294    0.6835       204\n",
      "      bacteroides     1.0000    0.5806    0.7347        31\n",
      "    brevundimonas     1.0000    0.2000    0.3333        10\n",
      "     burkholderia     1.0000    0.1026    0.1860        39\n",
      "    campylobacter     0.9765    0.6694    0.7943       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     1.0000    0.0588    0.1111        34\n",
      "      clostridium     0.7000    0.2917    0.4118        24\n",
      "      cronobacter     1.0000    0.0278    0.0541        36\n",
      "          dickeya     1.0000    0.5000    0.6667        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.7273    0.1127    0.1951       142\n",
      "     enterococcus     1.0000    0.5306    0.6933        49\n",
      "          erwinia     0.9852    0.4650    0.6318       286\n",
      "      escherichia     0.9205    0.4201    0.5769      1295\n",
      "   flavobacterium     1.0000    0.8542    0.9213        48\n",
      "         gordonia     1.0000    0.3208    0.4857        53\n",
      "       halorubrum     0.5000    0.3077    0.3810        13\n",
      "     helicobacter     1.0000    0.9375    0.9677        16\n",
      "       klebsiella     0.9641    0.4055    0.5709       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.5000    0.0435    0.0800        23\n",
      "      lactococcus     1.0000    0.6200    0.7654       100\n",
      "      leuconostoc     1.0000    0.2000    0.3333        10\n",
      "         listeria     0.9412    0.5714    0.7111        28\n",
      "   microbacterium     1.0000    0.4242    0.5957        33\n",
      "    mycobacterium     0.9864    0.8790    0.9296       248\n",
      "           others     0.2203    0.9777    0.3596       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9697    0.4507    0.6154        71\n",
      "     pelagibacter     1.0000    0.0606    0.1143        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     0.9500    0.3654    0.5278        52\n",
      "          proteus     1.0000    0.2500    0.4000        40\n",
      "      providencia     1.0000    0.0625    0.1176        16\n",
      "pseudoalteromonas     1.0000    0.0588    0.1111        17\n",
      "      pseudomonas     0.9973    0.7173    0.8345       513\n",
      "        ralstonia     1.0000    0.3810    0.5517        42\n",
      "     rheinheimera     1.0000    0.9716    0.9856       176\n",
      "        rhizobium     1.0000    0.4324    0.6038        74\n",
      "      rhodobacter     0.9565    0.6471    0.7719        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9397    0.5427    0.6881       632\n",
      "         serratia     0.8788    0.2990    0.4462        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.6071    0.1189    0.1988       143\n",
      "   staphylococcus     1.0000    0.8770    0.9345       244\n",
      " stenotrophomonas     1.0000    0.1463    0.2553        41\n",
      "    streptococcus     1.0000    0.4691    0.6387        81\n",
      "     streptomyces     1.0000    0.3721    0.5424        43\n",
      "    synechococcus     0.9950    0.7658    0.8655       521\n",
      "           vibrio     1.0000    0.4644    0.6342       463\n",
      "      xanthomonas     0.8824    0.2586    0.4000        58\n",
      "         yersinia     0.7727    0.3656    0.4964        93\n",
      "\n",
      "         accuracy                         0.5632      8116\n",
      "        macro avg     0.8701    0.3865    0.4841      8116\n",
      "     weighted avg     0.8566    0.5632    0.6018      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 70.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.1327    0.2342        98\n",
      "        aeromonas     1.0000    0.1368    0.2407        95\n",
      "    agrobacterium     1.0000    0.2500    0.4000        12\n",
      "     arthrobacter     1.0000    0.2692    0.4242        26\n",
      "         bacillus     0.9798    0.4755    0.6403       204\n",
      "      bacteroides     1.0000    0.4839    0.6522        31\n",
      "    brevundimonas     1.0000    0.1000    0.1818        10\n",
      "     burkholderia     1.0000    0.0513    0.0976        39\n",
      "    campylobacter     0.9730    0.5806    0.7273       124\n",
      "      caulobacter     1.0000    0.6000    0.7500        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.8333    0.2083    0.3333        24\n",
      "      cronobacter     1.0000    0.0278    0.0541        36\n",
      "          dickeya     1.0000    0.4423    0.6133        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.6667    0.0423    0.0795       142\n",
      "     enterococcus     1.0000    0.4694    0.6389        49\n",
      "          erwinia     0.9808    0.3566    0.5231       286\n",
      "      escherichia     0.9274    0.3452    0.5031      1295\n",
      "   flavobacterium     1.0000    0.8542    0.9213        48\n",
      "         gordonia     1.0000    0.2453    0.3939        53\n",
      "       halorubrum     0.4000    0.1538    0.2222        13\n",
      "     helicobacter     1.0000    0.8125    0.8966        16\n",
      "       klebsiella     0.9621    0.3199    0.4802       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.5000    0.0435    0.0800        23\n",
      "      lactococcus     1.0000    0.5400    0.7013       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     0.9333    0.5000    0.6512        28\n",
      "   microbacterium     1.0000    0.3333    0.5000        33\n",
      "    mycobacterium     0.9859    0.8468    0.9111       248\n",
      "           others     0.2000    0.9858    0.3326       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.3944    0.5657        71\n",
      "     pelagibacter     1.0000    0.0606    0.1143        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     1.0000    0.2500    0.4000        52\n",
      "          proteus     1.0000    0.1500    0.2609        40\n",
      "      providencia     1.0000    0.0625    0.1176        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     0.9970    0.6550    0.7906       513\n",
      "        ralstonia     1.0000    0.2857    0.4444        42\n",
      "     rheinheimera     1.0000    0.9659    0.9827       176\n",
      "        rhizobium     1.0000    0.2973    0.4583        74\n",
      "      rhodobacter     1.0000    0.6471    0.7857        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9479    0.4889    0.6451       632\n",
      "         serratia     0.8462    0.2268    0.3577        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.6087    0.0979    0.1687       143\n",
      "   staphylococcus     1.0000    0.8484    0.9180       244\n",
      " stenotrophomonas     1.0000    0.0976    0.1778        41\n",
      "    streptococcus     1.0000    0.3704    0.5405        81\n",
      "     streptomyces     1.0000    0.3023    0.4643        43\n",
      "    synechococcus     0.9947    0.7217    0.8365       521\n",
      "           vibrio     1.0000    0.4104    0.5819       463\n",
      "      xanthomonas     0.8571    0.2069    0.3333        58\n",
      "         yersinia     0.9655    0.3011    0.4590        93\n",
      "\n",
      "         accuracy                         0.5095      8116\n",
      "        macro avg     0.7891    0.3268    0.4152      8116\n",
      "     weighted avg     0.8467    0.5095    0.5454      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 80.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.0918    0.1682        98\n",
      "        aeromonas     1.0000    0.1158    0.2075        95\n",
      "    agrobacterium     1.0000    0.2500    0.4000        12\n",
      "     arthrobacter     1.0000    0.1923    0.3226        26\n",
      "         bacillus     1.0000    0.3480    0.5164       204\n",
      "      bacteroides     1.0000    0.3548    0.5238        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     1.0000    0.0513    0.0976        39\n",
      "    campylobacter     1.0000    0.4274    0.5989       124\n",
      "      caulobacter     1.0000    0.4400    0.6111        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.6667    0.0833    0.1481        24\n",
      "      cronobacter     1.0000    0.0278    0.0541        36\n",
      "          dickeya     1.0000    0.4231    0.5946        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.5000    0.0211    0.0405       142\n",
      "     enterococcus     1.0000    0.4694    0.6389        49\n",
      "          erwinia     0.9750    0.2727    0.4262       286\n",
      "      escherichia     0.9413    0.2602    0.4077      1295\n",
      "   flavobacterium     1.0000    0.7708    0.8706        48\n",
      "         gordonia     1.0000    0.2075    0.3438        53\n",
      "       halorubrum     0.3333    0.0769    0.1250        13\n",
      "     helicobacter     1.0000    0.7500    0.8571        16\n",
      "       klebsiella     0.9684    0.2317    0.3740       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.4900    0.6577       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.3571    0.5263        28\n",
      "   microbacterium     1.0000    0.3030    0.4651        33\n",
      "    mycobacterium     0.9856    0.8306    0.9015       248\n",
      "           others     0.1833    0.9929    0.3095       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.3099    0.4731        71\n",
      "     pelagibacter     1.0000    0.0303    0.0588        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     1.0000    0.2500    0.4000        52\n",
      "          proteus     1.0000    0.1250    0.2222        40\n",
      "      providencia     1.0000    0.0625    0.1176        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     1.0000    0.5770    0.7318       513\n",
      "        ralstonia     1.0000    0.1905    0.3200        42\n",
      "     rheinheimera     1.0000    0.9659    0.9827       176\n",
      "        rhizobium     1.0000    0.2162    0.3556        74\n",
      "      rhodobacter     1.0000    0.5882    0.7407        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9537    0.4241    0.5871       632\n",
      "         serratia     0.8421    0.1649    0.2759        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.4615    0.0420    0.0769       143\n",
      "   staphylococcus     1.0000    0.8074    0.8934       244\n",
      " stenotrophomonas     1.0000    0.0488    0.0930        41\n",
      "    streptococcus     1.0000    0.3210    0.4860        81\n",
      "     streptomyces     1.0000    0.1628    0.2800        43\n",
      "    synechococcus     0.9945    0.6929    0.8167       521\n",
      "           vibrio     1.0000    0.3456    0.5136       463\n",
      "      xanthomonas     0.9000    0.1552    0.2647        58\n",
      "         yersinia     1.0000    0.2688    0.4237        93\n",
      "\n",
      "         accuracy                         0.4550      8116\n",
      "        macro avg     0.7577    0.2783    0.3595      8116\n",
      "     weighted avg     0.8406    0.4550    0.4846      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 90.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.0612    0.1154        98\n",
      "        aeromonas     1.0000    0.0211    0.0412        95\n",
      "    agrobacterium     0.0000    0.0000    0.0000        12\n",
      "     arthrobacter     1.0000    0.1154    0.2069        26\n",
      "         bacillus     1.0000    0.1961    0.3279       204\n",
      "      bacteroides     1.0000    0.2258    0.3684        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     1.0000    0.0513    0.0976        39\n",
      "    campylobacter     1.0000    0.2903    0.4500       124\n",
      "      caulobacter     1.0000    0.2400    0.3871        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     1.0000    0.0833    0.1538        24\n",
      "      cronobacter     0.0000    0.0000    0.0000        36\n",
      "          dickeya     1.0000    0.3654    0.5352        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.6000    0.0211    0.0408       142\n",
      "     enterococcus     1.0000    0.3673    0.5373        49\n",
      "          erwinia     0.9844    0.2203    0.3600       286\n",
      "      escherichia     0.9472    0.1799    0.3024      1295\n",
      "   flavobacterium     1.0000    0.6667    0.8000        48\n",
      "         gordonia     1.0000    0.1132    0.2034        53\n",
      "       halorubrum     0.3333    0.0769    0.1250        13\n",
      "     helicobacter     1.0000    0.3125    0.4762        16\n",
      "       klebsiella     0.9836    0.1511    0.2620       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.3700    0.5401       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.2143    0.3529        28\n",
      "   microbacterium     1.0000    0.1212    0.2162        33\n",
      "    mycobacterium     0.9846    0.7742    0.8668       248\n",
      "           others     0.1653    0.9939    0.2834       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.2958    0.4565        71\n",
      "     pelagibacter     1.0000    0.0303    0.0588        33\n",
      "       prevotella     1.0000    0.7778    0.8750        18\n",
      "  prochlorococcus     1.0000    0.2308    0.3750        52\n",
      "          proteus     1.0000    0.1000    0.1818        40\n",
      "      providencia     0.0000    0.0000    0.0000        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     1.0000    0.4503    0.6210       513\n",
      "        ralstonia     1.0000    0.0952    0.1739        42\n",
      "     rheinheimera     1.0000    0.9489    0.9738       176\n",
      "        rhizobium     1.0000    0.1351    0.2381        74\n",
      "      rhodobacter     1.0000    0.4412    0.6122        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9533    0.3228    0.4823       632\n",
      "         serratia     0.8750    0.0722    0.1333        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.3333    0.0210    0.0395       143\n",
      "   staphylococcus     1.0000    0.6844    0.8127       244\n",
      " stenotrophomonas     0.0000    0.0000    0.0000        41\n",
      "    streptococcus     1.0000    0.1358    0.2391        81\n",
      "     streptomyces     1.0000    0.0930    0.1702        43\n",
      "    synechococcus     0.9970    0.6468    0.7846       521\n",
      "           vibrio     1.0000    0.2721    0.4278       463\n",
      "      xanthomonas     0.8333    0.0862    0.1563        58\n",
      "         yersinia     1.0000    0.2043    0.3393        93\n",
      "\n",
      "         accuracy                         0.3850      8116\n",
      "        macro avg     0.6948    0.2081    0.2746      8116\n",
      "     weighted avg     0.8280    0.3850    0.4020      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 100.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.0102    0.0202        98\n",
      "        aeromonas     0.0000    0.0000    0.0000        95\n",
      "    agrobacterium     0.0000    0.0000    0.0000        12\n",
      "     arthrobacter     0.0000    0.0000    0.0000        26\n",
      "         bacillus     1.0000    0.0098    0.0194       204\n",
      "      bacteroides     0.0000    0.0000    0.0000        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     0.0000    0.0000    0.0000        39\n",
      "    campylobacter     1.0000    0.0484    0.0923       124\n",
      "      caulobacter     1.0000    0.0400    0.0769        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.0000    0.0000    0.0000        24\n",
      "      cronobacter     0.0000    0.0000    0.0000        36\n",
      "          dickeya     1.0000    0.2885    0.4478        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.5000    0.0070    0.0139       142\n",
      "     enterococcus     1.0000    0.0408    0.0784        49\n",
      "          erwinia     1.0000    0.0490    0.0933       286\n",
      "      escherichia     0.9444    0.0656    0.1227      1295\n",
      "   flavobacterium     1.0000    0.6250    0.7692        48\n",
      "         gordonia     0.0000    0.0000    0.0000        53\n",
      "       halorubrum     0.0000    0.0000    0.0000        13\n",
      "     helicobacter     0.0000    0.0000    0.0000        16\n",
      "       klebsiella     1.0000    0.0327    0.0634       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.1900    0.3193       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.0714    0.1333        28\n",
      "   microbacterium     1.0000    0.0303    0.0588        33\n",
      "    mycobacterium     0.9812    0.6331    0.7696       248\n",
      "           others     0.1417    0.9959    0.2481       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.1268    0.2250        71\n",
      "     pelagibacter     0.0000    0.0000    0.0000        33\n",
      "       prevotella     1.0000    0.2778    0.4348        18\n",
      "  prochlorococcus     1.0000    0.1731    0.2951        52\n",
      "          proteus     1.0000    0.0500    0.0952        40\n",
      "      providencia     0.0000    0.0000    0.0000        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     1.0000    0.2865    0.4455       513\n",
      "        ralstonia     0.0000    0.0000    0.0000        42\n",
      "     rheinheimera     1.0000    0.9318    0.9647       176\n",
      "        rhizobium     1.0000    0.0135    0.0267        74\n",
      "      rhodobacter     1.0000    0.1176    0.2105        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9841    0.0981    0.1784       632\n",
      "         serratia     0.0000    0.0000    0.0000        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     1.0000    0.0070    0.0139       143\n",
      "   staphylococcus     1.0000    0.3320    0.4985       244\n",
      " stenotrophomonas     0.0000    0.0000    0.0000        41\n",
      "    streptococcus     0.0000    0.0000    0.0000        81\n",
      "     streptomyces     1.0000    0.0465    0.0889        43\n",
      "    synechococcus     0.9962    0.5048    0.6701       521\n",
      "           vibrio     1.0000    0.1361    0.2395       463\n",
      "      xanthomonas     0.5000    0.0172    0.0333        58\n",
      "         yersinia     1.0000    0.1183    0.2115        93\n",
      "\n",
      "         accuracy                         0.2656      8116\n",
      "        macro avg     0.5093    0.1080    0.1349      8116\n",
      "     weighted avg     0.7707    0.2656    0.2392      8116\n",
      "\n",
      "===================\n",
      "Finished\n",
      "==============\n",
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Confidence threshold k: 0.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.5294    0.6923        17\n",
      "    acinetobacter     0.9672    0.6020    0.7421        98\n",
      "        aeromonas     0.9219    0.6211    0.7421        95\n",
      "    agrobacterium     1.0000    0.4167    0.5882        12\n",
      "     arthrobacter     1.0000    0.6923    0.8182        26\n",
      "         bacillus     0.7277    0.8382    0.7790       204\n",
      "      bacteroides     0.9667    0.9355    0.9508        31\n",
      "    brevundimonas     1.0000    0.6000    0.7500        10\n",
      "     burkholderia     0.6818    0.3846    0.4918        39\n",
      "    campylobacter     0.8443    0.8306    0.8374       124\n",
      "      caulobacter     0.7407    0.8000    0.7692        25\n",
      "      citrobacter     0.3333    0.2941    0.3125        34\n",
      "      clostridium     0.4667    0.5833    0.5185        24\n",
      "      cronobacter     0.5600    0.3889    0.4590        36\n",
      "          dickeya     0.8636    0.7308    0.7917        52\n",
      "     edwardsiella     1.0000    0.3000    0.4615        10\n",
      "     enterobacter     0.4310    0.3521    0.3876       142\n",
      "     enterococcus     0.9362    0.8980    0.9167        49\n",
      "          erwinia     0.7085    0.8497    0.7727       286\n",
      "      escherichia     0.5197    0.8239    0.6374      1295\n",
      "   flavobacterium     0.9574    0.9375    0.9474        48\n",
      "         gordonia     0.9048    0.7170    0.8000        53\n",
      "       halorubrum     0.2917    0.5385    0.3784        13\n",
      "     helicobacter     0.8421    1.0000    0.9143        16\n",
      "       klebsiella     0.7117    0.7834    0.7458       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.7000    0.3043    0.4242        23\n",
      "      lactococcus     0.9425    0.8200    0.8770       100\n",
      "      leuconostoc     1.0000    0.9000    0.9474        10\n",
      "         listeria     0.8571    0.8571    0.8571        28\n",
      "   microbacterium     0.9600    0.7273    0.8276        33\n",
      "    mycobacterium     0.9787    0.9274    0.9524       248\n",
      "           others     0.0000    0.0000    0.0000       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.7869    0.6761    0.7273        71\n",
      "     pelagibacter     0.5758    0.5758    0.5758        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.7381    0.5962    0.6596        52\n",
      "          proteus     0.8077    0.5250    0.6364        40\n",
      "      providencia     0.5000    0.2500    0.3333        16\n",
      "pseudoalteromonas     0.6667    0.1176    0.2000        17\n",
      "      pseudomonas     0.6965    0.8947    0.7833       513\n",
      "        ralstonia     0.9000    0.4286    0.5806        42\n",
      "     rheinheimera     0.9943    0.9943    0.9943       176\n",
      "        rhizobium     0.9500    0.7703    0.8507        74\n",
      "      rhodobacter     0.8056    0.8529    0.8286        34\n",
      "      rhodococcus     1.0000    0.2308    0.3750        13\n",
      "       salmonella     0.8228    0.8006    0.8115       632\n",
      "         serratia     0.7600    0.5876    0.6628        97\n",
      "       shewanella     1.0000    0.1818    0.3077        11\n",
      "         shigella     0.4615    0.2517    0.3258       143\n",
      "   staphylococcus     0.7836    0.9795    0.8707       244\n",
      " stenotrophomonas     0.8750    0.5122    0.6462        41\n",
      "    streptococcus     0.8421    0.7901    0.8153        81\n",
      "     streptomyces     0.9143    0.7442    0.8205        43\n",
      "    synechococcus     0.7066    0.9616    0.8146       521\n",
      "           vibrio     0.6307    0.8596    0.7276       463\n",
      "      xanthomonas     0.6829    0.4828    0.5657        58\n",
      "         yersinia     0.4959    0.6559    0.5648        93\n",
      "\n",
      "         accuracy                         0.6897      8116\n",
      "        macro avg     0.7494    0.6212    0.6532      8116\n",
      "     weighted avg     0.6263    0.6897    0.6431      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 10.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2941    0.4545        17\n",
      "    acinetobacter     1.0000    0.5510    0.7105        98\n",
      "        aeromonas     0.9455    0.5474    0.6933        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.5769    0.7317        26\n",
      "         bacillus     0.9209    0.7990    0.8556       204\n",
      "      bacteroides     0.9667    0.9355    0.9508        31\n",
      "    brevundimonas     1.0000    0.5000    0.6667        10\n",
      "     burkholderia     0.9167    0.2821    0.4314        39\n",
      "    campylobacter     0.9615    0.8065    0.8772       124\n",
      "      caulobacter     1.0000    0.7600    0.8636        25\n",
      "      citrobacter     0.4375    0.2059    0.2800        34\n",
      "      clostridium     0.5909    0.5417    0.5652        24\n",
      "      cronobacter     0.6842    0.3611    0.4727        36\n",
      "          dickeya     0.9500    0.7308    0.8261        52\n",
      "     edwardsiella     1.0000    0.2000    0.3333        10\n",
      "     enterobacter     0.4479    0.3028    0.3613       142\n",
      "     enterococcus     1.0000    0.8571    0.9231        49\n",
      "          erwinia     0.8968    0.7902    0.8401       286\n",
      "      escherichia     0.7521    0.7614    0.7567      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.6792    0.8090        53\n",
      "       halorubrum     0.3333    0.5385    0.4118        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.8778    0.6877    0.7712       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.7500    0.2609    0.3871        23\n",
      "      lactococcus     1.0000    0.8100    0.8950       100\n",
      "      leuconostoc     1.0000    0.8000    0.8889        10\n",
      "         listeria     0.8889    0.8571    0.8727        28\n",
      "   microbacterium     0.9565    0.6667    0.7857        33\n",
      "    mycobacterium     0.9870    0.9194    0.9520       248\n",
      "           others     0.3704    0.7880    0.5039       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.8980    0.6197    0.7333        71\n",
      "     pelagibacter     0.8333    0.4545    0.5882        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9333    0.5385    0.6829        52\n",
      "          proteus     0.8636    0.4750    0.6129        40\n",
      "      providencia     0.6667    0.2500    0.3636        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9584    0.8538    0.9031       513\n",
      "        ralstonia     0.9444    0.4048    0.5667        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     0.9808    0.6892    0.8095        74\n",
      "      rhodobacter     0.8788    0.8529    0.8657        34\n",
      "      rhodococcus     1.0000    0.0769    0.1429        13\n",
      "       salmonella     0.8960    0.7358    0.8080       632\n",
      "         serratia     0.8438    0.5567    0.6708        97\n",
      "       shewanella     1.0000    0.1818    0.3077        11\n",
      "         shigella     0.5172    0.2098    0.2985       143\n",
      "   staphylococcus     0.9225    0.9754    0.9482       244\n",
      " stenotrophomonas     0.9412    0.3902    0.5517        41\n",
      "    streptococcus     0.9254    0.7654    0.8378        81\n",
      "     streptomyces     1.0000    0.6279    0.7714        43\n",
      "    synechococcus     0.8725    0.9060    0.8889       521\n",
      "           vibrio     0.9359    0.7257    0.8175       463\n",
      "      xanthomonas     0.8333    0.4310    0.5682        58\n",
      "         yersinia     0.5234    0.6022    0.5600        93\n",
      "\n",
      "         accuracy                         0.7360      8116\n",
      "        macro avg     0.8441    0.5841    0.6579      8116\n",
      "     weighted avg     0.8024    0.7360    0.7457      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 20.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2941    0.4545        17\n",
      "    acinetobacter     1.0000    0.4592    0.6294        98\n",
      "        aeromonas     0.9556    0.4526    0.6143        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.5385    0.7000        26\n",
      "         bacillus     0.9394    0.7598    0.8401       204\n",
      "      bacteroides     1.0000    0.8065    0.8929        31\n",
      "    brevundimonas     1.0000    0.4000    0.5714        10\n",
      "     burkholderia     1.0000    0.2564    0.4082        39\n",
      "    campylobacter     0.9604    0.7823    0.8622       124\n",
      "      caulobacter     1.0000    0.7200    0.8372        25\n",
      "      citrobacter     0.5000    0.1765    0.2609        34\n",
      "      clostridium     0.7222    0.5417    0.6190        24\n",
      "      cronobacter     0.7692    0.2778    0.4082        36\n",
      "          dickeya     0.9487    0.7115    0.8132        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.5278    0.2676    0.3551       142\n",
      "     enterococcus     1.0000    0.8163    0.8989        49\n",
      "          erwinia     0.9401    0.7133    0.8111       286\n",
      "      escherichia     0.8353    0.6934    0.7578      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.6226    0.7674        53\n",
      "       halorubrum     0.3529    0.4615    0.4000        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.9414    0.6071    0.7381       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.8333    0.2174    0.3448        23\n",
      "      lactococcus     1.0000    0.7900    0.8827       100\n",
      "      leuconostoc     1.0000    0.7000    0.8235        10\n",
      "         listeria     0.8889    0.8571    0.8727        28\n",
      "   microbacterium     1.0000    0.6667    0.8000        33\n",
      "    mycobacterium     0.9870    0.9194    0.9520       248\n",
      "           others     0.3230    0.9290    0.4793       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9111    0.5775    0.7069        71\n",
      "     pelagibacter     1.0000    0.3636    0.5333        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9333    0.5385    0.6829        52\n",
      "          proteus     0.8500    0.4250    0.5667        40\n",
      "      providencia     0.6000    0.1875    0.2857        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9907    0.8285    0.9023       513\n",
      "        ralstonia     0.9444    0.4048    0.5667        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     1.0000    0.6351    0.7769        74\n",
      "      rhodobacter     0.9286    0.7647    0.8387        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9214    0.6867    0.7869       632\n",
      "         serratia     0.8596    0.5052    0.6364        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5294    0.1888    0.2784       143\n",
      "   staphylococcus     0.9791    0.9590    0.9689       244\n",
      " stenotrophomonas     0.8889    0.1951    0.3200        41\n",
      "    streptococcus     0.9655    0.6914    0.8058        81\n",
      "     streptomyces     1.0000    0.5349    0.6970        43\n",
      "    synechococcus     0.9737    0.8541    0.9100       521\n",
      "           vibrio     0.9967    0.6587    0.7932       463\n",
      "      xanthomonas     0.9231    0.4138    0.5714        58\n",
      "         yersinia     0.5484    0.5484    0.5484        93\n",
      "\n",
      "         accuracy                         0.7091      8116\n",
      "        macro avg     0.8334    0.5394    0.6237      8116\n",
      "     weighted avg     0.8337    0.7091    0.7301      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 30.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2941    0.4545        17\n",
      "    acinetobacter     1.0000    0.4184    0.5899        98\n",
      "        aeromonas     0.9474    0.3789    0.5414        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.5000    0.6667        26\n",
      "         bacillus     0.9346    0.7010    0.8011       204\n",
      "      bacteroides     1.0000    0.7742    0.8727        31\n",
      "    brevundimonas     1.0000    0.4000    0.5714        10\n",
      "     burkholderia     1.0000    0.2308    0.3750        39\n",
      "    campylobacter     0.9596    0.7661    0.8520       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     0.7143    0.1471    0.2439        34\n",
      "      clostridium     0.7222    0.5417    0.6190        24\n",
      "      cronobacter     0.7500    0.1667    0.2727        36\n",
      "          dickeya     1.0000    0.6731    0.8046        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.6207    0.2535    0.3600       142\n",
      "     enterococcus     1.0000    0.7143    0.8333        49\n",
      "          erwinia     0.9536    0.6469    0.7708       286\n",
      "      escherichia     0.8695    0.6278    0.7291      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.5660    0.7229        53\n",
      "       halorubrum     0.3529    0.4615    0.4000        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.9496    0.5693    0.7118       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.8000    0.1739    0.2857        23\n",
      "      lactococcus     1.0000    0.7700    0.8701       100\n",
      "      leuconostoc     1.0000    0.6000    0.7500        10\n",
      "         listeria     0.9200    0.8214    0.8679        28\n",
      "   microbacterium     1.0000    0.5758    0.7308        33\n",
      "    mycobacterium     0.9870    0.9194    0.9520       248\n",
      "           others     0.2867    0.9473    0.4402       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9231    0.5070    0.6545        71\n",
      "     pelagibacter     1.0000    0.2424    0.3902        33\n",
      "       prevotella     1.0000    0.8889    0.9412        18\n",
      "  prochlorococcus     0.9259    0.4808    0.6329        52\n",
      "          proteus     0.8500    0.4250    0.5667        40\n",
      "      providencia     0.6000    0.1875    0.2857        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9952    0.8031    0.8889       513\n",
      "        ralstonia     1.0000    0.3810    0.5517        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     1.0000    0.5811    0.7350        74\n",
      "      rhodobacter     0.9231    0.7059    0.8000        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9300    0.6519    0.7665       632\n",
      "         serratia     0.8846    0.4742    0.6174        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5455    0.1678    0.2567       143\n",
      "   staphylococcus     0.9869    0.9262    0.9556       244\n",
      " stenotrophomonas     0.8889    0.1951    0.3200        41\n",
      "    streptococcus     0.9811    0.6420    0.7761        81\n",
      "     streptomyces     1.0000    0.4884    0.6562        43\n",
      "    synechococcus     0.9929    0.8100    0.8922       521\n",
      "           vibrio     1.0000    0.6048    0.7537       463\n",
      "      xanthomonas     0.9545    0.3621    0.5250        58\n",
      "         yersinia     0.5904    0.5269    0.5568        93\n",
      "\n",
      "         accuracy                         0.6741      8116\n",
      "        macro avg     0.8431    0.5053    0.5975      8116\n",
      "     weighted avg     0.8426    0.6741    0.7034      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 40.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2353    0.3810        17\n",
      "    acinetobacter     1.0000    0.3776    0.5481        98\n",
      "        aeromonas     0.9600    0.2526    0.4000        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.3846    0.5556        26\n",
      "         bacillus     0.9500    0.6520    0.7733       204\n",
      "      bacteroides     1.0000    0.6774    0.8077        31\n",
      "    brevundimonas     1.0000    0.3000    0.4615        10\n",
      "     burkholderia     1.0000    0.2051    0.3404        39\n",
      "    campylobacter     0.9592    0.7581    0.8468       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     0.5000    0.0588    0.1053        34\n",
      "      clostridium     0.6875    0.4583    0.5500        24\n",
      "      cronobacter     0.6000    0.0833    0.1463        36\n",
      "          dickeya     1.0000    0.5962    0.7470        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.6250    0.1761    0.2747       142\n",
      "     enterococcus     1.0000    0.6939    0.8193        49\n",
      "          erwinia     0.9714    0.5944    0.7375       286\n",
      "      escherichia     0.8806    0.5583    0.6834      1295\n",
      "   flavobacterium     1.0000    0.8958    0.9451        48\n",
      "         gordonia     1.0000    0.4151    0.5867        53\n",
      "       halorubrum     0.3571    0.3846    0.3704        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.9469    0.5390    0.6870       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.6667    0.0870    0.1538        23\n",
      "      lactococcus     1.0000    0.7400    0.8506       100\n",
      "      leuconostoc     1.0000    0.6000    0.7500        10\n",
      "         listeria     0.9130    0.7500    0.8235        28\n",
      "   microbacterium     1.0000    0.5152    0.6800        33\n",
      "    mycobacterium     0.9868    0.9073    0.9454       248\n",
      "           others     0.2635    0.9604    0.4135       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9231    0.5070    0.6545        71\n",
      "     pelagibacter     1.0000    0.1818    0.3077        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     0.9259    0.4808    0.6329        52\n",
      "          proteus     0.9286    0.3250    0.4815        40\n",
      "      providencia     0.6667    0.1250    0.2105        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9950    0.7797    0.8743       513\n",
      "        ralstonia     1.0000    0.3810    0.5517        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     1.0000    0.5811    0.7350        74\n",
      "      rhodobacter     0.9200    0.6765    0.7797        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9327    0.6361    0.7563       632\n",
      "         serratia     0.8776    0.4433    0.5890        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5385    0.1469    0.2308       143\n",
      "   staphylococcus     1.0000    0.8975    0.9460       244\n",
      " stenotrophomonas     0.8889    0.1951    0.3200        41\n",
      "    streptococcus     1.0000    0.6173    0.7634        81\n",
      "     streptomyces     1.0000    0.4186    0.5902        43\n",
      "    synechococcus     0.9952    0.8004    0.8872       521\n",
      "           vibrio     1.0000    0.5724    0.7280       463\n",
      "      xanthomonas     0.9444    0.2931    0.4474        58\n",
      "         yersinia     0.6301    0.4946    0.5542        93\n",
      "\n",
      "         accuracy                         0.6417      8116\n",
      "        macro avg     0.8379    0.4649    0.5594      8116\n",
      "     weighted avg     0.8422    0.6417    0.6742      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 50.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.1176    0.2105        17\n",
      "    acinetobacter     1.0000    0.3265    0.4923        98\n",
      "        aeromonas     0.9600    0.2526    0.4000        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.3462    0.5143        26\n",
      "         bacillus     0.9615    0.6127    0.7485       204\n",
      "      bacteroides     1.0000    0.6452    0.7843        31\n",
      "    brevundimonas     1.0000    0.2000    0.3333        10\n",
      "     burkholderia     1.0000    0.1538    0.2667        39\n",
      "    campylobacter     0.9674    0.7177    0.8241       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     0.5000    0.0588    0.1053        34\n",
      "      clostridium     0.8333    0.4167    0.5556        24\n",
      "      cronobacter     0.6667    0.0556    0.1026        36\n",
      "          dickeya     1.0000    0.5385    0.7000        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.6786    0.1338    0.2235       142\n",
      "     enterococcus     1.0000    0.5918    0.7436        49\n",
      "          erwinia     0.9804    0.5245    0.6834       286\n",
      "      escherichia     0.9001    0.4942    0.6381      1295\n",
      "   flavobacterium     1.0000    0.8958    0.9451        48\n",
      "         gordonia     1.0000    0.3774    0.5479        53\n",
      "       halorubrum     0.3077    0.3077    0.3077        13\n",
      "     helicobacter     1.0000    0.8750    0.9333        16\n",
      "       klebsiella     0.9495    0.4736    0.6319       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.5000    0.0435    0.0800        23\n",
      "      lactococcus     1.0000    0.6900    0.8166       100\n",
      "      leuconostoc     1.0000    0.5000    0.6667        10\n",
      "         listeria     0.9048    0.6786    0.7755        28\n",
      "   microbacterium     1.0000    0.4545    0.6250        33\n",
      "    mycobacterium     0.9864    0.8790    0.9296       248\n",
      "           others     0.2422    0.9675    0.3874       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9459    0.4930    0.6481        71\n",
      "     pelagibacter     1.0000    0.0909    0.1667        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     0.9200    0.4423    0.5974        52\n",
      "          proteus     1.0000    0.2750    0.4314        40\n",
      "      providencia     0.6667    0.1250    0.2105        16\n",
      "pseudoalteromonas     1.0000    0.0588    0.1111        17\n",
      "      pseudomonas     0.9949    0.7583    0.8606       513\n",
      "        ralstonia     1.0000    0.3810    0.5517        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     1.0000    0.4730    0.6422        74\n",
      "      rhodobacter     0.9200    0.6765    0.7797        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9415    0.6108    0.7409       632\n",
      "         serratia     0.8667    0.4021    0.5493        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5938    0.1329    0.2171       143\n",
      "   staphylococcus     1.0000    0.8811    0.9368       244\n",
      " stenotrophomonas     0.8571    0.1463    0.2500        41\n",
      "    streptococcus     1.0000    0.5802    0.7344        81\n",
      "     streptomyces     1.0000    0.4186    0.5902        43\n",
      "    synechococcus     0.9951    0.7850    0.8777       521\n",
      "           vibrio     1.0000    0.5184    0.6828       463\n",
      "      xanthomonas     0.9412    0.2759    0.4267        58\n",
      "         yersinia     0.6286    0.4731    0.5399        93\n",
      "\n",
      "         accuracy                         0.6071      8116\n",
      "        macro avg     0.8408    0.4276    0.5235      8116\n",
      "     weighted avg     0.8465    0.6071    0.6432      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 60.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.1176    0.2105        17\n",
      "    acinetobacter     1.0000    0.2551    0.4065        98\n",
      "        aeromonas     1.0000    0.2000    0.3333        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.3077    0.4706        26\n",
      "         bacillus     0.9754    0.5833    0.7301       204\n",
      "      bacteroides     1.0000    0.5806    0.7347        31\n",
      "    brevundimonas     1.0000    0.2000    0.3333        10\n",
      "     burkholderia     1.0000    0.1282    0.2273        39\n",
      "    campylobacter     0.9634    0.6371    0.7670       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.8571    0.2500    0.3871        24\n",
      "      cronobacter     1.0000    0.0278    0.0541        36\n",
      "          dickeya     1.0000    0.5000    0.6667        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.7391    0.1197    0.2061       142\n",
      "     enterococcus     1.0000    0.5306    0.6933        49\n",
      "          erwinia     0.9853    0.4685    0.6351       286\n",
      "      escherichia     0.9065    0.4193    0.5734      1295\n",
      "   flavobacterium     1.0000    0.8542    0.9213        48\n",
      "         gordonia     1.0000    0.2830    0.4412        53\n",
      "       halorubrum     0.5000    0.3077    0.3810        13\n",
      "     helicobacter     1.0000    0.7500    0.8571        16\n",
      "       klebsiella     0.9538    0.4156    0.5789       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.6500    0.7879       100\n",
      "      leuconostoc     1.0000    0.3000    0.4615        10\n",
      "         listeria     0.9000    0.6429    0.7500        28\n",
      "   microbacterium     1.0000    0.4242    0.5957        33\n",
      "    mycobacterium     0.9861    0.8589    0.9181       248\n",
      "           others     0.2227    0.9797    0.3629       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9706    0.4648    0.6286        71\n",
      "     pelagibacter     1.0000    0.0606    0.1143        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     0.9524    0.3846    0.5479        52\n",
      "          proteus     1.0000    0.2250    0.3673        40\n",
      "      providencia     0.5000    0.0625    0.1111        16\n",
      "pseudoalteromonas     1.0000    0.0588    0.1111        17\n",
      "      pseudomonas     0.9973    0.7310    0.8436       513\n",
      "        ralstonia     1.0000    0.3095    0.4727        42\n",
      "     rheinheimera     1.0000    0.9830    0.9914       176\n",
      "        rhizobium     1.0000    0.4189    0.5905        74\n",
      "      rhodobacter     0.9565    0.6471    0.7719        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9383    0.5538    0.6965       632\n",
      "         serratia     0.8571    0.3093    0.4545        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.6538    0.1189    0.2012       143\n",
      "   staphylococcus     1.0000    0.8689    0.9298       244\n",
      " stenotrophomonas     1.0000    0.1463    0.2553        41\n",
      "    streptococcus     1.0000    0.4938    0.6612        81\n",
      "     streptomyces     1.0000    0.3953    0.5667        43\n",
      "    synechococcus     0.9950    0.7601    0.8618       521\n",
      "           vibrio     1.0000    0.4773    0.6462       463\n",
      "      xanthomonas     0.9286    0.2241    0.3611        58\n",
      "         yersinia     0.7778    0.3763    0.5072        93\n",
      "\n",
      "         accuracy                         0.5667      8116\n",
      "        macro avg     0.8393    0.3859    0.4839      8116\n",
      "     weighted avg     0.8491    0.5667    0.6045      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 70.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.1735    0.2957        98\n",
      "        aeromonas     1.0000    0.1158    0.2075        95\n",
      "    agrobacterium     1.0000    0.2500    0.4000        12\n",
      "     arthrobacter     1.0000    0.2692    0.4242        26\n",
      "         bacillus     0.9712    0.4951    0.6558       204\n",
      "      bacteroides     1.0000    0.4516    0.6222        31\n",
      "    brevundimonas     1.0000    0.1000    0.1818        10\n",
      "     burkholderia     1.0000    0.0769    0.1429        39\n",
      "    campylobacter     0.9726    0.5726    0.7208       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.8000    0.1667    0.2759        24\n",
      "      cronobacter     1.0000    0.0278    0.0541        36\n",
      "          dickeya     1.0000    0.4423    0.6133        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.7000    0.0493    0.0921       142\n",
      "     enterococcus     1.0000    0.4694    0.6389        49\n",
      "          erwinia     0.9828    0.3986    0.5672       286\n",
      "      escherichia     0.9167    0.3483    0.5048      1295\n",
      "   flavobacterium     1.0000    0.8542    0.9213        48\n",
      "         gordonia     1.0000    0.2453    0.3939        53\n",
      "       halorubrum     0.4286    0.2308    0.3000        13\n",
      "     helicobacter     1.0000    0.4375    0.6087        16\n",
      "       klebsiella     0.9615    0.3149    0.4744       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.5400    0.7013       100\n",
      "      leuconostoc     1.0000    0.1000    0.1818        10\n",
      "         listeria     0.9333    0.5000    0.6512        28\n",
      "   microbacterium     1.0000    0.3333    0.5000        33\n",
      "    mycobacterium     0.9857    0.8347    0.9039       248\n",
      "           others     0.2012    0.9858    0.3343       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9697    0.4507    0.6154        71\n",
      "     pelagibacter     1.0000    0.0606    0.1143        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     1.0000    0.2500    0.4000        52\n",
      "          proteus     1.0000    0.1500    0.2609        40\n",
      "      providencia     1.0000    0.0625    0.1176        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     0.9970    0.6511    0.7877       513\n",
      "        ralstonia     1.0000    0.2619    0.4151        42\n",
      "     rheinheimera     1.0000    0.9773    0.9885       176\n",
      "        rhizobium     1.0000    0.2703    0.4255        74\n",
      "      rhodobacter     1.0000    0.6471    0.7857        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9410    0.4794    0.6352       632\n",
      "         serratia     0.8889    0.2474    0.3871        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.6522    0.1049    0.1807       143\n",
      "   staphylococcus     1.0000    0.8484    0.9180       244\n",
      " stenotrophomonas     1.0000    0.0976    0.1778        41\n",
      "    streptococcus     1.0000    0.4321    0.6034        81\n",
      "     streptomyces     1.0000    0.3488    0.5172        43\n",
      "    synechococcus     0.9948    0.7332    0.8442       521\n",
      "           vibrio     1.0000    0.4082    0.5798       463\n",
      "      xanthomonas     0.9091    0.1724    0.2899        58\n",
      "         yersinia     0.9667    0.3118    0.4715        93\n",
      "\n",
      "         accuracy                         0.5121      8116\n",
      "        macro avg     0.7995    0.3258    0.4165      8116\n",
      "     weighted avg     0.8461    0.5121    0.5482      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 80.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.1122    0.2018        98\n",
      "        aeromonas     1.0000    0.1053    0.1905        95\n",
      "    agrobacterium     1.0000    0.2500    0.4000        12\n",
      "     arthrobacter     1.0000    0.2308    0.3750        26\n",
      "         bacillus     0.9747    0.3775    0.5442       204\n",
      "      bacteroides     1.0000    0.3871    0.5581        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     1.0000    0.0513    0.0976        39\n",
      "    campylobacter     1.0000    0.4274    0.5989       124\n",
      "      caulobacter     1.0000    0.6000    0.7500        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.6667    0.0833    0.1481        24\n",
      "      cronobacter     1.0000    0.0278    0.0541        36\n",
      "          dickeya     1.0000    0.4423    0.6133        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.5000    0.0211    0.0405       142\n",
      "     enterococcus     1.0000    0.4694    0.6389        49\n",
      "          erwinia     0.9759    0.2832    0.4390       286\n",
      "      escherichia     0.9328    0.2571    0.4031      1295\n",
      "   flavobacterium     1.0000    0.8333    0.9091        48\n",
      "         gordonia     1.0000    0.1887    0.3175        53\n",
      "       halorubrum     0.3333    0.1538    0.2105        13\n",
      "     helicobacter     1.0000    0.1875    0.3158        16\n",
      "       klebsiella     0.9785    0.2292    0.3714       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.5000    0.6667       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.4286    0.6000        28\n",
      "   microbacterium     1.0000    0.3030    0.4651        33\n",
      "    mycobacterium     0.9854    0.8185    0.8943       248\n",
      "           others     0.1830    0.9899    0.3090       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.3239    0.4894        71\n",
      "     pelagibacter     1.0000    0.0303    0.0588        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     1.0000    0.2500    0.4000        52\n",
      "          proteus     1.0000    0.1250    0.2222        40\n",
      "      providencia     1.0000    0.0625    0.1176        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     1.0000    0.5789    0.7333       513\n",
      "        ralstonia     1.0000    0.2381    0.3846        42\n",
      "     rheinheimera     1.0000    0.9659    0.9827       176\n",
      "        rhizobium     1.0000    0.2297    0.3736        74\n",
      "      rhodobacter     1.0000    0.6471    0.7857        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9517    0.4051    0.5683       632\n",
      "         serratia     0.8889    0.1649    0.2783        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5000    0.0350    0.0654       143\n",
      "   staphylococcus     1.0000    0.8033    0.8909       244\n",
      " stenotrophomonas     1.0000    0.0488    0.0930        41\n",
      "    streptococcus     1.0000    0.3333    0.5000        81\n",
      "     streptomyces     1.0000    0.2326    0.3774        43\n",
      "    synechococcus     0.9945    0.6967    0.8194       521\n",
      "           vibrio     1.0000    0.3477    0.5160       463\n",
      "      xanthomonas     0.8889    0.1379    0.2388        58\n",
      "         yersinia     0.9615    0.2688    0.4202        93\n",
      "\n",
      "         accuracy                         0.4553      8116\n",
      "        macro avg     0.7579    0.2800    0.3616      8116\n",
      "     weighted avg     0.8397    0.4553    0.4851      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 90.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.0510    0.0971        98\n",
      "        aeromonas     1.0000    0.0421    0.0808        95\n",
      "    agrobacterium     0.0000    0.0000    0.0000        12\n",
      "     arthrobacter     1.0000    0.1923    0.3226        26\n",
      "         bacillus     0.9783    0.2206    0.3600       204\n",
      "      bacteroides     1.0000    0.0323    0.0625        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     1.0000    0.0256    0.0500        39\n",
      "    campylobacter     1.0000    0.2823    0.4403       124\n",
      "      caulobacter     1.0000    0.2000    0.3333        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     1.0000    0.0833    0.1538        24\n",
      "      cronobacter     0.0000    0.0000    0.0000        36\n",
      "          dickeya     1.0000    0.3654    0.5352        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.6000    0.0211    0.0408       142\n",
      "     enterococcus     1.0000    0.4286    0.6000        49\n",
      "          erwinia     1.0000    0.2098    0.3468       286\n",
      "      escherichia     0.9512    0.1807    0.3037      1295\n",
      "   flavobacterium     1.0000    0.6667    0.8000        48\n",
      "         gordonia     1.0000    0.1321    0.2333        53\n",
      "       halorubrum     0.3333    0.0769    0.1250        13\n",
      "     helicobacter     1.0000    0.0625    0.1176        16\n",
      "       klebsiella     0.9846    0.1612    0.2771       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.4500    0.6207       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.2500    0.4000        28\n",
      "   microbacterium     1.0000    0.1212    0.2162        33\n",
      "    mycobacterium     0.9845    0.7661    0.8617       248\n",
      "           others     0.1660    0.9939    0.2846       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.3099    0.4731        71\n",
      "     pelagibacter     1.0000    0.0303    0.0588        33\n",
      "       prevotella     1.0000    0.7222    0.8387        18\n",
      "  prochlorococcus     1.0000    0.2115    0.3492        52\n",
      "          proteus     1.0000    0.0750    0.1395        40\n",
      "      providencia     0.0000    0.0000    0.0000        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     1.0000    0.4717    0.6411       513\n",
      "        ralstonia     1.0000    0.0714    0.1333        42\n",
      "     rheinheimera     1.0000    0.9659    0.9827       176\n",
      "        rhizobium     1.0000    0.1351    0.2381        74\n",
      "      rhodobacter     1.0000    0.4118    0.5833        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9626    0.3259    0.4870       632\n",
      "         serratia     0.8571    0.0619    0.1154        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.4286    0.0210    0.0400       143\n",
      "   staphylococcus     1.0000    0.7090    0.8297       244\n",
      " stenotrophomonas     0.0000    0.0000    0.0000        41\n",
      "    streptococcus     1.0000    0.1975    0.3299        81\n",
      "     streptomyces     1.0000    0.1163    0.2083        43\n",
      "    synechococcus     0.9971    0.6526    0.7889       521\n",
      "           vibrio     1.0000    0.2743    0.4305       463\n",
      "      xanthomonas     0.8333    0.0862    0.1563        58\n",
      "         yersinia     0.9412    0.1720    0.2909        93\n",
      "\n",
      "         accuracy                         0.3890      8116\n",
      "        macro avg     0.6952    0.2040    0.2674      8116\n",
      "     weighted avg     0.8303    0.3890    0.4057      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 100.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.0102    0.0202        98\n",
      "        aeromonas     0.0000    0.0000    0.0000        95\n",
      "    agrobacterium     0.0000    0.0000    0.0000        12\n",
      "     arthrobacter     0.0000    0.0000    0.0000        26\n",
      "         bacillus     1.0000    0.0147    0.0290       204\n",
      "      bacteroides     0.0000    0.0000    0.0000        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     0.0000    0.0000    0.0000        39\n",
      "    campylobacter     1.0000    0.0403    0.0775       124\n",
      "      caulobacter     0.0000    0.0000    0.0000        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.0000    0.0000    0.0000        24\n",
      "      cronobacter     0.0000    0.0000    0.0000        36\n",
      "          dickeya     1.0000    0.2885    0.4478        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.0000    0.0000    0.0000       142\n",
      "     enterococcus     1.0000    0.0612    0.1154        49\n",
      "          erwinia     1.0000    0.0280    0.0544       286\n",
      "      escherichia     0.9479    0.0703    0.1308      1295\n",
      "   flavobacterium     1.0000    0.6250    0.7692        48\n",
      "         gordonia     0.0000    0.0000    0.0000        53\n",
      "       halorubrum     0.0000    0.0000    0.0000        13\n",
      "     helicobacter     0.0000    0.0000    0.0000        16\n",
      "       klebsiella     1.0000    0.0378    0.0728       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.1600    0.2759       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.0357    0.0690        28\n",
      "   microbacterium     1.0000    0.0606    0.1143        33\n",
      "    mycobacterium     0.9810    0.6250    0.7635       248\n",
      "           others     0.1414    0.9959    0.2476       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.1268    0.2250        71\n",
      "     pelagibacter     0.0000    0.0000    0.0000        33\n",
      "       prevotella     1.0000    0.0556    0.1053        18\n",
      "  prochlorococcus     1.0000    0.1731    0.2951        52\n",
      "          proteus     1.0000    0.0500    0.0952        40\n",
      "      providencia     0.0000    0.0000    0.0000        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     1.0000    0.2827    0.4407       513\n",
      "        ralstonia     0.0000    0.0000    0.0000        42\n",
      "     rheinheimera     1.0000    0.9318    0.9647       176\n",
      "        rhizobium     1.0000    0.0135    0.0267        74\n",
      "      rhodobacter     1.0000    0.1176    0.2105        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     1.0000    0.1076    0.1943       632\n",
      "         serratia     0.0000    0.0000    0.0000        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     1.0000    0.0070    0.0139       143\n",
      "   staphylococcus     1.0000    0.3115    0.4750       244\n",
      " stenotrophomonas     0.0000    0.0000    0.0000        41\n",
      "    streptococcus     0.0000    0.0000    0.0000        81\n",
      "     streptomyces     1.0000    0.0465    0.0889        43\n",
      "    synechococcus     0.9960    0.4798    0.6477       521\n",
      "           vibrio     1.0000    0.1490    0.2594       463\n",
      "      xanthomonas     0.5000    0.0172    0.0333        58\n",
      "         yersinia     1.0000    0.1290    0.2286        93\n",
      "\n",
      "         accuracy                         0.2638      8116\n",
      "        macro avg     0.4842    0.1026    0.1270      8116\n",
      "     weighted avg     0.7606    0.2638    0.2380      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished\n",
      "==============\n",
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Confidence threshold k: 0.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.9000    0.5294    0.6667        17\n",
      "    acinetobacter     0.9565    0.6735    0.7904        98\n",
      "        aeromonas     0.8267    0.6526    0.7294        95\n",
      "    agrobacterium     0.8333    0.4167    0.5556        12\n",
      "     arthrobacter     0.8333    0.7692    0.8000        26\n",
      "         bacillus     0.8128    0.8725    0.8416       204\n",
      "      bacteroides     0.9032    0.9032    0.9032        31\n",
      "    brevundimonas     1.0000    0.6000    0.7500        10\n",
      "     burkholderia     0.6452    0.5128    0.5714        39\n",
      "    campylobacter     0.8655    0.8306    0.8477       124\n",
      "      caulobacter     0.9545    0.8400    0.8936        25\n",
      "      citrobacter     0.3448    0.2941    0.3175        34\n",
      "      clostridium     0.5000    0.6667    0.5714        24\n",
      "      cronobacter     0.5217    0.3333    0.4068        36\n",
      "          dickeya     0.9048    0.7308    0.8085        52\n",
      "     edwardsiella     1.0000    0.3000    0.4615        10\n",
      "     enterobacter     0.4500    0.3803    0.4122       142\n",
      "     enterococcus     0.7925    0.8571    0.8235        49\n",
      "          erwinia     0.6322    0.8776    0.7350       286\n",
      "      escherichia     0.5646    0.8162    0.6675      1295\n",
      "   flavobacterium     0.9184    0.9375    0.9278        48\n",
      "         gordonia     0.8800    0.8302    0.8544        53\n",
      "       halorubrum     0.3200    0.6154    0.4211        13\n",
      "     helicobacter     0.9412    1.0000    0.9697        16\n",
      "       klebsiella     0.7210    0.8463    0.7787       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.7727    0.7391    0.7556        23\n",
      "      lactococcus     0.9247    0.8600    0.8912       100\n",
      "      leuconostoc     0.7500    0.9000    0.8182        10\n",
      "         listeria     0.8889    0.8571    0.8727        28\n",
      "   microbacterium     0.7714    0.8182    0.7941        33\n",
      "    mycobacterium     0.9520    0.9597    0.9558       248\n",
      "           others     0.0000    0.0000    0.0000       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.8909    0.6901    0.7778        71\n",
      "     pelagibacter     0.7353    0.7576    0.7463        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9143    0.6154    0.7356        52\n",
      "          proteus     0.8077    0.5250    0.6364        40\n",
      "      providencia     0.7143    0.3125    0.4348        16\n",
      "pseudoalteromonas     0.7500    0.1765    0.2857        17\n",
      "      pseudomonas     0.7567    0.8850    0.8158       513\n",
      "        ralstonia     0.7143    0.4762    0.5714        42\n",
      "     rheinheimera     1.0000    1.0000    1.0000       176\n",
      "        rhizobium     0.8667    0.8784    0.8725        74\n",
      "      rhodobacter     0.9375    0.8824    0.9091        34\n",
      "      rhodococcus     0.6000    0.2308    0.3333        13\n",
      "       salmonella     0.7982    0.8196    0.8087       632\n",
      "         serratia     0.7763    0.6082    0.6821        97\n",
      "       shewanella     1.0000    0.3636    0.5333        11\n",
      "         shigella     0.4250    0.2378    0.3049       143\n",
      "   staphylococcus     0.8511    0.9836    0.9125       244\n",
      " stenotrophomonas     0.6389    0.5610    0.5974        41\n",
      "    streptococcus     0.7931    0.8519    0.8214        81\n",
      "     streptomyces     0.9706    0.7674    0.8571        43\n",
      "    synechococcus     0.7891    0.9770    0.8731       521\n",
      "           vibrio     0.5813    0.9114    0.7098       463\n",
      "      xanthomonas     0.6875    0.5690    0.6226        58\n",
      "         yersinia     0.5000    0.6452    0.5634        93\n",
      "\n",
      "         accuracy                         0.7091      8116\n",
      "        macro avg     0.7387    0.6592    0.6774      8116\n",
      "     weighted avg     0.6357    0.7091    0.6610      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 10.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.3529    0.5217        17\n",
      "    acinetobacter     1.0000    0.5816    0.7355        98\n",
      "        aeromonas     0.9194    0.6000    0.7261        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.7308    0.8444        26\n",
      "         bacillus     0.9086    0.8284    0.8667       204\n",
      "      bacteroides     0.9231    0.7742    0.8421        31\n",
      "    brevundimonas     1.0000    0.5000    0.6667        10\n",
      "     burkholderia     0.9444    0.4359    0.5965        39\n",
      "    campylobacter     0.8899    0.7823    0.8326       124\n",
      "      caulobacter     1.0000    0.8000    0.8889        25\n",
      "      citrobacter     0.5000    0.2647    0.3462        34\n",
      "      clostridium     0.5769    0.6250    0.6000        24\n",
      "      cronobacter     0.6250    0.2778    0.3846        36\n",
      "          dickeya     0.9268    0.7308    0.8172        52\n",
      "     edwardsiella     1.0000    0.2000    0.3333        10\n",
      "     enterobacter     0.4747    0.3310    0.3900       142\n",
      "     enterococcus     0.9318    0.8367    0.8817        49\n",
      "          erwinia     0.8910    0.8287    0.8587       286\n",
      "      escherichia     0.7529    0.7552    0.7540      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     0.9767    0.7925    0.8750        53\n",
      "       halorubrum     0.3182    0.5385    0.4000        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.8822    0.7355    0.8022       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.8824    0.6522    0.7500        23\n",
      "      lactococcus     0.9759    0.8100    0.8852       100\n",
      "      leuconostoc     1.0000    0.9000    0.9474        10\n",
      "         listeria     0.8889    0.8571    0.8727        28\n",
      "   microbacterium     1.0000    0.7576    0.8621        33\n",
      "    mycobacterium     0.9708    0.9395    0.9549       248\n",
      "           others     0.4084    0.7759    0.5352       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9020    0.6479    0.7541        71\n",
      "     pelagibacter     0.8400    0.6364    0.7241        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9355    0.5577    0.6988        52\n",
      "          proteus     0.8333    0.5000    0.6250        40\n",
      "      providencia     0.6667    0.2500    0.3636        16\n",
      "pseudoalteromonas     1.0000    0.1765    0.3000        17\n",
      "      pseudomonas     0.9716    0.8655    0.9155       513\n",
      "        ralstonia     0.9000    0.4286    0.5806        42\n",
      "     rheinheimera     1.0000    1.0000    1.0000       176\n",
      "        rhizobium     0.9403    0.8514    0.8936        74\n",
      "      rhodobacter     0.9310    0.7941    0.8571        34\n",
      "      rhodococcus     1.0000    0.1538    0.2667        13\n",
      "       salmonella     0.8973    0.7468    0.8152       632\n",
      "         serratia     0.8529    0.5979    0.7030        97\n",
      "       shewanella     1.0000    0.2727    0.4286        11\n",
      "         shigella     0.4839    0.2098    0.2927       143\n",
      "   staphylococcus     0.9480    0.9713    0.9595       244\n",
      " stenotrophomonas     0.8182    0.4390    0.5714        41\n",
      "    streptococcus     0.8553    0.8025    0.8280        81\n",
      "     streptomyces     1.0000    0.7442    0.8533        43\n",
      "    synechococcus     0.9361    0.9559    0.9459       521\n",
      "           vibrio     0.8799    0.8229    0.8504       463\n",
      "      xanthomonas     0.7750    0.5345    0.6327        58\n",
      "         yersinia     0.5140    0.5914    0.5500        93\n",
      "\n",
      "         accuracy                         0.7585      8116\n",
      "        macro avg     0.8381    0.6231    0.6885      8116\n",
      "     weighted avg     0.8047    0.7585    0.7651      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 20.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2941    0.4545        17\n",
      "    acinetobacter     1.0000    0.5510    0.7105        98\n",
      "        aeromonas     0.9623    0.5368    0.6892        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.6538    0.7907        26\n",
      "         bacillus     0.9425    0.8039    0.8677       204\n",
      "      bacteroides     1.0000    0.7742    0.8727        31\n",
      "    brevundimonas     1.0000    0.5000    0.6667        10\n",
      "     burkholderia     0.9333    0.3590    0.5185        39\n",
      "    campylobacter     0.8889    0.7742    0.8276       124\n",
      "      caulobacter     1.0000    0.7600    0.8636        25\n",
      "      citrobacter     0.5000    0.1765    0.2609        34\n",
      "      clostridium     0.6500    0.5417    0.5909        24\n",
      "      cronobacter     0.6667    0.2222    0.3333        36\n",
      "          dickeya     0.9744    0.7308    0.8352        52\n",
      "     edwardsiella     1.0000    0.2000    0.3333        10\n",
      "     enterobacter     0.5821    0.2746    0.3732       142\n",
      "     enterococcus     1.0000    0.8163    0.8989        49\n",
      "          erwinia     0.9402    0.7692    0.8462       286\n",
      "      escherichia     0.8085    0.7042    0.7528      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.6981    0.8222        53\n",
      "       halorubrum     0.3000    0.4615    0.3636        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.9250    0.6524    0.7651       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.8462    0.4783    0.6111        23\n",
      "      lactococcus     0.9878    0.8100    0.8901       100\n",
      "      leuconostoc     1.0000    0.8000    0.8889        10\n",
      "         listeria     0.9231    0.8571    0.8889        28\n",
      "   microbacterium     1.0000    0.7273    0.8421        33\n",
      "    mycobacterium     0.9828    0.9234    0.9522       248\n",
      "           others     0.3515    0.8955    0.5049       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.8936    0.5915    0.7119        71\n",
      "     pelagibacter     1.0000    0.5152    0.6800        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9333    0.5385    0.6829        52\n",
      "          proteus     0.8571    0.4500    0.5902        40\n",
      "      providencia     0.6000    0.1875    0.2857        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9886    0.8460    0.9118       513\n",
      "        ralstonia     0.9412    0.3810    0.5424        42\n",
      "     rheinheimera     1.0000    1.0000    1.0000       176\n",
      "        rhizobium     0.9836    0.8108    0.8889        74\n",
      "      rhodobacter     0.9310    0.7941    0.8571        34\n",
      "      rhodococcus     1.0000    0.0769    0.1429        13\n",
      "       salmonella     0.9208    0.6994    0.7950       632\n",
      "         serratia     0.8772    0.5155    0.6494        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5600    0.1958    0.2902       143\n",
      "   staphylococcus     0.9752    0.9672    0.9712       244\n",
      " stenotrophomonas     0.8750    0.3415    0.4912        41\n",
      "    streptococcus     0.9118    0.7654    0.8322        81\n",
      "     streptomyces     1.0000    0.6977    0.8219        43\n",
      "    synechococcus     0.9641    0.9290    0.9462       521\n",
      "           vibrio     0.9548    0.7300    0.8274       463\n",
      "      xanthomonas     0.8438    0.4655    0.6000        58\n",
      "         yersinia     0.5532    0.5591    0.5561        93\n",
      "\n",
      "         accuracy                         0.7341      8116\n",
      "        macro avg     0.8429    0.5786    0.6566      8116\n",
      "     weighted avg     0.8288    0.7341    0.7496      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 30.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2941    0.4545        17\n",
      "    acinetobacter     1.0000    0.4898    0.6575        98\n",
      "        aeromonas     0.9574    0.4737    0.6338        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.6538    0.7907        26\n",
      "         bacillus     0.9503    0.7500    0.8384       204\n",
      "      bacteroides     1.0000    0.7419    0.8519        31\n",
      "    brevundimonas     1.0000    0.5000    0.6667        10\n",
      "     burkholderia     1.0000    0.3333    0.5000        39\n",
      "    campylobacter     0.8857    0.7500    0.8122       124\n",
      "      caulobacter     1.0000    0.7600    0.8636        25\n",
      "      citrobacter     0.5714    0.1176    0.1951        34\n",
      "      clostridium     0.7222    0.5417    0.6190        24\n",
      "      cronobacter     0.8000    0.2222    0.3478        36\n",
      "          dickeya     1.0000    0.6538    0.7907        52\n",
      "     edwardsiella     1.0000    0.2000    0.3333        10\n",
      "     enterobacter     0.6250    0.2465    0.3535       142\n",
      "     enterococcus     1.0000    0.7347    0.8471        49\n",
      "          erwinia     0.9575    0.7098    0.8153       286\n",
      "      escherichia     0.8560    0.6517    0.7400      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.6038    0.7529        53\n",
      "       halorubrum     0.3158    0.4615    0.3750        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.9457    0.6146    0.7450       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.8182    0.3913    0.5294        23\n",
      "      lactococcus     1.0000    0.7900    0.8827       100\n",
      "      leuconostoc     1.0000    0.6000    0.7500        10\n",
      "         listeria     0.9231    0.8571    0.8889        28\n",
      "   microbacterium     1.0000    0.7273    0.8421        33\n",
      "    mycobacterium     0.9871    0.9234    0.9542       248\n",
      "           others     0.3128    0.9371    0.4690       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9111    0.5775    0.7069        71\n",
      "     pelagibacter     1.0000    0.3333    0.5000        33\n",
      "       prevotella     1.0000    0.8889    0.9412        18\n",
      "  prochlorococcus     0.9333    0.5385    0.6829        52\n",
      "          proteus     0.8500    0.4250    0.5667        40\n",
      "      providencia     0.5000    0.1250    0.2000        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9929    0.8226    0.8998       513\n",
      "        ralstonia     1.0000    0.3810    0.5517        42\n",
      "     rheinheimera     1.0000    1.0000    1.0000       176\n",
      "        rhizobium     0.9825    0.7568    0.8550        74\n",
      "      rhodobacter     0.9259    0.7353    0.8197        34\n",
      "      rhodococcus     1.0000    0.0769    0.1429        13\n",
      "       salmonella     0.9256    0.6693    0.7769       632\n",
      "         serratia     0.8846    0.4742    0.6174        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5682    0.1748    0.2674       143\n",
      "   staphylococcus     0.9872    0.9508    0.9687       244\n",
      " stenotrophomonas     0.9167    0.2683    0.4151        41\n",
      "    streptococcus     0.9516    0.7284    0.8252        81\n",
      "     streptomyces     1.0000    0.5814    0.7353        43\n",
      "    synechococcus     0.9763    0.8695    0.9198       521\n",
      "           vibrio     0.9904    0.6717    0.8005       463\n",
      "      xanthomonas     0.8621    0.4310    0.5747        58\n",
      "         yersinia     0.5854    0.5161    0.5486        93\n",
      "\n",
      "         accuracy                         0.7039      8116\n",
      "        macro avg     0.8538    0.5443    0.6321      8116\n",
      "     weighted avg     0.8413    0.7039    0.7278      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 40.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2941    0.4545        17\n",
      "    acinetobacter     1.0000    0.4388    0.6099        98\n",
      "        aeromonas     0.9722    0.3684    0.5344        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.6538    0.7907        26\n",
      "         bacillus     0.9603    0.7108    0.8169       204\n",
      "      bacteroides     1.0000    0.7097    0.8302        31\n",
      "    brevundimonas     1.0000    0.3000    0.4615        10\n",
      "     burkholderia     1.0000    0.2564    0.4082        39\n",
      "    campylobacter     0.9271    0.7177    0.8091       124\n",
      "      caulobacter     1.0000    0.6800    0.8095        25\n",
      "      citrobacter     0.7500    0.0882    0.1579        34\n",
      "      clostridium     0.6667    0.4167    0.5128        24\n",
      "      cronobacter     0.6667    0.1111    0.1905        36\n",
      "          dickeya     1.0000    0.6346    0.7765        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.6222    0.1972    0.2995       142\n",
      "     enterococcus     1.0000    0.6939    0.8193        49\n",
      "          erwinia     0.9590    0.6538    0.7775       286\n",
      "      escherichia     0.8720    0.5892    0.7032      1295\n",
      "   flavobacterium     1.0000    0.8958    0.9451        48\n",
      "         gordonia     1.0000    0.5660    0.7229        53\n",
      "       halorubrum     0.4000    0.4615    0.4286        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.9542    0.5768    0.7190       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.7500    0.2609    0.3871        23\n",
      "      lactococcus     1.0000    0.7700    0.8701       100\n",
      "      leuconostoc     1.0000    0.6000    0.7500        10\n",
      "         listeria     0.9231    0.8571    0.8889        28\n",
      "   microbacterium     1.0000    0.6970    0.8214        33\n",
      "    mycobacterium     0.9870    0.9194    0.9520       248\n",
      "           others     0.2824    0.9523    0.4356       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9474    0.5070    0.6606        71\n",
      "     pelagibacter     1.0000    0.3030    0.4651        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     0.9259    0.4808    0.6329        52\n",
      "          proteus     0.9412    0.4000    0.5614        40\n",
      "      providencia     0.5000    0.1250    0.2000        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9951    0.7992    0.8865       513\n",
      "        ralstonia     1.0000    0.3571    0.5263        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     0.9800    0.6622    0.7903        74\n",
      "      rhodobacter     0.9231    0.7059    0.8000        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9404    0.6487    0.7678       632\n",
      "         serratia     0.9000    0.4639    0.6122        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5641    0.1538    0.2418       143\n",
      "   staphylococcus     0.9912    0.9221    0.9554       244\n",
      " stenotrophomonas     0.9000    0.2195    0.3529        41\n",
      "    streptococcus     0.9821    0.6790    0.8029        81\n",
      "     streptomyces     1.0000    0.5581    0.7164        43\n",
      "    synechococcus     0.9842    0.8349    0.9034       521\n",
      "           vibrio     1.0000    0.6263    0.7703       463\n",
      "      xanthomonas     0.8519    0.3966    0.5412        58\n",
      "         yersinia     0.6250    0.4839    0.5455        93\n",
      "\n",
      "         accuracy                         0.6705      8116\n",
      "        macro avg     0.8414    0.5047    0.5969      8116\n",
      "     weighted avg     0.8438    0.6705    0.7003      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 50.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2353    0.3810        17\n",
      "    acinetobacter     1.0000    0.3776    0.5481        98\n",
      "        aeromonas     1.0000    0.3053    0.4677        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.5769    0.7317        26\n",
      "         bacillus     0.9583    0.6765    0.7931       204\n",
      "      bacteroides     1.0000    0.6452    0.7843        31\n",
      "    brevundimonas     1.0000    0.2000    0.3333        10\n",
      "     burkholderia     1.0000    0.1538    0.2667        39\n",
      "    campylobacter     0.9326    0.6694    0.7793       124\n",
      "      caulobacter     1.0000    0.6800    0.8095        25\n",
      "      citrobacter     1.0000    0.0588    0.1111        34\n",
      "      clostridium     0.5833    0.2917    0.3889        24\n",
      "      cronobacter     0.6000    0.0833    0.1463        36\n",
      "          dickeya     1.0000    0.5385    0.7000        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.6667    0.1408    0.2326       142\n",
      "     enterococcus     1.0000    0.6327    0.7750        49\n",
      "          erwinia     0.9649    0.5769    0.7221       286\n",
      "      escherichia     0.8912    0.5251    0.6608      1295\n",
      "   flavobacterium     1.0000    0.8958    0.9451        48\n",
      "         gordonia     1.0000    0.5283    0.6914        53\n",
      "       halorubrum     0.3571    0.3846    0.3704        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.9552    0.5365    0.6871       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.6000    0.1304    0.2143        23\n",
      "      lactococcus     1.0000    0.7500    0.8571       100\n",
      "      leuconostoc     1.0000    0.6000    0.7500        10\n",
      "         listeria     0.9091    0.7143    0.8000        28\n",
      "   microbacterium     1.0000    0.6061    0.7547        33\n",
      "    mycobacterium     0.9868    0.9073    0.9454       248\n",
      "           others     0.2566    0.9594    0.4050       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9459    0.4930    0.6481        71\n",
      "     pelagibacter     1.0000    0.2727    0.4286        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     0.9200    0.4423    0.5974        52\n",
      "          proteus     1.0000    0.3250    0.4906        40\n",
      "      providencia     0.5000    0.1250    0.2000        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9950    0.7739    0.8706       513\n",
      "        ralstonia     1.0000    0.3571    0.5263        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     1.0000    0.5811    0.7350        74\n",
      "      rhodobacter     0.9231    0.7059    0.8000        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9415    0.6108    0.7409       632\n",
      "         serratia     0.8958    0.4433    0.5931        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5806    0.1259    0.2069       143\n",
      "   staphylococcus     0.9955    0.9016    0.9462       244\n",
      " stenotrophomonas     0.8750    0.1707    0.2857        41\n",
      "    streptococcus     0.9804    0.6173    0.7576        81\n",
      "     streptomyces     1.0000    0.4651    0.6349        43\n",
      "    synechococcus     0.9860    0.8138    0.8917       521\n",
      "           vibrio     1.0000    0.5810    0.7350       463\n",
      "      xanthomonas     0.9474    0.3103    0.4675        58\n",
      "         yersinia     0.6286    0.4731    0.5399        93\n",
      "\n",
      "         accuracy                         0.6338      8116\n",
      "        macro avg     0.8437    0.4635    0.5584      8116\n",
      "     weighted avg     0.8467    0.6338    0.6678      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 60.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.1176    0.2105        17\n",
      "    acinetobacter     1.0000    0.3163    0.4806        98\n",
      "        aeromonas     1.0000    0.2316    0.3761        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.5000    0.6667        26\n",
      "         bacillus     0.9638    0.6520    0.7778       204\n",
      "      bacteroides     1.0000    0.5806    0.7347        31\n",
      "    brevundimonas     1.0000    0.2000    0.3333        10\n",
      "     burkholderia     1.0000    0.1282    0.2273        39\n",
      "    campylobacter     0.9610    0.5968    0.7363       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     1.0000    0.0588    0.1111        34\n",
      "      clostridium     0.6667    0.1667    0.2667        24\n",
      "      cronobacter     1.0000    0.0556    0.1053        36\n",
      "          dickeya     1.0000    0.4808    0.6494        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.7619    0.1127    0.1963       142\n",
      "     enterococcus     1.0000    0.5102    0.6757        49\n",
      "          erwinia     0.9673    0.5175    0.6743       286\n",
      "      escherichia     0.9094    0.4494    0.6016      1295\n",
      "   flavobacterium     1.0000    0.8542    0.9213        48\n",
      "         gordonia     1.0000    0.4528    0.6234        53\n",
      "       halorubrum     0.3077    0.3077    0.3077        13\n",
      "     helicobacter     1.0000    0.8750    0.9333        16\n",
      "       klebsiella     0.9594    0.4761    0.6364       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.5000    0.0435    0.0800        23\n",
      "      lactococcus     1.0000    0.7000    0.8235       100\n",
      "      leuconostoc     1.0000    0.4000    0.5714        10\n",
      "         listeria     0.9000    0.6429    0.7500        28\n",
      "   microbacterium     1.0000    0.5152    0.6800        33\n",
      "    mycobacterium     0.9866    0.8911    0.9364       248\n",
      "           others     0.2325    0.9706    0.3751       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9706    0.4648    0.6286        71\n",
      "     pelagibacter     1.0000    0.2424    0.3902        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     0.9474    0.3462    0.5070        52\n",
      "          proteus     1.0000    0.2000    0.3333        40\n",
      "      providencia     0.6667    0.1250    0.2105        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9948    0.7407    0.8492       513\n",
      "        ralstonia     1.0000    0.3571    0.5263        42\n",
      "     rheinheimera     1.0000    0.9886    0.9943       176\n",
      "        rhizobium     1.0000    0.5541    0.7130        74\n",
      "      rhodobacter     0.9583    0.6765    0.7931        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9444    0.5649    0.7069       632\n",
      "         serratia     0.8611    0.3196    0.4662        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.6400    0.1119    0.1905       143\n",
      "   staphylococcus     1.0000    0.8770    0.9345       244\n",
      " stenotrophomonas     1.0000    0.1707    0.2917        41\n",
      "    streptococcus     1.0000    0.5802    0.7344        81\n",
      "     streptomyces     1.0000    0.4651    0.6349        43\n",
      "    synechococcus     0.9927    0.7793    0.8731       521\n",
      "           vibrio     1.0000    0.5356    0.6976       463\n",
      "      xanthomonas     0.9474    0.3103    0.4675        58\n",
      "         yersinia     0.7727    0.3656    0.4964        93\n",
      "\n",
      "         accuracy                         0.5915      8116\n",
      "        macro avg     0.8443    0.4153    0.5136      8116\n",
      "     weighted avg     0.8543    0.5915    0.6295      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 70.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.0588    0.1111        17\n",
      "    acinetobacter     1.0000    0.2143    0.3529        98\n",
      "        aeromonas     1.0000    0.1579    0.2727        95\n",
      "    agrobacterium     1.0000    0.2500    0.4000        12\n",
      "     arthrobacter     1.0000    0.4615    0.6316        26\n",
      "         bacillus     0.9754    0.5833    0.7301       204\n",
      "      bacteroides     1.0000    0.4839    0.6522        31\n",
      "    brevundimonas     1.0000    0.1000    0.1818        10\n",
      "     burkholderia     1.0000    0.1026    0.1860        39\n",
      "    campylobacter     0.9688    0.5000    0.6596       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.6667    0.1667    0.2667        24\n",
      "      cronobacter     1.0000    0.0278    0.0541        36\n",
      "          dickeya     1.0000    0.4423    0.6133        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.7000    0.0493    0.0921       142\n",
      "     enterococcus     1.0000    0.4694    0.6389        49\n",
      "          erwinia     0.9758    0.4231    0.5902       286\n",
      "      escherichia     0.9157    0.3776    0.5347      1295\n",
      "   flavobacterium     1.0000    0.8542    0.9213        48\n",
      "         gordonia     1.0000    0.3774    0.5479        53\n",
      "       halorubrum     0.3077    0.3077    0.3077        13\n",
      "     helicobacter     1.0000    0.7500    0.8571        16\n",
      "       klebsiella     0.9543    0.4207    0.5839       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.6300    0.7730       100\n",
      "      leuconostoc     1.0000    0.1000    0.1818        10\n",
      "         listeria     0.9333    0.5000    0.6512        28\n",
      "   microbacterium     1.0000    0.4242    0.5957        33\n",
      "    mycobacterium     0.9864    0.8790    0.9296       248\n",
      "           others     0.2094    0.9787    0.3450       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9677    0.4225    0.5882        71\n",
      "     pelagibacter     1.0000    0.0909    0.1667        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     0.9333    0.2692    0.4179        52\n",
      "          proteus     1.0000    0.1500    0.2609        40\n",
      "      providencia     1.0000    0.0625    0.1176        16\n",
      "pseudoalteromonas     1.0000    0.0588    0.1111        17\n",
      "      pseudomonas     0.9971    0.6784    0.8074       513\n",
      "        ralstonia     1.0000    0.2619    0.4151        42\n",
      "     rheinheimera     1.0000    0.9773    0.9885       176\n",
      "        rhizobium     1.0000    0.4189    0.5905        74\n",
      "      rhodobacter     1.0000    0.6471    0.7857        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9558    0.4794    0.6386       632\n",
      "         serratia     0.8148    0.2268    0.3548        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.6364    0.0979    0.1697       143\n",
      "   staphylococcus     1.0000    0.8443    0.9156       244\n",
      " stenotrophomonas     1.0000    0.1220    0.2174        41\n",
      "    streptococcus     1.0000    0.5062    0.6721        81\n",
      "     streptomyces     1.0000    0.4419    0.6129        43\n",
      "    synechococcus     0.9949    0.7486    0.8543       521\n",
      "           vibrio     1.0000    0.4492    0.6200       463\n",
      "      xanthomonas     0.9333    0.2414    0.3836        58\n",
      "         yersinia     0.9655    0.3011    0.4590        93\n",
      "\n",
      "         accuracy                         0.5372      8116\n",
      "        macro avg     0.8270    0.3569    0.4491      8116\n",
      "     weighted avg     0.8498    0.5372    0.5751      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 80.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.1633    0.2807        98\n",
      "        aeromonas     1.0000    0.1158    0.2075        95\n",
      "    agrobacterium     1.0000    0.2500    0.4000        12\n",
      "     arthrobacter     1.0000    0.2692    0.4242        26\n",
      "         bacillus     0.9802    0.4853    0.6492       204\n",
      "      bacteroides     1.0000    0.2903    0.4500        31\n",
      "    brevundimonas     1.0000    0.1000    0.1818        10\n",
      "     burkholderia     1.0000    0.0769    0.1429        39\n",
      "    campylobacter     1.0000    0.3710    0.5412       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.6667    0.0833    0.1481        24\n",
      "      cronobacter     1.0000    0.0278    0.0541        36\n",
      "          dickeya     1.0000    0.4423    0.6133        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.5714    0.0282    0.0537       142\n",
      "     enterococcus     1.0000    0.4490    0.6197        49\n",
      "          erwinia     0.9888    0.3077    0.4693       286\n",
      "      escherichia     0.9337    0.2934    0.4465      1295\n",
      "   flavobacterium     1.0000    0.8333    0.9091        48\n",
      "         gordonia     1.0000    0.2453    0.3939        53\n",
      "       halorubrum     0.2500    0.2308    0.2400        13\n",
      "     helicobacter     1.0000    0.6875    0.8148        16\n",
      "       klebsiella     0.9754    0.2997    0.4586       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.6100    0.7578       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.3929    0.5641        28\n",
      "   microbacterium     1.0000    0.3333    0.5000        33\n",
      "    mycobacterium     0.9861    0.8589    0.9181       248\n",
      "           others     0.1914    0.9838    0.3205       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.3662    0.5361        71\n",
      "     pelagibacter     1.0000    0.0303    0.0588        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     1.0000    0.2500    0.4000        52\n",
      "          proteus     1.0000    0.1500    0.2609        40\n",
      "      providencia     0.0000    0.0000    0.0000        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     0.9969    0.6316    0.7733       513\n",
      "        ralstonia     1.0000    0.2381    0.3846        42\n",
      "     rheinheimera     1.0000    0.9659    0.9827       176\n",
      "        rhizobium     1.0000    0.3108    0.4742        74\n",
      "      rhodobacter     1.0000    0.5882    0.7407        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9576    0.4288    0.5923       632\n",
      "         serratia     0.8571    0.1856    0.3051        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5789    0.0769    0.1358       143\n",
      "   staphylococcus     1.0000    0.8074    0.8934       244\n",
      " stenotrophomonas     1.0000    0.0732    0.1364        41\n",
      "    streptococcus     1.0000    0.4321    0.6034        81\n",
      "     streptomyces     1.0000    0.3721    0.5424        43\n",
      "    synechococcus     0.9947    0.7255    0.8391       521\n",
      "           vibrio     1.0000    0.3996    0.5710       463\n",
      "      xanthomonas     0.9231    0.2069    0.3380        58\n",
      "         yersinia     0.9615    0.2688    0.4202        93\n",
      "\n",
      "         accuracy                         0.4857      8116\n",
      "        macro avg     0.7596    0.3086    0.3939      8116\n",
      "     weighted avg     0.8432    0.4857    0.5198      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 90.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.0816    0.1509        98\n",
      "        aeromonas     1.0000    0.0316    0.0612        95\n",
      "    agrobacterium     0.0000    0.0000    0.0000        12\n",
      "     arthrobacter     1.0000    0.1154    0.2069        26\n",
      "         bacillus     1.0000    0.3137    0.4776       204\n",
      "      bacteroides     1.0000    0.0323    0.0625        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     1.0000    0.0513    0.0976        39\n",
      "    campylobacter     1.0000    0.1774    0.3014       124\n",
      "      caulobacter     1.0000    0.3200    0.4848        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.6667    0.0833    0.1481        24\n",
      "      cronobacter     0.0000    0.0000    0.0000        36\n",
      "          dickeya     1.0000    0.3654    0.5352        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.6000    0.0211    0.0408       142\n",
      "     enterococcus     1.0000    0.4490    0.6197        49\n",
      "          erwinia     1.0000    0.2343    0.3796       286\n",
      "      escherichia     0.9533    0.1892    0.3157      1295\n",
      "   flavobacterium     1.0000    0.7083    0.8293        48\n",
      "         gordonia     1.0000    0.1509    0.2623        53\n",
      "       halorubrum     0.2000    0.0769    0.1111        13\n",
      "     helicobacter     1.0000    0.1875    0.3158        16\n",
      "       klebsiella     0.9877    0.2015    0.3347       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.5700    0.7261       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.2500    0.4000        28\n",
      "   microbacterium     1.0000    0.1515    0.2632        33\n",
      "    mycobacterium     0.9858    0.8427    0.9087       248\n",
      "           others     0.1707    0.9909    0.2913       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.3099    0.4731        71\n",
      "     pelagibacter     1.0000    0.0303    0.0588        33\n",
      "       prevotella     1.0000    0.6667    0.8000        18\n",
      "  prochlorococcus     1.0000    0.2308    0.3750        52\n",
      "          proteus     1.0000    0.1000    0.1818        40\n",
      "      providencia     0.0000    0.0000    0.0000        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     1.0000    0.5224    0.6863       513\n",
      "        ralstonia     1.0000    0.0714    0.1333        42\n",
      "     rheinheimera     1.0000    0.9659    0.9827       176\n",
      "        rhizobium     1.0000    0.2162    0.3556        74\n",
      "      rhodobacter     1.0000    0.4118    0.5833        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9640    0.3386    0.5012       632\n",
      "         serratia     0.8889    0.0825    0.1509        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.3750    0.0210    0.0397       143\n",
      "   staphylococcus     1.0000    0.7090    0.8297       244\n",
      " stenotrophomonas     1.0000    0.0488    0.0930        41\n",
      "    streptococcus     1.0000    0.3210    0.4860        81\n",
      "     streptomyces     1.0000    0.2093    0.3462        43\n",
      "    synechococcus     0.9944    0.6795    0.8073       521\n",
      "           vibrio     1.0000    0.3132    0.4770       463\n",
      "      xanthomonas     0.8000    0.0690    0.1270        58\n",
      "         yersinia     0.9524    0.2151    0.3509        93\n",
      "\n",
      "         accuracy                         0.4103      8116\n",
      "        macro avg     0.7040    0.2225    0.2909      8116\n",
      "     weighted avg     0.8351    0.4103    0.4305      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 100.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.0102    0.0202        98\n",
      "        aeromonas     0.0000    0.0000    0.0000        95\n",
      "    agrobacterium     0.0000    0.0000    0.0000        12\n",
      "     arthrobacter     0.0000    0.0000    0.0000        26\n",
      "         bacillus     1.0000    0.0490    0.0935       204\n",
      "      bacteroides     0.0000    0.0000    0.0000        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     0.0000    0.0000    0.0000        39\n",
      "    campylobacter     1.0000    0.0484    0.0923       124\n",
      "      caulobacter     1.0000    0.0400    0.0769        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.0000    0.0000    0.0000        24\n",
      "      cronobacter     0.0000    0.0000    0.0000        36\n",
      "          dickeya     1.0000    0.2885    0.4478        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.0000    0.0000    0.0000       142\n",
      "     enterococcus     1.0000    0.0612    0.1154        49\n",
      "          erwinia     1.0000    0.0385    0.0741       286\n",
      "      escherichia     0.9457    0.0672    0.1255      1295\n",
      "   flavobacterium     1.0000    0.6250    0.7692        48\n",
      "         gordonia     0.0000    0.0000    0.0000        53\n",
      "       halorubrum     0.0000    0.0000    0.0000        13\n",
      "     helicobacter     0.0000    0.0000    0.0000        16\n",
      "       klebsiella     1.0000    0.0403    0.0775       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.2300    0.3740       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.0357    0.0690        28\n",
      "   microbacterium     0.0000    0.0000    0.0000        33\n",
      "    mycobacterium     0.9820    0.6613    0.7904       248\n",
      "           others     0.1426    0.9959    0.2496       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.1127    0.2025        71\n",
      "     pelagibacter     0.0000    0.0000    0.0000        33\n",
      "       prevotella     1.0000    0.0556    0.1053        18\n",
      "  prochlorococcus     1.0000    0.2115    0.3492        52\n",
      "          proteus     1.0000    0.0500    0.0952        40\n",
      "      providencia     0.0000    0.0000    0.0000        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     1.0000    0.3002    0.4618       513\n",
      "        ralstonia     0.0000    0.0000    0.0000        42\n",
      "     rheinheimera     1.0000    0.9318    0.9647       176\n",
      "        rhizobium     0.0000    0.0000    0.0000        74\n",
      "      rhodobacter     1.0000    0.1471    0.2564        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     1.0000    0.1282    0.2272       632\n",
      "         serratia     1.0000    0.0206    0.0404        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     1.0000    0.0070    0.0139       143\n",
      "   staphylococcus     1.0000    0.2951    0.4557       244\n",
      " stenotrophomonas     0.0000    0.0000    0.0000        41\n",
      "    streptococcus     1.0000    0.0247    0.0482        81\n",
      "     streptomyces     1.0000    0.0698    0.1304        43\n",
      "    synechococcus     0.9963    0.5125    0.6768       521\n",
      "           vibrio     1.0000    0.1490    0.2594       463\n",
      "      xanthomonas     0.5000    0.0172    0.0333        58\n",
      "         yersinia     1.0000    0.1183    0.2115        93\n",
      "\n",
      "         accuracy                         0.2716      8116\n",
      "        macro avg     0.5011    0.1075    0.1340      8116\n",
      "     weighted avg     0.7723    0.2716    0.2482      8116\n",
      "\n",
      "===================\n",
      "Finished\n",
      "==============\n",
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Confidence threshold k: 0.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.7500    0.3529    0.4800        17\n",
      "    acinetobacter     0.9342    0.7245    0.8161        98\n",
      "        aeromonas     0.8571    0.6316    0.7273        95\n",
      "    agrobacterium     0.8333    0.4167    0.5556        12\n",
      "     arthrobacter     0.7037    0.7308    0.7170        26\n",
      "         bacillus     0.7729    0.8676    0.8176       204\n",
      "      bacteroides     0.9032    0.9032    0.9032        31\n",
      "    brevundimonas     1.0000    0.5000    0.6667        10\n",
      "     burkholderia     0.6667    0.4615    0.5455        39\n",
      "    campylobacter     0.7969    0.8226    0.8095       124\n",
      "      caulobacter     0.9130    0.8400    0.8750        25\n",
      "      citrobacter     0.3750    0.3529    0.3636        34\n",
      "      clostridium     0.5161    0.6667    0.5818        24\n",
      "      cronobacter     0.5909    0.3611    0.4483        36\n",
      "          dickeya     0.9070    0.7500    0.8211        52\n",
      "     edwardsiella     1.0000    0.4000    0.5714        10\n",
      "     enterobacter     0.4240    0.3732    0.3970       142\n",
      "     enterococcus     0.8000    0.8980    0.8462        49\n",
      "          erwinia     0.6066    0.8951    0.7232       286\n",
      "      escherichia     0.5611    0.8232    0.6673      1295\n",
      "   flavobacterium     0.9783    0.9375    0.9574        48\n",
      "         gordonia     0.8036    0.8491    0.8257        53\n",
      "       halorubrum     0.4545    0.7692    0.5714        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.6870    0.8237    0.7491       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.8824    0.6522    0.7500        23\n",
      "      lactococcus     1.0000    0.8500    0.9189       100\n",
      "      leuconostoc     1.0000    0.9000    0.9474        10\n",
      "         listeria     0.8889    0.8571    0.8727        28\n",
      "   microbacterium     0.9615    0.7576    0.8475        33\n",
      "    mycobacterium     0.9312    0.9274    0.9293       248\n",
      "           others     0.0000    0.0000    0.0000       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.8167    0.6901    0.7481        71\n",
      "     pelagibacter     0.7097    0.6667    0.6875        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9429    0.6346    0.7586        52\n",
      "          proteus     0.8462    0.5500    0.6667        40\n",
      "      providencia     0.6250    0.3125    0.4167        16\n",
      "pseudoalteromonas     0.6667    0.1176    0.2000        17\n",
      "      pseudomonas     0.8035    0.8928    0.8458       513\n",
      "        ralstonia     0.7308    0.4524    0.5588        42\n",
      "     rheinheimera     1.0000    1.0000    1.0000       176\n",
      "        rhizobium     0.8462    0.8919    0.8684        74\n",
      "      rhodobacter     0.8529    0.8529    0.8529        34\n",
      "      rhodococcus     0.8000    0.3077    0.4444        13\n",
      "       salmonella     0.8009    0.8085    0.8047       632\n",
      "         serratia     0.7375    0.6082    0.6667        97\n",
      "       shewanella     1.0000    0.4545    0.6250        11\n",
      "         shigella     0.4487    0.2448    0.3167       143\n",
      "   staphylococcus     0.8520    0.9672    0.9060       244\n",
      " stenotrophomonas     0.9130    0.5122    0.6562        41\n",
      "    streptococcus     0.8095    0.8395    0.8242        81\n",
      "     streptomyces     0.9429    0.7674    0.8462        43\n",
      "    synechococcus     0.7216    0.9750    0.8294       521\n",
      "           vibrio     0.6129    0.8855    0.7244       463\n",
      "      xanthomonas     0.6739    0.5345    0.5962        58\n",
      "         yersinia     0.5310    0.6452    0.5825        93\n",
      "\n",
      "         accuracy                         0.7054      8116\n",
      "        macro avg     0.7489    0.6551    0.6797      8116\n",
      "     weighted avg     0.6332    0.7054    0.6572      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 10.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.3529    0.5217        17\n",
      "    acinetobacter     1.0000    0.6224    0.7673        98\n",
      "        aeromonas     0.9592    0.4947    0.6528        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.6923    0.8182        26\n",
      "         bacillus     0.8895    0.8284    0.8579       204\n",
      "      bacteroides     1.0000    0.8710    0.9310        31\n",
      "    brevundimonas     1.0000    0.5000    0.6667        10\n",
      "     burkholderia     1.0000    0.4103    0.5818        39\n",
      "    campylobacter     0.9259    0.8065    0.8621       124\n",
      "      caulobacter     1.0000    0.8400    0.9130        25\n",
      "      citrobacter     0.5263    0.2941    0.3774        34\n",
      "      clostridium     0.5600    0.5833    0.5714        24\n",
      "      cronobacter     0.6471    0.3056    0.4151        36\n",
      "          dickeya     0.9286    0.7500    0.8298        52\n",
      "     edwardsiella     1.0000    0.3000    0.4615        10\n",
      "     enterobacter     0.4737    0.3169    0.3797       142\n",
      "     enterococcus     0.9545    0.8571    0.9032        49\n",
      "          erwinia     0.8897    0.8182    0.8525       286\n",
      "      escherichia     0.7489    0.7714    0.7600      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     0.9750    0.7358    0.8387        53\n",
      "       halorubrum     0.3889    0.5385    0.4516        13\n",
      "     helicobacter     1.0000    0.9375    0.9677        16\n",
      "       klebsiella     0.8735    0.7305    0.7956       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.8667    0.5652    0.6842        23\n",
      "      lactococcus     1.0000    0.8200    0.9011       100\n",
      "      leuconostoc     1.0000    0.9000    0.9474        10\n",
      "         listeria     0.9231    0.8571    0.8889        28\n",
      "   microbacterium     1.0000    0.6970    0.8214        33\n",
      "    mycobacterium     0.9828    0.9194    0.9500       248\n",
      "           others     0.3935    0.7586    0.5182       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9020    0.6479    0.7541        71\n",
      "     pelagibacter     0.7273    0.4848    0.5818        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9375    0.5769    0.7143        52\n",
      "          proteus     0.8696    0.5000    0.6349        40\n",
      "      providencia     0.6667    0.2500    0.3636        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9610    0.8635    0.9097       513\n",
      "        ralstonia     0.9474    0.4286    0.5902        42\n",
      "     rheinheimera     1.0000    1.0000    1.0000       176\n",
      "        rhizobium     0.9104    0.8243    0.8652        74\n",
      "      rhodobacter     0.9310    0.7941    0.8571        34\n",
      "      rhodococcus     1.0000    0.0769    0.1429        13\n",
      "       salmonella     0.9021    0.7437    0.8153       632\n",
      "         serratia     0.8281    0.5464    0.6584        97\n",
      "       shewanella     1.0000    0.3636    0.5333        11\n",
      "         shigella     0.5161    0.2238    0.3122       143\n",
      "   staphylococcus     0.9514    0.9631    0.9572       244\n",
      " stenotrophomonas     0.9412    0.3902    0.5517        41\n",
      "    streptococcus     0.8873    0.7778    0.8289        81\n",
      "     streptomyces     1.0000    0.7442    0.8533        43\n",
      "    synechococcus     0.8748    0.9520    0.9118       521\n",
      "           vibrio     0.8828    0.7646    0.8194       463\n",
      "      xanthomonas     0.7879    0.4483    0.5714        58\n",
      "         yersinia     0.5481    0.6129    0.5787        93\n",
      "\n",
      "         accuracy                         0.7507      8116\n",
      "        macro avg     0.8454    0.6133    0.6838      8116\n",
      "     weighted avg     0.8015    0.7507    0.7577      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 20.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2941    0.4545        17\n",
      "    acinetobacter     1.0000    0.5204    0.6846        98\n",
      "        aeromonas     0.9565    0.4632    0.6241        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.6538    0.7907        26\n",
      "         bacillus     0.9415    0.7892    0.8587       204\n",
      "      bacteroides     1.0000    0.8710    0.9310        31\n",
      "    brevundimonas     1.0000    0.4000    0.5714        10\n",
      "     burkholderia     1.0000    0.3590    0.5283        39\n",
      "    campylobacter     0.9231    0.7742    0.8421       124\n",
      "      caulobacter     1.0000    0.8000    0.8889        25\n",
      "      citrobacter     0.5455    0.1765    0.2667        34\n",
      "      clostridium     0.5417    0.5417    0.5417        24\n",
      "      cronobacter     0.6364    0.1944    0.2979        36\n",
      "          dickeya     0.9429    0.6346    0.7586        52\n",
      "     edwardsiella     1.0000    0.2000    0.3333        10\n",
      "     enterobacter     0.5753    0.2958    0.3907       142\n",
      "     enterococcus     1.0000    0.8367    0.9111        49\n",
      "          erwinia     0.9353    0.7587    0.8378       286\n",
      "      escherichia     0.8150    0.7112    0.7596      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.6981    0.8222        53\n",
      "       halorubrum     0.3889    0.5385    0.4516        13\n",
      "     helicobacter     1.0000    0.8750    0.9333        16\n",
      "       klebsiella     0.9203    0.6398    0.7548       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.7778    0.3043    0.4375        23\n",
      "      lactococcus     1.0000    0.8100    0.8950       100\n",
      "      leuconostoc     1.0000    0.8000    0.8889        10\n",
      "         listeria     0.9231    0.8571    0.8889        28\n",
      "   microbacterium     1.0000    0.6667    0.8000        33\n",
      "    mycobacterium     0.9870    0.9153    0.9498       248\n",
      "           others     0.3421    0.8955    0.4951       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9048    0.5352    0.6726        71\n",
      "     pelagibacter     0.8667    0.3939    0.5417        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9310    0.5192    0.6667        52\n",
      "          proteus     0.8636    0.4750    0.6129        40\n",
      "      providencia     0.5000    0.1250    0.2000        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9863    0.8441    0.9097       513\n",
      "        ralstonia     0.9444    0.4048    0.5667        42\n",
      "     rheinheimera     1.0000    1.0000    1.0000       176\n",
      "        rhizobium     0.9344    0.7703    0.8444        74\n",
      "      rhodobacter     0.9286    0.7647    0.8387        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9172    0.7009    0.7946       632\n",
      "         serratia     0.8525    0.5361    0.6582        97\n",
      "       shewanella     1.0000    0.1818    0.3077        11\n",
      "         shigella     0.5577    0.2028    0.2974       143\n",
      "   staphylococcus     0.9873    0.9590    0.9730       244\n",
      " stenotrophomonas     0.9286    0.3171    0.4727        41\n",
      "    streptococcus     0.9385    0.7531    0.8356        81\n",
      "     streptomyces     1.0000    0.6047    0.7536        43\n",
      "    synechococcus     0.9432    0.8925    0.9172       521\n",
      "           vibrio     0.9642    0.6976    0.8095       463\n",
      "      xanthomonas     0.8387    0.4483    0.5843        58\n",
      "         yersinia     0.5579    0.5699    0.5638        93\n",
      "\n",
      "         accuracy                         0.7255      8116\n",
      "        macro avg     0.8389    0.5645    0.6451      8116\n",
      "     weighted avg     0.8267    0.7255    0.7424      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 30.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2941    0.4545        17\n",
      "    acinetobacter     1.0000    0.5102    0.6757        98\n",
      "        aeromonas     0.9730    0.3789    0.5455        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.6538    0.7907        26\n",
      "         bacillus     0.9563    0.7500    0.8407       204\n",
      "      bacteroides     1.0000    0.8387    0.9123        31\n",
      "    brevundimonas     1.0000    0.4000    0.5714        10\n",
      "     burkholderia     1.0000    0.2821    0.4400        39\n",
      "    campylobacter     0.9208    0.7500    0.8267       124\n",
      "      caulobacter     1.0000    0.7200    0.8372        25\n",
      "      citrobacter     0.5714    0.1176    0.1951        34\n",
      "      clostridium     0.7222    0.5417    0.6190        24\n",
      "      cronobacter     0.7778    0.1944    0.3111        36\n",
      "          dickeya     0.9697    0.6154    0.7529        52\n",
      "     edwardsiella     1.0000    0.2000    0.3333        10\n",
      "     enterobacter     0.6316    0.2535    0.3618       142\n",
      "     enterococcus     1.0000    0.7755    0.8736        49\n",
      "          erwinia     0.9469    0.6853    0.7951       286\n",
      "      escherichia     0.8574    0.6548    0.7426      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.6415    0.7816        53\n",
      "       halorubrum     0.3889    0.5385    0.4516        13\n",
      "     helicobacter     1.0000    0.8750    0.9333        16\n",
      "       klebsiella     0.9486    0.6045    0.7385       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.6667    0.1739    0.2759        23\n",
      "      lactococcus     1.0000    0.7800    0.8764       100\n",
      "      leuconostoc     1.0000    0.6000    0.7500        10\n",
      "         listeria     0.9231    0.8571    0.8889        28\n",
      "   microbacterium     1.0000    0.6364    0.7778        33\n",
      "    mycobacterium     0.9869    0.9113    0.9476       248\n",
      "           others     0.3065    0.9391    0.4622       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.8974    0.4930    0.6364        71\n",
      "     pelagibacter     1.0000    0.3333    0.5000        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9310    0.5192    0.6667        52\n",
      "          proteus     0.9444    0.4250    0.5862        40\n",
      "      providencia     0.5000    0.1250    0.2000        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9930    0.8265    0.9021       513\n",
      "        ralstonia     0.9412    0.3810    0.5424        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     0.9811    0.7027    0.8189        74\n",
      "      rhodobacter     0.9259    0.7353    0.8197        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9294    0.6661    0.7760       632\n",
      "         serratia     0.8571    0.4948    0.6275        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5435    0.1748    0.2646       143\n",
      "   staphylococcus     0.9871    0.9426    0.9644       244\n",
      " stenotrophomonas     0.8889    0.1951    0.3200        41\n",
      "    streptococcus     0.9836    0.7407    0.8451        81\n",
      "     streptomyces     1.0000    0.5349    0.6970        43\n",
      "    synechococcus     0.9758    0.8522    0.9098       521\n",
      "           vibrio     0.9901    0.6479    0.7833       463\n",
      "      xanthomonas     0.8276    0.4138    0.5517        58\n",
      "         yersinia     0.6000    0.5484    0.5730        93\n",
      "\n",
      "         accuracy                         0.6968      8116\n",
      "        macro avg     0.8347    0.5297    0.6169      8116\n",
      "     weighted avg     0.8389    0.6968    0.7212      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 40.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2941    0.4545        17\n",
      "    acinetobacter     1.0000    0.4694    0.6389        98\n",
      "        aeromonas     0.9677    0.3158    0.4762        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.5000    0.6667        26\n",
      "         bacillus     0.9586    0.6814    0.7966       204\n",
      "      bacteroides     1.0000    0.7097    0.8302        31\n",
      "    brevundimonas     1.0000    0.3000    0.4615        10\n",
      "     burkholderia     1.0000    0.2564    0.4082        39\n",
      "    campylobacter     0.9175    0.7177    0.8054       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     0.7500    0.0882    0.1579        34\n",
      "      clostridium     0.7059    0.5000    0.5854        24\n",
      "      cronobacter     0.6667    0.1111    0.1905        36\n",
      "          dickeya     1.0000    0.5962    0.7470        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.6512    0.1972    0.3027       142\n",
      "     enterococcus     1.0000    0.6327    0.7750        49\n",
      "          erwinia     0.9731    0.6329    0.7669       286\n",
      "      escherichia     0.8830    0.6000    0.7145      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.6038    0.7529        53\n",
      "       halorubrum     0.4000    0.4615    0.4286        13\n",
      "     helicobacter     1.0000    0.8750    0.9333        16\n",
      "       klebsiella     0.9496    0.5693    0.7118       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.8000    0.1739    0.2857        23\n",
      "      lactococcus     1.0000    0.7600    0.8636       100\n",
      "      leuconostoc     1.0000    0.5000    0.6667        10\n",
      "         listeria     0.9200    0.8214    0.8679        28\n",
      "   microbacterium     1.0000    0.6364    0.7778        33\n",
      "    mycobacterium     0.9868    0.9073    0.9454       248\n",
      "           others     0.2784    0.9533    0.4309       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9211    0.4930    0.6422        71\n",
      "     pelagibacter     1.0000    0.2727    0.4286        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9286    0.5000    0.6500        52\n",
      "          proteus     0.9375    0.3750    0.5357        40\n",
      "      providencia     0.5000    0.1250    0.2000        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9952    0.8012    0.8877       513\n",
      "        ralstonia     1.0000    0.3571    0.5263        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     1.0000    0.6216    0.7667        74\n",
      "      rhodobacter     0.9259    0.7353    0.8197        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9356    0.6440    0.7629       632\n",
      "         serratia     0.8800    0.4536    0.5986        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5128    0.1399    0.2198       143\n",
      "   staphylococcus     0.9957    0.9385    0.9662       244\n",
      " stenotrophomonas     0.8889    0.1951    0.3200        41\n",
      "    streptococcus     0.9818    0.6667    0.7941        81\n",
      "     streptomyces     1.0000    0.5116    0.6769        43\n",
      "    synechococcus     0.9885    0.8215    0.8973       521\n",
      "           vibrio     1.0000    0.6048    0.7537       463\n",
      "      xanthomonas     0.8846    0.3966    0.5476        58\n",
      "         yersinia     0.6267    0.5054    0.5595        93\n",
      "\n",
      "         accuracy                         0.6656      8116\n",
      "        macro avg     0.8426    0.4931    0.5865      8116\n",
      "     weighted avg     0.8449    0.6656    0.6960      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 50.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.1765    0.3000        17\n",
      "    acinetobacter     1.0000    0.3980    0.5693        98\n",
      "        aeromonas     0.9615    0.2632    0.4132        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.4231    0.5946        26\n",
      "         bacillus     0.9574    0.6618    0.7826       204\n",
      "      bacteroides     1.0000    0.7097    0.8302        31\n",
      "    brevundimonas     1.0000    0.3000    0.4615        10\n",
      "     burkholderia     1.0000    0.1795    0.3043        39\n",
      "    campylobacter     0.9231    0.6774    0.7814       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     0.6667    0.0588    0.1081        34\n",
      "      clostridium     0.7143    0.4167    0.5263        24\n",
      "      cronobacter     0.6000    0.0833    0.1463        36\n",
      "          dickeya     1.0000    0.5962    0.7470        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.6129    0.1338    0.2197       142\n",
      "     enterococcus     1.0000    0.5714    0.7273        49\n",
      "          erwinia     0.9821    0.5769    0.7269       286\n",
      "      escherichia     0.9000    0.5351    0.6712      1295\n",
      "   flavobacterium     1.0000    0.8958    0.9451        48\n",
      "         gordonia     1.0000    0.4717    0.6410        53\n",
      "       halorubrum     0.3571    0.3846    0.3704        13\n",
      "     helicobacter     1.0000    0.8750    0.9333        16\n",
      "       klebsiella     0.9548    0.5315    0.6828       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.7500    0.1304    0.2222        23\n",
      "      lactococcus     1.0000    0.7200    0.8372       100\n",
      "      leuconostoc     1.0000    0.5000    0.6667        10\n",
      "         listeria     0.9000    0.6429    0.7500        28\n",
      "   microbacterium     1.0000    0.5152    0.6800        33\n",
      "    mycobacterium     0.9868    0.9032    0.9432       248\n",
      "           others     0.2549    0.9604    0.4029       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9444    0.4789    0.6355        71\n",
      "     pelagibacter     1.0000    0.1818    0.3077        33\n",
      "       prevotella     1.0000    0.8889    0.9412        18\n",
      "  prochlorococcus     0.9200    0.4423    0.5974        52\n",
      "          proteus     1.0000    0.3500    0.5185        40\n",
      "      providencia     0.5000    0.1250    0.2000        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9950    0.7797    0.8743       513\n",
      "        ralstonia     1.0000    0.3571    0.5263        42\n",
      "     rheinheimera     1.0000    0.9886    0.9943       176\n",
      "        rhizobium     1.0000    0.5946    0.7458        74\n",
      "      rhodobacter     0.9231    0.7059    0.8000        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9380    0.5981    0.7304       632\n",
      "         serratia     0.8936    0.4330    0.5833        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5938    0.1329    0.2171       143\n",
      "   staphylococcus     0.9956    0.9180    0.9552       244\n",
      " stenotrophomonas     0.8750    0.1707    0.2857        41\n",
      "    streptococcus     1.0000    0.6173    0.7634        81\n",
      "     streptomyces     1.0000    0.4419    0.6129        43\n",
      "    synechococcus     0.9929    0.8023    0.8875       521\n",
      "           vibrio     1.0000    0.5702    0.7263       463\n",
      "      xanthomonas     0.9048    0.3276    0.4810        58\n",
      "         yersinia     0.6338    0.4839    0.5488        93\n",
      "\n",
      "         accuracy                         0.6313      8116\n",
      "        macro avg     0.8412    0.4555    0.5524      8116\n",
      "     weighted avg     0.8466    0.6313    0.6660      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 60.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.1176    0.2105        17\n",
      "    acinetobacter     1.0000    0.3061    0.4687        98\n",
      "        aeromonas     1.0000    0.2211    0.3621        95\n",
      "    agrobacterium     1.0000    0.2500    0.4000        12\n",
      "     arthrobacter     1.0000    0.4231    0.5946        26\n",
      "         bacillus     0.9612    0.6078    0.7447       204\n",
      "      bacteroides     1.0000    0.5161    0.6809        31\n",
      "    brevundimonas     1.0000    0.3000    0.4615        10\n",
      "     burkholderia     1.0000    0.1282    0.2273        39\n",
      "    campylobacter     0.9494    0.6048    0.7389       124\n",
      "      caulobacter     1.0000    0.6000    0.7500        25\n",
      "      citrobacter     0.6667    0.0588    0.1081        34\n",
      "      clostridium     0.8000    0.3333    0.4706        24\n",
      "      cronobacter     1.0000    0.0556    0.1053        36\n",
      "          dickeya     1.0000    0.4808    0.6494        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.7500    0.1268    0.2169       142\n",
      "     enterococcus     1.0000    0.4898    0.6575        49\n",
      "          erwinia     0.9801    0.5175    0.6773       286\n",
      "      escherichia     0.9143    0.4533    0.6061      1295\n",
      "   flavobacterium     1.0000    0.8542    0.9213        48\n",
      "         gordonia     1.0000    0.2830    0.4412        53\n",
      "       halorubrum     0.3077    0.3077    0.3077        13\n",
      "     helicobacter     1.0000    0.8125    0.8966        16\n",
      "       klebsiella     0.9485    0.4635    0.6227       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.6667    0.0870    0.1538        23\n",
      "      lactococcus     1.0000    0.6900    0.8166       100\n",
      "      leuconostoc     1.0000    0.3000    0.4615        10\n",
      "         listeria     0.9375    0.5357    0.6818        28\n",
      "   microbacterium     1.0000    0.3939    0.5652        33\n",
      "    mycobacterium     0.9865    0.8831    0.9319       248\n",
      "           others     0.2286    0.9696    0.3700       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9706    0.4648    0.6286        71\n",
      "     pelagibacter     1.0000    0.1515    0.2632        33\n",
      "       prevotella     1.0000    0.8889    0.9412        18\n",
      "  prochlorococcus     0.9524    0.3846    0.5479        52\n",
      "          proteus     1.0000    0.2250    0.3673        40\n",
      "      providencia     0.5000    0.0625    0.1111        16\n",
      "pseudoalteromonas     1.0000    0.0588    0.1111        17\n",
      "      pseudomonas     0.9947    0.7329    0.8440       513\n",
      "        ralstonia     1.0000    0.3095    0.4727        42\n",
      "     rheinheimera     1.0000    0.9830    0.9914       176\n",
      "        rhizobium     1.0000    0.4730    0.6422        74\n",
      "      rhodobacter     0.9565    0.6471    0.7719        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9462    0.5570    0.7012       632\n",
      "         serratia     0.8684    0.3402    0.4889        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.6296    0.1189    0.2000       143\n",
      "   staphylococcus     0.9953    0.8689    0.9278       244\n",
      " stenotrophomonas     1.0000    0.1463    0.2553        41\n",
      "    streptococcus     1.0000    0.5556    0.7143        81\n",
      "     streptomyces     1.0000    0.3721    0.5424        43\n",
      "    synechococcus     0.9951    0.7831    0.8765       521\n",
      "           vibrio     1.0000    0.5097    0.6753       463\n",
      "      xanthomonas     0.8889    0.2759    0.4211        58\n",
      "         yersinia     0.7826    0.3871    0.5180        93\n",
      "\n",
      "         accuracy                         0.5833      8116\n",
      "        macro avg     0.8403    0.3977    0.4968      8116\n",
      "     weighted avg     0.8531    0.5833    0.6221      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 70.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.0588    0.1111        17\n",
      "    acinetobacter     1.0000    0.2143    0.3529        98\n",
      "        aeromonas     1.0000    0.1579    0.2727        95\n",
      "    agrobacterium     1.0000    0.2500    0.4000        12\n",
      "     arthrobacter     1.0000    0.3846    0.5556        26\n",
      "         bacillus     0.9735    0.5392    0.6940       204\n",
      "      bacteroides     1.0000    0.5161    0.6809        31\n",
      "    brevundimonas     1.0000    0.1000    0.1818        10\n",
      "     burkholderia     1.0000    0.1026    0.1860        39\n",
      "    campylobacter     0.9697    0.5161    0.6737       124\n",
      "      caulobacter     1.0000    0.6000    0.7500        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.6667    0.0833    0.1481        24\n",
      "      cronobacter     1.0000    0.0278    0.0541        36\n",
      "          dickeya     1.0000    0.4423    0.6133        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.7333    0.0775    0.1401       142\n",
      "     enterococcus     1.0000    0.4694    0.6389        49\n",
      "          erwinia     0.9756    0.4196    0.5868       286\n",
      "      escherichia     0.9114    0.3892    0.5455      1295\n",
      "   flavobacterium     1.0000    0.8542    0.9213        48\n",
      "         gordonia     1.0000    0.2642    0.4179        53\n",
      "       halorubrum     0.3333    0.1538    0.2105        13\n",
      "     helicobacter     1.0000    0.8125    0.8966        16\n",
      "       klebsiella     0.9571    0.3929    0.5571       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.6300    0.7730       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     0.9286    0.4643    0.6190        28\n",
      "   microbacterium     1.0000    0.3030    0.4651        33\n",
      "    mycobacterium     0.9862    0.8629    0.9204       248\n",
      "           others     0.2091    0.9838    0.3448       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9677    0.4225    0.5882        71\n",
      "     pelagibacter     1.0000    0.0303    0.0588        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     1.0000    0.2500    0.4000        52\n",
      "          proteus     1.0000    0.1750    0.2979        40\n",
      "      providencia     1.0000    0.0625    0.1176        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     0.9971    0.6706    0.8019       513\n",
      "        ralstonia     1.0000    0.2619    0.4151        42\n",
      "     rheinheimera     1.0000    0.9659    0.9827       176\n",
      "        rhizobium     1.0000    0.3243    0.4898        74\n",
      "      rhodobacter     1.0000    0.6471    0.7857        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9517    0.4984    0.6542       632\n",
      "         serratia     0.8519    0.2371    0.3710        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.6190    0.0909    0.1585       143\n",
      "   staphylococcus     1.0000    0.8361    0.9107       244\n",
      " stenotrophomonas     1.0000    0.0976    0.1778        41\n",
      "    streptococcus     1.0000    0.4691    0.6387        81\n",
      "     streptomyces     1.0000    0.3721    0.5424        43\n",
      "    synechococcus     0.9948    0.7390    0.8480       521\n",
      "           vibrio     1.0000    0.4557    0.6261       463\n",
      "      xanthomonas     0.8667    0.2241    0.3562        58\n",
      "         yersinia     0.9697    0.3441    0.5079        93\n",
      "\n",
      "         accuracy                         0.5338      8116\n",
      "        macro avg     0.7943    0.3403    0.4297      8116\n",
      "     weighted avg     0.8462    0.5338    0.5713      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 80.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.1531    0.2655        98\n",
      "        aeromonas     1.0000    0.1053    0.1905        95\n",
      "    agrobacterium     1.0000    0.2500    0.4000        12\n",
      "     arthrobacter     1.0000    0.2692    0.4242        26\n",
      "         bacillus     0.9789    0.4559    0.6221       204\n",
      "      bacteroides     1.0000    0.2903    0.4500        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     1.0000    0.0513    0.0976        39\n",
      "    campylobacter     1.0000    0.3871    0.5581       124\n",
      "      caulobacter     1.0000    0.5600    0.7179        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.6667    0.0833    0.1481        24\n",
      "      cronobacter     1.0000    0.0278    0.0541        36\n",
      "          dickeya     1.0000    0.4423    0.6133        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.5000    0.0211    0.0405       142\n",
      "     enterococcus     1.0000    0.4694    0.6389        49\n",
      "          erwinia     0.9767    0.2937    0.4516       286\n",
      "      escherichia     0.9287    0.3019    0.4557      1295\n",
      "   flavobacterium     1.0000    0.8333    0.9091        48\n",
      "         gordonia     1.0000    0.2264    0.3692        53\n",
      "       halorubrum     0.0000    0.0000    0.0000        13\n",
      "     helicobacter     1.0000    0.7500    0.8571        16\n",
      "       klebsiella     0.9756    0.3023    0.4615       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.5700    0.7261       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.3929    0.5641        28\n",
      "   microbacterium     1.0000    0.3030    0.4651        33\n",
      "    mycobacterium     0.9859    0.8468    0.9111       248\n",
      "           others     0.1902    0.9888    0.3191       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.3521    0.5208        71\n",
      "     pelagibacter     1.0000    0.0303    0.0588        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     1.0000    0.2500    0.4000        52\n",
      "          proteus     1.0000    0.1250    0.2222        40\n",
      "      providencia     1.0000    0.0625    0.1176        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     0.9968    0.6140    0.7600       513\n",
      "        ralstonia     1.0000    0.2381    0.3846        42\n",
      "     rheinheimera     1.0000    0.9659    0.9827       176\n",
      "        rhizobium     1.0000    0.2432    0.3913        74\n",
      "      rhodobacter     1.0000    0.5294    0.6923        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9539    0.4256    0.5886       632\n",
      "         serratia     0.8500    0.1753    0.2906        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.3636    0.0280    0.0519       143\n",
      "   staphylococcus     1.0000    0.8033    0.8909       244\n",
      " stenotrophomonas     1.0000    0.0732    0.1364        41\n",
      "    streptococcus     1.0000    0.3951    0.5664        81\n",
      "     streptomyces     1.0000    0.3023    0.4643        43\n",
      "    synechococcus     0.9947    0.7198    0.8352       521\n",
      "           vibrio     1.0000    0.3888    0.5599       463\n",
      "      xanthomonas     0.9091    0.1724    0.2899        58\n",
      "         yersinia     0.9600    0.2581    0.4068        93\n",
      "\n",
      "         accuracy                         0.4792      8116\n",
      "        macro avg     0.7497    0.2943    0.3768      8116\n",
      "     weighted avg     0.8366    0.4792    0.5118      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 90.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.0612    0.1154        98\n",
      "        aeromonas     1.0000    0.0211    0.0412        95\n",
      "    agrobacterium     0.0000    0.0000    0.0000        12\n",
      "     arthrobacter     1.0000    0.1154    0.2069        26\n",
      "         bacillus     0.9815    0.2598    0.4109       204\n",
      "      bacteroides     1.0000    0.2581    0.4103        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     1.0000    0.0513    0.0976        39\n",
      "    campylobacter     1.0000    0.2742    0.4304       124\n",
      "      caulobacter     1.0000    0.3200    0.4848        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.6667    0.0833    0.1481        24\n",
      "      cronobacter     0.0000    0.0000    0.0000        36\n",
      "          dickeya     1.0000    0.3654    0.5352        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.6000    0.0211    0.0408       142\n",
      "     enterococcus     1.0000    0.3265    0.4923        49\n",
      "          erwinia     0.9844    0.2203    0.3600       286\n",
      "      escherichia     0.9509    0.1946    0.3231      1295\n",
      "   flavobacterium     1.0000    0.7292    0.8434        48\n",
      "         gordonia     1.0000    0.1698    0.2903        53\n",
      "       halorubrum     0.0000    0.0000    0.0000        13\n",
      "     helicobacter     1.0000    0.5000    0.6667        16\n",
      "       klebsiella     0.9865    0.1839    0.3100       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.4900    0.6577       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.2857    0.4444        28\n",
      "   microbacterium     1.0000    0.1212    0.2162        33\n",
      "    mycobacterium     0.9854    0.8185    0.8943       248\n",
      "           others     0.1703    0.9939    0.2908       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.3239    0.4894        71\n",
      "     pelagibacter     1.0000    0.0303    0.0588        33\n",
      "       prevotella     1.0000    0.7222    0.8387        18\n",
      "  prochlorococcus     1.0000    0.2308    0.3750        52\n",
      "          proteus     1.0000    0.1000    0.1818        40\n",
      "      providencia     0.0000    0.0000    0.0000        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     1.0000    0.5205    0.6846       513\n",
      "        ralstonia     1.0000    0.0952    0.1739        42\n",
      "     rheinheimera     1.0000    0.9602    0.9797       176\n",
      "        rhizobium     1.0000    0.1757    0.2989        74\n",
      "      rhodobacter     1.0000    0.4118    0.5833        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9730    0.3418    0.5059       632\n",
      "         serratia     0.8889    0.0825    0.1509        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.2500    0.0140    0.0265       143\n",
      "   staphylococcus     1.0000    0.6926    0.8184       244\n",
      " stenotrophomonas     0.0000    0.0000    0.0000        41\n",
      "    streptococcus     1.0000    0.3210    0.4860        81\n",
      "     streptomyces     1.0000    0.1860    0.3137        43\n",
      "    synechococcus     0.9944    0.6756    0.8046       521\n",
      "           vibrio     1.0000    0.3002    0.4618       463\n",
      "      xanthomonas     0.8333    0.0862    0.1563        58\n",
      "         yersinia     0.9444    0.1828    0.3063        93\n",
      "\n",
      "         accuracy                         0.4069      8116\n",
      "        macro avg     0.6815    0.2257    0.2950      8116\n",
      "     weighted avg     0.8268    0.4069    0.4270      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 100.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.0204    0.0400        98\n",
      "        aeromonas     0.0000    0.0000    0.0000        95\n",
      "    agrobacterium     0.0000    0.0000    0.0000        12\n",
      "     arthrobacter     0.0000    0.0000    0.0000        26\n",
      "         bacillus     1.0000    0.0294    0.0571       204\n",
      "      bacteroides     0.0000    0.0000    0.0000        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     0.0000    0.0000    0.0000        39\n",
      "    campylobacter     1.0000    0.0565    0.1069       124\n",
      "      caulobacter     0.0000    0.0000    0.0000        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.0000    0.0000    0.0000        24\n",
      "      cronobacter     0.0000    0.0000    0.0000        36\n",
      "          dickeya     1.0000    0.2885    0.4478        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.0000    0.0000    0.0000       142\n",
      "     enterococcus     1.0000    0.0408    0.0784        49\n",
      "          erwinia     1.0000    0.0385    0.0741       286\n",
      "      escherichia     0.9439    0.0780    0.1441      1295\n",
      "   flavobacterium     1.0000    0.6250    0.7692        48\n",
      "         gordonia     0.0000    0.0000    0.0000        53\n",
      "       halorubrum     0.0000    0.0000    0.0000        13\n",
      "     helicobacter     0.0000    0.0000    0.0000        16\n",
      "       klebsiella     0.9474    0.0453    0.0865       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.1700    0.2906       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.1071    0.1935        28\n",
      "   microbacterium     1.0000    0.0303    0.0588        33\n",
      "    mycobacterium     0.9819    0.6573    0.7874       248\n",
      "           others     0.1431    0.9959    0.2502       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.1268    0.2250        71\n",
      "     pelagibacter     0.0000    0.0000    0.0000        33\n",
      "       prevotella     1.0000    0.0556    0.1053        18\n",
      "  prochlorococcus     1.0000    0.2115    0.3492        52\n",
      "          proteus     1.0000    0.0500    0.0952        40\n",
      "      providencia     0.0000    0.0000    0.0000        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     1.0000    0.2943    0.4548       513\n",
      "        ralstonia     0.0000    0.0000    0.0000        42\n",
      "     rheinheimera     1.0000    0.9318    0.9647       176\n",
      "        rhizobium     1.0000    0.0135    0.0267        74\n",
      "      rhodobacter     1.0000    0.1176    0.2105        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9677    0.1424    0.2483       632\n",
      "         serratia     0.0000    0.0000    0.0000        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     1.0000    0.0070    0.0139       143\n",
      "   staphylococcus     1.0000    0.3279    0.4938       244\n",
      " stenotrophomonas     0.0000    0.0000    0.0000        41\n",
      "    streptococcus     1.0000    0.0247    0.0482        81\n",
      "     streptomyces     1.0000    0.0698    0.1304        43\n",
      "    synechococcus     0.9962    0.5010    0.6667       521\n",
      "           vibrio     1.0000    0.1490    0.2594       463\n",
      "      xanthomonas     0.5000    0.0172    0.0333        58\n",
      "         yersinia     1.0000    0.1183    0.2115        93\n",
      "\n",
      "         accuracy                         0.2734      8116\n",
      "        macro avg     0.4997    0.1075    0.1343      8116\n",
      "     weighted avg     0.7651    0.2734    0.2518      8116\n",
      "\n",
      "===================\n",
      "Finished\n",
      "==============\n",
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Confidence threshold k: 0.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.7000    0.4118    0.5185        17\n",
      "    acinetobacter     0.9306    0.6837    0.7882        98\n",
      "        aeromonas     0.8873    0.6632    0.7590        95\n",
      "    agrobacterium     0.6667    0.3333    0.4444        12\n",
      "     arthrobacter     0.7778    0.8077    0.7925        26\n",
      "         bacillus     0.7672    0.8725    0.8165       204\n",
      "      bacteroides     1.0000    0.9355    0.9667        31\n",
      "    brevundimonas     1.0000    0.5000    0.6667        10\n",
      "     burkholderia     0.6538    0.4359    0.5231        39\n",
      "    campylobacter     0.8016    0.8145    0.8080       124\n",
      "      caulobacter     1.0000    0.8400    0.9130        25\n",
      "      citrobacter     0.3871    0.3529    0.3692        34\n",
      "      clostridium     0.5161    0.6667    0.5818        24\n",
      "      cronobacter     0.6400    0.4444    0.5246        36\n",
      "          dickeya     0.9070    0.7500    0.8211        52\n",
      "     edwardsiella     0.8000    0.4000    0.5333        10\n",
      "     enterobacter     0.4957    0.4085    0.4479       142\n",
      "     enterococcus     0.8462    0.8980    0.8713        49\n",
      "          erwinia     0.6225    0.8881    0.7320       286\n",
      "      escherichia     0.5809    0.8263    0.6822      1295\n",
      "   flavobacterium     0.9375    0.9375    0.9375        48\n",
      "         gordonia     0.7925    0.7925    0.7925        53\n",
      "       halorubrum     0.3913    0.6923    0.5000        13\n",
      "     helicobacter     0.8889    1.0000    0.9412        16\n",
      "       klebsiella     0.7054    0.8262    0.7610       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.8235    0.6087    0.7000        23\n",
      "      lactococcus     0.9663    0.8600    0.9101       100\n",
      "      leuconostoc     1.0000    0.9000    0.9474        10\n",
      "         listeria     0.8889    0.8571    0.8727        28\n",
      "   microbacterium     0.8929    0.7576    0.8197        33\n",
      "    mycobacterium     0.8498    0.9355    0.8906       248\n",
      "           others     0.0000    0.0000    0.0000       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.8065    0.7042    0.7519        71\n",
      "     pelagibacter     0.6250    0.6061    0.6154        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9429    0.6346    0.7586        52\n",
      "          proteus     0.9167    0.5500    0.6875        40\n",
      "      providencia     0.4444    0.2500    0.3200        16\n",
      "pseudoalteromonas     0.7500    0.1765    0.2857        17\n",
      "      pseudomonas     0.7765    0.8869    0.8280       513\n",
      "        ralstonia     0.7778    0.5000    0.6087        42\n",
      "     rheinheimera     1.0000    1.0000    1.0000       176\n",
      "        rhizobium     0.8481    0.9054    0.8758        74\n",
      "      rhodobacter     0.8108    0.8824    0.8451        34\n",
      "      rhodococcus     0.8000    0.3077    0.4444        13\n",
      "       salmonella     0.8333    0.8070    0.8199       632\n",
      "         serratia     0.8228    0.6701    0.7386        97\n",
      "       shewanella     1.0000    0.3636    0.5333        11\n",
      "         shigella     0.4321    0.2448    0.3125       143\n",
      "   staphylococcus     0.7895    0.9836    0.8759       244\n",
      " stenotrophomonas     0.8519    0.5610    0.6765        41\n",
      "    streptococcus     0.8193    0.8395    0.8293        81\n",
      "     streptomyces     0.9189    0.7907    0.8500        43\n",
      "    synechococcus     0.7173    0.9693    0.8245       521\n",
      "           vibrio     0.5940    0.8942    0.7138       463\n",
      "      xanthomonas     0.6739    0.5345    0.5962        58\n",
      "         yersinia     0.5172    0.6452    0.5742        93\n",
      "\n",
      "         accuracy                         0.7081      8116\n",
      "        macro avg     0.7387    0.6568    0.6775      8116\n",
      "     weighted avg     0.6338    0.7081    0.6592      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 10.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.3529    0.5217        17\n",
      "    acinetobacter     0.9836    0.6122    0.7547        98\n",
      "        aeromonas     0.8833    0.5579    0.6839        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.6923    0.8182        26\n",
      "         bacillus     0.9101    0.8431    0.8753       204\n",
      "      bacteroides     1.0000    0.9355    0.9667        31\n",
      "    brevundimonas     1.0000    0.4000    0.5714        10\n",
      "     burkholderia     0.8421    0.4103    0.5517        39\n",
      "    campylobacter     0.9159    0.7903    0.8485       124\n",
      "      caulobacter     1.0000    0.8400    0.9130        25\n",
      "      citrobacter     0.5294    0.2647    0.3529        34\n",
      "      clostridium     0.5600    0.5833    0.5714        24\n",
      "      cronobacter     0.7647    0.3611    0.4906        36\n",
      "          dickeya     0.9268    0.7308    0.8172        52\n",
      "     edwardsiella     1.0000    0.3000    0.4615        10\n",
      "     enterobacter     0.5053    0.3380    0.4051       142\n",
      "     enterococcus     0.9535    0.8367    0.8913        49\n",
      "          erwinia     0.9087    0.8357    0.8707       286\n",
      "      escherichia     0.7530    0.7653    0.7591      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.7170    0.8352        53\n",
      "       halorubrum     0.4091    0.6923    0.5143        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.8480    0.7305    0.7848       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.8182    0.3913    0.5294        23\n",
      "      lactococcus     1.0000    0.8100    0.8950       100\n",
      "      leuconostoc     1.0000    0.9000    0.9474        10\n",
      "         listeria     0.8889    0.8571    0.8727        28\n",
      "   microbacterium     1.0000    0.7273    0.8421        33\n",
      "    mycobacterium     0.9664    0.9274    0.9465       248\n",
      "           others     0.3944    0.7363    0.5136       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.8448    0.6901    0.7597        71\n",
      "     pelagibacter     0.8235    0.4242    0.5600        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9375    0.5769    0.7143        52\n",
      "          proteus     0.9167    0.5500    0.6875        40\n",
      "      providencia     0.6667    0.2500    0.3636        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9675    0.8694    0.9158       513\n",
      "        ralstonia     0.9500    0.4524    0.6129        42\n",
      "     rheinheimera     1.0000    1.0000    1.0000       176\n",
      "        rhizobium     0.9683    0.8243    0.8905        74\n",
      "      rhodobacter     0.9091    0.8824    0.8955        34\n",
      "      rhodococcus     1.0000    0.0769    0.1429        13\n",
      "       salmonella     0.8994    0.7358    0.8094       632\n",
      "         serratia     0.8310    0.6082    0.7024        97\n",
      "       shewanella     1.0000    0.3636    0.5333        11\n",
      "         shigella     0.4545    0.2098    0.2871       143\n",
      "   staphylococcus     0.9514    0.9631    0.9572       244\n",
      " stenotrophomonas     0.9412    0.3902    0.5517        41\n",
      "    streptococcus     0.9286    0.8025    0.8609        81\n",
      "     streptomyces     1.0000    0.7209    0.8378        43\n",
      "    synechococcus     0.8624    0.9386    0.8989       521\n",
      "           vibrio     0.8578    0.8078    0.8320       463\n",
      "      xanthomonas     0.7317    0.5172    0.6061        58\n",
      "         yersinia     0.5333    0.6022    0.5657        93\n",
      "\n",
      "         accuracy                         0.7520      8116\n",
      "        macro avg     0.8430    0.6191    0.6854      8116\n",
      "     weighted avg     0.7980    0.7520    0.7585      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 20.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.3529    0.5217        17\n",
      "    acinetobacter     1.0000    0.5510    0.7105        98\n",
      "        aeromonas     0.9423    0.5158    0.6667        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.6923    0.8182        26\n",
      "         bacillus     0.9364    0.7941    0.8594       204\n",
      "      bacteroides     1.0000    0.7742    0.8727        31\n",
      "    brevundimonas     1.0000    0.4000    0.5714        10\n",
      "     burkholderia     1.0000    0.3590    0.5283        39\n",
      "    campylobacter     0.9314    0.7661    0.8407       124\n",
      "      caulobacter     1.0000    0.7600    0.8636        25\n",
      "      citrobacter     0.4615    0.1765    0.2553        34\n",
      "      clostridium     0.6667    0.5833    0.6222        24\n",
      "      cronobacter     0.7692    0.2778    0.4082        36\n",
      "          dickeya     0.9737    0.7115    0.8222        52\n",
      "     edwardsiella     1.0000    0.3000    0.4615        10\n",
      "     enterobacter     0.5676    0.2958    0.3889       142\n",
      "     enterococcus     1.0000    0.8367    0.9111        49\n",
      "          erwinia     0.9404    0.7727    0.8484       286\n",
      "      escherichia     0.8289    0.7181    0.7695      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.6604    0.7955        53\n",
      "       halorubrum     0.3158    0.4615    0.3750        13\n",
      "     helicobacter     1.0000    0.9375    0.9677        16\n",
      "       klebsiella     0.9395    0.6650    0.7788       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.7778    0.3043    0.4375        23\n",
      "      lactococcus     1.0000    0.8100    0.8950       100\n",
      "      leuconostoc     1.0000    0.8000    0.8889        10\n",
      "         listeria     0.9231    0.8571    0.8889        28\n",
      "   microbacterium     1.0000    0.6970    0.8214        33\n",
      "    mycobacterium     0.9744    0.9194    0.9461       248\n",
      "           others     0.3495    0.8986    0.5033       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9130    0.5915    0.7179        71\n",
      "     pelagibacter     0.9286    0.3939    0.5532        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9333    0.5385    0.6829        52\n",
      "          proteus     0.9048    0.4750    0.6230        40\n",
      "      providencia     0.6000    0.1875    0.2857        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9864    0.8480    0.9119       513\n",
      "        ralstonia     0.9412    0.3810    0.5424        42\n",
      "     rheinheimera     1.0000    1.0000    1.0000       176\n",
      "        rhizobium     0.9667    0.7838    0.8657        74\n",
      "      rhodobacter     0.9333    0.8235    0.8750        34\n",
      "      rhodococcus     1.0000    0.0769    0.1429        13\n",
      "       salmonella     0.9207    0.6978    0.7939       632\n",
      "         serratia     0.8525    0.5361    0.6582        97\n",
      "       shewanella     1.0000    0.2727    0.4286        11\n",
      "         shigella     0.5185    0.1958    0.2843       143\n",
      "   staphylococcus     0.9832    0.9590    0.9710       244\n",
      " stenotrophomonas     0.9167    0.2683    0.4151        41\n",
      "    streptococcus     0.9538    0.7654    0.8493        81\n",
      "     streptomyces     1.0000    0.6744    0.8056        43\n",
      "    synechococcus     0.9438    0.9021    0.9225       521\n",
      "           vibrio     0.9545    0.7257    0.8245       463\n",
      "      xanthomonas     0.8387    0.4483    0.5843        58\n",
      "         yersinia     0.5699    0.5699    0.5699        93\n",
      "\n",
      "         accuracy                         0.7341      8116\n",
      "        macro avg     0.8620    0.5779    0.6609      8116\n",
      "     weighted avg     0.8327    0.7341    0.7507      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 30.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2941    0.4545        17\n",
      "    acinetobacter     1.0000    0.4898    0.6575        98\n",
      "        aeromonas     0.9524    0.4211    0.5839        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.6538    0.7907        26\n",
      "         bacillus     0.9497    0.7402    0.8320       204\n",
      "      bacteroides     1.0000    0.7097    0.8302        31\n",
      "    brevundimonas     1.0000    0.4000    0.5714        10\n",
      "     burkholderia     1.0000    0.3077    0.4706        39\n",
      "    campylobacter     0.9300    0.7500    0.8304       124\n",
      "      caulobacter     1.0000    0.7200    0.8372        25\n",
      "      citrobacter     0.7143    0.1471    0.2439        34\n",
      "      clostridium     0.7222    0.5417    0.6190        24\n",
      "      cronobacter     0.7143    0.1389    0.2326        36\n",
      "          dickeya     1.0000    0.6346    0.7765        52\n",
      "     edwardsiella     1.0000    0.2000    0.3333        10\n",
      "     enterobacter     0.6290    0.2746    0.3824       142\n",
      "     enterococcus     1.0000    0.8163    0.8989        49\n",
      "          erwinia     0.9537    0.7203    0.8207       286\n",
      "      escherichia     0.8599    0.6494    0.7400      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.6226    0.7674        53\n",
      "       halorubrum     0.3529    0.4615    0.4000        13\n",
      "     helicobacter     1.0000    0.8750    0.9333        16\n",
      "       klebsiella     0.9492    0.6121    0.7443       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.7500    0.2609    0.3871        23\n",
      "      lactococcus     1.0000    0.7800    0.8764       100\n",
      "      leuconostoc     1.0000    0.6000    0.7500        10\n",
      "         listeria     0.9231    0.8571    0.8889        28\n",
      "   microbacterium     1.0000    0.6667    0.8000        33\n",
      "    mycobacterium     0.9869    0.9113    0.9476       248\n",
      "           others     0.3090    0.9381    0.4648       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9070    0.5493    0.6842        71\n",
      "     pelagibacter     1.0000    0.3636    0.5333        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9286    0.5000    0.6500        52\n",
      "          proteus     0.9000    0.4500    0.6000        40\n",
      "      providencia     0.5000    0.1250    0.2000        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9953    0.8304    0.9054       513\n",
      "        ralstonia     0.9375    0.3571    0.5172        42\n",
      "     rheinheimera     1.0000    1.0000    1.0000       176\n",
      "        rhizobium     0.9623    0.6892    0.8031        74\n",
      "      rhodobacter     0.9259    0.7353    0.8197        34\n",
      "      rhodococcus     1.0000    0.0769    0.1429        13\n",
      "       salmonella     0.9261    0.6741    0.7802       632\n",
      "         serratia     0.8621    0.5155    0.6452        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5333    0.1678    0.2553       143\n",
      "   staphylococcus     0.9871    0.9385    0.9622       244\n",
      " stenotrophomonas     0.9000    0.2195    0.3529        41\n",
      "    streptococcus     0.9672    0.7284    0.8310        81\n",
      "     streptomyces     1.0000    0.6047    0.7536        43\n",
      "    synechococcus     0.9780    0.8541    0.9119       521\n",
      "           vibrio     0.9806    0.6566    0.7865       463\n",
      "      xanthomonas     0.8571    0.4138    0.5581        58\n",
      "         yersinia     0.6098    0.5376    0.5714        93\n",
      "\n",
      "         accuracy                         0.7001      8116\n",
      "        macro avg     0.8535    0.5342    0.6234      8116\n",
      "     weighted avg     0.8411    0.7001    0.7246      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 40.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2941    0.4545        17\n",
      "    acinetobacter     1.0000    0.4490    0.6197        98\n",
      "        aeromonas     0.9737    0.3895    0.5564        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.5769    0.7317        26\n",
      "         bacillus     0.9586    0.6814    0.7966       204\n",
      "      bacteroides     1.0000    0.7097    0.8302        31\n",
      "    brevundimonas     1.0000    0.3000    0.4615        10\n",
      "     burkholderia     1.0000    0.2051    0.3404        39\n",
      "    campylobacter     0.9271    0.7177    0.8091       124\n",
      "      caulobacter     1.0000    0.6800    0.8095        25\n",
      "      citrobacter     0.5000    0.0588    0.1053        34\n",
      "      clostridium     0.7059    0.5000    0.5854        24\n",
      "      cronobacter     0.6000    0.0833    0.1463        36\n",
      "          dickeya     1.0000    0.5962    0.7470        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.6809    0.2254    0.3386       142\n",
      "     enterococcus     1.0000    0.6735    0.8049        49\n",
      "          erwinia     0.9588    0.6503    0.7750       286\n",
      "      escherichia     0.8794    0.5915    0.7073      1295\n",
      "   flavobacterium     1.0000    0.8958    0.9451        48\n",
      "         gordonia     1.0000    0.5472    0.7073        53\n",
      "       halorubrum     0.3529    0.4615    0.4000        13\n",
      "     helicobacter     1.0000    0.8750    0.9333        16\n",
      "       klebsiella     0.9456    0.5693    0.7107       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.6667    0.1739    0.2759        23\n",
      "      lactococcus     1.0000    0.7600    0.8636       100\n",
      "      leuconostoc     1.0000    0.6000    0.7500        10\n",
      "         listeria     0.9167    0.7857    0.8462        28\n",
      "   microbacterium     1.0000    0.6364    0.7778        33\n",
      "    mycobacterium     0.9868    0.9073    0.9454       248\n",
      "           others     0.2788    0.9473    0.4308       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9211    0.4930    0.6422        71\n",
      "     pelagibacter     1.0000    0.2424    0.3902        33\n",
      "       prevotella     1.0000    0.8889    0.9412        18\n",
      "  prochlorococcus     0.9231    0.4615    0.6154        52\n",
      "          proteus     0.9375    0.3750    0.5357        40\n",
      "      providencia     0.5000    0.1250    0.2000        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9952    0.8109    0.8937       513\n",
      "        ralstonia     1.0000    0.3571    0.5263        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     1.0000    0.6757    0.8065        74\n",
      "      rhodobacter     0.9231    0.7059    0.8000        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9335    0.6440    0.7622       632\n",
      "         serratia     0.8704    0.4845    0.6225        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5750    0.1608    0.2514       143\n",
      "   staphylococcus     0.9913    0.9303    0.9598       244\n",
      " stenotrophomonas     0.8889    0.1951    0.3200        41\n",
      "    streptococcus     0.9821    0.6790    0.8029        81\n",
      "     streptomyces     1.0000    0.5349    0.6970        43\n",
      "    synechococcus     0.9840    0.8253    0.8977       521\n",
      "           vibrio     0.9930    0.6134    0.7583       463\n",
      "      xanthomonas     0.8889    0.4138    0.5647        58\n",
      "         yersinia     0.6528    0.5054    0.5697        93\n",
      "\n",
      "         accuracy                         0.6668      8116\n",
      "        macro avg     0.8355    0.4951    0.5873      8116\n",
      "     weighted avg     0.8429    0.6668    0.6973      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 50.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.1765    0.3000        17\n",
      "    acinetobacter     1.0000    0.4082    0.5797        98\n",
      "        aeromonas     0.9630    0.2737    0.4262        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.5385    0.7000        26\n",
      "         bacillus     0.9638    0.6520    0.7778       204\n",
      "      bacteroides     1.0000    0.7097    0.8302        31\n",
      "    brevundimonas     1.0000    0.3000    0.4615        10\n",
      "     burkholderia     1.0000    0.1795    0.3043        39\n",
      "    campylobacter     0.9333    0.6774    0.7850       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     0.6667    0.0588    0.1081        34\n",
      "      clostridium     0.6667    0.4167    0.5128        24\n",
      "      cronobacter     0.6000    0.0833    0.1463        36\n",
      "          dickeya     1.0000    0.5000    0.6667        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.7586    0.1549    0.2573       142\n",
      "     enterococcus     1.0000    0.5510    0.7105        49\n",
      "          erwinia     0.9769    0.5909    0.7364       286\n",
      "      escherichia     0.8930    0.5220    0.6589      1295\n",
      "   flavobacterium     1.0000    0.8958    0.9451        48\n",
      "         gordonia     1.0000    0.5094    0.6750        53\n",
      "       halorubrum     0.4000    0.4615    0.4286        13\n",
      "     helicobacter     1.0000    0.6875    0.8148        16\n",
      "       klebsiella     0.9502    0.5290    0.6796       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.6667    0.0870    0.1538        23\n",
      "      lactococcus     1.0000    0.7400    0.8506       100\n",
      "      leuconostoc     1.0000    0.5000    0.6667        10\n",
      "         listeria     0.9091    0.7143    0.8000        28\n",
      "   microbacterium     1.0000    0.5152    0.6800        33\n",
      "    mycobacterium     0.9868    0.9073    0.9454       248\n",
      "           others     0.2534    0.9594    0.4009       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9459    0.4930    0.6481        71\n",
      "     pelagibacter     1.0000    0.0606    0.1143        33\n",
      "       prevotella     1.0000    0.8889    0.9412        18\n",
      "  prochlorococcus     0.9200    0.4423    0.5974        52\n",
      "          proteus     1.0000    0.3250    0.4906        40\n",
      "      providencia     0.5000    0.1250    0.2000        16\n",
      "pseudoalteromonas     1.0000    0.0588    0.1111        17\n",
      "      pseudomonas     0.9951    0.7856    0.8780       513\n",
      "        ralstonia     1.0000    0.3571    0.5263        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     1.0000    0.5541    0.7130        74\n",
      "      rhodobacter     0.9231    0.7059    0.8000        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9410    0.6060    0.7372       632\n",
      "         serratia     0.8936    0.4330    0.5833        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5758    0.1329    0.2159       143\n",
      "   staphylococcus     0.9955    0.9016    0.9462       244\n",
      " stenotrophomonas     0.8750    0.1707    0.2857        41\n",
      "    streptococcus     0.9800    0.6049    0.7481        81\n",
      "     streptomyces     1.0000    0.4651    0.6349        43\n",
      "    synechococcus     0.9929    0.8081    0.8910       521\n",
      "           vibrio     1.0000    0.5659    0.7228       463\n",
      "      xanthomonas     0.8696    0.3448    0.4938        58\n",
      "         yersinia     0.6774    0.4516    0.5419        93\n",
      "\n",
      "         accuracy                         0.6295      8116\n",
      "        macro avg     0.8419    0.4517    0.5472      8116\n",
      "     weighted avg     0.8475    0.6295    0.6640      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 60.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.1765    0.3000        17\n",
      "    acinetobacter     1.0000    0.3265    0.4923        98\n",
      "        aeromonas     1.0000    0.2421    0.3898        95\n",
      "    agrobacterium     1.0000    0.2500    0.4000        12\n",
      "     arthrobacter     1.0000    0.5000    0.6667        26\n",
      "         bacillus     0.9771    0.6275    0.7642       204\n",
      "      bacteroides     1.0000    0.6452    0.7843        31\n",
      "    brevundimonas     1.0000    0.2000    0.3333        10\n",
      "     burkholderia     1.0000    0.1282    0.2273        39\n",
      "    campylobacter     0.9318    0.6613    0.7736       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.7000    0.2917    0.4118        24\n",
      "      cronobacter     0.5000    0.0278    0.0526        36\n",
      "          dickeya     1.0000    0.4615    0.6316        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.7727    0.1197    0.2073       142\n",
      "     enterococcus     1.0000    0.4898    0.6575        49\n",
      "          erwinia     0.9796    0.5035    0.6651       286\n",
      "      escherichia     0.9098    0.4440    0.5968      1295\n",
      "   flavobacterium     1.0000    0.8542    0.9213        48\n",
      "         gordonia     1.0000    0.3774    0.5479        53\n",
      "       halorubrum     0.3077    0.3077    0.3077        13\n",
      "     helicobacter     1.0000    0.6250    0.7692        16\n",
      "       klebsiella     0.9521    0.4509    0.6120       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.7300    0.8439       100\n",
      "      leuconostoc     1.0000    0.4000    0.5714        10\n",
      "         listeria     0.8889    0.5714    0.6957        28\n",
      "   microbacterium     1.0000    0.4848    0.6531        33\n",
      "    mycobacterium     0.9866    0.8911    0.9364       248\n",
      "           others     0.2298    0.9726    0.3718       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9706    0.4648    0.6286        71\n",
      "     pelagibacter     1.0000    0.0606    0.1143        33\n",
      "       prevotella     1.0000    0.8889    0.9412        18\n",
      "  prochlorococcus     0.9524    0.3846    0.5479        52\n",
      "          proteus     1.0000    0.2500    0.4000        40\n",
      "      providencia     0.5000    0.0625    0.1111        16\n",
      "pseudoalteromonas     1.0000    0.0588    0.1111        17\n",
      "      pseudomonas     0.9947    0.7290    0.8414       513\n",
      "        ralstonia     1.0000    0.3095    0.4727        42\n",
      "     rheinheimera     1.0000    0.9830    0.9914       176\n",
      "        rhizobium     1.0000    0.4324    0.6038        74\n",
      "      rhodobacter     1.0000    0.6765    0.8070        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9413    0.5585    0.7011       632\n",
      "         serratia     0.8857    0.3196    0.4697        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.6667    0.1259    0.2118       143\n",
      "   staphylococcus     1.0000    0.8689    0.9298       244\n",
      " stenotrophomonas     1.0000    0.1707    0.2917        41\n",
      "    streptococcus     1.0000    0.5432    0.7040        81\n",
      "     streptomyces     1.0000    0.4186    0.5902        43\n",
      "    synechococcus     0.9951    0.7831    0.8765       521\n",
      "           vibrio     1.0000    0.5227    0.6865       463\n",
      "      xanthomonas     0.9444    0.2931    0.4474        58\n",
      "         yersinia     0.7872    0.3978    0.5286        93\n",
      "\n",
      "         accuracy                         0.5849      8116\n",
      "        macro avg     0.8097    0.4017    0.4978      8116\n",
      "     weighted avg     0.8471    0.5849    0.6227      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 70.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.0588    0.1111        17\n",
      "    acinetobacter     1.0000    0.2347    0.3802        98\n",
      "        aeromonas     1.0000    0.1895    0.3186        95\n",
      "    agrobacterium     1.0000    0.2500    0.4000        12\n",
      "     arthrobacter     1.0000    0.5000    0.6667        26\n",
      "         bacillus     0.9828    0.5588    0.7125       204\n",
      "      bacteroides     1.0000    0.5806    0.7347        31\n",
      "    brevundimonas     1.0000    0.1000    0.1818        10\n",
      "     burkholderia     1.0000    0.1026    0.1860        39\n",
      "    campylobacter     0.9559    0.5242    0.6771       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.6667    0.1667    0.2667        24\n",
      "      cronobacter     1.0000    0.0278    0.0541        36\n",
      "          dickeya     1.0000    0.4423    0.6133        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.7500    0.0634    0.1169       142\n",
      "     enterococcus     1.0000    0.4694    0.6389        49\n",
      "          erwinia     0.9828    0.3986    0.5672       286\n",
      "      escherichia     0.9194    0.3699    0.5275      1295\n",
      "   flavobacterium     1.0000    0.8542    0.9213        48\n",
      "         gordonia     1.0000    0.2453    0.3939        53\n",
      "       halorubrum     0.2500    0.2308    0.2400        13\n",
      "     helicobacter     1.0000    0.3750    0.5455        16\n",
      "       klebsiella     0.9568    0.3904    0.5546       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.6200    0.7654       100\n",
      "      leuconostoc     1.0000    0.1000    0.1818        10\n",
      "         listeria     0.9286    0.4643    0.6190        28\n",
      "   microbacterium     1.0000    0.3939    0.5652        33\n",
      "    mycobacterium     0.9862    0.8669    0.9227       248\n",
      "           others     0.2077    0.9787    0.3427       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.4366    0.6078        71\n",
      "     pelagibacter     1.0000    0.0303    0.0588        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     1.0000    0.2500    0.4000        52\n",
      "          proteus     1.0000    0.1750    0.2979        40\n",
      "      providencia     1.0000    0.0625    0.1176        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     0.9971    0.6745    0.8047       513\n",
      "        ralstonia     1.0000    0.2619    0.4151        42\n",
      "     rheinheimera     1.0000    0.9773    0.9885       176\n",
      "        rhizobium     1.0000    0.3649    0.5347        74\n",
      "      rhodobacter     1.0000    0.6471    0.7857        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9485    0.4953    0.6507       632\n",
      "         serratia     0.8571    0.2474    0.3840        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.6400    0.1119    0.1905       143\n",
      "   staphylococcus     1.0000    0.8320    0.9083       244\n",
      " stenotrophomonas     1.0000    0.0976    0.1778        41\n",
      "    streptococcus     1.0000    0.4815    0.6500        81\n",
      "     streptomyces     1.0000    0.3953    0.5667        43\n",
      "    synechococcus     0.9948    0.7390    0.8480       521\n",
      "           vibrio     1.0000    0.4687    0.6382       463\n",
      "      xanthomonas     0.9375    0.2586    0.4054        58\n",
      "         yersinia     0.9688    0.3333    0.4960        93\n",
      "\n",
      "         accuracy                         0.5330      8116\n",
      "        macro avg     0.8124    0.3453    0.4377      8116\n",
      "     weighted avg     0.8499    0.5330    0.5716      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 80.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.0588    0.1111        17\n",
      "    acinetobacter     1.0000    0.1429    0.2500        98\n",
      "        aeromonas     1.0000    0.1263    0.2243        95\n",
      "    agrobacterium     1.0000    0.1667    0.2857        12\n",
      "     arthrobacter     1.0000    0.2692    0.4242        26\n",
      "         bacillus     0.9785    0.4461    0.6128       204\n",
      "      bacteroides     1.0000    0.4516    0.6222        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     1.0000    0.0769    0.1429        39\n",
      "    campylobacter     1.0000    0.4274    0.5989       124\n",
      "      caulobacter     1.0000    0.4400    0.6111        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.6667    0.0833    0.1481        24\n",
      "      cronobacter     1.0000    0.0278    0.0541        36\n",
      "          dickeya     1.0000    0.4423    0.6133        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.6667    0.0423    0.0795       142\n",
      "     enterococcus     1.0000    0.4694    0.6389        49\n",
      "          erwinia     0.9794    0.3322    0.4961       286\n",
      "      escherichia     0.9313    0.2826    0.4336      1295\n",
      "   flavobacterium     1.0000    0.7917    0.8837        48\n",
      "         gordonia     1.0000    0.2075    0.3438        53\n",
      "       halorubrum     0.6000    0.2308    0.3333        13\n",
      "     helicobacter     1.0000    0.3125    0.4762        16\n",
      "       klebsiella     0.9683    0.3073    0.4665       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.5500    0.7097       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     0.9091    0.3571    0.5128        28\n",
      "   microbacterium     1.0000    0.3333    0.5000        33\n",
      "    mycobacterium     0.9858    0.8387    0.9063       248\n",
      "           others     0.1901    0.9899    0.3190       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.3662    0.5361        71\n",
      "     pelagibacter     1.0000    0.0303    0.0588        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     1.0000    0.2500    0.4000        52\n",
      "          proteus     1.0000    0.1250    0.2222        40\n",
      "      providencia     0.0000    0.0000    0.0000        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     0.9968    0.6082    0.7554       513\n",
      "        ralstonia     1.0000    0.2381    0.3846        42\n",
      "     rheinheimera     1.0000    0.9716    0.9856       176\n",
      "        rhizobium     1.0000    0.2973    0.4583        74\n",
      "      rhodobacter     1.0000    0.6176    0.7636        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9512    0.4320    0.5941       632\n",
      "         serratia     0.8421    0.1649    0.2759        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5789    0.0769    0.1358       143\n",
      "   staphylococcus     1.0000    0.7869    0.8807       244\n",
      " stenotrophomonas     1.0000    0.0732    0.1364        41\n",
      "    streptococcus     1.0000    0.4074    0.5789        81\n",
      "     streptomyces     1.0000    0.3023    0.4643        43\n",
      "    synechococcus     0.9946    0.7102    0.8287       521\n",
      "           vibrio     1.0000    0.3823    0.5531       463\n",
      "      xanthomonas     0.9167    0.1897    0.3143        58\n",
      "         yersinia     0.9630    0.2796    0.4333        93\n",
      "\n",
      "         accuracy                         0.4786      8116\n",
      "        macro avg     0.7647    0.2940    0.3808      8116\n",
      "     weighted avg     0.8440    0.4786    0.5129      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 90.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.0816    0.1509        98\n",
      "        aeromonas     1.0000    0.0421    0.0808        95\n",
      "    agrobacterium     0.0000    0.0000    0.0000        12\n",
      "     arthrobacter     1.0000    0.1154    0.2069        26\n",
      "         bacillus     0.9808    0.2500    0.3984       204\n",
      "      bacteroides     1.0000    0.1290    0.2286        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     1.0000    0.0513    0.0976        39\n",
      "    campylobacter     1.0000    0.2500    0.4000       124\n",
      "      caulobacter     1.0000    0.2800    0.4375        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.6667    0.0833    0.1481        24\n",
      "      cronobacter     0.0000    0.0000    0.0000        36\n",
      "          dickeya     1.0000    0.3654    0.5352        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.6000    0.0211    0.0408       142\n",
      "     enterococcus     1.0000    0.4082    0.5797        49\n",
      "          erwinia     1.0000    0.2378    0.3842       286\n",
      "      escherichia     0.9494    0.1884    0.3144      1295\n",
      "   flavobacterium     1.0000    0.7292    0.8434        48\n",
      "         gordonia     1.0000    0.1509    0.2623        53\n",
      "       halorubrum     0.5000    0.1538    0.2353        13\n",
      "     helicobacter     1.0000    0.1250    0.2222        16\n",
      "       klebsiella     0.9870    0.1914    0.3207       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.4700    0.6395       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.2500    0.4000        28\n",
      "   microbacterium     1.0000    0.2424    0.3902        33\n",
      "    mycobacterium     0.9854    0.8145    0.8918       248\n",
      "           others     0.1698    0.9929    0.2901       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.2958    0.4565        71\n",
      "     pelagibacter     1.0000    0.0303    0.0588        33\n",
      "       prevotella     1.0000    0.7778    0.8750        18\n",
      "  prochlorococcus     1.0000    0.2308    0.3750        52\n",
      "          proteus     1.0000    0.1000    0.1818        40\n",
      "      providencia     0.0000    0.0000    0.0000        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     1.0000    0.5224    0.6863       513\n",
      "        ralstonia     1.0000    0.1190    0.2128        42\n",
      "     rheinheimera     1.0000    0.9545    0.9767       176\n",
      "        rhizobium     1.0000    0.1351    0.2381        74\n",
      "      rhodobacter     1.0000    0.5294    0.6923        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9773    0.3402    0.5047       632\n",
      "         serratia     0.8571    0.0619    0.1154        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.2857    0.0140    0.0267       143\n",
      "   staphylococcus     1.0000    0.6721    0.8039       244\n",
      " stenotrophomonas     1.0000    0.0244    0.0476        41\n",
      "    streptococcus     1.0000    0.3210    0.4860        81\n",
      "     streptomyces     1.0000    0.2093    0.3462        43\n",
      "    synechococcus     0.9943    0.6718    0.8018       521\n",
      "           vibrio     1.0000    0.3110    0.4745       463\n",
      "      xanthomonas     0.8333    0.0862    0.1563        58\n",
      "         yersinia     1.0000    0.2043    0.3393        93\n",
      "\n",
      "         accuracy                         0.4059      8116\n",
      "        macro avg     0.7083    0.2243    0.2941      8116\n",
      "     weighted avg     0.8342    0.4059    0.4266      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 100.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.0102    0.0202        98\n",
      "        aeromonas     0.0000    0.0000    0.0000        95\n",
      "    agrobacterium     0.0000    0.0000    0.0000        12\n",
      "     arthrobacter     0.0000    0.0000    0.0000        26\n",
      "         bacillus     1.0000    0.0392    0.0755       204\n",
      "      bacteroides     0.0000    0.0000    0.0000        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     0.0000    0.0000    0.0000        39\n",
      "    campylobacter     1.0000    0.0403    0.0775       124\n",
      "      caulobacter     1.0000    0.0400    0.0769        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.0000    0.0000    0.0000        24\n",
      "      cronobacter     0.0000    0.0000    0.0000        36\n",
      "          dickeya     1.0000    0.2885    0.4478        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.0000    0.0000    0.0000       142\n",
      "     enterococcus     1.0000    0.0408    0.0784        49\n",
      "          erwinia     1.0000    0.0280    0.0544       286\n",
      "      escherichia     0.9462    0.0680    0.1268      1295\n",
      "   flavobacterium     1.0000    0.6250    0.7692        48\n",
      "         gordonia     0.0000    0.0000    0.0000        53\n",
      "       halorubrum     0.0000    0.0000    0.0000        13\n",
      "     helicobacter     0.0000    0.0000    0.0000        16\n",
      "       klebsiella     1.0000    0.0479    0.0913       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.2400    0.3871       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.1071    0.1935        28\n",
      "   microbacterium     1.0000    0.0303    0.0588        33\n",
      "    mycobacterium     0.9817    0.6492    0.7816       248\n",
      "           others     0.1429    0.9959    0.2499       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.1408    0.2469        71\n",
      "     pelagibacter     0.0000    0.0000    0.0000        33\n",
      "       prevotella     1.0000    0.2778    0.4348        18\n",
      "  prochlorococcus     1.0000    0.2115    0.3492        52\n",
      "          proteus     1.0000    0.0500    0.0952        40\n",
      "      providencia     0.0000    0.0000    0.0000        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     1.0000    0.3002    0.4618       513\n",
      "        ralstonia     0.0000    0.0000    0.0000        42\n",
      "     rheinheimera     1.0000    0.9318    0.9647       176\n",
      "        rhizobium     0.0000    0.0000    0.0000        74\n",
      "      rhodobacter     1.0000    0.2353    0.3810        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9785    0.1440    0.2510       632\n",
      "         serratia     1.0000    0.0206    0.0404        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.6667    0.0140    0.0274       143\n",
      "   staphylococcus     1.0000    0.3033    0.4654       244\n",
      " stenotrophomonas     0.0000    0.0000    0.0000        41\n",
      "    streptococcus     1.0000    0.0123    0.0244        81\n",
      "     streptomyces     1.0000    0.0465    0.0889        43\n",
      "    synechococcus     0.9961    0.4914    0.6581       521\n",
      "           vibrio     1.0000    0.1490    0.2594       463\n",
      "      xanthomonas     0.5000    0.0172    0.0333        58\n",
      "         yersinia     1.0000    0.1183    0.2115        93\n",
      "\n",
      "         accuracy                         0.2724      8116\n",
      "        macro avg     0.5121    0.1138    0.1438      8116\n",
      "     weighted avg     0.7689    0.2724    0.2504      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished\n",
      "==============\n",
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Confidence threshold k: 0.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.4118    0.5833        17\n",
      "    acinetobacter     0.8767    0.6531    0.7485        98\n",
      "        aeromonas     0.9508    0.6105    0.7436        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.6923    0.8182        26\n",
      "         bacillus     0.7576    0.8578    0.8046       204\n",
      "      bacteroides     1.0000    0.9355    0.9667        31\n",
      "    brevundimonas     1.0000    0.5000    0.6667        10\n",
      "     burkholderia     0.6667    0.4103    0.5079        39\n",
      "    campylobacter     0.8559    0.7661    0.8085       124\n",
      "      caulobacter     0.8400    0.8400    0.8400        25\n",
      "      citrobacter     0.5000    0.2941    0.3704        34\n",
      "      clostridium     0.4242    0.5833    0.4912        24\n",
      "      cronobacter     0.6000    0.4167    0.4918        36\n",
      "          dickeya     0.9048    0.7308    0.8085        52\n",
      "     edwardsiella     1.0000    0.4000    0.5714        10\n",
      "     enterobacter     0.4872    0.4014    0.4402       142\n",
      "     enterococcus     0.9565    0.8980    0.9263        49\n",
      "          erwinia     0.7023    0.8497    0.7690       286\n",
      "      escherichia     0.5056    0.8734    0.6404      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     0.8125    0.7358    0.7723        53\n",
      "       halorubrum     0.4762    0.7692    0.5882        13\n",
      "     helicobacter     0.9412    1.0000    0.9697        16\n",
      "       klebsiella     0.6856    0.7909    0.7345       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.7273    0.3478    0.4706        23\n",
      "      lactococcus     0.9773    0.8600    0.9149       100\n",
      "      leuconostoc     1.0000    0.9000    0.9474        10\n",
      "         listeria     0.8889    0.8571    0.8727        28\n",
      "   microbacterium     0.8889    0.7273    0.8000        33\n",
      "    mycobacterium     0.8846    0.9274    0.9055       248\n",
      "           others     0.0000    0.0000    0.0000       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9038    0.6620    0.7642        71\n",
      "     pelagibacter     0.5484    0.5152    0.5312        33\n",
      "       prevotella     0.8947    0.9444    0.9189        18\n",
      "  prochlorococcus     0.8649    0.6154    0.7191        52\n",
      "          proteus     0.8519    0.5750    0.6866        40\n",
      "      providencia     0.5714    0.2500    0.3478        16\n",
      "pseudoalteromonas     0.6667    0.1176    0.2000        17\n",
      "      pseudomonas     0.6594    0.8869    0.7564       513\n",
      "        ralstonia     0.8571    0.4286    0.5714        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     0.9310    0.7297    0.8182        74\n",
      "      rhodobacter     0.8333    0.8824    0.8571        34\n",
      "      rhodococcus     1.0000    0.3077    0.4706        13\n",
      "       salmonella     0.8466    0.7595    0.8007       632\n",
      "         serratia     0.7901    0.6598    0.7191        97\n",
      "       shewanella     1.0000    0.3636    0.5333        11\n",
      "         shigella     0.4857    0.2378    0.3192       143\n",
      "   staphylococcus     0.7264    0.9795    0.8342       244\n",
      " stenotrophomonas     0.8500    0.4146    0.5574        41\n",
      "    streptococcus     0.8592    0.7531    0.8026        81\n",
      "     streptomyces     0.9143    0.7442    0.8205        43\n",
      "    synechococcus     0.7313    0.9405    0.8228       521\n",
      "           vibrio     0.7297    0.8164    0.7706       463\n",
      "      xanthomonas     0.6591    0.5000    0.5686        58\n",
      "         yersinia     0.7403    0.6129    0.6706        93\n",
      "\n",
      "         accuracy                         0.6920      8116\n",
      "        macro avg     0.7665    0.6272    0.6661      8116\n",
      "     weighted avg     0.6330    0.6920    0.6456      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 10.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2941    0.4545        17\n",
      "    acinetobacter     1.0000    0.5408    0.7020        98\n",
      "        aeromonas     0.9615    0.5263    0.6803        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.6154    0.7619        26\n",
      "         bacillus     0.8871    0.8088    0.8462       204\n",
      "      bacteroides     1.0000    0.9355    0.9667        31\n",
      "    brevundimonas     1.0000    0.4000    0.5714        10\n",
      "     burkholderia     0.9333    0.3590    0.5185        39\n",
      "    campylobacter     0.9495    0.7581    0.8430       124\n",
      "      caulobacter     1.0000    0.8400    0.9130        25\n",
      "      citrobacter     0.4615    0.1765    0.2553        34\n",
      "      clostridium     0.5200    0.5417    0.5306        24\n",
      "      cronobacter     0.6667    0.3333    0.4444        36\n",
      "          dickeya     0.9737    0.7115    0.8222        52\n",
      "     edwardsiella     1.0000    0.2000    0.3333        10\n",
      "     enterobacter     0.5444    0.3451    0.4224       142\n",
      "     enterococcus     0.9767    0.8571    0.9130        49\n",
      "          erwinia     0.8963    0.7552    0.8197       286\n",
      "      escherichia     0.7518    0.7954    0.7730      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     0.9744    0.7170    0.8261        53\n",
      "       halorubrum     0.4000    0.4615    0.4286        13\n",
      "     helicobacter     0.9412    1.0000    0.9697        16\n",
      "       klebsiella     0.8812    0.6725    0.7629       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.7500    0.2609    0.3871        23\n",
      "      lactococcus     1.0000    0.8200    0.9011       100\n",
      "      leuconostoc     1.0000    0.8000    0.8889        10\n",
      "         listeria     0.8889    0.8571    0.8727        28\n",
      "   microbacterium     0.9545    0.6364    0.7636        33\n",
      "    mycobacterium     0.9870    0.9194    0.9520       248\n",
      "           others     0.3581    0.7779    0.4904       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.8958    0.6056    0.7227        71\n",
      "     pelagibacter     0.9333    0.4242    0.5833        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9375    0.5769    0.7143        52\n",
      "          proteus     0.8696    0.5000    0.6349        40\n",
      "      providencia     0.6667    0.2500    0.3636        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9187    0.8596    0.8882       513\n",
      "        ralstonia     0.9444    0.4048    0.5667        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     0.9804    0.6757    0.8000        74\n",
      "      rhodobacter     0.9000    0.7941    0.8438        34\n",
      "      rhodococcus     1.0000    0.0769    0.1429        13\n",
      "       salmonella     0.9091    0.7120    0.7986       632\n",
      "         serratia     0.8361    0.5258    0.6456        97\n",
      "       shewanella     1.0000    0.1818    0.3077        11\n",
      "         shigella     0.5088    0.2028    0.2900       143\n",
      "   staphylococcus     0.8623    0.9754    0.9154       244\n",
      " stenotrophomonas     0.9333    0.3415    0.5000        41\n",
      "    streptococcus     0.9677    0.7407    0.8392        81\n",
      "     streptomyces     1.0000    0.6047    0.7536        43\n",
      "    synechococcus     0.9061    0.9079    0.9070       521\n",
      "           vibrio     0.9499    0.6955    0.8030       463\n",
      "      xanthomonas     0.8182    0.4655    0.5934        58\n",
      "         yersinia     0.8060    0.5806    0.6750        93\n",
      "\n",
      "         accuracy                         0.7340      8116\n",
      "        macro avg     0.8509    0.5787    0.6568      8116\n",
      "     weighted avg     0.8052    0.7340    0.7443      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 20.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2353    0.3810        17\n",
      "    acinetobacter     1.0000    0.4694    0.6389        98\n",
      "        aeromonas     0.9524    0.4211    0.5839        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.5385    0.7000        26\n",
      "         bacillus     0.9268    0.7451    0.8261       204\n",
      "      bacteroides     1.0000    0.9355    0.9667        31\n",
      "    brevundimonas     1.0000    0.4000    0.5714        10\n",
      "     burkholderia     0.9333    0.3590    0.5185        39\n",
      "    campylobacter     0.9681    0.7339    0.8349       124\n",
      "      caulobacter     1.0000    0.7600    0.8636        25\n",
      "      citrobacter     0.6250    0.1471    0.2381        34\n",
      "      clostridium     0.5455    0.5000    0.5217        24\n",
      "      cronobacter     0.7143    0.2778    0.4000        36\n",
      "          dickeya     1.0000    0.7115    0.8315        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.6471    0.3099    0.4190       142\n",
      "     enterococcus     0.9756    0.8163    0.8889        49\n",
      "          erwinia     0.9474    0.6923    0.8000       286\n",
      "      escherichia     0.8217    0.7297    0.7730      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.6604    0.7955        53\n",
      "       halorubrum     0.4000    0.4615    0.4286        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.9344    0.6096    0.7378       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.6667    0.1739    0.2759        23\n",
      "      lactococcus     1.0000    0.7800    0.8764       100\n",
      "      leuconostoc     1.0000    0.7000    0.8235        10\n",
      "         listeria     0.8889    0.8571    0.8727        28\n",
      "   microbacterium     1.0000    0.5758    0.7308        33\n",
      "    mycobacterium     0.9870    0.9153    0.9498       248\n",
      "           others     0.3216    0.9097    0.4752       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9070    0.5493    0.6842        71\n",
      "     pelagibacter     1.0000    0.3030    0.4651        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9310    0.5192    0.6667        52\n",
      "          proteus     0.9000    0.4500    0.6000        40\n",
      "      providencia     0.5000    0.1250    0.2000        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9749    0.8343    0.8992       513\n",
      "        ralstonia     0.9444    0.4048    0.5667        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     0.9800    0.6622    0.7903        74\n",
      "      rhodobacter     0.9259    0.7353    0.8197        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9206    0.6788    0.7814       632\n",
      "         serratia     0.8305    0.5052    0.6282        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5745    0.1888    0.2842       143\n",
      "   staphylococcus     0.9323    0.9590    0.9455       244\n",
      " stenotrophomonas     0.9167    0.2683    0.4151        41\n",
      "    streptococcus     1.0000    0.6914    0.8175        81\n",
      "     streptomyces     1.0000    0.5116    0.6769        43\n",
      "    synechococcus     0.9657    0.8637    0.9119       521\n",
      "           vibrio     0.9967    0.6479    0.7853       463\n",
      "      xanthomonas     0.8333    0.4310    0.5682        58\n",
      "         yersinia     0.8525    0.5591    0.6753        93\n",
      "\n",
      "         accuracy                         0.7113      8116\n",
      "        macro avg     0.8329    0.5380    0.6226      8116\n",
      "     weighted avg     0.8330    0.7113    0.7314      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 30.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2353    0.3810        17\n",
      "    acinetobacter     1.0000    0.4184    0.5899        98\n",
      "        aeromonas     0.9394    0.3263    0.4844        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.5000    0.6667        26\n",
      "         bacillus     0.9346    0.7010    0.8011       204\n",
      "      bacteroides     1.0000    0.8065    0.8929        31\n",
      "    brevundimonas     1.0000    0.4000    0.5714        10\n",
      "     burkholderia     1.0000    0.2821    0.4400        39\n",
      "    campylobacter     0.9663    0.6935    0.8075       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     0.6000    0.0882    0.1538        34\n",
      "      clostridium     0.6923    0.3750    0.4865        24\n",
      "      cronobacter     0.7778    0.1944    0.3111        36\n",
      "          dickeya     1.0000    0.6538    0.7907        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.6379    0.2606    0.3700       142\n",
      "     enterococcus     1.0000    0.7143    0.8333        49\n",
      "          erwinia     0.9577    0.6329    0.7621       286\n",
      "      escherichia     0.8525    0.6649    0.7471      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.5660    0.7229        53\n",
      "       halorubrum     0.4000    0.4615    0.4286        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.9456    0.5693    0.7107       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.8000    0.1739    0.2857        23\n",
      "      lactococcus     1.0000    0.7700    0.8701       100\n",
      "      leuconostoc     1.0000    0.6000    0.7500        10\n",
      "         listeria     0.8800    0.7857    0.8302        28\n",
      "   microbacterium     1.0000    0.5455    0.7059        33\n",
      "    mycobacterium     0.9870    0.9153    0.9498       248\n",
      "           others     0.2914    0.9483    0.4458       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9048    0.5352    0.6726        71\n",
      "     pelagibacter     1.0000    0.2424    0.3902        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9286    0.5000    0.6500        52\n",
      "          proteus     0.9000    0.4500    0.6000        40\n",
      "      providencia     0.6667    0.1250    0.2105        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9880    0.8031    0.8860       513\n",
      "        ralstonia     0.9412    0.3810    0.5424        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     0.9796    0.6486    0.7805        74\n",
      "      rhodobacter     0.9231    0.7059    0.8000        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9348    0.6582    0.7725       632\n",
      "         serratia     0.8627    0.4536    0.5946        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.6154    0.1678    0.2637       143\n",
      "   staphylococcus     0.9913    0.9385    0.9642       244\n",
      " stenotrophomonas     0.8889    0.1951    0.3200        41\n",
      "    streptococcus     1.0000    0.6543    0.7910        81\n",
      "     streptomyces     1.0000    0.4419    0.6129        43\n",
      "    synechococcus     0.9863    0.8273    0.8998       521\n",
      "           vibrio     1.0000    0.6112    0.7587       463\n",
      "      xanthomonas     0.8846    0.3966    0.5476        58\n",
      "         yersinia     0.8421    0.5161    0.6400        93\n",
      "\n",
      "         accuracy                         0.6804      8116\n",
      "        macro avg     0.8458    0.5017    0.5948      8116\n",
      "     weighted avg     0.8431    0.6804    0.7076      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 40.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2353    0.3810        17\n",
      "    acinetobacter     1.0000    0.3776    0.5481        98\n",
      "        aeromonas     0.9677    0.3158    0.4762        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.4615    0.6316        26\n",
      "         bacillus     0.9504    0.6569    0.7768       204\n",
      "      bacteroides     1.0000    0.7419    0.8519        31\n",
      "    brevundimonas     1.0000    0.3000    0.4615        10\n",
      "     burkholderia     1.0000    0.2308    0.3750        39\n",
      "    campylobacter     0.9655    0.6774    0.7962       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     0.6000    0.0882    0.1538        34\n",
      "      clostridium     0.7273    0.3333    0.4571        24\n",
      "      cronobacter     0.6667    0.1111    0.1905        36\n",
      "          dickeya     1.0000    0.5769    0.7317        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.6591    0.2042    0.3118       142\n",
      "     enterococcus     1.0000    0.6939    0.8193        49\n",
      "          erwinia     0.9609    0.6014    0.7398       286\n",
      "      escherichia     0.8826    0.6039    0.7171      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.4528    0.6234        53\n",
      "       halorubrum     0.3571    0.3846    0.3704        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.9452    0.5214    0.6721       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.8000    0.1739    0.2857        23\n",
      "      lactococcus     1.0000    0.7300    0.8439       100\n",
      "      leuconostoc     1.0000    0.5000    0.6667        10\n",
      "         listeria     0.9130    0.7500    0.8235        28\n",
      "   microbacterium     1.0000    0.5152    0.6800        33\n",
      "    mycobacterium     0.9869    0.9113    0.9476       248\n",
      "           others     0.2652    0.9584    0.4155       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9211    0.4930    0.6422        71\n",
      "     pelagibacter     1.0000    0.2121    0.3500        33\n",
      "       prevotella     1.0000    0.8889    0.9412        18\n",
      "  prochlorococcus     0.9259    0.4808    0.6329        52\n",
      "          proteus     0.9375    0.3750    0.5357        40\n",
      "      providencia     0.5000    0.0625    0.1111        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9878    0.7875    0.8764       513\n",
      "        ralstonia     1.0000    0.3810    0.5517        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     1.0000    0.5270    0.6903        74\n",
      "      rhodobacter     0.9231    0.7059    0.8000        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9408    0.6282    0.7533       632\n",
      "         serratia     0.8936    0.4330    0.5833        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.6176    0.1469    0.2373       143\n",
      "   staphylococcus     0.9955    0.9057    0.9485       244\n",
      " stenotrophomonas     0.8889    0.1951    0.3200        41\n",
      "    streptococcus     1.0000    0.5926    0.7442        81\n",
      "     streptomyces     1.0000    0.4186    0.5902        43\n",
      "    synechococcus     0.9952    0.8023    0.8884       521\n",
      "           vibrio     1.0000    0.5529    0.7121       463\n",
      "      xanthomonas     0.8500    0.2931    0.4359        58\n",
      "         yersinia     0.9020    0.4946    0.6389        93\n",
      "\n",
      "         accuracy                         0.6479      8116\n",
      "        macro avg     0.8462    0.4679    0.5656      8116\n",
      "     weighted avg     0.8480    0.6479    0.6810      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 50.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.1765    0.3000        17\n",
      "    acinetobacter     1.0000    0.3571    0.5263        98\n",
      "        aeromonas     0.9615    0.2632    0.4132        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.3462    0.5143        26\n",
      "         bacillus     0.9627    0.6324    0.7633       204\n",
      "      bacteroides     1.0000    0.6774    0.8077        31\n",
      "    brevundimonas     1.0000    0.2000    0.3333        10\n",
      "     burkholderia     1.0000    0.1795    0.3043        39\n",
      "    campylobacter     0.9625    0.6210    0.7549       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     0.5000    0.0588    0.1053        34\n",
      "      clostridium     0.6667    0.2500    0.3636        24\n",
      "      cronobacter     0.6667    0.0556    0.1026        36\n",
      "          dickeya     1.0000    0.5385    0.7000        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.6774    0.1479    0.2428       142\n",
      "     enterococcus     1.0000    0.6122    0.7595        49\n",
      "          erwinia     0.9804    0.5245    0.6834       286\n",
      "      escherichia     0.8934    0.5305    0.6657      1295\n",
      "   flavobacterium     1.0000    0.8958    0.9451        48\n",
      "         gordonia     1.0000    0.4151    0.5867        53\n",
      "       halorubrum     0.3077    0.3077    0.3077        13\n",
      "     helicobacter     1.0000    0.8750    0.9333        16\n",
      "       klebsiella     0.9403    0.4761    0.6321       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.6667    0.0870    0.1538        23\n",
      "      lactococcus     1.0000    0.6900    0.8166       100\n",
      "      leuconostoc     1.0000    0.4000    0.5714        10\n",
      "         listeria     0.9474    0.6429    0.7660        28\n",
      "   microbacterium     1.0000    0.4545    0.6250        33\n",
      "    mycobacterium     0.9867    0.8952    0.9387       248\n",
      "           others     0.2439    0.9665    0.3895       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9412    0.4507    0.6095        71\n",
      "     pelagibacter     1.0000    0.2121    0.3500        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     0.9200    0.4423    0.5974        52\n",
      "          proteus     1.0000    0.3250    0.4906        40\n",
      "      providencia     0.5000    0.0625    0.1111        16\n",
      "pseudoalteromonas     1.0000    0.0588    0.1111        17\n",
      "      pseudomonas     0.9949    0.7583    0.8606       513\n",
      "        ralstonia     1.0000    0.3810    0.5517        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     1.0000    0.4865    0.6545        74\n",
      "      rhodobacter     0.9200    0.6765    0.7797        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9491    0.5902    0.7278       632\n",
      "         serratia     0.8913    0.4227    0.5734        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.6667    0.1259    0.2118       143\n",
      "   staphylococcus     1.0000    0.8811    0.9368       244\n",
      " stenotrophomonas     0.8750    0.1707    0.2857        41\n",
      "    streptococcus     1.0000    0.5432    0.7040        81\n",
      "     streptomyces     1.0000    0.4186    0.5902        43\n",
      "    synechococcus     0.9952    0.7927    0.8825       521\n",
      "           vibrio     1.0000    0.5184    0.6828       463\n",
      "      xanthomonas     0.8947    0.2931    0.4416        58\n",
      "         yersinia     0.9000    0.4839    0.6294        93\n",
      "\n",
      "         accuracy                         0.6126      8116\n",
      "        macro avg     0.8443    0.4283    0.5281      8116\n",
      "     weighted avg     0.8499    0.6126    0.6497      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 60.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.2653    0.4194        98\n",
      "        aeromonas     1.0000    0.1789    0.3036        95\n",
      "    agrobacterium     1.0000    0.2500    0.4000        12\n",
      "     arthrobacter     1.0000    0.3077    0.4706        26\n",
      "         bacillus     0.9750    0.5735    0.7222       204\n",
      "      bacteroides     1.0000    0.4839    0.6522        31\n",
      "    brevundimonas     1.0000    0.2000    0.3333        10\n",
      "     burkholderia     1.0000    0.1282    0.2273        39\n",
      "    campylobacter     0.9583    0.5565    0.7041       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.7143    0.2083    0.3226        24\n",
      "      cronobacter     1.0000    0.0556    0.1053        36\n",
      "          dickeya     1.0000    0.5000    0.6667        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.7391    0.1197    0.2061       142\n",
      "     enterococcus     1.0000    0.5102    0.6757        49\n",
      "          erwinia     0.9856    0.4790    0.6447       286\n",
      "      escherichia     0.9098    0.4517    0.6037      1295\n",
      "   flavobacterium     1.0000    0.8542    0.9213        48\n",
      "         gordonia     1.0000    0.3208    0.4857        53\n",
      "       halorubrum     0.3333    0.1538    0.2105        13\n",
      "     helicobacter     1.0000    0.5625    0.7200        16\n",
      "       klebsiella     0.9551    0.4282    0.5913       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     1.0000    0.0435    0.0833        23\n",
      "      lactococcus     1.0000    0.6600    0.7952       100\n",
      "      leuconostoc     1.0000    0.2000    0.3333        10\n",
      "         listeria     0.9375    0.5357    0.6818        28\n",
      "   microbacterium     1.0000    0.4545    0.6250        33\n",
      "    mycobacterium     0.9862    0.8629    0.9204       248\n",
      "           others     0.2225    0.9828    0.3628       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9677    0.4225    0.5882        71\n",
      "     pelagibacter     1.0000    0.0606    0.1143        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     0.9500    0.3654    0.5278        52\n",
      "          proteus     1.0000    0.2250    0.3673        40\n",
      "      providencia     1.0000    0.0625    0.1176        16\n",
      "pseudoalteromonas     1.0000    0.0588    0.1111        17\n",
      "      pseudomonas     0.9973    0.7212    0.8371       513\n",
      "        ralstonia     1.0000    0.3333    0.5000        42\n",
      "     rheinheimera     1.0000    0.9886    0.9943       176\n",
      "        rhizobium     1.0000    0.4324    0.6038        74\n",
      "      rhodobacter     0.9583    0.6765    0.7931        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9511    0.5237    0.6755       632\n",
      "         serratia     0.8750    0.2887    0.4341        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.7826    0.1259    0.2169       143\n",
      "   staphylococcus     1.0000    0.8566    0.9227       244\n",
      " stenotrophomonas     1.0000    0.1463    0.2553        41\n",
      "    streptococcus     1.0000    0.4691    0.6387        81\n",
      "     streptomyces     1.0000    0.3721    0.5424        43\n",
      "    synechococcus     0.9950    0.7620    0.8630       521\n",
      "           vibrio     1.0000    0.4773    0.6462       463\n",
      "      xanthomonas     0.9444    0.2931    0.4474        58\n",
      "         yersinia     0.9730    0.3871    0.5538        93\n",
      "\n",
      "         accuracy                         0.5673      8116\n",
      "        macro avg     0.8493    0.3720    0.4714      8116\n",
      "     weighted avg     0.8565    0.5673    0.6064      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 70.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.1939    0.3248        98\n",
      "        aeromonas     1.0000    0.1368    0.2407        95\n",
      "    agrobacterium     1.0000    0.2500    0.4000        12\n",
      "     arthrobacter     1.0000    0.2308    0.3750        26\n",
      "         bacillus     0.9896    0.4657    0.6333       204\n",
      "      bacteroides     1.0000    0.3871    0.5581        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     1.0000    0.0769    0.1429        39\n",
      "    campylobacter     0.9667    0.4677    0.6304       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.7500    0.1250    0.2143        24\n",
      "      cronobacter     1.0000    0.0556    0.1053        36\n",
      "          dickeya     1.0000    0.4423    0.6133        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.6667    0.0423    0.0795       142\n",
      "     enterococcus     1.0000    0.4694    0.6389        49\n",
      "          erwinia     0.9823    0.3881    0.5564       286\n",
      "      escherichia     0.9198    0.3722    0.5300      1295\n",
      "   flavobacterium     1.0000    0.8542    0.9213        48\n",
      "         gordonia     1.0000    0.2642    0.4179        53\n",
      "       halorubrum     0.2500    0.0769    0.1176        13\n",
      "     helicobacter     1.0000    0.4375    0.6087        16\n",
      "       klebsiella     0.9630    0.3275    0.4887       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.6100    0.7578       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     0.9333    0.5000    0.6512        28\n",
      "   microbacterium     1.0000    0.3030    0.4651        33\n",
      "    mycobacterium     0.9860    0.8548    0.9158       248\n",
      "           others     0.2011    0.9868    0.3341       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.3380    0.5053        71\n",
      "     pelagibacter     1.0000    0.0606    0.1143        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     1.0000    0.2500    0.4000        52\n",
      "          proteus     1.0000    0.1500    0.2609        40\n",
      "      providencia     1.0000    0.0625    0.1176        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     0.9971    0.6667    0.7991       513\n",
      "        ralstonia     1.0000    0.2619    0.4151        42\n",
      "     rheinheimera     1.0000    0.9773    0.9885       176\n",
      "        rhizobium     1.0000    0.3108    0.4742        74\n",
      "      rhodobacter     1.0000    0.6471    0.7857        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9472    0.4541    0.6139       632\n",
      "         serratia     0.8462    0.2268    0.3577        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.7368    0.0979    0.1728       143\n",
      "   staphylococcus     1.0000    0.8320    0.9083       244\n",
      " stenotrophomonas     1.0000    0.0976    0.1778        41\n",
      "    streptococcus     1.0000    0.3704    0.5405        81\n",
      "     streptomyces     1.0000    0.3488    0.5172        43\n",
      "    synechococcus     0.9947    0.7255    0.8391       521\n",
      "           vibrio     1.0000    0.4039    0.5754       463\n",
      "      xanthomonas     0.9412    0.2759    0.4267        58\n",
      "         yersinia     0.9697    0.3441    0.5079        93\n",
      "\n",
      "         accuracy                         0.5122      8116\n",
      "        macro avg     0.7634    0.3168    0.4052      8116\n",
      "     weighted avg     0.8456    0.5122    0.5490      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 80.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.1224    0.2182        98\n",
      "        aeromonas     1.0000    0.0947    0.1731        95\n",
      "    agrobacterium     1.0000    0.2500    0.4000        12\n",
      "     arthrobacter     1.0000    0.2308    0.3750        26\n",
      "         bacillus     0.9873    0.3824    0.5512       204\n",
      "      bacteroides     1.0000    0.3548    0.5238        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     1.0000    0.0513    0.0976        39\n",
      "    campylobacter     0.9583    0.3710    0.5349       124\n",
      "      caulobacter     1.0000    0.4800    0.6486        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.6667    0.0833    0.1481        24\n",
      "      cronobacter     1.0000    0.0278    0.0541        36\n",
      "          dickeya     1.0000    0.4423    0.6133        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.5000    0.0211    0.0405       142\n",
      "     enterococcus     1.0000    0.4694    0.6389        49\n",
      "          erwinia     0.9759    0.2832    0.4390       286\n",
      "      escherichia     0.9284    0.2903    0.4424      1295\n",
      "   flavobacterium     1.0000    0.8125    0.8966        48\n",
      "         gordonia     1.0000    0.2075    0.3438        53\n",
      "       halorubrum     0.3333    0.0769    0.1250        13\n",
      "     helicobacter     1.0000    0.3750    0.5455        16\n",
      "       klebsiella     0.9684    0.2317    0.3740       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.5000    0.6667       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.4286    0.6000        28\n",
      "   microbacterium     1.0000    0.2727    0.4286        33\n",
      "    mycobacterium     0.9858    0.8387    0.9063       248\n",
      "           others     0.1847    0.9919    0.3115       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.2958    0.4565        71\n",
      "     pelagibacter     1.0000    0.0303    0.0588        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     1.0000    0.2500    0.4000        52\n",
      "          proteus     1.0000    0.1250    0.2222        40\n",
      "      providencia     1.0000    0.0625    0.1176        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     1.0000    0.5887    0.7411       513\n",
      "        ralstonia     1.0000    0.2381    0.3846        42\n",
      "     rheinheimera     1.0000    0.9659    0.9827       176\n",
      "        rhizobium     1.0000    0.2297    0.3736        74\n",
      "      rhodobacter     1.0000    0.5588    0.7170        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9544    0.3972    0.5609       632\n",
      "         serratia     0.8235    0.1443    0.2456        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.7647    0.0909    0.1625       143\n",
      "   staphylococcus     1.0000    0.7992    0.8884       244\n",
      " stenotrophomonas     1.0000    0.0488    0.0930        41\n",
      "    streptococcus     1.0000    0.2840    0.4423        81\n",
      "     streptomyces     1.0000    0.2791    0.4364        43\n",
      "    synechococcus     0.9945    0.6910    0.8154       521\n",
      "           vibrio     1.0000    0.3391    0.5065       463\n",
      "      xanthomonas     0.9000    0.1552    0.2647        58\n",
      "         yersinia     0.9643    0.2903    0.4463        93\n",
      "\n",
      "         accuracy                         0.4597      8116\n",
      "        macro avg     0.7609    0.2778    0.3614      8116\n",
      "     weighted avg     0.8426    0.4597    0.4912      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 90.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.0408    0.0784        98\n",
      "        aeromonas     1.0000    0.0211    0.0412        95\n",
      "    agrobacterium     0.0000    0.0000    0.0000        12\n",
      "     arthrobacter     1.0000    0.1154    0.2069        26\n",
      "         bacillus     1.0000    0.2157    0.3548       204\n",
      "      bacteroides     1.0000    0.0645    0.1212        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     1.0000    0.0256    0.0500        39\n",
      "    campylobacter     0.9375    0.2419    0.3846       124\n",
      "      caulobacter     1.0000    0.2000    0.3333        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     1.0000    0.0833    0.1538        24\n",
      "      cronobacter     0.0000    0.0000    0.0000        36\n",
      "          dickeya     1.0000    0.3654    0.5352        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.6000    0.0211    0.0408       142\n",
      "     enterococcus     1.0000    0.3061    0.4687        49\n",
      "          erwinia     1.0000    0.2203    0.3610       286\n",
      "      escherichia     0.9368    0.1946    0.3223      1295\n",
      "   flavobacterium     1.0000    0.6875    0.8148        48\n",
      "         gordonia     1.0000    0.1321    0.2333        53\n",
      "       halorubrum     0.0000    0.0000    0.0000        13\n",
      "     helicobacter     1.0000    0.0625    0.1176        16\n",
      "       klebsiella     0.9828    0.1436    0.2505       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.4200    0.5915       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.3214    0.4865        28\n",
      "   microbacterium     1.0000    0.1515    0.2632        33\n",
      "    mycobacterium     0.9846    0.7742    0.8668       248\n",
      "           others     0.1662    0.9949    0.2848       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.2958    0.4565        71\n",
      "     pelagibacter     1.0000    0.0303    0.0588        33\n",
      "       prevotella     1.0000    0.7222    0.8387        18\n",
      "  prochlorococcus     1.0000    0.2115    0.3492        52\n",
      "          proteus     1.0000    0.0750    0.1395        40\n",
      "      providencia     0.0000    0.0000    0.0000        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     1.0000    0.4737    0.6429       513\n",
      "        ralstonia     1.0000    0.0714    0.1333        42\n",
      "     rheinheimera     1.0000    0.9659    0.9827       176\n",
      "        rhizobium     1.0000    0.1351    0.2381        74\n",
      "      rhodobacter     1.0000    0.4412    0.6122        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9567    0.3149    0.4738       632\n",
      "         serratia     0.8333    0.0515    0.0971        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.8000    0.0280    0.0541       143\n",
      "   staphylococcus     1.0000    0.6926    0.8184       244\n",
      " stenotrophomonas     1.0000    0.0488    0.0930        41\n",
      "    streptococcus     1.0000    0.1358    0.2391        81\n",
      "     streptomyces     1.0000    0.1860    0.3137        43\n",
      "    synechococcus     0.9942    0.6564    0.7908       521\n",
      "           vibrio     1.0000    0.2829    0.4411       463\n",
      "      xanthomonas     0.8333    0.0862    0.1563        58\n",
      "         yersinia     1.0000    0.1828    0.3091        93\n",
      "\n",
      "         accuracy                         0.3887      8116\n",
      "        macro avg     0.7123    0.2016    0.2644      8116\n",
      "     weighted avg     0.8383    0.3887    0.4052      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 100.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.0102    0.0202        98\n",
      "        aeromonas     0.0000    0.0000    0.0000        95\n",
      "    agrobacterium     0.0000    0.0000    0.0000        12\n",
      "     arthrobacter     0.0000    0.0000    0.0000        26\n",
      "         bacillus     1.0000    0.0098    0.0194       204\n",
      "      bacteroides     0.0000    0.0000    0.0000        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     0.0000    0.0000    0.0000        39\n",
      "    campylobacter     1.0000    0.0403    0.0775       124\n",
      "      caulobacter     0.0000    0.0000    0.0000        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.0000    0.0000    0.0000        24\n",
      "      cronobacter     0.0000    0.0000    0.0000        36\n",
      "          dickeya     1.0000    0.2115    0.3492        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.0000    0.0000    0.0000       142\n",
      "     enterococcus     1.0000    0.0612    0.1154        49\n",
      "          erwinia     1.0000    0.0385    0.0741       286\n",
      "      escherichia     0.9510    0.0749    0.1389      1295\n",
      "   flavobacterium     1.0000    0.6250    0.7692        48\n",
      "         gordonia     0.0000    0.0000    0.0000        53\n",
      "       halorubrum     0.0000    0.0000    0.0000        13\n",
      "     helicobacter     0.0000    0.0000    0.0000        16\n",
      "       klebsiella     1.0000    0.0327    0.0634       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.1700    0.2906       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.0357    0.0690        28\n",
      "   microbacterium     0.0000    0.0000    0.0000        33\n",
      "    mycobacterium     0.9806    0.6129    0.7543       248\n",
      "           others     0.1406    0.9959    0.2464       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.0845    0.1558        71\n",
      "     pelagibacter     0.0000    0.0000    0.0000        33\n",
      "       prevotella     0.0000    0.0000    0.0000        18\n",
      "  prochlorococcus     1.0000    0.2115    0.3492        52\n",
      "          proteus     0.0000    0.0000    0.0000        40\n",
      "      providencia     0.0000    0.0000    0.0000        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     1.0000    0.2554    0.4068       513\n",
      "        ralstonia     0.0000    0.0000    0.0000        42\n",
      "     rheinheimera     1.0000    0.9205    0.9586       176\n",
      "        rhizobium     0.0000    0.0000    0.0000        74\n",
      "      rhodobacter     1.0000    0.1176    0.2105        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9861    0.1123    0.2017       632\n",
      "         serratia     0.0000    0.0000    0.0000        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.5000    0.0070    0.0138       143\n",
      "   staphylococcus     1.0000    0.2500    0.4000       244\n",
      " stenotrophomonas     0.0000    0.0000    0.0000        41\n",
      "    streptococcus     0.0000    0.0000    0.0000        81\n",
      "     streptomyces     1.0000    0.0465    0.0889        43\n",
      "    synechococcus     0.9960    0.4760    0.6442       521\n",
      "           vibrio     1.0000    0.1404    0.2462       463\n",
      "      xanthomonas     0.6667    0.0345    0.0656        58\n",
      "         yersinia     1.0000    0.1290    0.2286        93\n",
      "\n",
      "         accuracy                         0.2589      8116\n",
      "        macro avg     0.4105    0.0967    0.1179      8116\n",
      "     weighted avg     0.7320    0.2589    0.2320      8116\n",
      "\n",
      "===================\n",
      "Finished\n",
      "==============\n"
     ]
    }
   ],
   "source": [
    "models = list(constants.PLM_EMBEDDINGS_CSV.keys())\n",
    "\n",
    "# len(models) - 1 to exclude benchmark model\n",
    "for i in range(len(models) - 1):\n",
    "    save_feature_importance = False\n",
    "    if models[i] == 'PROTT5':\n",
    "        save_feature_importance = True\n",
    "        \n",
    "    util.classify(models[i], save_feature_importance)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "232caf3e",
   "metadata": {},
   "source": [
    "Display training and test dataset statistics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e53eba5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructing training and test sets...\n"
     ]
    }
   ],
   "source": [
    "plm = 'PROTT5'\n",
    "\n",
    "# Load data\n",
    "rbp_embeddings = pd.read_csv(f'{constants.INPHARED}/{constants.DATA}/{constants.PLM_EMBEDDINGS_CSV[plm]}', \n",
    "                             low_memory = False)\n",
    "rbp_embeddings['Modification Date'] = pd.to_datetime(rbp_embeddings['Modification Date'])\n",
    "\n",
    "# Get only the top 25% hosts\n",
    "all_counts = rbp_embeddings['Host'].value_counts()\n",
    "TOP_X_PERCENT = 0.25\n",
    "top_x = math.floor(all_counts.shape[0] * TOP_X_PERCENT)\n",
    "\n",
    "top_genus = set()\n",
    "genus_counts = all_counts.index\n",
    "for entry in genus_counts[:top_x]:\n",
    "    top_genus.add(entry)\n",
    "\n",
    "# Construct the training and test sets\n",
    "print(\"Constructing training and test sets...\")\n",
    "\n",
    "rbp_embeddings_top = rbp_embeddings[rbp_embeddings['Host'].isin(top_genus)]\n",
    "\n",
    "counts, X_train, X_test, y_train, y_test = util.random_train_test_split(rbp_embeddings_top, 'Host',\n",
    "                                                                        embeddings_size = rbp_embeddings.shape[1] - constants.INPHARED_EXTRA_COLS)\n",
    "\n",
    "counts_df = pd.DataFrame(counts, columns = ['Genus', f'Train', f'Test', 'Total'])\n",
    "\n",
    "unknown_hosts_X, unknown_hosts_y = util.get_unknown_hosts(rbp_embeddings[~rbp_embeddings['Host'].isin(top_genus)], 'Host',\n",
    "                                                          embeddings_size = rbp_embeddings.shape[1] - constants.INPHARED_EXTRA_COLS)\n",
    "\n",
    "X_test = X_test.append(unknown_hosts_X)\n",
    "y_test = y_test.append(unknown_hosts_y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "87e50ee4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genus</th>\n",
       "      <th>Train</th>\n",
       "      <th>Test</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>escherichia</td>\n",
       "      <td>3021</td>\n",
       "      <td>1295</td>\n",
       "      <td>4316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>salmonella</td>\n",
       "      <td>1474</td>\n",
       "      <td>632</td>\n",
       "      <td>2106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>synechococcus</td>\n",
       "      <td>1216</td>\n",
       "      <td>521</td>\n",
       "      <td>1737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pseudomonas</td>\n",
       "      <td>1196</td>\n",
       "      <td>513</td>\n",
       "      <td>1709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>vibrio</td>\n",
       "      <td>1079</td>\n",
       "      <td>463</td>\n",
       "      <td>1542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>klebsiella</td>\n",
       "      <td>926</td>\n",
       "      <td>397</td>\n",
       "      <td>1323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>erwinia</td>\n",
       "      <td>667</td>\n",
       "      <td>286</td>\n",
       "      <td>953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>mycobacterium</td>\n",
       "      <td>578</td>\n",
       "      <td>248</td>\n",
       "      <td>826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>staphylococcus</td>\n",
       "      <td>568</td>\n",
       "      <td>244</td>\n",
       "      <td>812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>bacillus</td>\n",
       "      <td>475</td>\n",
       "      <td>204</td>\n",
       "      <td>679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>rheinheimera</td>\n",
       "      <td>409</td>\n",
       "      <td>176</td>\n",
       "      <td>585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>shigella</td>\n",
       "      <td>332</td>\n",
       "      <td>143</td>\n",
       "      <td>475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>enterobacter</td>\n",
       "      <td>330</td>\n",
       "      <td>142</td>\n",
       "      <td>472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>campylobacter</td>\n",
       "      <td>290</td>\n",
       "      <td>124</td>\n",
       "      <td>414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>lactococcus</td>\n",
       "      <td>233</td>\n",
       "      <td>100</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>acinetobacter</td>\n",
       "      <td>230</td>\n",
       "      <td>98</td>\n",
       "      <td>328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>serratia</td>\n",
       "      <td>226</td>\n",
       "      <td>97</td>\n",
       "      <td>323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>aeromonas</td>\n",
       "      <td>223</td>\n",
       "      <td>95</td>\n",
       "      <td>318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>yersinia</td>\n",
       "      <td>218</td>\n",
       "      <td>93</td>\n",
       "      <td>311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>streptococcus</td>\n",
       "      <td>190</td>\n",
       "      <td>81</td>\n",
       "      <td>271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>rhizobium</td>\n",
       "      <td>174</td>\n",
       "      <td>74</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>pectobacterium</td>\n",
       "      <td>167</td>\n",
       "      <td>71</td>\n",
       "      <td>238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>xanthomonas</td>\n",
       "      <td>136</td>\n",
       "      <td>58</td>\n",
       "      <td>194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>gordonia</td>\n",
       "      <td>123</td>\n",
       "      <td>53</td>\n",
       "      <td>176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>prochlorococcus</td>\n",
       "      <td>123</td>\n",
       "      <td>52</td>\n",
       "      <td>175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>dickeya</td>\n",
       "      <td>121</td>\n",
       "      <td>52</td>\n",
       "      <td>173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>enterococcus</td>\n",
       "      <td>114</td>\n",
       "      <td>49</td>\n",
       "      <td>163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>flavobacterium</td>\n",
       "      <td>113</td>\n",
       "      <td>48</td>\n",
       "      <td>161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>streptomyces</td>\n",
       "      <td>101</td>\n",
       "      <td>43</td>\n",
       "      <td>144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ralstonia</td>\n",
       "      <td>99</td>\n",
       "      <td>42</td>\n",
       "      <td>141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>stenotrophomonas</td>\n",
       "      <td>97</td>\n",
       "      <td>41</td>\n",
       "      <td>138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>proteus</td>\n",
       "      <td>92</td>\n",
       "      <td>40</td>\n",
       "      <td>132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>burkholderia</td>\n",
       "      <td>90</td>\n",
       "      <td>39</td>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>cronobacter</td>\n",
       "      <td>84</td>\n",
       "      <td>36</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>rhodobacter</td>\n",
       "      <td>79</td>\n",
       "      <td>34</td>\n",
       "      <td>113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>citrobacter</td>\n",
       "      <td>78</td>\n",
       "      <td>34</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>microbacterium</td>\n",
       "      <td>76</td>\n",
       "      <td>33</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>pelagibacter</td>\n",
       "      <td>76</td>\n",
       "      <td>33</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>bacteroides</td>\n",
       "      <td>72</td>\n",
       "      <td>31</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>listeria</td>\n",
       "      <td>67</td>\n",
       "      <td>28</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>arthrobacter</td>\n",
       "      <td>61</td>\n",
       "      <td>26</td>\n",
       "      <td>87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>caulobacter</td>\n",
       "      <td>60</td>\n",
       "      <td>25</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>clostridium</td>\n",
       "      <td>55</td>\n",
       "      <td>24</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>lactobacillus</td>\n",
       "      <td>54</td>\n",
       "      <td>23</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>prevotella</td>\n",
       "      <td>43</td>\n",
       "      <td>18</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>achromobacter</td>\n",
       "      <td>41</td>\n",
       "      <td>17</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>pseudoalteromonas</td>\n",
       "      <td>39</td>\n",
       "      <td>17</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>pantoea</td>\n",
       "      <td>38</td>\n",
       "      <td>16</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>providencia</td>\n",
       "      <td>38</td>\n",
       "      <td>16</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>helicobacter</td>\n",
       "      <td>36</td>\n",
       "      <td>16</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>rhodococcus</td>\n",
       "      <td>32</td>\n",
       "      <td>13</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>halorubrum</td>\n",
       "      <td>29</td>\n",
       "      <td>13</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>agrobacterium</td>\n",
       "      <td>29</td>\n",
       "      <td>12</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>shewanella</td>\n",
       "      <td>26</td>\n",
       "      <td>11</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>leuconostoc</td>\n",
       "      <td>24</td>\n",
       "      <td>10</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>kosakonia</td>\n",
       "      <td>24</td>\n",
       "      <td>10</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>edwardsiella</td>\n",
       "      <td>22</td>\n",
       "      <td>10</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>brevundimonas</td>\n",
       "      <td>22</td>\n",
       "      <td>10</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Genus  Train  Test  Total\n",
       "0         escherichia   3021  1295   4316\n",
       "1          salmonella   1474   632   2106\n",
       "2       synechococcus   1216   521   1737\n",
       "3         pseudomonas   1196   513   1709\n",
       "4              vibrio   1079   463   1542\n",
       "5          klebsiella    926   397   1323\n",
       "6             erwinia    667   286    953\n",
       "7       mycobacterium    578   248    826\n",
       "8      staphylococcus    568   244    812\n",
       "9            bacillus    475   204    679\n",
       "10       rheinheimera    409   176    585\n",
       "11           shigella    332   143    475\n",
       "12       enterobacter    330   142    472\n",
       "13      campylobacter    290   124    414\n",
       "14        lactococcus    233   100    333\n",
       "15      acinetobacter    230    98    328\n",
       "16           serratia    226    97    323\n",
       "17          aeromonas    223    95    318\n",
       "18           yersinia    218    93    311\n",
       "19      streptococcus    190    81    271\n",
       "20          rhizobium    174    74    248\n",
       "21     pectobacterium    167    71    238\n",
       "22        xanthomonas    136    58    194\n",
       "23           gordonia    123    53    176\n",
       "24    prochlorococcus    123    52    175\n",
       "25            dickeya    121    52    173\n",
       "26       enterococcus    114    49    163\n",
       "27     flavobacterium    113    48    161\n",
       "28       streptomyces    101    43    144\n",
       "29          ralstonia     99    42    141\n",
       "30   stenotrophomonas     97    41    138\n",
       "31            proteus     92    40    132\n",
       "32       burkholderia     90    39    129\n",
       "33        cronobacter     84    36    120\n",
       "34        rhodobacter     79    34    113\n",
       "35        citrobacter     78    34    112\n",
       "36     microbacterium     76    33    109\n",
       "37       pelagibacter     76    33    109\n",
       "38        bacteroides     72    31    103\n",
       "39           listeria     67    28     95\n",
       "40       arthrobacter     61    26     87\n",
       "41        caulobacter     60    25     85\n",
       "42        clostridium     55    24     79\n",
       "43      lactobacillus     54    23     77\n",
       "44         prevotella     43    18     61\n",
       "45      achromobacter     41    17     58\n",
       "46  pseudoalteromonas     39    17     56\n",
       "47            pantoea     38    16     54\n",
       "48        providencia     38    16     54\n",
       "49       helicobacter     36    16     52\n",
       "50        rhodococcus     32    13     45\n",
       "51         halorubrum     29    13     42\n",
       "52      agrobacterium     29    12     41\n",
       "53         shewanella     26    11     37\n",
       "54        leuconostoc     24    10     34\n",
       "55          kosakonia     24    10     34\n",
       "56       edwardsiella     22    10     32\n",
       "57      brevundimonas     22    10     32"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "86e0c98d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training Set:\t 16636\n",
      "Test Set:\t 8116\n"
     ]
    }
   ],
   "source": [
    "print('Training Set:\\t', counts_df['Train'].sum())\n",
    "print('Test Set:\\t', len(X_test))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "300ab12e",
   "metadata": {},
   "source": [
    "Investigate the extent of class imbalance by calculating the percentage of RBPs associated with the top 25% hosts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3eb1730a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9601648351648352"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_df['Total'].sum() / rbp_embeddings.shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "927d1bdf",
   "metadata": {},
   "source": [
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2e141c0",
   "metadata": {},
   "source": [
    "## Part III: Benchmarking"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca612f9f",
   "metadata": {},
   "source": [
    "Compare the performance of our model with the state-of-the-art phage-host interaction prediction tool by [Boeckaerts <i>et al.</i> (2021)](https://www.nature.com/articles/s41598-021-81063-4)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "72f76ef8",
   "metadata": {},
   "outputs": [],
   "source": [
    "rbp_embeddings = pd.read_csv(f'{constants.INPHARED}/{constants.DATA}/{constants.INPHARED_RBP_DATA}')\n",
    "rbp_embeddings['Modification Date'] = pd.to_datetime(rbp_embeddings['Modification Date'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76064481",
   "metadata": {},
   "source": [
    "The code in the cell below is taken from https://github.com/dimiboeckaerts/BacteriophageHostPrediction\n",
    "\n",
    "**Reference**:\n",
    "Boeckaerts, D., Stock, M., Criel, B., Gerstmans, H., De Baets, B., & Briers, Y. (2021). Predicting bacteriophage hosts based on sequences of annotated receptor-binding proteins. *Scientific Reports, 11*(1), 1467. doi:10.1038/s41598-021-81063-4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "95e377ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein ID</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Description</th>\n",
       "      <th>Classification</th>\n",
       "      <th>Genome Length (bp)</th>\n",
       "      <th>Jumbophage</th>\n",
       "      <th>molGC (%)</th>\n",
       "      <th>Molecule</th>\n",
       "      <th>Modification Date</th>\n",
       "      <th>Number CDS</th>\n",
       "      <th>Positive Strand (%)</th>\n",
       "      <th>Negative Strand (%)</th>\n",
       "      <th>Coding Capacity(%)</th>\n",
       "      <th>Low Coding Capacity Warning</th>\n",
       "      <th>tRNAs</th>\n",
       "      <th>Host</th>\n",
       "      <th>Lowest Taxa</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Sub-family</th>\n",
       "      <th>Family</th>\n",
       "      <th>Order</th>\n",
       "      <th>Class</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Realm</th>\n",
       "      <th>...</th>\n",
       "      <th>CTDT20</th>\n",
       "      <th>CTDT21</th>\n",
       "      <th>CTDT22</th>\n",
       "      <th>CTDT23</th>\n",
       "      <th>CTDT24</th>\n",
       "      <th>CTDT25</th>\n",
       "      <th>CTDT26</th>\n",
       "      <th>CTDT27</th>\n",
       "      <th>CTDT28</th>\n",
       "      <th>CTDT29</th>\n",
       "      <th>CTDT30</th>\n",
       "      <th>CTDT31</th>\n",
       "      <th>CTDT32</th>\n",
       "      <th>CTDT33</th>\n",
       "      <th>CTDT34</th>\n",
       "      <th>CTDT35</th>\n",
       "      <th>CTDT36</th>\n",
       "      <th>CTDT37</th>\n",
       "      <th>CTDT38</th>\n",
       "      <th>CTDT39</th>\n",
       "      <th>Z1</th>\n",
       "      <th>Z2</th>\n",
       "      <th>Z3</th>\n",
       "      <th>Z4</th>\n",
       "      <th>Z5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BAF36105.1</td>\n",
       "      <td>AB231700</td>\n",
       "      <td>Microcystis virus Ma-LMM01</td>\n",
       "      <td>Microcystis virus Ma-LMM01 Fukuivirus Caudovir...</td>\n",
       "      <td>162109</td>\n",
       "      <td>False</td>\n",
       "      <td>45.953</td>\n",
       "      <td>DNA</td>\n",
       "      <td>2021-07-14</td>\n",
       "      <td>189</td>\n",
       "      <td>34.391534</td>\n",
       "      <td>65.608466</td>\n",
       "      <td>93.542616</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>microcystis</td>\n",
       "      <td>Fukuivirus</td>\n",
       "      <td>Fukuivirus</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Caudoviricetes</td>\n",
       "      <td>Uroviricota</td>\n",
       "      <td>Heunggongvirae</td>\n",
       "      <td>Duplodnaviria</td>\n",
       "      <td>...</td>\n",
       "      <td>0.289209</td>\n",
       "      <td>0.220144</td>\n",
       "      <td>0.271942</td>\n",
       "      <td>0.185612</td>\n",
       "      <td>0.211511</td>\n",
       "      <td>0.208633</td>\n",
       "      <td>0.246043</td>\n",
       "      <td>0.205755</td>\n",
       "      <td>0.264748</td>\n",
       "      <td>0.155396</td>\n",
       "      <td>0.241727</td>\n",
       "      <td>0.155396</td>\n",
       "      <td>0.025899</td>\n",
       "      <td>0.181295</td>\n",
       "      <td>0.208633</td>\n",
       "      <td>0.235971</td>\n",
       "      <td>0.201439</td>\n",
       "      <td>0.292086</td>\n",
       "      <td>0.212950</td>\n",
       "      <td>0.145324</td>\n",
       "      <td>0.162356</td>\n",
       "      <td>-0.329641</td>\n",
       "      <td>-0.204770</td>\n",
       "      <td>-0.290445</td>\n",
       "      <td>0.273951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BAF36110.1</td>\n",
       "      <td>AB231700</td>\n",
       "      <td>Microcystis virus Ma-LMM01</td>\n",
       "      <td>Microcystis virus Ma-LMM01 Fukuivirus Caudovir...</td>\n",
       "      <td>162109</td>\n",
       "      <td>False</td>\n",
       "      <td>45.953</td>\n",
       "      <td>DNA</td>\n",
       "      <td>2021-07-14</td>\n",
       "      <td>189</td>\n",
       "      <td>34.391534</td>\n",
       "      <td>65.608466</td>\n",
       "      <td>93.542616</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>microcystis</td>\n",
       "      <td>Fukuivirus</td>\n",
       "      <td>Fukuivirus</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Caudoviricetes</td>\n",
       "      <td>Uroviricota</td>\n",
       "      <td>Heunggongvirae</td>\n",
       "      <td>Duplodnaviria</td>\n",
       "      <td>...</td>\n",
       "      <td>0.266467</td>\n",
       "      <td>0.260479</td>\n",
       "      <td>0.350299</td>\n",
       "      <td>0.134731</td>\n",
       "      <td>0.125749</td>\n",
       "      <td>0.266467</td>\n",
       "      <td>0.212575</td>\n",
       "      <td>0.212575</td>\n",
       "      <td>0.347305</td>\n",
       "      <td>0.101796</td>\n",
       "      <td>0.158683</td>\n",
       "      <td>0.104790</td>\n",
       "      <td>0.002994</td>\n",
       "      <td>0.140719</td>\n",
       "      <td>0.236527</td>\n",
       "      <td>0.182635</td>\n",
       "      <td>0.206587</td>\n",
       "      <td>0.266467</td>\n",
       "      <td>0.239521</td>\n",
       "      <td>0.146707</td>\n",
       "      <td>-0.129463</td>\n",
       "      <td>-0.544507</td>\n",
       "      <td>-0.264687</td>\n",
       "      <td>-0.483254</td>\n",
       "      <td>0.406597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BAF36131.1</td>\n",
       "      <td>AB231700</td>\n",
       "      <td>Microcystis virus Ma-LMM01</td>\n",
       "      <td>Microcystis virus Ma-LMM01 Fukuivirus Caudovir...</td>\n",
       "      <td>162109</td>\n",
       "      <td>False</td>\n",
       "      <td>45.953</td>\n",
       "      <td>DNA</td>\n",
       "      <td>2021-07-14</td>\n",
       "      <td>189</td>\n",
       "      <td>34.391534</td>\n",
       "      <td>65.608466</td>\n",
       "      <td>93.542616</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>microcystis</td>\n",
       "      <td>Fukuivirus</td>\n",
       "      <td>Fukuivirus</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Caudoviricetes</td>\n",
       "      <td>Uroviricota</td>\n",
       "      <td>Heunggongvirae</td>\n",
       "      <td>Duplodnaviria</td>\n",
       "      <td>...</td>\n",
       "      <td>0.274648</td>\n",
       "      <td>0.232394</td>\n",
       "      <td>0.380282</td>\n",
       "      <td>0.112676</td>\n",
       "      <td>0.126761</td>\n",
       "      <td>0.246479</td>\n",
       "      <td>0.204225</td>\n",
       "      <td>0.246479</td>\n",
       "      <td>0.401408</td>\n",
       "      <td>0.091549</td>\n",
       "      <td>0.147887</td>\n",
       "      <td>0.084507</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.183099</td>\n",
       "      <td>0.267606</td>\n",
       "      <td>0.260563</td>\n",
       "      <td>0.218310</td>\n",
       "      <td>0.323944</td>\n",
       "      <td>0.232394</td>\n",
       "      <td>0.112676</td>\n",
       "      <td>0.035315</td>\n",
       "      <td>-0.836783</td>\n",
       "      <td>-0.170000</td>\n",
       "      <td>-0.553846</td>\n",
       "      <td>0.415035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BAF36132.1</td>\n",
       "      <td>AB231700</td>\n",
       "      <td>Microcystis virus Ma-LMM01</td>\n",
       "      <td>Microcystis virus Ma-LMM01 Fukuivirus Caudovir...</td>\n",
       "      <td>162109</td>\n",
       "      <td>False</td>\n",
       "      <td>45.953</td>\n",
       "      <td>DNA</td>\n",
       "      <td>2021-07-14</td>\n",
       "      <td>189</td>\n",
       "      <td>34.391534</td>\n",
       "      <td>65.608466</td>\n",
       "      <td>93.542616</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>microcystis</td>\n",
       "      <td>Fukuivirus</td>\n",
       "      <td>Fukuivirus</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Caudoviricetes</td>\n",
       "      <td>Uroviricota</td>\n",
       "      <td>Heunggongvirae</td>\n",
       "      <td>Duplodnaviria</td>\n",
       "      <td>...</td>\n",
       "      <td>0.161417</td>\n",
       "      <td>0.307087</td>\n",
       "      <td>0.314961</td>\n",
       "      <td>0.153543</td>\n",
       "      <td>0.173228</td>\n",
       "      <td>0.232283</td>\n",
       "      <td>0.181102</td>\n",
       "      <td>0.228346</td>\n",
       "      <td>0.279528</td>\n",
       "      <td>0.133858</td>\n",
       "      <td>0.192913</td>\n",
       "      <td>0.102362</td>\n",
       "      <td>0.007874</td>\n",
       "      <td>0.102362</td>\n",
       "      <td>0.173228</td>\n",
       "      <td>0.216535</td>\n",
       "      <td>0.259843</td>\n",
       "      <td>0.228346</td>\n",
       "      <td>0.244094</td>\n",
       "      <td>0.173228</td>\n",
       "      <td>-0.017569</td>\n",
       "      <td>-0.509137</td>\n",
       "      <td>-0.090314</td>\n",
       "      <td>-0.423137</td>\n",
       "      <td>0.288863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BAF36193.1</td>\n",
       "      <td>AB231700</td>\n",
       "      <td>Microcystis virus Ma-LMM01</td>\n",
       "      <td>Microcystis virus Ma-LMM01 Fukuivirus Caudovir...</td>\n",
       "      <td>162109</td>\n",
       "      <td>False</td>\n",
       "      <td>45.953</td>\n",
       "      <td>DNA</td>\n",
       "      <td>2021-07-14</td>\n",
       "      <td>189</td>\n",
       "      <td>34.391534</td>\n",
       "      <td>65.608466</td>\n",
       "      <td>93.542616</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>microcystis</td>\n",
       "      <td>Fukuivirus</td>\n",
       "      <td>Fukuivirus</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Caudoviricetes</td>\n",
       "      <td>Uroviricota</td>\n",
       "      <td>Heunggongvirae</td>\n",
       "      <td>Duplodnaviria</td>\n",
       "      <td>...</td>\n",
       "      <td>0.262931</td>\n",
       "      <td>0.254310</td>\n",
       "      <td>0.317888</td>\n",
       "      <td>0.156250</td>\n",
       "      <td>0.135776</td>\n",
       "      <td>0.274784</td>\n",
       "      <td>0.214440</td>\n",
       "      <td>0.177802</td>\n",
       "      <td>0.330819</td>\n",
       "      <td>0.123922</td>\n",
       "      <td>0.168103</td>\n",
       "      <td>0.136853</td>\n",
       "      <td>0.018319</td>\n",
       "      <td>0.164871</td>\n",
       "      <td>0.226293</td>\n",
       "      <td>0.230603</td>\n",
       "      <td>0.226293</td>\n",
       "      <td>0.246767</td>\n",
       "      <td>0.262931</td>\n",
       "      <td>0.134698</td>\n",
       "      <td>-0.100990</td>\n",
       "      <td>-0.675307</td>\n",
       "      <td>-0.202917</td>\n",
       "      <td>-0.397621</td>\n",
       "      <td>0.265425</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows √ó 254 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Protein ID Accession                 Description  \\\n",
       "0  BAF36105.1  AB231700  Microcystis virus Ma-LMM01   \n",
       "1  BAF36110.1  AB231700  Microcystis virus Ma-LMM01   \n",
       "2  BAF36131.1  AB231700  Microcystis virus Ma-LMM01   \n",
       "3  BAF36132.1  AB231700  Microcystis virus Ma-LMM01   \n",
       "4  BAF36193.1  AB231700  Microcystis virus Ma-LMM01   \n",
       "\n",
       "                                      Classification  Genome Length (bp)  \\\n",
       "0  Microcystis virus Ma-LMM01 Fukuivirus Caudovir...              162109   \n",
       "1  Microcystis virus Ma-LMM01 Fukuivirus Caudovir...              162109   \n",
       "2  Microcystis virus Ma-LMM01 Fukuivirus Caudovir...              162109   \n",
       "3  Microcystis virus Ma-LMM01 Fukuivirus Caudovir...              162109   \n",
       "4  Microcystis virus Ma-LMM01 Fukuivirus Caudovir...              162109   \n",
       "\n",
       "   Jumbophage  molGC (%) Molecule Modification Date  Number CDS  \\\n",
       "0       False     45.953      DNA        2021-07-14         189   \n",
       "1       False     45.953      DNA        2021-07-14         189   \n",
       "2       False     45.953      DNA        2021-07-14         189   \n",
       "3       False     45.953      DNA        2021-07-14         189   \n",
       "4       False     45.953      DNA        2021-07-14         189   \n",
       "\n",
       "   Positive Strand (%)  Negative Strand (%)  Coding Capacity(%)  \\\n",
       "0            34.391534            65.608466           93.542616   \n",
       "1            34.391534            65.608466           93.542616   \n",
       "2            34.391534            65.608466           93.542616   \n",
       "3            34.391534            65.608466           93.542616   \n",
       "4            34.391534            65.608466           93.542616   \n",
       "\n",
       "  Low Coding Capacity Warning  tRNAs         Host Lowest Taxa       Genus  \\\n",
       "0                         NaN      2  microcystis  Fukuivirus  Fukuivirus   \n",
       "1                         NaN      2  microcystis  Fukuivirus  Fukuivirus   \n",
       "2                         NaN      2  microcystis  Fukuivirus  Fukuivirus   \n",
       "3                         NaN      2  microcystis  Fukuivirus  Fukuivirus   \n",
       "4                         NaN      2  microcystis  Fukuivirus  Fukuivirus   \n",
       "\n",
       "     Sub-family        Family         Order           Class       Phylum  \\\n",
       "0  Unclassified  Unclassified  Unclassified  Caudoviricetes  Uroviricota   \n",
       "1  Unclassified  Unclassified  Unclassified  Caudoviricetes  Uroviricota   \n",
       "2  Unclassified  Unclassified  Unclassified  Caudoviricetes  Uroviricota   \n",
       "3  Unclassified  Unclassified  Unclassified  Caudoviricetes  Uroviricota   \n",
       "4  Unclassified  Unclassified  Unclassified  Caudoviricetes  Uroviricota   \n",
       "\n",
       "          Kingdom          Realm  ...    CTDT20    CTDT21    CTDT22    CTDT23  \\\n",
       "0  Heunggongvirae  Duplodnaviria  ...  0.289209  0.220144  0.271942  0.185612   \n",
       "1  Heunggongvirae  Duplodnaviria  ...  0.266467  0.260479  0.350299  0.134731   \n",
       "2  Heunggongvirae  Duplodnaviria  ...  0.274648  0.232394  0.380282  0.112676   \n",
       "3  Heunggongvirae  Duplodnaviria  ...  0.161417  0.307087  0.314961  0.153543   \n",
       "4  Heunggongvirae  Duplodnaviria  ...  0.262931  0.254310  0.317888  0.156250   \n",
       "\n",
       "     CTDT24    CTDT25    CTDT26    CTDT27    CTDT28    CTDT29    CTDT30  \\\n",
       "0  0.211511  0.208633  0.246043  0.205755  0.264748  0.155396  0.241727   \n",
       "1  0.125749  0.266467  0.212575  0.212575  0.347305  0.101796  0.158683   \n",
       "2  0.126761  0.246479  0.204225  0.246479  0.401408  0.091549  0.147887   \n",
       "3  0.173228  0.232283  0.181102  0.228346  0.279528  0.133858  0.192913   \n",
       "4  0.135776  0.274784  0.214440  0.177802  0.330819  0.123922  0.168103   \n",
       "\n",
       "     CTDT31    CTDT32    CTDT33    CTDT34    CTDT35    CTDT36    CTDT37  \\\n",
       "0  0.155396  0.025899  0.181295  0.208633  0.235971  0.201439  0.292086   \n",
       "1  0.104790  0.002994  0.140719  0.236527  0.182635  0.206587  0.266467   \n",
       "2  0.084507  0.000000  0.183099  0.267606  0.260563  0.218310  0.323944   \n",
       "3  0.102362  0.007874  0.102362  0.173228  0.216535  0.259843  0.228346   \n",
       "4  0.136853  0.018319  0.164871  0.226293  0.230603  0.226293  0.246767   \n",
       "\n",
       "     CTDT38    CTDT39        Z1        Z2        Z3        Z4        Z5  \n",
       "0  0.212950  0.145324  0.162356 -0.329641 -0.204770 -0.290445  0.273951  \n",
       "1  0.239521  0.146707 -0.129463 -0.544507 -0.264687 -0.483254  0.406597  \n",
       "2  0.232394  0.112676  0.035315 -0.836783 -0.170000 -0.553846  0.415035  \n",
       "3  0.244094  0.173228 -0.017569 -0.509137 -0.090314 -0.423137  0.288863  \n",
       "4  0.262931  0.134698 -0.100990 -0.675307 -0.202917 -0.397621  0.265425  \n",
       "\n",
       "[5 rows x 254 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_list = list(rbp_embeddings.iloc[:,-1])\n",
    "dna_feats = RBP_f.dna_features(dna_list)\n",
    "\n",
    "protein_list = list(rbp_embeddings.iloc[:,-2])\n",
    "protein_feats = RBP_f.protein_features(protein_list)\n",
    "\n",
    "# protein features: CTD & Z-scale\n",
    "extra_feats = np.zeros((rbp_embeddings.shape[0], 47))\n",
    "\n",
    "for i,item in enumerate(protein_list):\n",
    "    feature_lst = []\n",
    "    feature_lst  += RBP_f.CTDC(item)\n",
    "    feature_lst += RBP_f.CTDT(item)\n",
    "    feature_lst += RBP_f.zscale(item)\n",
    "    extra_feats[i,:] = feature_lst\n",
    "    \n",
    "extra_feats_df = pd.DataFrame(extra_feats, columns=['CTDC1', 'CTDC2', 'CTDC3', 'CTDT1', 'CTDT2', 'CTDT3', \n",
    "                        'CTDT4', 'CTDT5', 'CTDT6', 'CTDT7', 'CTDT8', 'CTDT9', 'CTDT10', 'CTDT11', 'CTDT12', 'CTDT13', \n",
    "                        'CTDT14', 'CTDT15', 'CTDT16', 'CTDT17', 'CTDT18', 'CTDT19', 'CTDT20', 'CTDT21', 'CTDT22',\n",
    "                        'CTDT23', 'CTDT24', 'CTDT25', 'CTDT26', 'CTDT27', 'CTDT28', 'CTDT29', 'CTDT30', 'CTDT31', \n",
    "                        'CTDT32', 'CTDT33', 'CTDT34', 'CTDT35', 'CTDT36', 'CTDT37', 'CTDT38', 'CTDT39', 'Z1', 'Z2',\n",
    "                        'Z3', 'Z4', 'Z5'])\n",
    "\n",
    "# all features, not splitted\n",
    "features = pd.concat([dna_feats, protein_feats, extra_feats_df], axis=1)\n",
    "rbp_embeddings = pd.concat([rbp_embeddings, features], axis = 1)\n",
    "rbp_embeddings.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c533cde5",
   "metadata": {},
   "source": [
    "Generate the phage-host-features CSV file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a08f0e8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "rbp_embeddings.to_csv(os.path.join(f'{constants.INPHARED}/{constants.DATA}', constants.PLM_EMBEDDINGS_CSV['BOECKAERTS']), \n",
    "                      index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "230318e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "rbp_embeddings = pd.read_csv(f'{constants.INPHARED}/{constants.DATA}/{constants.PLM_EMBEDDINGS_CSV[\"BOECKAERTS\"]}',\n",
    "                             low_memory = False)\n",
    "features = rbp_embeddings.columns\n",
    "features = features[constants.INPHARED_EXTRA_COLS:]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1060c22",
   "metadata": {},
   "source": [
    "Train the model by [Boeckaerts <i>et al.</i> (2021)](https://www.nature.com/articles/s41598-021-81063-4) on our dataset, and display the results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3d6cbef9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Confidence threshold k: 0.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.3529    0.5217        17\n",
      "    acinetobacter     0.9839    0.6224    0.7625        98\n",
      "        aeromonas     0.9444    0.5368    0.6846        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.5000    0.6667        26\n",
      "         bacillus     0.7155    0.8382    0.7720       204\n",
      "      bacteroides     1.0000    0.9355    0.9667        31\n",
      "    brevundimonas     1.0000    0.5000    0.6667        10\n",
      "     burkholderia     0.7188    0.5897    0.6479        39\n",
      "    campylobacter     0.8240    0.8306    0.8273       124\n",
      "      caulobacter     0.8400    0.8400    0.8400        25\n",
      "      citrobacter     0.5000    0.2647    0.3462        34\n",
      "      clostridium     0.4000    0.5833    0.4746        24\n",
      "      cronobacter     0.7368    0.3889    0.5091        36\n",
      "          dickeya     0.9500    0.7308    0.8261        52\n",
      "     edwardsiella     1.0000    0.4000    0.5714        10\n",
      "     enterobacter     0.5049    0.3662    0.4245       142\n",
      "     enterococcus     0.9556    0.8776    0.9149        49\n",
      "          erwinia     0.7926    0.8951    0.8407       286\n",
      "      escherichia     0.5178    0.8857    0.6536      1295\n",
      "   flavobacterium     0.9375    0.9375    0.9375        48\n",
      "         gordonia     0.6885    0.7925    0.7368        53\n",
      "       halorubrum     0.3333    0.6923    0.4500        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.7708    0.7708    0.7708       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.7500    0.2609    0.3871        23\n",
      "      lactococcus     0.9149    0.8600    0.8866       100\n",
      "      leuconostoc     1.0000    0.9000    0.9474        10\n",
      "         listeria     0.9231    0.8571    0.8889        28\n",
      "   microbacterium     0.8519    0.6970    0.7667        33\n",
      "    mycobacterium     0.8962    0.9395    0.9173       248\n",
      "           others     0.0000    0.0000    0.0000       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.8776    0.6056    0.7167        71\n",
      "     pelagibacter     0.6818    0.9091    0.7792        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.8250    0.6346    0.7174        52\n",
      "          proteus     0.8214    0.5750    0.6765        40\n",
      "      providencia     0.6667    0.2500    0.3636        16\n",
      "pseudoalteromonas     1.0000    0.1765    0.3000        17\n",
      "      pseudomonas     0.5624    0.9318    0.7014       513\n",
      "        ralstonia     0.9444    0.4048    0.5667        42\n",
      "     rheinheimera     1.0000    1.0000    1.0000       176\n",
      "        rhizobium     0.8485    0.7568    0.8000        74\n",
      "      rhodobacter     0.9091    0.8824    0.8955        34\n",
      "      rhodococcus     1.0000    0.1538    0.2667        13\n",
      "       salmonella     0.8873    0.7722    0.8257       632\n",
      "         serratia     0.8358    0.5773    0.6829        97\n",
      "       shewanella     1.0000    0.4545    0.6250        11\n",
      "         shigella     0.5152    0.2378    0.3254       143\n",
      "   staphylococcus     0.7047    0.9877    0.8225       244\n",
      " stenotrophomonas     0.9048    0.4634    0.6129        41\n",
      "    streptococcus     0.8889    0.7901    0.8366        81\n",
      "     streptomyces     0.8095    0.7907    0.8000        43\n",
      "    synechococcus     0.8435    0.9520    0.8945       521\n",
      "           vibrio     0.7640    0.8812    0.8185       463\n",
      "      xanthomonas     0.8286    0.5000    0.6237        58\n",
      "         yersinia     0.6705    0.6344    0.6519        93\n",
      "\n",
      "         accuracy                         0.7033      8116\n",
      "        macro avg     0.7837    0.6313    0.6675      8116\n",
      "     weighted avg     0.6498    0.7033    0.6564      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 10.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2941    0.4545        17\n",
      "    acinetobacter     1.0000    0.5000    0.6667        98\n",
      "        aeromonas     0.9545    0.4421    0.6043        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.4231    0.5946        26\n",
      "         bacillus     0.8901    0.7941    0.8394       204\n",
      "      bacteroides     1.0000    0.9355    0.9667        31\n",
      "    brevundimonas     1.0000    0.5000    0.6667        10\n",
      "     burkholderia     0.9000    0.4615    0.6102        39\n",
      "    campylobacter     0.9189    0.8226    0.8681       124\n",
      "      caulobacter     1.0000    0.8000    0.8889        25\n",
      "      citrobacter     0.6154    0.2353    0.3404        34\n",
      "      clostridium     0.4194    0.5417    0.4727        24\n",
      "      cronobacter     0.7500    0.3333    0.4615        36\n",
      "          dickeya     1.0000    0.6731    0.8046        52\n",
      "     edwardsiella     1.0000    0.2000    0.3333        10\n",
      "     enterobacter     0.5769    0.3169    0.4091       142\n",
      "     enterococcus     0.9545    0.8571    0.9032        49\n",
      "          erwinia     0.9438    0.8217    0.8785       286\n",
      "      escherichia     0.7306    0.8124    0.7693      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     0.9474    0.6792    0.7912        53\n",
      "       halorubrum     0.3478    0.6154    0.4444        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.9223    0.6877    0.7879       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.7500    0.1304    0.2222        23\n",
      "      lactococcus     0.9878    0.8100    0.8901       100\n",
      "      leuconostoc     1.0000    0.8000    0.8889        10\n",
      "         listeria     0.9200    0.8214    0.8679        28\n",
      "   microbacterium     1.0000    0.6667    0.8000        33\n",
      "    mycobacterium     0.9830    0.9315    0.9565       248\n",
      "           others     0.3699    0.7008    0.4842       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9318    0.5775    0.7130        71\n",
      "     pelagibacter     0.8750    0.8485    0.8615        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9412    0.6154    0.7442        52\n",
      "          proteus     0.8333    0.5000    0.6250        40\n",
      "      providencia     0.6000    0.1875    0.2857        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.7607    0.8986    0.8239       513\n",
      "        ralstonia     1.0000    0.4048    0.5763        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     0.9608    0.6622    0.7840        74\n",
      "      rhodobacter     0.9310    0.7941    0.8571        34\n",
      "      rhodococcus     1.0000    0.0769    0.1429        13\n",
      "       salmonella     0.9248    0.7389    0.8215       632\n",
      "         serratia     0.8393    0.4845    0.6144        97\n",
      "       shewanella     1.0000    0.0909    0.1667        11\n",
      "         shigella     0.5283    0.1958    0.2857       143\n",
      "   staphylococcus     0.8074    0.9795    0.8852       244\n",
      " stenotrophomonas     0.9231    0.2927    0.4444        41\n",
      "    streptococcus     0.9492    0.6914    0.8000        81\n",
      "     streptomyces     0.8889    0.7442    0.8101        43\n",
      "    synechococcus     0.9538    0.9117    0.9323       521\n",
      "           vibrio     0.9167    0.7840    0.8452       463\n",
      "      xanthomonas     0.8667    0.4483    0.5909        58\n",
      "         yersinia     0.6627    0.5914    0.6250        93\n",
      "\n",
      "         accuracy                         0.7395      8116\n",
      "        macro avg     0.8471    0.5840    0.6534      8116\n",
      "     weighted avg     0.7965    0.7395    0.7447      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 20.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2941    0.4545        17\n",
      "    acinetobacter     1.0000    0.4388    0.6099        98\n",
      "        aeromonas     0.9487    0.3895    0.5522        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.3846    0.5556        26\n",
      "         bacillus     0.9346    0.7010    0.8011       204\n",
      "      bacteroides     1.0000    0.8710    0.9310        31\n",
      "    brevundimonas     1.0000    0.5000    0.6667        10\n",
      "     burkholderia     0.9375    0.3846    0.5455        39\n",
      "    campylobacter     0.9510    0.7823    0.8584       124\n",
      "      caulobacter     1.0000    0.7600    0.8636        25\n",
      "      citrobacter     0.6364    0.2059    0.3111        34\n",
      "      clostridium     0.5000    0.5000    0.5000        24\n",
      "      cronobacter     0.8182    0.2500    0.3830        36\n",
      "          dickeya     1.0000    0.6731    0.8046        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.6032    0.2676    0.3707       142\n",
      "     enterococcus     0.9744    0.7755    0.8636        49\n",
      "          erwinia     0.9595    0.7448    0.8386       286\n",
      "      escherichia     0.8272    0.7243    0.7723      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.5849    0.7381        53\n",
      "       halorubrum     0.3158    0.4615    0.3750        13\n",
      "     helicobacter     1.0000    1.0000    1.0000        16\n",
      "       klebsiella     0.9476    0.6373    0.7620       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.7500    0.1304    0.2222        23\n",
      "      lactococcus     1.0000    0.7900    0.8827       100\n",
      "      leuconostoc     1.0000    0.7000    0.8235        10\n",
      "         listeria     0.9167    0.7857    0.8462        28\n",
      "   microbacterium     1.0000    0.6364    0.7778        33\n",
      "    mycobacterium     0.9828    0.9234    0.9522       248\n",
      "           others     0.3207    0.8692    0.4686       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9318    0.5775    0.7130        71\n",
      "     pelagibacter     1.0000    0.7273    0.8421        33\n",
      "       prevotella     1.0000    0.9444    0.9714        18\n",
      "  prochlorococcus     0.9355    0.5577    0.6988        52\n",
      "          proteus     0.8261    0.4750    0.6032        40\n",
      "      providencia     0.6000    0.1875    0.2857        16\n",
      "pseudoalteromonas     1.0000    0.1176    0.2105        17\n",
      "      pseudomonas     0.9039    0.8616    0.8822       513\n",
      "        ralstonia     1.0000    0.4048    0.5763        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     0.9792    0.6351    0.7705        74\n",
      "      rhodobacter     0.9286    0.7647    0.8387        34\n",
      "      rhodococcus     1.0000    0.0769    0.1429        13\n",
      "       salmonella     0.9375    0.7120    0.8094       632\n",
      "         serratia     0.8654    0.4639    0.6040        97\n",
      "       shewanella     1.0000    0.0909    0.1667        11\n",
      "         shigella     0.5349    0.1608    0.2473       143\n",
      "   staphylococcus     0.9011    0.9713    0.9349       244\n",
      " stenotrophomonas     0.8889    0.1951    0.3200        41\n",
      "    streptococcus     0.9623    0.6296    0.7612        81\n",
      "     streptomyces     0.9688    0.7209    0.8267        43\n",
      "    synechococcus     0.9808    0.8829    0.9293       521\n",
      "           vibrio     0.9598    0.6695    0.7888       463\n",
      "      xanthomonas     0.8889    0.4138    0.5647        58\n",
      "         yersinia     0.6795    0.5699    0.6199        93\n",
      "\n",
      "         accuracy                         0.7148      8116\n",
      "        macro avg     0.8644    0.5482    0.6320      8116\n",
      "     weighted avg     0.8299    0.7148    0.7333      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 30.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.2353    0.3810        17\n",
      "    acinetobacter     1.0000    0.3878    0.5588        98\n",
      "        aeromonas     0.9375    0.3158    0.4724        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.3462    0.5143        26\n",
      "         bacillus     0.9500    0.6520    0.7733       204\n",
      "      bacteroides     1.0000    0.8387    0.9123        31\n",
      "    brevundimonas     1.0000    0.4000    0.5714        10\n",
      "     burkholderia     1.0000    0.2051    0.3404        39\n",
      "    campylobacter     0.9500    0.7661    0.8482       124\n",
      "      caulobacter     1.0000    0.7200    0.8372        25\n",
      "      citrobacter     0.6667    0.1765    0.2791        34\n",
      "      clostridium     0.5882    0.4167    0.4878        24\n",
      "      cronobacter     0.7143    0.1389    0.2326        36\n",
      "          dickeya     1.0000    0.6346    0.7765        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.7021    0.2324    0.3492       142\n",
      "     enterococcus     0.9722    0.7143    0.8235        49\n",
      "          erwinia     0.9652    0.6783    0.7967       286\n",
      "      escherichia     0.8666    0.6371    0.7343      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.4151    0.5867        53\n",
      "       halorubrum     0.4000    0.4615    0.4286        13\n",
      "     helicobacter     1.0000    0.8750    0.9333        16\n",
      "       klebsiella     0.9553    0.5919    0.7309       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     1.0000    0.0870    0.1600        23\n",
      "      lactococcus     1.0000    0.7700    0.8701       100\n",
      "      leuconostoc     1.0000    0.6000    0.7500        10\n",
      "         listeria     0.9130    0.7500    0.8235        28\n",
      "   microbacterium     1.0000    0.6364    0.7778        33\n",
      "    mycobacterium     0.9870    0.9153    0.9498       248\n",
      "           others     0.2911    0.9432    0.4449       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9286    0.5493    0.6903        71\n",
      "     pelagibacter     1.0000    0.6667    0.8000        33\n",
      "       prevotella     1.0000    0.8889    0.9412        18\n",
      "  prochlorococcus     0.9310    0.5192    0.6667        52\n",
      "          proteus     0.8235    0.3500    0.4912        40\n",
      "      providencia     0.5000    0.1250    0.2000        16\n",
      "pseudoalteromonas     1.0000    0.0588    0.1111        17\n",
      "      pseudomonas     0.9817    0.8382    0.9043       513\n",
      "        ralstonia     1.0000    0.3810    0.5517        42\n",
      "     rheinheimera     1.0000    0.9943    0.9972       176\n",
      "        rhizobium     1.0000    0.6081    0.7563        74\n",
      "      rhodobacter     0.9259    0.7353    0.8197        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9457    0.6883    0.7967       632\n",
      "         serratia     0.8600    0.4433    0.5850        97\n",
      "       shewanella     1.0000    0.0909    0.1667        11\n",
      "         shigella     0.5526    0.1469    0.2320       143\n",
      "   staphylococcus     0.9592    0.9631    0.9611       244\n",
      " stenotrophomonas     0.8750    0.1707    0.2857        41\n",
      "    streptococcus     1.0000    0.5926    0.7442        81\n",
      "     streptomyces     1.0000    0.6512    0.7887        43\n",
      "    synechococcus     0.9867    0.8541    0.9156       521\n",
      "           vibrio     0.9929    0.6026    0.7500       463\n",
      "      xanthomonas     0.9500    0.3276    0.4872        58\n",
      "         yersinia     0.7042    0.5376    0.6098        93\n",
      "\n",
      "         accuracy                         0.6826      8116\n",
      "        macro avg     0.8606    0.5033    0.5940      8116\n",
      "     weighted avg     0.8457    0.6826    0.7084      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 40.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.1765    0.3000        17\n",
      "    acinetobacter     1.0000    0.2857    0.4444        98\n",
      "        aeromonas     0.9630    0.2737    0.4262        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.3462    0.5143        26\n",
      "         bacillus     0.9531    0.5980    0.7349       204\n",
      "      bacteroides     1.0000    0.8065    0.8929        31\n",
      "    brevundimonas     1.0000    0.4000    0.5714        10\n",
      "     burkholderia     1.0000    0.1282    0.2273        39\n",
      "    campylobacter     0.9588    0.7500    0.8416       124\n",
      "      caulobacter     1.0000    0.6400    0.7805        25\n",
      "      citrobacter     0.6000    0.0882    0.1538        34\n",
      "      clostridium     0.6154    0.3333    0.4324        24\n",
      "      cronobacter     0.8333    0.1389    0.2381        36\n",
      "          dickeya     1.0000    0.6154    0.7619        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.7222    0.1831    0.2921       142\n",
      "     enterococcus     0.9697    0.6531    0.7805        49\n",
      "          erwinia     0.9718    0.6014    0.7430       286\n",
      "      escherichia     0.8850    0.5529    0.6806      1295\n",
      "   flavobacterium     1.0000    0.9375    0.9677        48\n",
      "         gordonia     1.0000    0.3585    0.5278        53\n",
      "       halorubrum     0.3571    0.3846    0.3704        13\n",
      "     helicobacter     1.0000    0.8750    0.9333        16\n",
      "       klebsiella     0.9621    0.5113    0.6678       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     1.0000    0.0435    0.0833        23\n",
      "      lactococcus     1.0000    0.7200    0.8372       100\n",
      "      leuconostoc     1.0000    0.6000    0.7500        10\n",
      "         listeria     0.9130    0.7500    0.8235        28\n",
      "   microbacterium     1.0000    0.6364    0.7778        33\n",
      "    mycobacterium     0.9868    0.9032    0.9432       248\n",
      "           others     0.2615    0.9604    0.4110       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9231    0.5070    0.6545        71\n",
      "     pelagibacter     1.0000    0.5455    0.7059        33\n",
      "       prevotella     1.0000    0.8889    0.9412        18\n",
      "  prochlorococcus     0.9259    0.4808    0.6329        52\n",
      "          proteus     0.8667    0.3250    0.4727        40\n",
      "      providencia     0.5000    0.0625    0.1111        16\n",
      "pseudoalteromonas     1.0000    0.0588    0.1111        17\n",
      "      pseudomonas     0.9928    0.8012    0.8867       513\n",
      "        ralstonia     1.0000    0.3810    0.5517        42\n",
      "     rheinheimera     1.0000    0.9886    0.9943       176\n",
      "        rhizobium     1.0000    0.5541    0.7130        74\n",
      "      rhodobacter     0.9259    0.7353    0.8197        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9498    0.6582    0.7776       632\n",
      "         serratia     0.8913    0.4227    0.5734        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.6207    0.1259    0.2093       143\n",
      "   staphylococcus     0.9912    0.9262    0.9576       244\n",
      " stenotrophomonas     0.8750    0.1707    0.2857        41\n",
      "    streptococcus     1.0000    0.5556    0.7143        81\n",
      "     streptomyces     1.0000    0.5581    0.7164        43\n",
      "    synechococcus     0.9908    0.8253    0.9005       521\n",
      "           vibrio     1.0000    0.5486    0.7085       463\n",
      "      xanthomonas     0.9412    0.2759    0.4267        58\n",
      "         yersinia     0.7101    0.5269    0.6049        93\n",
      "\n",
      "         accuracy                         0.6421      8116\n",
      "        macro avg     0.8484    0.4679    0.5604      8116\n",
      "     weighted avg     0.8497    0.6421    0.6739      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 50.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     1.0000    0.0588    0.1111        17\n",
      "    acinetobacter     1.0000    0.2755    0.4320        98\n",
      "        aeromonas     1.0000    0.2526    0.4034        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.3077    0.4706        26\n",
      "         bacillus     0.9655    0.5490    0.7000       204\n",
      "      bacteroides     1.0000    0.6774    0.8077        31\n",
      "    brevundimonas     1.0000    0.2000    0.3333        10\n",
      "     burkholderia     1.0000    0.1282    0.2273        39\n",
      "    campylobacter     0.9775    0.7016    0.8169       124\n",
      "      caulobacter     1.0000    0.6000    0.7500        25\n",
      "      citrobacter     0.7500    0.0882    0.1579        34\n",
      "      clostridium     0.6000    0.2500    0.3529        24\n",
      "      cronobacter     0.8000    0.1111    0.1951        36\n",
      "          dickeya     1.0000    0.5000    0.6667        52\n",
      "     edwardsiella     1.0000    0.1000    0.1818        10\n",
      "     enterobacter     0.7500    0.1479    0.2471       142\n",
      "     enterococcus     1.0000    0.5714    0.7273        49\n",
      "          erwinia     0.9929    0.4860    0.6526       286\n",
      "      escherichia     0.9000    0.4865    0.6316      1295\n",
      "   flavobacterium     1.0000    0.8958    0.9451        48\n",
      "         gordonia     1.0000    0.3208    0.4857        53\n",
      "       halorubrum     0.5000    0.3077    0.3810        13\n",
      "     helicobacter     1.0000    0.8125    0.8966        16\n",
      "       klebsiella     0.9727    0.4484    0.6138       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     1.0000    0.0435    0.0833        23\n",
      "      lactococcus     1.0000    0.6900    0.8166       100\n",
      "      leuconostoc     1.0000    0.3000    0.4615        10\n",
      "         listeria     0.9474    0.6429    0.7660        28\n",
      "   microbacterium     1.0000    0.5455    0.7059        33\n",
      "    mycobacterium     0.9867    0.8952    0.9387       248\n",
      "           others     0.2385    0.9736    0.3832       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9714    0.4789    0.6415        71\n",
      "     pelagibacter     1.0000    0.4242    0.5957        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     0.9200    0.4423    0.5974        52\n",
      "          proteus     1.0000    0.3000    0.4615        40\n",
      "      providencia     0.5000    0.0625    0.1111        16\n",
      "pseudoalteromonas     1.0000    0.0588    0.1111        17\n",
      "      pseudomonas     0.9949    0.7544    0.8581       513\n",
      "        ralstonia     1.0000    0.3571    0.5263        42\n",
      "     rheinheimera     1.0000    0.9886    0.9943       176\n",
      "        rhizobium     1.0000    0.4459    0.6168        74\n",
      "      rhodobacter     0.9231    0.7059    0.8000        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9520    0.5965    0.7335       632\n",
      "         serratia     0.8780    0.3711    0.5217        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.7500    0.1049    0.1840       143\n",
      "   staphylococcus     1.0000    0.8975    0.9460       244\n",
      " stenotrophomonas     1.0000    0.1463    0.2553        41\n",
      "    streptococcus     1.0000    0.5185    0.6829        81\n",
      "     streptomyces     1.0000    0.5116    0.6769        43\n",
      "    synechococcus     0.9930    0.8138    0.8945       521\n",
      "           vibrio     1.0000    0.5270    0.6902       463\n",
      "      xanthomonas     0.9333    0.2414    0.3836        58\n",
      "         yersinia     0.7231    0.5054    0.5949        93\n",
      "\n",
      "         accuracy                         0.6024      8116\n",
      "        macro avg     0.8630    0.4201    0.5191      8116\n",
      "     weighted avg     0.8576    0.6024    0.6389      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 60.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.1837    0.3103        98\n",
      "        aeromonas     1.0000    0.2211    0.3621        95\n",
      "    agrobacterium     1.0000    0.3333    0.5000        12\n",
      "     arthrobacter     1.0000    0.2308    0.3750        26\n",
      "         bacillus     0.9691    0.4608    0.6246       204\n",
      "      bacteroides     1.0000    0.4839    0.6522        31\n",
      "    brevundimonas     1.0000    0.1000    0.1818        10\n",
      "     burkholderia     1.0000    0.1026    0.1860        39\n",
      "    campylobacter     0.9762    0.6613    0.7885       124\n",
      "      caulobacter     1.0000    0.6000    0.7500        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.6250    0.2083    0.3125        24\n",
      "      cronobacter     1.0000    0.0278    0.0541        36\n",
      "          dickeya     1.0000    0.4423    0.6133        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.8571    0.1268    0.2209       142\n",
      "     enterococcus     1.0000    0.5102    0.6757        49\n",
      "          erwinia     0.9921    0.4371    0.6068       286\n",
      "      escherichia     0.9227    0.4147    0.5722      1295\n",
      "   flavobacterium     1.0000    0.8958    0.9451        48\n",
      "         gordonia     1.0000    0.2642    0.4179        53\n",
      "       halorubrum     0.5000    0.3077    0.3810        13\n",
      "     helicobacter     1.0000    0.6875    0.8148        16\n",
      "       klebsiella     0.9669    0.3678    0.5328       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.6500    0.7879       100\n",
      "      leuconostoc     1.0000    0.2000    0.3333        10\n",
      "         listeria     0.9444    0.6071    0.7391        28\n",
      "   microbacterium     1.0000    0.4545    0.6250        33\n",
      "    mycobacterium     0.9863    0.8710    0.9251       248\n",
      "           others     0.2160    0.9797    0.3539       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9688    0.4366    0.6019        71\n",
      "     pelagibacter     1.0000    0.3333    0.5000        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     1.0000    0.3654    0.5352        52\n",
      "          proteus     1.0000    0.2500    0.4000        40\n",
      "      providencia     1.0000    0.0625    0.1176        16\n",
      "pseudoalteromonas     1.0000    0.0588    0.1111        17\n",
      "      pseudomonas     0.9945    0.7057    0.8255       513\n",
      "        ralstonia     1.0000    0.3333    0.5000        42\n",
      "     rheinheimera     1.0000    0.9830    0.9914       176\n",
      "        rhizobium     1.0000    0.3919    0.5631        74\n",
      "      rhodobacter     0.9583    0.6765    0.7931        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9507    0.5190    0.6714       632\n",
      "         serratia     0.8571    0.3093    0.4545        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.7222    0.0909    0.1615       143\n",
      "   staphylococcus     1.0000    0.8689    0.9298       244\n",
      " stenotrophomonas     1.0000    0.1463    0.2553        41\n",
      "    streptococcus     1.0000    0.4321    0.6034        81\n",
      "     streptomyces     1.0000    0.4419    0.6129        43\n",
      "    synechococcus     0.9951    0.7831    0.8765       521\n",
      "           vibrio     1.0000    0.4708    0.6402       463\n",
      "      xanthomonas     0.9286    0.2241    0.3611        58\n",
      "         yersinia     0.9750    0.4194    0.5865        93\n",
      "\n",
      "         accuracy                         0.5548      8116\n",
      "        macro avg     0.8187    0.3723    0.4685      8116\n",
      "     weighted avg     0.8555    0.5548    0.5935      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 70.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.1633    0.2807        98\n",
      "        aeromonas     1.0000    0.1474    0.2569        95\n",
      "    agrobacterium     1.0000    0.2500    0.4000        12\n",
      "     arthrobacter     1.0000    0.1923    0.3226        26\n",
      "         bacillus     0.9718    0.3382    0.5018       204\n",
      "      bacteroides     1.0000    0.4194    0.5909        31\n",
      "    brevundimonas     1.0000    0.1000    0.1818        10\n",
      "     burkholderia     1.0000    0.1026    0.1860        39\n",
      "    campylobacter     0.9722    0.5645    0.7143       124\n",
      "      caulobacter     1.0000    0.6000    0.7500        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.7500    0.1250    0.2143        24\n",
      "      cronobacter     1.0000    0.0278    0.0541        36\n",
      "          dickeya     1.0000    0.4231    0.5946        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.7692    0.0704    0.1290       142\n",
      "     enterococcus     1.0000    0.4490    0.6197        49\n",
      "          erwinia     0.9896    0.3322    0.4974       286\n",
      "      escherichia     0.9284    0.3405    0.4983      1295\n",
      "   flavobacterium     1.0000    0.8542    0.9213        48\n",
      "         gordonia     1.0000    0.2075    0.3438        53\n",
      "       halorubrum     0.3333    0.1538    0.2105        13\n",
      "     helicobacter     1.0000    0.4375    0.6087        16\n",
      "       klebsiella     0.9737    0.2796    0.4344       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.5300    0.6928       100\n",
      "      leuconostoc     1.0000    0.1000    0.1818        10\n",
      "         listeria     0.9333    0.5000    0.6512        28\n",
      "   microbacterium     1.0000    0.3333    0.5000        33\n",
      "    mycobacterium     0.9862    0.8669    0.9227       248\n",
      "           others     0.1960    0.9858    0.3269       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     0.9655    0.3944    0.5600        71\n",
      "     pelagibacter     1.0000    0.2727    0.4286        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     1.0000    0.2500    0.4000        52\n",
      "          proteus     1.0000    0.1250    0.2222        40\n",
      "      providencia     0.0000    0.0000    0.0000        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     0.9970    0.6452    0.7834       513\n",
      "        ralstonia     1.0000    0.2619    0.4151        42\n",
      "     rheinheimera     1.0000    0.9773    0.9885       176\n",
      "        rhizobium     1.0000    0.2703    0.4255        74\n",
      "      rhodobacter     1.0000    0.6471    0.7857        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9550    0.4367    0.5993       632\n",
      "         serratia     0.8571    0.2474    0.3840        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.7059    0.0839    0.1500       143\n",
      "   staphylococcus     1.0000    0.8484    0.9180       244\n",
      " stenotrophomonas     1.0000    0.0976    0.1778        41\n",
      "    streptococcus     1.0000    0.3457    0.5138        81\n",
      "     streptomyces     1.0000    0.2791    0.4364        43\n",
      "    synechococcus     0.9948    0.7313    0.8429       521\n",
      "           vibrio     1.0000    0.4082    0.5798       463\n",
      "      xanthomonas     0.9000    0.1552    0.2647        58\n",
      "         yersinia     0.9722    0.3763    0.5426        93\n",
      "\n",
      "         accuracy                         0.4984      8116\n",
      "        macro avg     0.7822    0.3149    0.4053      8116\n",
      "     weighted avg     0.8488    0.4984    0.5341      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 80.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.0612    0.1154        98\n",
      "        aeromonas     1.0000    0.0737    0.1373        95\n",
      "    agrobacterium     1.0000    0.1667    0.2857        12\n",
      "     arthrobacter     1.0000    0.1923    0.3226        26\n",
      "         bacillus     0.9783    0.2206    0.3600       204\n",
      "      bacteroides     1.0000    0.3871    0.5581        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     1.0000    0.0513    0.0976        39\n",
      "    campylobacter     0.9692    0.5081    0.6667       124\n",
      "      caulobacter     1.0000    0.5200    0.6842        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.6667    0.0833    0.1481        24\n",
      "      cronobacter     1.0000    0.0278    0.0541        36\n",
      "          dickeya     1.0000    0.4038    0.5753        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.5714    0.0282    0.0537       142\n",
      "     enterococcus     1.0000    0.4082    0.5797        49\n",
      "          erwinia     0.9863    0.2517    0.4011       286\n",
      "      escherichia     0.9460    0.2571    0.4044      1295\n",
      "   flavobacterium     1.0000    0.8333    0.9091        48\n",
      "         gordonia     1.0000    0.1509    0.2623        53\n",
      "       halorubrum     0.2500    0.0769    0.1176        13\n",
      "     helicobacter     1.0000    0.2500    0.4000        16\n",
      "       klebsiella     0.9753    0.1990    0.3305       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.4600    0.6301       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.4643    0.6341        28\n",
      "   microbacterium     1.0000    0.2727    0.4286        33\n",
      "    mycobacterium     0.9859    0.8468    0.9111       248\n",
      "           others     0.1799    0.9899    0.3045       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.3803    0.5510        71\n",
      "     pelagibacter     1.0000    0.0909    0.1667        33\n",
      "       prevotella     1.0000    0.8333    0.9091        18\n",
      "  prochlorococcus     1.0000    0.2500    0.4000        52\n",
      "          proteus     1.0000    0.1250    0.2222        40\n",
      "      providencia     0.0000    0.0000    0.0000        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     0.9965    0.5614    0.7182       513\n",
      "        ralstonia     1.0000    0.2143    0.3529        42\n",
      "     rheinheimera     1.0000    0.9773    0.9885       176\n",
      "        rhizobium     1.0000    0.1757    0.2989        74\n",
      "      rhodobacter     1.0000    0.4706    0.6400        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9563    0.3813    0.5452       632\n",
      "         serratia     0.8571    0.1856    0.3051        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.8000    0.0559    0.1046       143\n",
      "   staphylococcus     1.0000    0.7787    0.8756       244\n",
      " stenotrophomonas     0.0000    0.0000    0.0000        41\n",
      "    streptococcus     1.0000    0.2840    0.4423        81\n",
      "     streptomyces     1.0000    0.2326    0.3774        43\n",
      "    synechococcus     0.9946    0.7102    0.8287       521\n",
      "           vibrio     1.0000    0.3477    0.5160       463\n",
      "      xanthomonas     0.8333    0.0862    0.1563        58\n",
      "         yersinia     1.0000    0.3333    0.5000        93\n",
      "\n",
      "         accuracy                         0.4450      8116\n",
      "        macro avg     0.7279    0.2654    0.3436      8116\n",
      "     weighted avg     0.8405    0.4450    0.4720      8116\n",
      "\n",
      "===================\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confidence threshold k: 90.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     1.0000    0.0204    0.0400        98\n",
      "        aeromonas     1.0000    0.0316    0.0612        95\n",
      "    agrobacterium     0.0000    0.0000    0.0000        12\n",
      "     arthrobacter     1.0000    0.0769    0.1429        26\n",
      "         bacillus     1.0000    0.1127    0.2026       204\n",
      "      bacteroides     1.0000    0.3548    0.5238        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     1.0000    0.0513    0.0976        39\n",
      "    campylobacter     0.9459    0.2823    0.4348       124\n",
      "      caulobacter     1.0000    0.2000    0.3333        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     1.0000    0.0833    0.1538        24\n",
      "      cronobacter     0.0000    0.0000    0.0000        36\n",
      "          dickeya     1.0000    0.3654    0.5352        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.6000    0.0211    0.0408       142\n",
      "     enterococcus     1.0000    0.2653    0.4194        49\n",
      "          erwinia     1.0000    0.1434    0.2508       286\n",
      "      escherichia     0.9522    0.1691    0.2872      1295\n",
      "   flavobacterium     1.0000    0.7500    0.8571        48\n",
      "         gordonia     1.0000    0.0755    0.1404        53\n",
      "       halorubrum     0.5000    0.0769    0.1333        13\n",
      "     helicobacter     1.0000    0.1250    0.2222        16\n",
      "       klebsiella     1.0000    0.1385    0.2434       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.4100    0.5816       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.3214    0.4865        28\n",
      "   microbacterium     1.0000    0.1818    0.3077        33\n",
      "    mycobacterium     0.9854    0.8185    0.8943       248\n",
      "           others     0.1635    0.9949    0.2809       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.3239    0.4894        71\n",
      "     pelagibacter     1.0000    0.0606    0.1143        33\n",
      "       prevotella     1.0000    0.7778    0.8750        18\n",
      "  prochlorococcus     1.0000    0.2115    0.3492        52\n",
      "          proteus     1.0000    0.0500    0.0952        40\n",
      "      providencia     0.0000    0.0000    0.0000        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     1.0000    0.4542    0.6247       513\n",
      "        ralstonia     1.0000    0.1429    0.2500        42\n",
      "     rheinheimera     1.0000    0.9545    0.9767       176\n",
      "        rhizobium     1.0000    0.0676    0.1266        74\n",
      "      rhodobacter     1.0000    0.4118    0.5833        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9714    0.2690    0.4213       632\n",
      "         serratia     0.8333    0.0515    0.0971        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     0.6667    0.0140    0.0274       143\n",
      "   staphylococcus     1.0000    0.6516    0.7891       244\n",
      " stenotrophomonas     0.0000    0.0000    0.0000        41\n",
      "    streptococcus     1.0000    0.1975    0.3299        81\n",
      "     streptomyces     1.0000    0.1395    0.2449        43\n",
      "    synechococcus     0.9943    0.6679    0.7991       521\n",
      "           vibrio     1.0000    0.3024    0.4643       463\n",
      "      xanthomonas     0.8000    0.0690    0.1270        58\n",
      "         yersinia     1.0000    0.2473    0.3966        93\n",
      "\n",
      "         accuracy                         0.3781      8116\n",
      "        macro avg     0.7019    0.2057    0.2687      8116\n",
      "     weighted avg     0.8358    0.3781    0.3897      8116\n",
      "\n",
      "===================\n",
      "Confidence threshold k: 100.0%\n",
      "                   precision    recall  f1-score   support\n",
      "\n",
      "    achromobacter     0.0000    0.0000    0.0000        17\n",
      "    acinetobacter     0.0000    0.0000    0.0000        98\n",
      "        aeromonas     0.0000    0.0000    0.0000        95\n",
      "    agrobacterium     0.0000    0.0000    0.0000        12\n",
      "     arthrobacter     0.0000    0.0000    0.0000        26\n",
      "         bacillus     1.0000    0.0049    0.0098       204\n",
      "      bacteroides     0.0000    0.0000    0.0000        31\n",
      "    brevundimonas     0.0000    0.0000    0.0000        10\n",
      "     burkholderia     0.0000    0.0000    0.0000        39\n",
      "    campylobacter     1.0000    0.0484    0.0923       124\n",
      "      caulobacter     0.0000    0.0000    0.0000        25\n",
      "      citrobacter     0.0000    0.0000    0.0000        34\n",
      "      clostridium     0.0000    0.0000    0.0000        24\n",
      "      cronobacter     0.0000    0.0000    0.0000        36\n",
      "          dickeya     1.0000    0.2692    0.4242        52\n",
      "     edwardsiella     0.0000    0.0000    0.0000        10\n",
      "     enterobacter     0.0000    0.0000    0.0000       142\n",
      "     enterococcus     0.0000    0.0000    0.0000        49\n",
      "          erwinia     1.0000    0.0140    0.0276       286\n",
      "      escherichia     0.9500    0.0587    0.1105      1295\n",
      "   flavobacterium     1.0000    0.6250    0.7692        48\n",
      "         gordonia     0.0000    0.0000    0.0000        53\n",
      "       halorubrum     0.0000    0.0000    0.0000        13\n",
      "     helicobacter     0.0000    0.0000    0.0000        16\n",
      "       klebsiella     1.0000    0.0277    0.0539       397\n",
      "        kosakonia     0.0000    0.0000    0.0000        10\n",
      "    lactobacillus     0.0000    0.0000    0.0000        23\n",
      "      lactococcus     1.0000    0.1500    0.2609       100\n",
      "      leuconostoc     0.0000    0.0000    0.0000        10\n",
      "         listeria     1.0000    0.0714    0.1333        28\n",
      "   microbacterium     0.0000    0.0000    0.0000        33\n",
      "    mycobacterium     0.9810    0.6250    0.7635       248\n",
      "           others     0.1389    0.9959    0.2437       986\n",
      "          pantoea     0.0000    0.0000    0.0000        16\n",
      "   pectobacterium     1.0000    0.1549    0.2683        71\n",
      "     pelagibacter     0.0000    0.0000    0.0000        33\n",
      "       prevotella     0.0000    0.0000    0.0000        18\n",
      "  prochlorococcus     1.0000    0.1346    0.2373        52\n",
      "          proteus     0.0000    0.0000    0.0000        40\n",
      "      providencia     0.0000    0.0000    0.0000        16\n",
      "pseudoalteromonas     0.0000    0.0000    0.0000        17\n",
      "      pseudomonas     1.0000    0.1696    0.2900       513\n",
      "        ralstonia     0.0000    0.0000    0.0000        42\n",
      "     rheinheimera     1.0000    0.9261    0.9617       176\n",
      "        rhizobium     0.0000    0.0000    0.0000        74\n",
      "      rhodobacter     1.0000    0.1471    0.2564        34\n",
      "      rhodococcus     0.0000    0.0000    0.0000        13\n",
      "       salmonella     0.9841    0.0981    0.1784       632\n",
      "         serratia     0.0000    0.0000    0.0000        97\n",
      "       shewanella     0.0000    0.0000    0.0000        11\n",
      "         shigella     1.0000    0.0070    0.0139       143\n",
      "   staphylococcus     1.0000    0.2746    0.4309       244\n",
      " stenotrophomonas     0.0000    0.0000    0.0000        41\n",
      "    streptococcus     1.0000    0.0247    0.0482        81\n",
      "     streptomyces     1.0000    0.0233    0.0455        43\n",
      "    synechococcus     0.9956    0.4376    0.6080       521\n",
      "           vibrio     1.0000    0.1598    0.2756       463\n",
      "      xanthomonas     0.5000    0.0172    0.0333        58\n",
      "         yersinia     1.0000    0.1183    0.2115        93\n",
      "\n",
      "         accuracy                         0.2484      8116\n",
      "        macro avg     0.3991    0.0946    0.1144      8116\n",
      "     weighted avg     0.7309    0.2484    0.2161      8116\n",
      "\n",
      "===================\n",
      "['A_freq', 'GC', 'C_freq', 'TTA', 'TTA_b', 'K', 'AAA', 'AGC_b', 'G_freq', 'ATA', 'ACG_b', 'pI', 'GGC_b', 'Z4', 'CTDC2', 'mol_weight', 'AGC', 'GTC_b', 'ACA', 'ATC_b', 'D', 'ATA_b', 'GGA', 'GAC', 'TAC', 'T_freq', 'protein_length', 'CTDT33', 'Z1', 'GGA_b', 'frac_uncharged_polar', 'E', 'P', 'ACG', 'ACA_b', 'frac_sulfur', 'frac_hydrophobic', 'ACC_b', 'CTDT31', 'GAA_b', 'ACC', 'GAA', 'flexibility', 'CAA_b', 'CTDT5', 'G', 'CAA', 'frac_positive', 'GTA', 'GAT', 'frac_negative', 'CTDC1', 'CAG', 'GGG_b', 'GTC', 'GCA', 'CTT_b', 'GCA_b', 'GCG', 'GAG', 'CTG_b', 'GTA_b', 'AGA_b', 'AAC', 'AAT', 'R', 'GGC', 'GGT', 'GCC_b', 'GAC_b', 'frac_amide', 'TTG', 'A', 'GGG', 'CTT', 'frac_alcohol', 'I', 'CTDT38', 'Z2', 'Q', 'AAG_b', 'CTC_b', 'GAT_b', 'CTG', 'CTA_b', 'GAG_b', 'GTG', 'TCA_b', 'CAG_b', 'CTA', 'GTG_b', 'CTDT11', 'CTDT12', 'TAT_b', 'AAA_b', 'M', 'ATT', 'ATT_b', 'ACT_b', 'ATC', 'CTDT24', 'TTC', 'TTG_b', 'AA_frac_sheet', 'GCC', 'CGT', 'Z3', 'F', 'CTC', 'TAC_b', 'frac_aliphatic', 'TCG_b', 'ATG', 'CTDT18', 'CTDT10', 'L', 'TTT', 'CCG', 'AAT_b', 'CTDT25', 'TCA', 'CTDT3', 'Z5', 'AA_frac_helix', 'S', 'CCA', 'W', 'CTDT32', 'CCT', 'AAG', 'AAC_b', 'CTDT8', 'T', 'CCC', 'CTDT7', 'CTDT37', 'TCC', 'TGG', 'GCT', 'CCT_b', 'GGT_b', 'TCG', 'TCT_b', 'CTDT21', 'GTT', 'N', 'TAT', 'GCG_b', 'CTDT19', 'CTDT34', 'GTT_b', 'CTDT39', 'AGT_b', 'CTDT36', 'frac_polar', 'CTDC3', 'CTDT4', 'CTDT30', 'CCG_b', 'CGT_b', 'CTDT26', 'TCC_b', 'TTT_b', 'ACT', 'CTDT28', 'AA_frac_turn', 'CTDT14', 'CTDT6', 'CTDT13', 'TTC_b', 'Y', 'AGA', 'CTDT2', 'CTDT23', 'CTDT22', 'TCT', 'GCT_b', 'CTDT1', 'CGC', 'aromaticity', 'CTDT17', 'CCA_b', 'CTDT16', 'C', 'instability', 'CTDT27', 'CGG', 'CTDT29', 'H', 'CTDT9', 'CTDT20', 'CGC_b', 'CGG_b', 'AGT', 'AGG', 'V', 'CGA', 'CTDT35', 'CTDT15', 'CGA_b', 'CAC', 'CCC_b', 'AGG_b', 'TAA', 'TGC', 'CAT', 'TGT', 'CAC_b', 'CAT_b', 'TAG', 'TGC_b', 'TGA', 'TGT_b', 'TAG_b', 'TAA_b', 'TGA_b', 'TGG_b', 'ATG_b']\n"
     ]
    }
   ],
   "source": [
    "important_features = util.classify(models[-1], display_feature_importance = True, feature_columns = list(features))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "309a5ac4",
   "metadata": {},
   "source": [
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b31144fb",
   "metadata": {},
   "source": [
    "## Part IV: Handpicked Features + Embeddings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73194c49",
   "metadata": {},
   "source": [
    "### Features with Highest Importance (Global)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5589f37",
   "metadata": {},
   "source": [
    "Investigate the performance if the ProtT5 embeddings are combined with the sequence properties that registered the highest Gini importance after training the phage-host interaction prediction tool by [Boeckaerts <i>et al.</i> (2021)](https://www.nature.com/articles/s41598-021-81063-4) on our dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b2788afd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n"
     ]
    }
   ],
   "source": [
    "for i in range(5):\n",
    "    util.classify_handpicked_embeddings([important_features[i]], important_features[i])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "104b69f1",
   "metadata": {},
   "source": [
    "Investigate the performance if the ProtT5 embeddings are combined with the **<i>n</i> sequence properties** that registered the highest Gini importance after training the phage-host interaction prediction tool by [Boeckaerts <i>et al.</i> (2021)](https://www.nature.com/articles/s41598-021-81063-4)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "abf04cb0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n"
     ]
    }
   ],
   "source": [
    "for i in range(1, 6):\n",
    "    util.classify_handpicked_embeddings(important_features[:i], f'handcrafted_{i}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c83f3531",
   "metadata": {},
   "source": [
    "### Protein Features with Highest Importance"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "364713b4",
   "metadata": {},
   "source": [
    "Investigate the performance if the ProtT5 embeddings are combined with the **protein** sequence properties that registered the highest Gini importance after training the phage-host interaction prediction tool by [Boeckaerts <i>et al.</i> (2021)](https://www.nature.com/articles/s41598-021-81063-4) on our dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3d30dc3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "important_protein_features = ['K', 'pI', 'Z4', 'CTDC2', 'mol_weight']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "fa321d83",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n"
     ]
    }
   ],
   "source": [
    "for i in range(5):\n",
    "    util.classify_handpicked_embeddings([important_protein_features[i]], important_protein_features[i])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ced2129",
   "metadata": {},
   "source": [
    "Investigate the performance if the ProtT5 embeddings are combined with the **<i>n</i> protein sequence properties** that registered the highest Gini importance after training the phage-host interaction prediction tool by [Boeckaerts <i>et al.</i> (2021)](https://www.nature.com/articles/s41598-021-81063-4)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "913a3287",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n",
      "Constructing training and test sets...\n",
      "Training the model...\n",
      "Saving evaluation results...\n"
     ]
    }
   ],
   "source": [
    "for i in range(1, 6):\n",
    "    util.classify_handpicked_embeddings(important_protein_features[:i], f'handcrafted_protein_{i}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f09e9c5",
   "metadata": {},
   "source": [
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "711b0318",
   "metadata": {},
   "source": [
    "## Part V: Results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d11997ca",
   "metadata": {},
   "source": [
    "Load the pickled files storing the results of evaluating the performance of the models."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ff872681",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_results = []\n",
    "for model in models:\n",
    "    with open(constants.PLM_RESULTS[model], 'rb') as f:\n",
    "        model_results.append(pickle.load(f))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6d4b89f",
   "metadata": {},
   "source": [
    "Declare constant variables for readability of subsequent code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c26e9498",
   "metadata": {},
   "outputs": [],
   "source": [
    "CLASS = 0\n",
    "MICRO = 1\n",
    "MACRO = 2\n",
    "WEIGHTED = 3\n",
    "\n",
    "PRECISION = 0\n",
    "RECALL = 1\n",
    "F1 = 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11395e7d",
   "metadata": {},
   "source": [
    "The shaded cells in the subsequent tables correspond to the highest scores (i.e., best performance in terms of the specified evaluation metric)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b5efe1c",
   "metadata": {},
   "source": [
    "### Protein Language Models vs Handcrafted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "537f1395",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Weighted Precision\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_73e06_row1_col6, #T_73e06_row2_col2, #T_73e06_row3_col10, #T_73e06_row5_col7, #T_73e06_row5_col8, #T_73e06_row6_col1, #T_73e06_row6_col9, #T_73e06_row7_col0, #T_73e06_row7_col3, #T_73e06_row7_col4, #T_73e06_row7_col5 {\n",
       "  background-color: lightgreen;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_73e06_\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th class=\"col_heading level0 col0\" >0%</th>\n",
       "      <th class=\"col_heading level0 col1\" >10%</th>\n",
       "      <th class=\"col_heading level0 col2\" >20%</th>\n",
       "      <th class=\"col_heading level0 col3\" >30%</th>\n",
       "      <th class=\"col_heading level0 col4\" >40%</th>\n",
       "      <th class=\"col_heading level0 col5\" >50%</th>\n",
       "      <th class=\"col_heading level0 col6\" >60%</th>\n",
       "      <th class=\"col_heading level0 col7\" >70%</th>\n",
       "      <th class=\"col_heading level0 col8\" >80%</th>\n",
       "      <th class=\"col_heading level0 col9\" >90%</th>\n",
       "      <th class=\"col_heading level0 col10\" >100%</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_73e06_level0_row0\" class=\"row_heading level0 row0\" >PROTTRANSBERT</th>\n",
       "      <td id=\"T_73e06_row0_col0\" class=\"data row0 col0\" >63.26%</td>\n",
       "      <td id=\"T_73e06_row0_col1\" class=\"data row0 col1\" >79.31%</td>\n",
       "      <td id=\"T_73e06_row0_col2\" class=\"data row0 col2\" >82.93%</td>\n",
       "      <td id=\"T_73e06_row0_col3\" class=\"data row0 col3\" >83.89%</td>\n",
       "      <td id=\"T_73e06_row0_col4\" class=\"data row0 col4\" >84.12%</td>\n",
       "      <td id=\"T_73e06_row0_col5\" class=\"data row0 col5\" >84.52%</td>\n",
       "      <td id=\"T_73e06_row0_col6\" class=\"data row0 col6\" >84.96%</td>\n",
       "      <td id=\"T_73e06_row0_col7\" class=\"data row0 col7\" >84.55%</td>\n",
       "      <td id=\"T_73e06_row0_col8\" class=\"data row0 col8\" >84.20%</td>\n",
       "      <td id=\"T_73e06_row0_col9\" class=\"data row0 col9\" >83.69%</td>\n",
       "      <td id=\"T_73e06_row0_col10\" class=\"data row0 col10\" >73.74%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_73e06_level0_row1\" class=\"row_heading level0 row1\" >PROTXLNET</th>\n",
       "      <td id=\"T_73e06_row1_col0\" class=\"data row1 col0\" >62.18%</td>\n",
       "      <td id=\"T_73e06_row1_col1\" class=\"data row1 col1\" >79.28%</td>\n",
       "      <td id=\"T_73e06_row1_col2\" class=\"data row1 col2\" >82.92%</td>\n",
       "      <td id=\"T_73e06_row1_col3\" class=\"data row1 col3\" >83.93%</td>\n",
       "      <td id=\"T_73e06_row1_col4\" class=\"data row1 col4\" >84.23%</td>\n",
       "      <td id=\"T_73e06_row1_col5\" class=\"data row1 col5\" >84.61%</td>\n",
       "      <td id=\"T_73e06_row1_col6\" class=\"data row1 col6\" >85.66%</td>\n",
       "      <td id=\"T_73e06_row1_col7\" class=\"data row1 col7\" >84.67%</td>\n",
       "      <td id=\"T_73e06_row1_col8\" class=\"data row1 col8\" >84.06%</td>\n",
       "      <td id=\"T_73e06_row1_col9\" class=\"data row1 col9\" >82.80%</td>\n",
       "      <td id=\"T_73e06_row1_col10\" class=\"data row1 col10\" >77.07%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_73e06_level0_row2\" class=\"row_heading level0 row2\" >PROTTRANSALBERT</th>\n",
       "      <td id=\"T_73e06_row2_col0\" class=\"data row2 col0\" >62.63%</td>\n",
       "      <td id=\"T_73e06_row2_col1\" class=\"data row2 col1\" >80.24%</td>\n",
       "      <td id=\"T_73e06_row2_col2\" class=\"data row2 col2\" >83.37%</td>\n",
       "      <td id=\"T_73e06_row2_col3\" class=\"data row2 col3\" >84.26%</td>\n",
       "      <td id=\"T_73e06_row2_col4\" class=\"data row2 col4\" >84.22%</td>\n",
       "      <td id=\"T_73e06_row2_col5\" class=\"data row2 col5\" >84.65%</td>\n",
       "      <td id=\"T_73e06_row2_col6\" class=\"data row2 col6\" >84.91%</td>\n",
       "      <td id=\"T_73e06_row2_col7\" class=\"data row2 col7\" >84.61%</td>\n",
       "      <td id=\"T_73e06_row2_col8\" class=\"data row2 col8\" >83.97%</td>\n",
       "      <td id=\"T_73e06_row2_col9\" class=\"data row2 col9\" >83.03%</td>\n",
       "      <td id=\"T_73e06_row2_col10\" class=\"data row2 col10\" >76.06%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_73e06_level0_row3\" class=\"row_heading level0 row3\" >PROTT5</th>\n",
       "      <td id=\"T_73e06_row3_col0\" class=\"data row3 col0\" >63.57%</td>\n",
       "      <td id=\"T_73e06_row3_col1\" class=\"data row3 col1\" >80.47%</td>\n",
       "      <td id=\"T_73e06_row3_col2\" class=\"data row3 col2\" >82.88%</td>\n",
       "      <td id=\"T_73e06_row3_col3\" class=\"data row3 col3\" >84.13%</td>\n",
       "      <td id=\"T_73e06_row3_col4\" class=\"data row3 col4\" >84.38%</td>\n",
       "      <td id=\"T_73e06_row3_col5\" class=\"data row3 col5\" >84.67%</td>\n",
       "      <td id=\"T_73e06_row3_col6\" class=\"data row3 col6\" >85.43%</td>\n",
       "      <td id=\"T_73e06_row3_col7\" class=\"data row3 col7\" >84.98%</td>\n",
       "      <td id=\"T_73e06_row3_col8\" class=\"data row3 col8\" >84.32%</td>\n",
       "      <td id=\"T_73e06_row3_col9\" class=\"data row3 col9\" >83.51%</td>\n",
       "      <td id=\"T_73e06_row3_col10\" class=\"data row3 col10\" >77.23%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_73e06_level0_row4\" class=\"row_heading level0 row4\" >ESM</th>\n",
       "      <td id=\"T_73e06_row4_col0\" class=\"data row4 col0\" >63.32%</td>\n",
       "      <td id=\"T_73e06_row4_col1\" class=\"data row4 col1\" >80.15%</td>\n",
       "      <td id=\"T_73e06_row4_col2\" class=\"data row4 col2\" >82.67%</td>\n",
       "      <td id=\"T_73e06_row4_col3\" class=\"data row4 col3\" >83.89%</td>\n",
       "      <td id=\"T_73e06_row4_col4\" class=\"data row4 col4\" >84.49%</td>\n",
       "      <td id=\"T_73e06_row4_col5\" class=\"data row4 col5\" >84.66%</td>\n",
       "      <td id=\"T_73e06_row4_col6\" class=\"data row4 col6\" >85.31%</td>\n",
       "      <td id=\"T_73e06_row4_col7\" class=\"data row4 col7\" >84.62%</td>\n",
       "      <td id=\"T_73e06_row4_col8\" class=\"data row4 col8\" >83.66%</td>\n",
       "      <td id=\"T_73e06_row4_col9\" class=\"data row4 col9\" >82.68%</td>\n",
       "      <td id=\"T_73e06_row4_col10\" class=\"data row4 col10\" >76.51%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_73e06_level0_row5\" class=\"row_heading level0 row5\" >ESM1B</th>\n",
       "      <td id=\"T_73e06_row5_col0\" class=\"data row5 col0\" >63.38%</td>\n",
       "      <td id=\"T_73e06_row5_col1\" class=\"data row5 col1\" >79.80%</td>\n",
       "      <td id=\"T_73e06_row5_col2\" class=\"data row5 col2\" >83.27%</td>\n",
       "      <td id=\"T_73e06_row5_col3\" class=\"data row5 col3\" >84.11%</td>\n",
       "      <td id=\"T_73e06_row5_col4\" class=\"data row5 col4\" >84.29%</td>\n",
       "      <td id=\"T_73e06_row5_col5\" class=\"data row5 col5\" >84.75%</td>\n",
       "      <td id=\"T_73e06_row5_col6\" class=\"data row5 col6\" >84.71%</td>\n",
       "      <td id=\"T_73e06_row5_col7\" class=\"data row5 col7\" >84.99%</td>\n",
       "      <td id=\"T_73e06_row5_col8\" class=\"data row5 col8\" >84.40%</td>\n",
       "      <td id=\"T_73e06_row5_col9\" class=\"data row5 col9\" >83.42%</td>\n",
       "      <td id=\"T_73e06_row5_col10\" class=\"data row5 col10\" >76.89%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_73e06_level0_row6\" class=\"row_heading level0 row6\" >SEQVEC</th>\n",
       "      <td id=\"T_73e06_row6_col0\" class=\"data row6 col0\" >63.30%</td>\n",
       "      <td id=\"T_73e06_row6_col1\" class=\"data row6 col1\" >80.52%</td>\n",
       "      <td id=\"T_73e06_row6_col2\" class=\"data row6 col2\" >83.30%</td>\n",
       "      <td id=\"T_73e06_row6_col3\" class=\"data row6 col3\" >84.31%</td>\n",
       "      <td id=\"T_73e06_row6_col4\" class=\"data row6 col4\" >84.80%</td>\n",
       "      <td id=\"T_73e06_row6_col5\" class=\"data row6 col5\" >84.99%</td>\n",
       "      <td id=\"T_73e06_row6_col6\" class=\"data row6 col6\" >85.65%</td>\n",
       "      <td id=\"T_73e06_row6_col7\" class=\"data row6 col7\" >84.56%</td>\n",
       "      <td id=\"T_73e06_row6_col8\" class=\"data row6 col8\" >84.26%</td>\n",
       "      <td id=\"T_73e06_row6_col9\" class=\"data row6 col9\" >83.83%</td>\n",
       "      <td id=\"T_73e06_row6_col10\" class=\"data row6 col10\" >73.20%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_73e06_level0_row7\" class=\"row_heading level0 row7\" >BOECKAERTS</th>\n",
       "      <td id=\"T_73e06_row7_col0\" class=\"data row7 col0\" >64.98%</td>\n",
       "      <td id=\"T_73e06_row7_col1\" class=\"data row7 col1\" >79.65%</td>\n",
       "      <td id=\"T_73e06_row7_col2\" class=\"data row7 col2\" >82.99%</td>\n",
       "      <td id=\"T_73e06_row7_col3\" class=\"data row7 col3\" >84.57%</td>\n",
       "      <td id=\"T_73e06_row7_col4\" class=\"data row7 col4\" >84.97%</td>\n",
       "      <td id=\"T_73e06_row7_col5\" class=\"data row7 col5\" >85.76%</td>\n",
       "      <td id=\"T_73e06_row7_col6\" class=\"data row7 col6\" >85.55%</td>\n",
       "      <td id=\"T_73e06_row7_col7\" class=\"data row7 col7\" >84.88%</td>\n",
       "      <td id=\"T_73e06_row7_col8\" class=\"data row7 col8\" >84.05%</td>\n",
       "      <td id=\"T_73e06_row7_col9\" class=\"data row7 col9\" >83.58%</td>\n",
       "      <td id=\"T_73e06_row7_col10\" class=\"data row7 col10\" >73.09%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x2249967c348>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = []\n",
    "for model in models:\n",
    "    result = []\n",
    "    for threshold in range(0, 11):\n",
    "        metric = \"{:.2f}\".format(model_results[models.index(model)][threshold][WEIGHTED][PRECISION] * 100)\n",
    "        \n",
    "        result.append(f'{metric}%')\n",
    "        \n",
    "    results.append(result)\n",
    "\n",
    "print(\"Weighted Precision\")\n",
    "results_df = pd.DataFrame(results, columns = [str(_) + '%' for _ in range(0, 101, 10)], index = models)\n",
    "results_df.style.highlight_max(color = 'lightgreen', axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a8b10706",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Weighted Recall\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_a2b8f_row3_col0, #T_a2b8f_row3_col1, #T_a2b8f_row3_col2, #T_a2b8f_row3_col3, #T_a2b8f_row3_col4, #T_a2b8f_row3_col5, #T_a2b8f_row3_col6, #T_a2b8f_row3_col7, #T_a2b8f_row3_col8, #T_a2b8f_row3_col9, #T_a2b8f_row4_col10, #T_a2b8f_row5_col2 {\n",
       "  background-color: lightgreen;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_a2b8f_\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th class=\"col_heading level0 col0\" >0%</th>\n",
       "      <th class=\"col_heading level0 col1\" >10%</th>\n",
       "      <th class=\"col_heading level0 col2\" >20%</th>\n",
       "      <th class=\"col_heading level0 col3\" >30%</th>\n",
       "      <th class=\"col_heading level0 col4\" >40%</th>\n",
       "      <th class=\"col_heading level0 col5\" >50%</th>\n",
       "      <th class=\"col_heading level0 col6\" >60%</th>\n",
       "      <th class=\"col_heading level0 col7\" >70%</th>\n",
       "      <th class=\"col_heading level0 col8\" >80%</th>\n",
       "      <th class=\"col_heading level0 col9\" >90%</th>\n",
       "      <th class=\"col_heading level0 col10\" >100%</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_a2b8f_level0_row0\" class=\"row_heading level0 row0\" >PROTTRANSBERT</th>\n",
       "      <td id=\"T_a2b8f_row0_col0\" class=\"data row0 col0\" >69.70%</td>\n",
       "      <td id=\"T_a2b8f_row0_col1\" class=\"data row0 col1\" >73.57%</td>\n",
       "      <td id=\"T_a2b8f_row0_col2\" class=\"data row0 col2\" >71.66%</td>\n",
       "      <td id=\"T_a2b8f_row0_col3\" class=\"data row0 col3\" >68.36%</td>\n",
       "      <td id=\"T_a2b8f_row0_col4\" class=\"data row0 col4\" >64.82%</td>\n",
       "      <td id=\"T_a2b8f_row0_col5\" class=\"data row0 col5\" >61.27%</td>\n",
       "      <td id=\"T_a2b8f_row0_col6\" class=\"data row0 col6\" >56.65%</td>\n",
       "      <td id=\"T_a2b8f_row0_col7\" class=\"data row0 col7\" >51.60%</td>\n",
       "      <td id=\"T_a2b8f_row0_col8\" class=\"data row0 col8\" >46.16%</td>\n",
       "      <td id=\"T_a2b8f_row0_col9\" class=\"data row0 col9\" >39.48%</td>\n",
       "      <td id=\"T_a2b8f_row0_col10\" class=\"data row0 col10\" >27.02%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_a2b8f_level0_row1\" class=\"row_heading level0 row1\" >PROTXLNET</th>\n",
       "      <td id=\"T_a2b8f_row1_col0\" class=\"data row1 col0\" >68.31%</td>\n",
       "      <td id=\"T_a2b8f_row1_col1\" class=\"data row1 col1\" >72.12%</td>\n",
       "      <td id=\"T_a2b8f_row1_col2\" class=\"data row1 col2\" >69.99%</td>\n",
       "      <td id=\"T_a2b8f_row1_col3\" class=\"data row1 col3\" >67.11%</td>\n",
       "      <td id=\"T_a2b8f_row1_col4\" class=\"data row1 col4\" >64.17%</td>\n",
       "      <td id=\"T_a2b8f_row1_col5\" class=\"data row1 col5\" >60.46%</td>\n",
       "      <td id=\"T_a2b8f_row1_col6\" class=\"data row1 col6\" >56.32%</td>\n",
       "      <td id=\"T_a2b8f_row1_col7\" class=\"data row1 col7\" >50.95%</td>\n",
       "      <td id=\"T_a2b8f_row1_col8\" class=\"data row1 col8\" >45.50%</td>\n",
       "      <td id=\"T_a2b8f_row1_col9\" class=\"data row1 col9\" >38.50%</td>\n",
       "      <td id=\"T_a2b8f_row1_col10\" class=\"data row1 col10\" >26.56%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_a2b8f_level0_row2\" class=\"row_heading level0 row2\" >PROTTRANSALBERT</th>\n",
       "      <td id=\"T_a2b8f_row2_col0\" class=\"data row2 col0\" >68.97%</td>\n",
       "      <td id=\"T_a2b8f_row2_col1\" class=\"data row2 col1\" >73.60%</td>\n",
       "      <td id=\"T_a2b8f_row2_col2\" class=\"data row2 col2\" >70.91%</td>\n",
       "      <td id=\"T_a2b8f_row2_col3\" class=\"data row2 col3\" >67.41%</td>\n",
       "      <td id=\"T_a2b8f_row2_col4\" class=\"data row2 col4\" >64.17%</td>\n",
       "      <td id=\"T_a2b8f_row2_col5\" class=\"data row2 col5\" >60.71%</td>\n",
       "      <td id=\"T_a2b8f_row2_col6\" class=\"data row2 col6\" >56.67%</td>\n",
       "      <td id=\"T_a2b8f_row2_col7\" class=\"data row2 col7\" >51.21%</td>\n",
       "      <td id=\"T_a2b8f_row2_col8\" class=\"data row2 col8\" >45.53%</td>\n",
       "      <td id=\"T_a2b8f_row2_col9\" class=\"data row2 col9\" >38.90%</td>\n",
       "      <td id=\"T_a2b8f_row2_col10\" class=\"data row2 col10\" >26.38%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_a2b8f_level0_row3\" class=\"row_heading level0 row3\" >PROTT5</th>\n",
       "      <td id=\"T_a2b8f_row3_col0\" class=\"data row3 col0\" >70.91%</td>\n",
       "      <td id=\"T_a2b8f_row3_col1\" class=\"data row3 col1\" >75.85%</td>\n",
       "      <td id=\"T_a2b8f_row3_col2\" class=\"data row3 col2\" >73.41%</td>\n",
       "      <td id=\"T_a2b8f_row3_col3\" class=\"data row3 col3\" >70.39%</td>\n",
       "      <td id=\"T_a2b8f_row3_col4\" class=\"data row3 col4\" >67.05%</td>\n",
       "      <td id=\"T_a2b8f_row3_col5\" class=\"data row3 col5\" >63.38%</td>\n",
       "      <td id=\"T_a2b8f_row3_col6\" class=\"data row3 col6\" >59.15%</td>\n",
       "      <td id=\"T_a2b8f_row3_col7\" class=\"data row3 col7\" >53.72%</td>\n",
       "      <td id=\"T_a2b8f_row3_col8\" class=\"data row3 col8\" >48.57%</td>\n",
       "      <td id=\"T_a2b8f_row3_col9\" class=\"data row3 col9\" >41.03%</td>\n",
       "      <td id=\"T_a2b8f_row3_col10\" class=\"data row3 col10\" >27.16%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_a2b8f_level0_row4\" class=\"row_heading level0 row4\" >ESM</th>\n",
       "      <td id=\"T_a2b8f_row4_col0\" class=\"data row4 col0\" >70.54%</td>\n",
       "      <td id=\"T_a2b8f_row4_col1\" class=\"data row4 col1\" >75.07%</td>\n",
       "      <td id=\"T_a2b8f_row4_col2\" class=\"data row4 col2\" >72.55%</td>\n",
       "      <td id=\"T_a2b8f_row4_col3\" class=\"data row4 col3\" >69.68%</td>\n",
       "      <td id=\"T_a2b8f_row4_col4\" class=\"data row4 col4\" >66.56%</td>\n",
       "      <td id=\"T_a2b8f_row4_col5\" class=\"data row4 col5\" >63.13%</td>\n",
       "      <td id=\"T_a2b8f_row4_col6\" class=\"data row4 col6\" >58.33%</td>\n",
       "      <td id=\"T_a2b8f_row4_col7\" class=\"data row4 col7\" >53.38%</td>\n",
       "      <td id=\"T_a2b8f_row4_col8\" class=\"data row4 col8\" >47.92%</td>\n",
       "      <td id=\"T_a2b8f_row4_col9\" class=\"data row4 col9\" >40.69%</td>\n",
       "      <td id=\"T_a2b8f_row4_col10\" class=\"data row4 col10\" >27.34%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_a2b8f_level0_row5\" class=\"row_heading level0 row5\" >ESM1B</th>\n",
       "      <td id=\"T_a2b8f_row5_col0\" class=\"data row5 col0\" >70.81%</td>\n",
       "      <td id=\"T_a2b8f_row5_col1\" class=\"data row5 col1\" >75.20%</td>\n",
       "      <td id=\"T_a2b8f_row5_col2\" class=\"data row5 col2\" >73.41%</td>\n",
       "      <td id=\"T_a2b8f_row5_col3\" class=\"data row5 col3\" >70.01%</td>\n",
       "      <td id=\"T_a2b8f_row5_col4\" class=\"data row5 col4\" >66.68%</td>\n",
       "      <td id=\"T_a2b8f_row5_col5\" class=\"data row5 col5\" >62.95%</td>\n",
       "      <td id=\"T_a2b8f_row5_col6\" class=\"data row5 col6\" >58.49%</td>\n",
       "      <td id=\"T_a2b8f_row5_col7\" class=\"data row5 col7\" >53.30%</td>\n",
       "      <td id=\"T_a2b8f_row5_col8\" class=\"data row5 col8\" >47.86%</td>\n",
       "      <td id=\"T_a2b8f_row5_col9\" class=\"data row5 col9\" >40.59%</td>\n",
       "      <td id=\"T_a2b8f_row5_col10\" class=\"data row5 col10\" >27.24%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_a2b8f_level0_row6\" class=\"row_heading level0 row6\" >SEQVEC</th>\n",
       "      <td id=\"T_a2b8f_row6_col0\" class=\"data row6 col0\" >69.20%</td>\n",
       "      <td id=\"T_a2b8f_row6_col1\" class=\"data row6 col1\" >73.40%</td>\n",
       "      <td id=\"T_a2b8f_row6_col2\" class=\"data row6 col2\" >71.13%</td>\n",
       "      <td id=\"T_a2b8f_row6_col3\" class=\"data row6 col3\" >68.04%</td>\n",
       "      <td id=\"T_a2b8f_row6_col4\" class=\"data row6 col4\" >64.79%</td>\n",
       "      <td id=\"T_a2b8f_row6_col5\" class=\"data row6 col5\" >61.26%</td>\n",
       "      <td id=\"T_a2b8f_row6_col6\" class=\"data row6 col6\" >56.73%</td>\n",
       "      <td id=\"T_a2b8f_row6_col7\" class=\"data row6 col7\" >51.22%</td>\n",
       "      <td id=\"T_a2b8f_row6_col8\" class=\"data row6 col8\" >45.97%</td>\n",
       "      <td id=\"T_a2b8f_row6_col9\" class=\"data row6 col9\" >38.87%</td>\n",
       "      <td id=\"T_a2b8f_row6_col10\" class=\"data row6 col10\" >25.89%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_a2b8f_level0_row7\" class=\"row_heading level0 row7\" >BOECKAERTS</th>\n",
       "      <td id=\"T_a2b8f_row7_col0\" class=\"data row7 col0\" >70.33%</td>\n",
       "      <td id=\"T_a2b8f_row7_col1\" class=\"data row7 col1\" >73.95%</td>\n",
       "      <td id=\"T_a2b8f_row7_col2\" class=\"data row7 col2\" >71.48%</td>\n",
       "      <td id=\"T_a2b8f_row7_col3\" class=\"data row7 col3\" >68.26%</td>\n",
       "      <td id=\"T_a2b8f_row7_col4\" class=\"data row7 col4\" >64.21%</td>\n",
       "      <td id=\"T_a2b8f_row7_col5\" class=\"data row7 col5\" >60.24%</td>\n",
       "      <td id=\"T_a2b8f_row7_col6\" class=\"data row7 col6\" >55.48%</td>\n",
       "      <td id=\"T_a2b8f_row7_col7\" class=\"data row7 col7\" >49.84%</td>\n",
       "      <td id=\"T_a2b8f_row7_col8\" class=\"data row7 col8\" >44.50%</td>\n",
       "      <td id=\"T_a2b8f_row7_col9\" class=\"data row7 col9\" >37.81%</td>\n",
       "      <td id=\"T_a2b8f_row7_col10\" class=\"data row7 col10\" >24.84%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x224c44108c8>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = []\n",
    "for model in models:\n",
    "    result = []\n",
    "    for threshold in range(0, 11):\n",
    "        metric = \"{:.2f}\".format(model_results[models.index(model)][threshold][WEIGHTED][RECALL] * 100)\n",
    "        \n",
    "        result.append(f'{metric}%')\n",
    "        \n",
    "    results.append(result)\n",
    "\n",
    "print(\"Weighted Recall\")\n",
    "results_df = pd.DataFrame(results, columns = [str(_) + '%' for _ in range(0, 101, 10)], index = models)\n",
    "results_df.style.highlight_max(color = 'lightgreen', axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e7958491",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Weighted F1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_27436_row3_col0, #T_27436_row3_col1, #T_27436_row3_col3, #T_27436_row3_col4, #T_27436_row3_col5, #T_27436_row3_col6, #T_27436_row3_col7, #T_27436_row3_col8, #T_27436_row3_col9, #T_27436_row4_col10, #T_27436_row5_col2 {\n",
       "  background-color: lightgreen;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_27436_\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th class=\"col_heading level0 col0\" >0%</th>\n",
       "      <th class=\"col_heading level0 col1\" >10%</th>\n",
       "      <th class=\"col_heading level0 col2\" >20%</th>\n",
       "      <th class=\"col_heading level0 col3\" >30%</th>\n",
       "      <th class=\"col_heading level0 col4\" >40%</th>\n",
       "      <th class=\"col_heading level0 col5\" >50%</th>\n",
       "      <th class=\"col_heading level0 col6\" >60%</th>\n",
       "      <th class=\"col_heading level0 col7\" >70%</th>\n",
       "      <th class=\"col_heading level0 col8\" >80%</th>\n",
       "      <th class=\"col_heading level0 col9\" >90%</th>\n",
       "      <th class=\"col_heading level0 col10\" >100%</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_27436_level0_row0\" class=\"row_heading level0 row0\" >PROTTRANSBERT</th>\n",
       "      <td id=\"T_27436_row0_col0\" class=\"data row0 col0\" >65.00%</td>\n",
       "      <td id=\"T_27436_row0_col1\" class=\"data row0 col1\" >74.39%</td>\n",
       "      <td id=\"T_27436_row0_col2\" class=\"data row0 col2\" >73.61%</td>\n",
       "      <td id=\"T_27436_row0_col3\" class=\"data row0 col3\" >71.07%</td>\n",
       "      <td id=\"T_27436_row0_col4\" class=\"data row0 col4\" >68.05%</td>\n",
       "      <td id=\"T_27436_row0_col5\" class=\"data row0 col5\" >64.87%</td>\n",
       "      <td id=\"T_27436_row0_col6\" class=\"data row0 col6\" >60.58%</td>\n",
       "      <td id=\"T_27436_row0_col7\" class=\"data row0 col7\" >55.27%</td>\n",
       "      <td id=\"T_27436_row0_col8\" class=\"data row0 col8\" >49.29%</td>\n",
       "      <td id=\"T_27436_row0_col9\" class=\"data row0 col9\" >41.07%</td>\n",
       "      <td id=\"T_27436_row0_col10\" class=\"data row0 col10\" >24.59%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_27436_level0_row1\" class=\"row_heading level0 row1\" >PROTXLNET</th>\n",
       "      <td id=\"T_27436_row1_col0\" class=\"data row1 col0\" >63.71%</td>\n",
       "      <td id=\"T_27436_row1_col1\" class=\"data row1 col1\" >73.31%</td>\n",
       "      <td id=\"T_27436_row1_col2\" class=\"data row1 col2\" >72.18%</td>\n",
       "      <td id=\"T_27436_row1_col3\" class=\"data row1 col3\" >69.93%</td>\n",
       "      <td id=\"T_27436_row1_col4\" class=\"data row1 col4\" >67.44%</td>\n",
       "      <td id=\"T_27436_row1_col5\" class=\"data row1 col5\" >64.11%</td>\n",
       "      <td id=\"T_27436_row1_col6\" class=\"data row1 col6\" >60.18%</td>\n",
       "      <td id=\"T_27436_row1_col7\" class=\"data row1 col7\" >54.54%</td>\n",
       "      <td id=\"T_27436_row1_col8\" class=\"data row1 col8\" >48.46%</td>\n",
       "      <td id=\"T_27436_row1_col9\" class=\"data row1 col9\" >40.20%</td>\n",
       "      <td id=\"T_27436_row1_col10\" class=\"data row1 col10\" >23.92%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_27436_level0_row2\" class=\"row_heading level0 row2\" >PROTTRANSALBERT</th>\n",
       "      <td id=\"T_27436_row2_col0\" class=\"data row2 col0\" >64.31%</td>\n",
       "      <td id=\"T_27436_row2_col1\" class=\"data row2 col1\" >74.57%</td>\n",
       "      <td id=\"T_27436_row2_col2\" class=\"data row2 col2\" >73.01%</td>\n",
       "      <td id=\"T_27436_row2_col3\" class=\"data row2 col3\" >70.34%</td>\n",
       "      <td id=\"T_27436_row2_col4\" class=\"data row2 col4\" >67.42%</td>\n",
       "      <td id=\"T_27436_row2_col5\" class=\"data row2 col5\" >64.32%</td>\n",
       "      <td id=\"T_27436_row2_col6\" class=\"data row2 col6\" >60.45%</td>\n",
       "      <td id=\"T_27436_row2_col7\" class=\"data row2 col7\" >54.82%</td>\n",
       "      <td id=\"T_27436_row2_col8\" class=\"data row2 col8\" >48.51%</td>\n",
       "      <td id=\"T_27436_row2_col9\" class=\"data row2 col9\" >40.57%</td>\n",
       "      <td id=\"T_27436_row2_col10\" class=\"data row2 col10\" >23.80%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_27436_level0_row3\" class=\"row_heading level0 row3\" >PROTT5</th>\n",
       "      <td id=\"T_27436_row3_col0\" class=\"data row3 col0\" >66.10%</td>\n",
       "      <td id=\"T_27436_row3_col1\" class=\"data row3 col1\" >76.51%</td>\n",
       "      <td id=\"T_27436_row3_col2\" class=\"data row3 col2\" >74.96%</td>\n",
       "      <td id=\"T_27436_row3_col3\" class=\"data row3 col3\" >72.78%</td>\n",
       "      <td id=\"T_27436_row3_col4\" class=\"data row3 col4\" >70.03%</td>\n",
       "      <td id=\"T_27436_row3_col5\" class=\"data row3 col5\" >66.78%</td>\n",
       "      <td id=\"T_27436_row3_col6\" class=\"data row3 col6\" >62.95%</td>\n",
       "      <td id=\"T_27436_row3_col7\" class=\"data row3 col7\" >57.51%</td>\n",
       "      <td id=\"T_27436_row3_col8\" class=\"data row3 col8\" >51.98%</td>\n",
       "      <td id=\"T_27436_row3_col9\" class=\"data row3 col9\" >43.05%</td>\n",
       "      <td id=\"T_27436_row3_col10\" class=\"data row3 col10\" >24.82%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_27436_level0_row4\" class=\"row_heading level0 row4\" >ESM</th>\n",
       "      <td id=\"T_27436_row4_col0\" class=\"data row4 col0\" >65.72%</td>\n",
       "      <td id=\"T_27436_row4_col1\" class=\"data row4 col1\" >75.77%</td>\n",
       "      <td id=\"T_27436_row4_col2\" class=\"data row4 col2\" >74.24%</td>\n",
       "      <td id=\"T_27436_row4_col3\" class=\"data row4 col3\" >72.12%</td>\n",
       "      <td id=\"T_27436_row4_col4\" class=\"data row4 col4\" >69.60%</td>\n",
       "      <td id=\"T_27436_row4_col5\" class=\"data row4 col5\" >66.60%</td>\n",
       "      <td id=\"T_27436_row4_col6\" class=\"data row4 col6\" >62.21%</td>\n",
       "      <td id=\"T_27436_row4_col7\" class=\"data row4 col7\" >57.13%</td>\n",
       "      <td id=\"T_27436_row4_col8\" class=\"data row4 col8\" >51.18%</td>\n",
       "      <td id=\"T_27436_row4_col9\" class=\"data row4 col9\" >42.70%</td>\n",
       "      <td id=\"T_27436_row4_col10\" class=\"data row4 col10\" >25.18%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_27436_level0_row5\" class=\"row_heading level0 row5\" >ESM1B</th>\n",
       "      <td id=\"T_27436_row5_col0\" class=\"data row5 col0\" >65.92%</td>\n",
       "      <td id=\"T_27436_row5_col1\" class=\"data row5 col1\" >75.85%</td>\n",
       "      <td id=\"T_27436_row5_col2\" class=\"data row5 col2\" >75.07%</td>\n",
       "      <td id=\"T_27436_row5_col3\" class=\"data row5 col3\" >72.46%</td>\n",
       "      <td id=\"T_27436_row5_col4\" class=\"data row5 col4\" >69.73%</td>\n",
       "      <td id=\"T_27436_row5_col5\" class=\"data row5 col5\" >66.40%</td>\n",
       "      <td id=\"T_27436_row5_col6\" class=\"data row5 col6\" >62.27%</td>\n",
       "      <td id=\"T_27436_row5_col7\" class=\"data row5 col7\" >57.16%</td>\n",
       "      <td id=\"T_27436_row5_col8\" class=\"data row5 col8\" >51.29%</td>\n",
       "      <td id=\"T_27436_row5_col9\" class=\"data row5 col9\" >42.66%</td>\n",
       "      <td id=\"T_27436_row5_col10\" class=\"data row5 col10\" >25.04%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_27436_level0_row6\" class=\"row_heading level0 row6\" >SEQVEC</th>\n",
       "      <td id=\"T_27436_row6_col0\" class=\"data row6 col0\" >64.56%</td>\n",
       "      <td id=\"T_27436_row6_col1\" class=\"data row6 col1\" >74.43%</td>\n",
       "      <td id=\"T_27436_row6_col2\" class=\"data row6 col2\" >73.14%</td>\n",
       "      <td id=\"T_27436_row6_col3\" class=\"data row6 col3\" >70.76%</td>\n",
       "      <td id=\"T_27436_row6_col4\" class=\"data row6 col4\" >68.10%</td>\n",
       "      <td id=\"T_27436_row6_col5\" class=\"data row6 col5\" >64.97%</td>\n",
       "      <td id=\"T_27436_row6_col6\" class=\"data row6 col6\" >60.64%</td>\n",
       "      <td id=\"T_27436_row6_col7\" class=\"data row6 col7\" >54.90%</td>\n",
       "      <td id=\"T_27436_row6_col8\" class=\"data row6 col8\" >49.12%</td>\n",
       "      <td id=\"T_27436_row6_col9\" class=\"data row6 col9\" >40.52%</td>\n",
       "      <td id=\"T_27436_row6_col10\" class=\"data row6 col10\" >23.20%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_27436_level0_row7\" class=\"row_heading level0 row7\" >BOECKAERTS</th>\n",
       "      <td id=\"T_27436_row7_col0\" class=\"data row7 col0\" >65.64%</td>\n",
       "      <td id=\"T_27436_row7_col1\" class=\"data row7 col1\" >74.47%</td>\n",
       "      <td id=\"T_27436_row7_col2\" class=\"data row7 col2\" >73.33%</td>\n",
       "      <td id=\"T_27436_row7_col3\" class=\"data row7 col3\" >70.84%</td>\n",
       "      <td id=\"T_27436_row7_col4\" class=\"data row7 col4\" >67.39%</td>\n",
       "      <td id=\"T_27436_row7_col5\" class=\"data row7 col5\" >63.89%</td>\n",
       "      <td id=\"T_27436_row7_col6\" class=\"data row7 col6\" >59.35%</td>\n",
       "      <td id=\"T_27436_row7_col7\" class=\"data row7 col7\" >53.41%</td>\n",
       "      <td id=\"T_27436_row7_col8\" class=\"data row7 col8\" >47.20%</td>\n",
       "      <td id=\"T_27436_row7_col9\" class=\"data row7 col9\" >38.97%</td>\n",
       "      <td id=\"T_27436_row7_col10\" class=\"data row7 col10\" >21.61%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x224c4234288>"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = []\n",
    "for model in models:\n",
    "    result = []\n",
    "    for threshold in range(0, 11):\n",
    "        metric = \"{:.2f}\".format(model_results[models.index(model)][threshold][WEIGHTED][F1] * 100)\n",
    "        \n",
    "        result.append(f'{metric}%')\n",
    "        \n",
    "    results.append(result)\n",
    "\n",
    "print(\"Weighted F1\")\n",
    "results_df = pd.DataFrame(results, columns = [str(_) + '%' for _ in range(0, 101, 10)], index = models)\n",
    "results_df.style.highlight_max(color = 'lightgreen', axis = 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6c3cae9",
   "metadata": {},
   "source": [
    "### Global: ProtT5 + Handcrafted (Individually)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "234aa8f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_results = []\n",
    "for model in important_features[:5]:\n",
    "    with open(f'{constants.TEMP_RESULTS}/prott5_{model}.pickle', 'rb') as f:\n",
    "        model_results.append(pickle.load(f))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "445a4df3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Weighted Precision\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_e7f32_row0_col7, #T_e7f32_row1_col5, #T_e7f32_row1_col6, #T_e7f32_row1_col9, #T_e7f32_row2_col0, #T_e7f32_row2_col3, #T_e7f32_row3_col4, #T_e7f32_row3_col10, #T_e7f32_row4_col1, #T_e7f32_row4_col2, #T_e7f32_row4_col8 {\n",
       "  background-color: lightgreen;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_e7f32_\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th class=\"col_heading level0 col0\" >0%</th>\n",
       "      <th class=\"col_heading level0 col1\" >10%</th>\n",
       "      <th class=\"col_heading level0 col2\" >20%</th>\n",
       "      <th class=\"col_heading level0 col3\" >30%</th>\n",
       "      <th class=\"col_heading level0 col4\" >40%</th>\n",
       "      <th class=\"col_heading level0 col5\" >50%</th>\n",
       "      <th class=\"col_heading level0 col6\" >60%</th>\n",
       "      <th class=\"col_heading level0 col7\" >70%</th>\n",
       "      <th class=\"col_heading level0 col8\" >80%</th>\n",
       "      <th class=\"col_heading level0 col9\" >90%</th>\n",
       "      <th class=\"col_heading level0 col10\" >100%</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_e7f32_level0_row0\" class=\"row_heading level0 row0\" >A_freq</th>\n",
       "      <td id=\"T_e7f32_row0_col0\" class=\"data row0 col0\" >63.80%</td>\n",
       "      <td id=\"T_e7f32_row0_col1\" class=\"data row0 col1\" >80.24%</td>\n",
       "      <td id=\"T_e7f32_row0_col2\" class=\"data row0 col2\" >83.29%</td>\n",
       "      <td id=\"T_e7f32_row0_col3\" class=\"data row0 col3\" >84.03%</td>\n",
       "      <td id=\"T_e7f32_row0_col4\" class=\"data row0 col4\" >84.42%</td>\n",
       "      <td id=\"T_e7f32_row0_col5\" class=\"data row0 col5\" >84.74%</td>\n",
       "      <td id=\"T_e7f32_row0_col6\" class=\"data row0 col6\" >85.86%</td>\n",
       "      <td id=\"T_e7f32_row0_col7\" class=\"data row0 col7\" >85.05%</td>\n",
       "      <td id=\"T_e7f32_row0_col8\" class=\"data row0 col8\" >84.48%</td>\n",
       "      <td id=\"T_e7f32_row0_col9\" class=\"data row0 col9\" >83.47%</td>\n",
       "      <td id=\"T_e7f32_row0_col10\" class=\"data row0 col10\" >76.52%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e7f32_level0_row1\" class=\"row_heading level0 row1\" >GC</th>\n",
       "      <td id=\"T_e7f32_row1_col0\" class=\"data row1 col0\" >63.94%</td>\n",
       "      <td id=\"T_e7f32_row1_col1\" class=\"data row1 col1\" >80.30%</td>\n",
       "      <td id=\"T_e7f32_row1_col2\" class=\"data row1 col2\" >83.21%</td>\n",
       "      <td id=\"T_e7f32_row1_col3\" class=\"data row1 col3\" >84.05%</td>\n",
       "      <td id=\"T_e7f32_row1_col4\" class=\"data row1 col4\" >84.50%</td>\n",
       "      <td id=\"T_e7f32_row1_col5\" class=\"data row1 col5\" >84.84%</td>\n",
       "      <td id=\"T_e7f32_row1_col6\" class=\"data row1 col6\" >85.89%</td>\n",
       "      <td id=\"T_e7f32_row1_col7\" class=\"data row1 col7\" >84.93%</td>\n",
       "      <td id=\"T_e7f32_row1_col8\" class=\"data row1 col8\" >84.22%</td>\n",
       "      <td id=\"T_e7f32_row1_col9\" class=\"data row1 col9\" >84.10%</td>\n",
       "      <td id=\"T_e7f32_row1_col10\" class=\"data row1 col10\" >75.47%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e7f32_level0_row2\" class=\"row_heading level0 row2\" >C_freq</th>\n",
       "      <td id=\"T_e7f32_row2_col0\" class=\"data row2 col0\" >64.08%</td>\n",
       "      <td id=\"T_e7f32_row2_col1\" class=\"data row2 col1\" >80.61%</td>\n",
       "      <td id=\"T_e7f32_row2_col2\" class=\"data row2 col2\" >83.05%</td>\n",
       "      <td id=\"T_e7f32_row2_col3\" class=\"data row2 col3\" >84.20%</td>\n",
       "      <td id=\"T_e7f32_row2_col4\" class=\"data row2 col4\" >84.49%</td>\n",
       "      <td id=\"T_e7f32_row2_col5\" class=\"data row2 col5\" >84.69%</td>\n",
       "      <td id=\"T_e7f32_row2_col6\" class=\"data row2 col6\" >85.44%</td>\n",
       "      <td id=\"T_e7f32_row2_col7\" class=\"data row2 col7\" >84.80%</td>\n",
       "      <td id=\"T_e7f32_row2_col8\" class=\"data row2 col8\" >84.43%</td>\n",
       "      <td id=\"T_e7f32_row2_col9\" class=\"data row2 col9\" >83.79%</td>\n",
       "      <td id=\"T_e7f32_row2_col10\" class=\"data row2 col10\" >77.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e7f32_level0_row3\" class=\"row_heading level0 row3\" >TTA</th>\n",
       "      <td id=\"T_e7f32_row3_col0\" class=\"data row3 col0\" >63.83%</td>\n",
       "      <td id=\"T_e7f32_row3_col1\" class=\"data row3 col1\" >80.22%</td>\n",
       "      <td id=\"T_e7f32_row3_col2\" class=\"data row3 col2\" >83.17%</td>\n",
       "      <td id=\"T_e7f32_row3_col3\" class=\"data row3 col3\" >84.06%</td>\n",
       "      <td id=\"T_e7f32_row3_col4\" class=\"data row3 col4\" >84.57%</td>\n",
       "      <td id=\"T_e7f32_row3_col5\" class=\"data row3 col5\" >84.73%</td>\n",
       "      <td id=\"T_e7f32_row3_col6\" class=\"data row3 col6\" >85.46%</td>\n",
       "      <td id=\"T_e7f32_row3_col7\" class=\"data row3 col7\" >84.81%</td>\n",
       "      <td id=\"T_e7f32_row3_col8\" class=\"data row3 col8\" >84.34%</td>\n",
       "      <td id=\"T_e7f32_row3_col9\" class=\"data row3 col9\" >83.60%</td>\n",
       "      <td id=\"T_e7f32_row3_col10\" class=\"data row3 col10\" >77.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e7f32_level0_row4\" class=\"row_heading level0 row4\" >TTA_b</th>\n",
       "      <td id=\"T_e7f32_row4_col0\" class=\"data row4 col0\" >63.56%</td>\n",
       "      <td id=\"T_e7f32_row4_col1\" class=\"data row4 col1\" >80.65%</td>\n",
       "      <td id=\"T_e7f32_row4_col2\" class=\"data row4 col2\" >83.30%</td>\n",
       "      <td id=\"T_e7f32_row4_col3\" class=\"data row4 col3\" >83.86%</td>\n",
       "      <td id=\"T_e7f32_row4_col4\" class=\"data row4 col4\" >84.47%</td>\n",
       "      <td id=\"T_e7f32_row4_col5\" class=\"data row4 col5\" >84.77%</td>\n",
       "      <td id=\"T_e7f32_row4_col6\" class=\"data row4 col6\" >85.69%</td>\n",
       "      <td id=\"T_e7f32_row4_col7\" class=\"data row4 col7\" >84.78%</td>\n",
       "      <td id=\"T_e7f32_row4_col8\" class=\"data row4 col8\" >84.89%</td>\n",
       "      <td id=\"T_e7f32_row4_col9\" class=\"data row4 col9\" >84.01%</td>\n",
       "      <td id=\"T_e7f32_row4_col10\" class=\"data row4 col10\" >75.16%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x224c2fdc6c8>"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = []\n",
    "for model in important_features[:5]:\n",
    "    result = []\n",
    "    for threshold in range(0, 11):\n",
    "        metric = \"{:.2f}\".format(model_results[important_features[:5].index(model)][threshold][WEIGHTED][PRECISION] * 100)\n",
    "        result.append(f'{metric}%')\n",
    "\n",
    "    results.append(result)\n",
    "        \n",
    "print(\"Weighted Precision\")\n",
    "results_df = pd.DataFrame(results, columns = [str(_) + '%' for _ in range(0, 101, 10)], index = important_features[:5])\n",
    "results_df.style.highlight_max(color = 'lightgreen', axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "62d9bc79",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Weighted Recall\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_c6c5f_row1_col1, #T_c6c5f_row1_col6, #T_c6c5f_row1_col7, #T_c6c5f_row1_col8, #T_c6c5f_row1_col9, #T_c6c5f_row1_col10, #T_c6c5f_row2_col0, #T_c6c5f_row2_col2, #T_c6c5f_row2_col3, #T_c6c5f_row2_col4, #T_c6c5f_row2_col5 {\n",
       "  background-color: lightgreen;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_c6c5f_\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th class=\"col_heading level0 col0\" >0%</th>\n",
       "      <th class=\"col_heading level0 col1\" >10%</th>\n",
       "      <th class=\"col_heading level0 col2\" >20%</th>\n",
       "      <th class=\"col_heading level0 col3\" >30%</th>\n",
       "      <th class=\"col_heading level0 col4\" >40%</th>\n",
       "      <th class=\"col_heading level0 col5\" >50%</th>\n",
       "      <th class=\"col_heading level0 col6\" >60%</th>\n",
       "      <th class=\"col_heading level0 col7\" >70%</th>\n",
       "      <th class=\"col_heading level0 col8\" >80%</th>\n",
       "      <th class=\"col_heading level0 col9\" >90%</th>\n",
       "      <th class=\"col_heading level0 col10\" >100%</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_c6c5f_level0_row0\" class=\"row_heading level0 row0\" >A_freq</th>\n",
       "      <td id=\"T_c6c5f_row0_col0\" class=\"data row0 col0\" >71.39%</td>\n",
       "      <td id=\"T_c6c5f_row0_col1\" class=\"data row0 col1\" >76.27%</td>\n",
       "      <td id=\"T_c6c5f_row0_col2\" class=\"data row0 col2\" >74.05%</td>\n",
       "      <td id=\"T_c6c5f_row0_col3\" class=\"data row0 col3\" >70.93%</td>\n",
       "      <td id=\"T_c6c5f_row0_col4\" class=\"data row0 col4\" >67.76%</td>\n",
       "      <td id=\"T_c6c5f_row0_col5\" class=\"data row0 col5\" >64.11%</td>\n",
       "      <td id=\"T_c6c5f_row0_col6\" class=\"data row0 col6\" >59.71%</td>\n",
       "      <td id=\"T_c6c5f_row0_col7\" class=\"data row0 col7\" >54.51%</td>\n",
       "      <td id=\"T_c6c5f_row0_col8\" class=\"data row0 col8\" >49.14%</td>\n",
       "      <td id=\"T_c6c5f_row0_col9\" class=\"data row0 col9\" >41.29%</td>\n",
       "      <td id=\"T_c6c5f_row0_col10\" class=\"data row0 col10\" >27.16%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c6c5f_level0_row1\" class=\"row_heading level0 row1\" >GC</th>\n",
       "      <td id=\"T_c6c5f_row1_col0\" class=\"data row1 col0\" >71.48%</td>\n",
       "      <td id=\"T_c6c5f_row1_col1\" class=\"data row1 col1\" >76.50%</td>\n",
       "      <td id=\"T_c6c5f_row1_col2\" class=\"data row1 col2\" >74.30%</td>\n",
       "      <td id=\"T_c6c5f_row1_col3\" class=\"data row1 col3\" >71.17%</td>\n",
       "      <td id=\"T_c6c5f_row1_col4\" class=\"data row1 col4\" >67.89%</td>\n",
       "      <td id=\"T_c6c5f_row1_col5\" class=\"data row1 col5\" >64.42%</td>\n",
       "      <td id=\"T_c6c5f_row1_col6\" class=\"data row1 col6\" >60.20%</td>\n",
       "      <td id=\"T_c6c5f_row1_col7\" class=\"data row1 col7\" >54.77%</td>\n",
       "      <td id=\"T_c6c5f_row1_col8\" class=\"data row1 col8\" >49.15%</td>\n",
       "      <td id=\"T_c6c5f_row1_col9\" class=\"data row1 col9\" >41.74%</td>\n",
       "      <td id=\"T_c6c5f_row1_col10\" class=\"data row1 col10\" >27.70%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c6c5f_level0_row2\" class=\"row_heading level0 row2\" >C_freq</th>\n",
       "      <td id=\"T_c6c5f_row2_col0\" class=\"data row2 col0\" >71.51%</td>\n",
       "      <td id=\"T_c6c5f_row2_col1\" class=\"data row2 col1\" >76.38%</td>\n",
       "      <td id=\"T_c6c5f_row2_col2\" class=\"data row2 col2\" >74.32%</td>\n",
       "      <td id=\"T_c6c5f_row2_col3\" class=\"data row2 col3\" >71.18%</td>\n",
       "      <td id=\"T_c6c5f_row2_col4\" class=\"data row2 col4\" >68.05%</td>\n",
       "      <td id=\"T_c6c5f_row2_col5\" class=\"data row2 col5\" >64.58%</td>\n",
       "      <td id=\"T_c6c5f_row2_col6\" class=\"data row2 col6\" >59.91%</td>\n",
       "      <td id=\"T_c6c5f_row2_col7\" class=\"data row2 col7\" >54.62%</td>\n",
       "      <td id=\"T_c6c5f_row2_col8\" class=\"data row2 col8\" >48.98%</td>\n",
       "      <td id=\"T_c6c5f_row2_col9\" class=\"data row2 col9\" >41.38%</td>\n",
       "      <td id=\"T_c6c5f_row2_col10\" class=\"data row2 col10\" >27.24%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c6c5f_level0_row3\" class=\"row_heading level0 row3\" >TTA</th>\n",
       "      <td id=\"T_c6c5f_row3_col0\" class=\"data row3 col0\" >71.30%</td>\n",
       "      <td id=\"T_c6c5f_row3_col1\" class=\"data row3 col1\" >75.86%</td>\n",
       "      <td id=\"T_c6c5f_row3_col2\" class=\"data row3 col2\" >73.55%</td>\n",
       "      <td id=\"T_c6c5f_row3_col3\" class=\"data row3 col3\" >70.58%</td>\n",
       "      <td id=\"T_c6c5f_row3_col4\" class=\"data row3 col4\" >67.36%</td>\n",
       "      <td id=\"T_c6c5f_row3_col5\" class=\"data row3 col5\" >63.85%</td>\n",
       "      <td id=\"T_c6c5f_row3_col6\" class=\"data row3 col6\" >59.51%</td>\n",
       "      <td id=\"T_c6c5f_row3_col7\" class=\"data row3 col7\" >53.77%</td>\n",
       "      <td id=\"T_c6c5f_row3_col8\" class=\"data row3 col8\" >48.39%</td>\n",
       "      <td id=\"T_c6c5f_row3_col9\" class=\"data row3 col9\" >41.01%</td>\n",
       "      <td id=\"T_c6c5f_row3_col10\" class=\"data row3 col10\" >27.23%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_c6c5f_level0_row4\" class=\"row_heading level0 row4\" >TTA_b</th>\n",
       "      <td id=\"T_c6c5f_row4_col0\" class=\"data row4 col0\" >71.11%</td>\n",
       "      <td id=\"T_c6c5f_row4_col1\" class=\"data row4 col1\" >76.21%</td>\n",
       "      <td id=\"T_c6c5f_row4_col2\" class=\"data row4 col2\" >73.77%</td>\n",
       "      <td id=\"T_c6c5f_row4_col3\" class=\"data row4 col3\" >70.65%</td>\n",
       "      <td id=\"T_c6c5f_row4_col4\" class=\"data row4 col4\" >67.11%</td>\n",
       "      <td id=\"T_c6c5f_row4_col5\" class=\"data row4 col5\" >63.79%</td>\n",
       "      <td id=\"T_c6c5f_row4_col6\" class=\"data row4 col6\" >59.41%</td>\n",
       "      <td id=\"T_c6c5f_row4_col7\" class=\"data row4 col7\" >53.78%</td>\n",
       "      <td id=\"T_c6c5f_row4_col8\" class=\"data row4 col8\" >48.32%</td>\n",
       "      <td id=\"T_c6c5f_row4_col9\" class=\"data row4 col9\" >40.99%</td>\n",
       "      <td id=\"T_c6c5f_row4_col10\" class=\"data row4 col10\" >27.07%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x224c3561208>"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = []\n",
    "for model in important_features[:5]:\n",
    "    result = []\n",
    "    for threshold in range(0, 11):\n",
    "        metric = \"{:.2f}\".format(model_results[important_features[:5].index(model)][threshold][WEIGHTED][RECALL] * 100)\n",
    "        result.append(f'{metric}%')\n",
    "\n",
    "    results.append(result)\n",
    "        \n",
    "print(\"Weighted Recall\")\n",
    "results_df = pd.DataFrame(results, columns = [str(_) + '%' for _ in range(0, 101, 10)], index = important_features[:5])\n",
    "results_df.style.highlight_max(color = 'lightgreen', axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ff0b9602",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Weighted F1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_e63b5_row1_col6, #T_e63b5_row1_col7, #T_e63b5_row1_col8, #T_e63b5_row1_col9, #T_e63b5_row1_col10, #T_e63b5_row2_col0, #T_e63b5_row2_col1, #T_e63b5_row2_col2, #T_e63b5_row2_col3, #T_e63b5_row2_col4, #T_e63b5_row2_col5 {\n",
       "  background-color: lightgreen;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_e63b5_\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th class=\"col_heading level0 col0\" >0%</th>\n",
       "      <th class=\"col_heading level0 col1\" >10%</th>\n",
       "      <th class=\"col_heading level0 col2\" >20%</th>\n",
       "      <th class=\"col_heading level0 col3\" >30%</th>\n",
       "      <th class=\"col_heading level0 col4\" >40%</th>\n",
       "      <th class=\"col_heading level0 col5\" >50%</th>\n",
       "      <th class=\"col_heading level0 col6\" >60%</th>\n",
       "      <th class=\"col_heading level0 col7\" >70%</th>\n",
       "      <th class=\"col_heading level0 col8\" >80%</th>\n",
       "      <th class=\"col_heading level0 col9\" >90%</th>\n",
       "      <th class=\"col_heading level0 col10\" >100%</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_e63b5_level0_row0\" class=\"row_heading level0 row0\" >A_freq</th>\n",
       "      <td id=\"T_e63b5_row0_col0\" class=\"data row0 col0\" >66.41%</td>\n",
       "      <td id=\"T_e63b5_row0_col1\" class=\"data row0 col1\" >76.78%</td>\n",
       "      <td id=\"T_e63b5_row0_col2\" class=\"data row0 col2\" >75.58%</td>\n",
       "      <td id=\"T_e63b5_row0_col3\" class=\"data row0 col3\" >73.23%</td>\n",
       "      <td id=\"T_e63b5_row0_col4\" class=\"data row0 col4\" >70.63%</td>\n",
       "      <td id=\"T_e63b5_row0_col5\" class=\"data row0 col5\" >67.47%</td>\n",
       "      <td id=\"T_e63b5_row0_col6\" class=\"data row0 col6\" >63.49%</td>\n",
       "      <td id=\"T_e63b5_row0_col7\" class=\"data row0 col7\" >58.36%</td>\n",
       "      <td id=\"T_e63b5_row0_col8\" class=\"data row0 col8\" >52.56%</td>\n",
       "      <td id=\"T_e63b5_row0_col9\" class=\"data row0 col9\" >43.37%</td>\n",
       "      <td id=\"T_e63b5_row0_col10\" class=\"data row0 col10\" >24.93%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e63b5_level0_row1\" class=\"row_heading level0 row1\" >GC</th>\n",
       "      <td id=\"T_e63b5_row1_col0\" class=\"data row1 col0\" >66.49%</td>\n",
       "      <td id=\"T_e63b5_row1_col1\" class=\"data row1 col1\" >76.92%</td>\n",
       "      <td id=\"T_e63b5_row1_col2\" class=\"data row1 col2\" >75.76%</td>\n",
       "      <td id=\"T_e63b5_row1_col3\" class=\"data row1 col3\" >73.42%</td>\n",
       "      <td id=\"T_e63b5_row1_col4\" class=\"data row1 col4\" >70.75%</td>\n",
       "      <td id=\"T_e63b5_row1_col5\" class=\"data row1 col5\" >67.74%</td>\n",
       "      <td id=\"T_e63b5_row1_col6\" class=\"data row1 col6\" >63.91%</td>\n",
       "      <td id=\"T_e63b5_row1_col7\" class=\"data row1 col7\" >58.60%</td>\n",
       "      <td id=\"T_e63b5_row1_col8\" class=\"data row1 col8\" >52.60%</td>\n",
       "      <td id=\"T_e63b5_row1_col9\" class=\"data row1 col9\" >43.99%</td>\n",
       "      <td id=\"T_e63b5_row1_col10\" class=\"data row1 col10\" >25.57%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e63b5_level0_row2\" class=\"row_heading level0 row2\" >C_freq</th>\n",
       "      <td id=\"T_e63b5_row2_col0\" class=\"data row2 col0\" >66.63%</td>\n",
       "      <td id=\"T_e63b5_row2_col1\" class=\"data row2 col1\" >76.93%</td>\n",
       "      <td id=\"T_e63b5_row2_col2\" class=\"data row2 col2\" >75.77%</td>\n",
       "      <td id=\"T_e63b5_row2_col3\" class=\"data row2 col3\" >73.43%</td>\n",
       "      <td id=\"T_e63b5_row2_col4\" class=\"data row2 col4\" >70.83%</td>\n",
       "      <td id=\"T_e63b5_row2_col5\" class=\"data row2 col5\" >67.87%</td>\n",
       "      <td id=\"T_e63b5_row2_col6\" class=\"data row2 col6\" >63.57%</td>\n",
       "      <td id=\"T_e63b5_row2_col7\" class=\"data row2 col7\" >58.47%</td>\n",
       "      <td id=\"T_e63b5_row2_col8\" class=\"data row2 col8\" >52.44%</td>\n",
       "      <td id=\"T_e63b5_row2_col9\" class=\"data row2 col9\" >43.48%</td>\n",
       "      <td id=\"T_e63b5_row2_col10\" class=\"data row2 col10\" >25.09%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e63b5_level0_row3\" class=\"row_heading level0 row3\" >TTA</th>\n",
       "      <td id=\"T_e63b5_row3_col0\" class=\"data row3 col0\" >66.39%</td>\n",
       "      <td id=\"T_e63b5_row3_col1\" class=\"data row3 col1\" >76.50%</td>\n",
       "      <td id=\"T_e63b5_row3_col2\" class=\"data row3 col2\" >75.15%</td>\n",
       "      <td id=\"T_e63b5_row3_col3\" class=\"data row3 col3\" >72.99%</td>\n",
       "      <td id=\"T_e63b5_row3_col4\" class=\"data row3 col4\" >70.33%</td>\n",
       "      <td id=\"T_e63b5_row3_col5\" class=\"data row3 col5\" >67.26%</td>\n",
       "      <td id=\"T_e63b5_row3_col6\" class=\"data row3 col6\" >63.25%</td>\n",
       "      <td id=\"T_e63b5_row3_col7\" class=\"data row3 col7\" >57.60%</td>\n",
       "      <td id=\"T_e63b5_row3_col8\" class=\"data row3 col8\" >51.75%</td>\n",
       "      <td id=\"T_e63b5_row3_col9\" class=\"data row3 col9\" >43.01%</td>\n",
       "      <td id=\"T_e63b5_row3_col10\" class=\"data row3 col10\" >24.97%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_e63b5_level0_row4\" class=\"row_heading level0 row4\" >TTA_b</th>\n",
       "      <td id=\"T_e63b5_row4_col0\" class=\"data row4 col0\" >66.19%</td>\n",
       "      <td id=\"T_e63b5_row4_col1\" class=\"data row4 col1\" >76.81%</td>\n",
       "      <td id=\"T_e63b5_row4_col2\" class=\"data row4 col2\" >75.33%</td>\n",
       "      <td id=\"T_e63b5_row4_col3\" class=\"data row4 col3\" >72.99%</td>\n",
       "      <td id=\"T_e63b5_row4_col4\" class=\"data row4 col4\" >70.07%</td>\n",
       "      <td id=\"T_e63b5_row4_col5\" class=\"data row4 col5\" >67.18%</td>\n",
       "      <td id=\"T_e63b5_row4_col6\" class=\"data row4 col6\" >63.22%</td>\n",
       "      <td id=\"T_e63b5_row4_col7\" class=\"data row4 col7\" >57.56%</td>\n",
       "      <td id=\"T_e63b5_row4_col8\" class=\"data row4 col8\" >51.70%</td>\n",
       "      <td id=\"T_e63b5_row4_col9\" class=\"data row4 col9\" >43.05%</td>\n",
       "      <td id=\"T_e63b5_row4_col10\" class=\"data row4 col10\" >24.63%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x224c2e41bc8>"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = []\n",
    "for model in important_features[:5]:\n",
    "    result = []\n",
    "    for threshold in range(0, 11):\n",
    "        metric = \"{:.2f}\".format(model_results[important_features[:5].index(model)][threshold][WEIGHTED][F1] * 100)\n",
    "        result.append(f'{metric}%')\n",
    "\n",
    "    results.append(result)\n",
    "        \n",
    "print(\"Weighted F1\")\n",
    "results_df = pd.DataFrame(results, columns = [str(_) + '%' for _ in range(0, 101, 10)], index = important_features[:5])\n",
    "results_df.style.highlight_max(color = 'lightgreen', axis = 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "080ef532",
   "metadata": {},
   "source": [
    "### Protein Features Only: ProtT5 + Handcrafted (Individually)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "66c694e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_results = []\n",
    "for model in important_protein_features[:5]:\n",
    "    with open(f'{constants.TEMP_RESULTS}/prott5_{model}.pickle', 'rb') as f:\n",
    "        model_results.append(pickle.load(f))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "2643d64b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Weighted Precision\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_28bfc_row0_col5, #T_28bfc_row0_col6, #T_28bfc_row0_col9, #T_28bfc_row0_col10, #T_28bfc_row1_col1, #T_28bfc_row1_col3, #T_28bfc_row1_col4, #T_28bfc_row1_col7, #T_28bfc_row1_col8, #T_28bfc_row3_col5, #T_28bfc_row4_col0, #T_28bfc_row4_col2 {\n",
       "  background-color: lightgreen;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_28bfc_\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th class=\"col_heading level0 col0\" >0%</th>\n",
       "      <th class=\"col_heading level0 col1\" >10%</th>\n",
       "      <th class=\"col_heading level0 col2\" >20%</th>\n",
       "      <th class=\"col_heading level0 col3\" >30%</th>\n",
       "      <th class=\"col_heading level0 col4\" >40%</th>\n",
       "      <th class=\"col_heading level0 col5\" >50%</th>\n",
       "      <th class=\"col_heading level0 col6\" >60%</th>\n",
       "      <th class=\"col_heading level0 col7\" >70%</th>\n",
       "      <th class=\"col_heading level0 col8\" >80%</th>\n",
       "      <th class=\"col_heading level0 col9\" >90%</th>\n",
       "      <th class=\"col_heading level0 col10\" >100%</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_28bfc_level0_row0\" class=\"row_heading level0 row0\" >K</th>\n",
       "      <td id=\"T_28bfc_row0_col0\" class=\"data row0 col0\" >63.62%</td>\n",
       "      <td id=\"T_28bfc_row0_col1\" class=\"data row0 col1\" >80.09%</td>\n",
       "      <td id=\"T_28bfc_row0_col2\" class=\"data row0 col2\" >83.08%</td>\n",
       "      <td id=\"T_28bfc_row0_col3\" class=\"data row0 col3\" >84.27%</td>\n",
       "      <td id=\"T_28bfc_row0_col4\" class=\"data row0 col4\" >84.53%</td>\n",
       "      <td id=\"T_28bfc_row0_col5\" class=\"data row0 col5\" >84.62%</td>\n",
       "      <td id=\"T_28bfc_row0_col6\" class=\"data row0 col6\" >85.54%</td>\n",
       "      <td id=\"T_28bfc_row0_col7\" class=\"data row0 col7\" >84.87%</td>\n",
       "      <td id=\"T_28bfc_row0_col8\" class=\"data row0 col8\" >84.24%</td>\n",
       "      <td id=\"T_28bfc_row0_col9\" class=\"data row0 col9\" >83.98%</td>\n",
       "      <td id=\"T_28bfc_row0_col10\" class=\"data row0 col10\" >78.06%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_28bfc_level0_row1\" class=\"row_heading level0 row1\" >pI</th>\n",
       "      <td id=\"T_28bfc_row1_col0\" class=\"data row1 col0\" >63.82%</td>\n",
       "      <td id=\"T_28bfc_row1_col1\" class=\"data row1 col1\" >80.51%</td>\n",
       "      <td id=\"T_28bfc_row1_col2\" class=\"data row1 col2\" >83.13%</td>\n",
       "      <td id=\"T_28bfc_row1_col3\" class=\"data row1 col3\" >84.28%</td>\n",
       "      <td id=\"T_28bfc_row1_col4\" class=\"data row1 col4\" >84.55%</td>\n",
       "      <td id=\"T_28bfc_row1_col5\" class=\"data row1 col5\" >84.46%</td>\n",
       "      <td id=\"T_28bfc_row1_col6\" class=\"data row1 col6\" >85.28%</td>\n",
       "      <td id=\"T_28bfc_row1_col7\" class=\"data row1 col7\" >85.34%</td>\n",
       "      <td id=\"T_28bfc_row1_col8\" class=\"data row1 col8\" >84.40%</td>\n",
       "      <td id=\"T_28bfc_row1_col9\" class=\"data row1 col9\" >83.71%</td>\n",
       "      <td id=\"T_28bfc_row1_col10\" class=\"data row1 col10\" >77.09%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_28bfc_level0_row2\" class=\"row_heading level0 row2\" >Z4</th>\n",
       "      <td id=\"T_28bfc_row2_col0\" class=\"data row2 col0\" >63.69%</td>\n",
       "      <td id=\"T_28bfc_row2_col1\" class=\"data row2 col1\" >80.09%</td>\n",
       "      <td id=\"T_28bfc_row2_col2\" class=\"data row2 col2\" >83.18%</td>\n",
       "      <td id=\"T_28bfc_row2_col3\" class=\"data row2 col3\" >83.99%</td>\n",
       "      <td id=\"T_28bfc_row2_col4\" class=\"data row2 col4\" >84.33%</td>\n",
       "      <td id=\"T_28bfc_row2_col5\" class=\"data row2 col5\" >84.60%</td>\n",
       "      <td id=\"T_28bfc_row2_col6\" class=\"data row2 col6\" >85.37%</td>\n",
       "      <td id=\"T_28bfc_row2_col7\" class=\"data row2 col7\" >84.77%</td>\n",
       "      <td id=\"T_28bfc_row2_col8\" class=\"data row2 col8\" >84.32%</td>\n",
       "      <td id=\"T_28bfc_row2_col9\" class=\"data row2 col9\" >83.79%</td>\n",
       "      <td id=\"T_28bfc_row2_col10\" class=\"data row2 col10\" >77.44%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_28bfc_level0_row3\" class=\"row_heading level0 row3\" >CTDC2</th>\n",
       "      <td id=\"T_28bfc_row3_col0\" class=\"data row3 col0\" >63.76%</td>\n",
       "      <td id=\"T_28bfc_row3_col1\" class=\"data row3 col1\" >79.91%</td>\n",
       "      <td id=\"T_28bfc_row3_col2\" class=\"data row3 col2\" >83.14%</td>\n",
       "      <td id=\"T_28bfc_row3_col3\" class=\"data row3 col3\" >84.07%</td>\n",
       "      <td id=\"T_28bfc_row3_col4\" class=\"data row3 col4\" >84.28%</td>\n",
       "      <td id=\"T_28bfc_row3_col5\" class=\"data row3 col5\" >84.62%</td>\n",
       "      <td id=\"T_28bfc_row3_col6\" class=\"data row3 col6\" >84.95%</td>\n",
       "      <td id=\"T_28bfc_row3_col7\" class=\"data row3 col7\" >84.96%</td>\n",
       "      <td id=\"T_28bfc_row3_col8\" class=\"data row3 col8\" >84.34%</td>\n",
       "      <td id=\"T_28bfc_row3_col9\" class=\"data row3 col9\" >83.47%</td>\n",
       "      <td id=\"T_28bfc_row3_col10\" class=\"data row3 col10\" >76.82%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_28bfc_level0_row4\" class=\"row_heading level0 row4\" >mol_weight</th>\n",
       "      <td id=\"T_28bfc_row4_col0\" class=\"data row4 col0\" >63.84%</td>\n",
       "      <td id=\"T_28bfc_row4_col1\" class=\"data row4 col1\" >80.20%</td>\n",
       "      <td id=\"T_28bfc_row4_col2\" class=\"data row4 col2\" >83.24%</td>\n",
       "      <td id=\"T_28bfc_row4_col3\" class=\"data row4 col3\" >84.14%</td>\n",
       "      <td id=\"T_28bfc_row4_col4\" class=\"data row4 col4\" >84.45%</td>\n",
       "      <td id=\"T_28bfc_row4_col5\" class=\"data row4 col5\" >84.46%</td>\n",
       "      <td id=\"T_28bfc_row4_col6\" class=\"data row4 col6\" >85.17%</td>\n",
       "      <td id=\"T_28bfc_row4_col7\" class=\"data row4 col7\" >85.06%</td>\n",
       "      <td id=\"T_28bfc_row4_col8\" class=\"data row4 col8\" >84.33%</td>\n",
       "      <td id=\"T_28bfc_row4_col9\" class=\"data row4 col9\" >83.26%</td>\n",
       "      <td id=\"T_28bfc_row4_col10\" class=\"data row4 col10\" >75.83%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x224c2e41108>"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = []\n",
    "for model in important_protein_features[:5]:\n",
    "    result = []\n",
    "    for threshold in range(0, 11):\n",
    "        metric = \"{:.2f}\".format(model_results[important_protein_features[:5].index(model)][threshold][WEIGHTED][PRECISION] * 100)\n",
    "        result.append(f'{metric}%')\n",
    "\n",
    "    results.append(result)\n",
    "        \n",
    "print(\"Weighted Precision\")\n",
    "results_df = pd.DataFrame(results, columns = [str(_) + '%' for _ in range(0, 101, 10)], index = important_protein_features[:5])\n",
    "results_df.style.highlight_max(color = 'lightgreen', axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "affe8435",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Weighted Recall\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_ce6e2_row0_col2, #T_ce6e2_row0_col3, #T_ce6e2_row0_col6, #T_ce6e2_row0_col8, #T_ce6e2_row0_col9, #T_ce6e2_row1_col0, #T_ce6e2_row1_col1, #T_ce6e2_row2_col4, #T_ce6e2_row3_col5, #T_ce6e2_row3_col10, #T_ce6e2_row4_col6, #T_ce6e2_row4_col7 {\n",
       "  background-color: lightgreen;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_ce6e2_\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th class=\"col_heading level0 col0\" >0%</th>\n",
       "      <th class=\"col_heading level0 col1\" >10%</th>\n",
       "      <th class=\"col_heading level0 col2\" >20%</th>\n",
       "      <th class=\"col_heading level0 col3\" >30%</th>\n",
       "      <th class=\"col_heading level0 col4\" >40%</th>\n",
       "      <th class=\"col_heading level0 col5\" >50%</th>\n",
       "      <th class=\"col_heading level0 col6\" >60%</th>\n",
       "      <th class=\"col_heading level0 col7\" >70%</th>\n",
       "      <th class=\"col_heading level0 col8\" >80%</th>\n",
       "      <th class=\"col_heading level0 col9\" >90%</th>\n",
       "      <th class=\"col_heading level0 col10\" >100%</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_ce6e2_level0_row0\" class=\"row_heading level0 row0\" >K</th>\n",
       "      <td id=\"T_ce6e2_row0_col0\" class=\"data row0 col0\" >70.92%</td>\n",
       "      <td id=\"T_ce6e2_row0_col1\" class=\"data row0 col1\" >75.73%</td>\n",
       "      <td id=\"T_ce6e2_row0_col2\" class=\"data row0 col2\" >73.58%</td>\n",
       "      <td id=\"T_ce6e2_row0_col3\" class=\"data row0 col3\" >70.68%</td>\n",
       "      <td id=\"T_ce6e2_row0_col4\" class=\"data row0 col4\" >67.20%</td>\n",
       "      <td id=\"T_ce6e2_row0_col5\" class=\"data row0 col5\" >63.64%</td>\n",
       "      <td id=\"T_ce6e2_row0_col6\" class=\"data row0 col6\" >59.22%</td>\n",
       "      <td id=\"T_ce6e2_row0_col7\" class=\"data row0 col7\" >53.67%</td>\n",
       "      <td id=\"T_ce6e2_row0_col8\" class=\"data row0 col8\" >48.63%</td>\n",
       "      <td id=\"T_ce6e2_row0_col9\" class=\"data row0 col9\" >41.13%</td>\n",
       "      <td id=\"T_ce6e2_row0_col10\" class=\"data row0 col10\" >27.23%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_ce6e2_level0_row1\" class=\"row_heading level0 row1\" >pI</th>\n",
       "      <td id=\"T_ce6e2_row1_col0\" class=\"data row1 col0\" >71.00%</td>\n",
       "      <td id=\"T_ce6e2_row1_col1\" class=\"data row1 col1\" >75.91%</td>\n",
       "      <td id=\"T_ce6e2_row1_col2\" class=\"data row1 col2\" >73.41%</td>\n",
       "      <td id=\"T_ce6e2_row1_col3\" class=\"data row1 col3\" >70.45%</td>\n",
       "      <td id=\"T_ce6e2_row1_col4\" class=\"data row1 col4\" >67.04%</td>\n",
       "      <td id=\"T_ce6e2_row1_col5\" class=\"data row1 col5\" >63.39%</td>\n",
       "      <td id=\"T_ce6e2_row1_col6\" class=\"data row1 col6\" >59.13%</td>\n",
       "      <td id=\"T_ce6e2_row1_col7\" class=\"data row1 col7\" >53.55%</td>\n",
       "      <td id=\"T_ce6e2_row1_col8\" class=\"data row1 col8\" >48.44%</td>\n",
       "      <td id=\"T_ce6e2_row1_col9\" class=\"data row1 col9\" >40.93%</td>\n",
       "      <td id=\"T_ce6e2_row1_col10\" class=\"data row1 col10\" >27.51%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_ce6e2_level0_row2\" class=\"row_heading level0 row2\" >Z4</th>\n",
       "      <td id=\"T_ce6e2_row2_col0\" class=\"data row2 col0\" >70.81%</td>\n",
       "      <td id=\"T_ce6e2_row2_col1\" class=\"data row2 col1\" >75.64%</td>\n",
       "      <td id=\"T_ce6e2_row2_col2\" class=\"data row2 col2\" >73.35%</td>\n",
       "      <td id=\"T_ce6e2_row2_col3\" class=\"data row2 col3\" >70.59%</td>\n",
       "      <td id=\"T_ce6e2_row2_col4\" class=\"data row2 col4\" >67.42%</td>\n",
       "      <td id=\"T_ce6e2_row2_col5\" class=\"data row2 col5\" >63.54%</td>\n",
       "      <td id=\"T_ce6e2_row2_col6\" class=\"data row2 col6\" >59.09%</td>\n",
       "      <td id=\"T_ce6e2_row2_col7\" class=\"data row2 col7\" >53.63%</td>\n",
       "      <td id=\"T_ce6e2_row2_col8\" class=\"data row2 col8\" >48.18%</td>\n",
       "      <td id=\"T_ce6e2_row2_col9\" class=\"data row2 col9\" >41.01%</td>\n",
       "      <td id=\"T_ce6e2_row2_col10\" class=\"data row2 col10\" >27.23%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_ce6e2_level0_row3\" class=\"row_heading level0 row3\" >CTDC2</th>\n",
       "      <td id=\"T_ce6e2_row3_col0\" class=\"data row3 col0\" >70.87%</td>\n",
       "      <td id=\"T_ce6e2_row3_col1\" class=\"data row3 col1\" >75.68%</td>\n",
       "      <td id=\"T_ce6e2_row3_col2\" class=\"data row3 col2\" >73.51%</td>\n",
       "      <td id=\"T_ce6e2_row3_col3\" class=\"data row3 col3\" >70.49%</td>\n",
       "      <td id=\"T_ce6e2_row3_col4\" class=\"data row3 col4\" >67.18%</td>\n",
       "      <td id=\"T_ce6e2_row3_col5\" class=\"data row3 col5\" >63.71%</td>\n",
       "      <td id=\"T_ce6e2_row3_col6\" class=\"data row3 col6\" >59.01%</td>\n",
       "      <td id=\"T_ce6e2_row3_col7\" class=\"data row3 col7\" >53.82%</td>\n",
       "      <td id=\"T_ce6e2_row3_col8\" class=\"data row3 col8\" >48.36%</td>\n",
       "      <td id=\"T_ce6e2_row3_col9\" class=\"data row3 col9\" >40.88%</td>\n",
       "      <td id=\"T_ce6e2_row3_col10\" class=\"data row3 col10\" >27.58%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_ce6e2_level0_row4\" class=\"row_heading level0 row4\" >mol_weight</th>\n",
       "      <td id=\"T_ce6e2_row4_col0\" class=\"data row4 col0\" >70.92%</td>\n",
       "      <td id=\"T_ce6e2_row4_col1\" class=\"data row4 col1\" >75.76%</td>\n",
       "      <td id=\"T_ce6e2_row4_col2\" class=\"data row4 col2\" >73.36%</td>\n",
       "      <td id=\"T_ce6e2_row4_col3\" class=\"data row4 col3\" >70.52%</td>\n",
       "      <td id=\"T_ce6e2_row4_col4\" class=\"data row4 col4\" >67.26%</td>\n",
       "      <td id=\"T_ce6e2_row4_col5\" class=\"data row4 col5\" >63.27%</td>\n",
       "      <td id=\"T_ce6e2_row4_col6\" class=\"data row4 col6\" >59.22%</td>\n",
       "      <td id=\"T_ce6e2_row4_col7\" class=\"data row4 col7\" >53.87%</td>\n",
       "      <td id=\"T_ce6e2_row4_col8\" class=\"data row4 col8\" >48.46%</td>\n",
       "      <td id=\"T_ce6e2_row4_col9\" class=\"data row4 col9\" >40.86%</td>\n",
       "      <td id=\"T_ce6e2_row4_col10\" class=\"data row4 col10\" >27.34%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x224c44ce048>"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = []\n",
    "for model in important_protein_features[:5]:\n",
    "    result = []\n",
    "    for threshold in range(0, 11):\n",
    "        metric = \"{:.2f}\".format(model_results[important_protein_features[:5].index(model)][threshold][WEIGHTED][RECALL] * 100)\n",
    "        result.append(f'{metric}%')\n",
    "\n",
    "    results.append(result)\n",
    "        \n",
    "print(\"Weighted Recall\")\n",
    "results_df = pd.DataFrame(results, columns = [str(_) + '%' for _ in range(0, 101, 10)], index = important_protein_features[:5])\n",
    "results_df.style.highlight_max(color = 'lightgreen', axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "8481c043",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Weighted F1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_57991_row0_col2, #T_57991_row0_col3, #T_57991_row0_col6, #T_57991_row0_col8, #T_57991_row0_col9, #T_57991_row1_col0, #T_57991_row1_col1, #T_57991_row2_col4, #T_57991_row3_col5, #T_57991_row3_col10, #T_57991_row4_col7 {\n",
       "  background-color: lightgreen;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_57991_\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th class=\"col_heading level0 col0\" >0%</th>\n",
       "      <th class=\"col_heading level0 col1\" >10%</th>\n",
       "      <th class=\"col_heading level0 col2\" >20%</th>\n",
       "      <th class=\"col_heading level0 col3\" >30%</th>\n",
       "      <th class=\"col_heading level0 col4\" >40%</th>\n",
       "      <th class=\"col_heading level0 col5\" >50%</th>\n",
       "      <th class=\"col_heading level0 col6\" >60%</th>\n",
       "      <th class=\"col_heading level0 col7\" >70%</th>\n",
       "      <th class=\"col_heading level0 col8\" >80%</th>\n",
       "      <th class=\"col_heading level0 col9\" >90%</th>\n",
       "      <th class=\"col_heading level0 col10\" >100%</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_57991_level0_row0\" class=\"row_heading level0 row0\" >K</th>\n",
       "      <td id=\"T_57991_row0_col0\" class=\"data row0 col0\" >66.10%</td>\n",
       "      <td id=\"T_57991_row0_col1\" class=\"data row0 col1\" >76.32%</td>\n",
       "      <td id=\"T_57991_row0_col2\" class=\"data row0 col2\" >75.15%</td>\n",
       "      <td id=\"T_57991_row0_col3\" class=\"data row0 col3\" >73.03%</td>\n",
       "      <td id=\"T_57991_row0_col4\" class=\"data row0 col4\" >70.18%</td>\n",
       "      <td id=\"T_57991_row0_col5\" class=\"data row0 col5\" >66.96%</td>\n",
       "      <td id=\"T_57991_row0_col6\" class=\"data row0 col6\" >63.00%</td>\n",
       "      <td id=\"T_57991_row0_col7\" class=\"data row0 col7\" >57.49%</td>\n",
       "      <td id=\"T_57991_row0_col8\" class=\"data row0 col8\" >52.00%</td>\n",
       "      <td id=\"T_57991_row0_col9\" class=\"data row0 col9\" >43.13%</td>\n",
       "      <td id=\"T_57991_row0_col10\" class=\"data row0 col10\" >25.07%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_57991_level0_row1\" class=\"row_heading level0 row1\" >pI</th>\n",
       "      <td id=\"T_57991_row1_col0\" class=\"data row1 col0\" >66.24%</td>\n",
       "      <td id=\"T_57991_row1_col1\" class=\"data row1 col1\" >76.57%</td>\n",
       "      <td id=\"T_57991_row1_col2\" class=\"data row1 col2\" >75.02%</td>\n",
       "      <td id=\"T_57991_row1_col3\" class=\"data row1 col3\" >72.86%</td>\n",
       "      <td id=\"T_57991_row1_col4\" class=\"data row1 col4\" >70.02%</td>\n",
       "      <td id=\"T_57991_row1_col5\" class=\"data row1 col5\" >66.76%</td>\n",
       "      <td id=\"T_57991_row1_col6\" class=\"data row1 col6\" >62.90%</td>\n",
       "      <td id=\"T_57991_row1_col7\" class=\"data row1 col7\" >57.36%</td>\n",
       "      <td id=\"T_57991_row1_col8\" class=\"data row1 col8\" >51.76%</td>\n",
       "      <td id=\"T_57991_row1_col9\" class=\"data row1 col9\" >43.01%</td>\n",
       "      <td id=\"T_57991_row1_col10\" class=\"data row1 col10\" >25.27%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_57991_level0_row2\" class=\"row_heading level0 row2\" >Z4</th>\n",
       "      <td id=\"T_57991_row2_col0\" class=\"data row2 col0\" >66.03%</td>\n",
       "      <td id=\"T_57991_row2_col1\" class=\"data row2 col1\" >76.32%</td>\n",
       "      <td id=\"T_57991_row2_col2\" class=\"data row2 col2\" >74.95%</td>\n",
       "      <td id=\"T_57991_row2_col3\" class=\"data row2 col3\" >72.92%</td>\n",
       "      <td id=\"T_57991_row2_col4\" class=\"data row2 col4\" >70.31%</td>\n",
       "      <td id=\"T_57991_row2_col5\" class=\"data row2 col5\" >66.91%</td>\n",
       "      <td id=\"T_57991_row2_col6\" class=\"data row2 col6\" >62.86%</td>\n",
       "      <td id=\"T_57991_row2_col7\" class=\"data row2 col7\" >57.40%</td>\n",
       "      <td id=\"T_57991_row2_col8\" class=\"data row2 col8\" >51.48%</td>\n",
       "      <td id=\"T_57991_row2_col9\" class=\"data row2 col9\" >43.06%</td>\n",
       "      <td id=\"T_57991_row2_col10\" class=\"data row2 col10\" >24.97%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_57991_level0_row3\" class=\"row_heading level0 row3\" >CTDC2</th>\n",
       "      <td id=\"T_57991_row3_col0\" class=\"data row3 col0\" >66.10%</td>\n",
       "      <td id=\"T_57991_row3_col1\" class=\"data row3 col1\" >76.26%</td>\n",
       "      <td id=\"T_57991_row3_col2\" class=\"data row3 col2\" >75.12%</td>\n",
       "      <td id=\"T_57991_row3_col3\" class=\"data row3 col3\" >72.84%</td>\n",
       "      <td id=\"T_57991_row3_col4\" class=\"data row3 col4\" >70.14%</td>\n",
       "      <td id=\"T_57991_row3_col5\" class=\"data row3 col5\" >67.09%</td>\n",
       "      <td id=\"T_57991_row3_col6\" class=\"data row3 col6\" >62.78%</td>\n",
       "      <td id=\"T_57991_row3_col7\" class=\"data row3 col7\" >57.65%</td>\n",
       "      <td id=\"T_57991_row3_col8\" class=\"data row3 col8\" >51.64%</td>\n",
       "      <td id=\"T_57991_row3_col9\" class=\"data row3 col9\" >42.91%</td>\n",
       "      <td id=\"T_57991_row3_col10\" class=\"data row3 col10\" >25.42%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_57991_level0_row4\" class=\"row_heading level0 row4\" >mol_weight</th>\n",
       "      <td id=\"T_57991_row4_col0\" class=\"data row4 col0\" >66.16%</td>\n",
       "      <td id=\"T_57991_row4_col1\" class=\"data row4 col1\" >76.35%</td>\n",
       "      <td id=\"T_57991_row4_col2\" class=\"data row4 col2\" >75.02%</td>\n",
       "      <td id=\"T_57991_row4_col3\" class=\"data row4 col3\" >72.94%</td>\n",
       "      <td id=\"T_57991_row4_col4\" class=\"data row4 col4\" >70.26%</td>\n",
       "      <td id=\"T_57991_row4_col5\" class=\"data row4 col5\" >66.66%</td>\n",
       "      <td id=\"T_57991_row4_col6\" class=\"data row4 col6\" >62.98%</td>\n",
       "      <td id=\"T_57991_row4_col7\" class=\"data row4 col7\" >57.74%</td>\n",
       "      <td id=\"T_57991_row4_col8\" class=\"data row4 col8\" >51.79%</td>\n",
       "      <td id=\"T_57991_row4_col9\" class=\"data row4 col9\" >42.87%</td>\n",
       "      <td id=\"T_57991_row4_col10\" class=\"data row4 col10\" >25.15%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x224c424c1c8>"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = []\n",
    "for model in important_protein_features[:5]:\n",
    "    result = []\n",
    "    for threshold in range(0, 11):\n",
    "        metric = \"{:.2f}\".format(model_results[important_protein_features[:5].index(model)][threshold][WEIGHTED][F1] * 100)\n",
    "        result.append(f'{metric}%')\n",
    "\n",
    "    results.append(result)\n",
    "        \n",
    "print(\"Weighted F1\")\n",
    "results_df = pd.DataFrame(results, columns = [str(_) + '%' for _ in range(0, 101, 10)], index = important_protein_features[:5])\n",
    "results_df.style.highlight_max(color = 'lightgreen', axis = 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57549e71",
   "metadata": {},
   "source": [
    "### Global: ProtT5 + Handcrafted (Top n)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "b2894425",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_results = []\n",
    "for i in range(1, 6):\n",
    "    with open(f'{constants.TEMP_RESULTS}/prott5_handcrafted_{i}.pickle', 'rb') as f:\n",
    "        model_results.append(pickle.load(f))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "5cc4fd56",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Weighted Precision\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_f71d1_row0_col2, #T_f71d1_row1_col4, #T_f71d1_row2_col6, #T_f71d1_row2_col9, #T_f71d1_row2_col10, #T_f71d1_row3_col1, #T_f71d1_row3_col3, #T_f71d1_row4_col0, #T_f71d1_row4_col5, #T_f71d1_row4_col7, #T_f71d1_row4_col8 {\n",
       "  background-color: lightgreen;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_f71d1_\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th class=\"col_heading level0 col0\" >0%</th>\n",
       "      <th class=\"col_heading level0 col1\" >10%</th>\n",
       "      <th class=\"col_heading level0 col2\" >20%</th>\n",
       "      <th class=\"col_heading level0 col3\" >30%</th>\n",
       "      <th class=\"col_heading level0 col4\" >40%</th>\n",
       "      <th class=\"col_heading level0 col5\" >50%</th>\n",
       "      <th class=\"col_heading level0 col6\" >60%</th>\n",
       "      <th class=\"col_heading level0 col7\" >70%</th>\n",
       "      <th class=\"col_heading level0 col8\" >80%</th>\n",
       "      <th class=\"col_heading level0 col9\" >90%</th>\n",
       "      <th class=\"col_heading level0 col10\" >100%</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_f71d1_level0_row0\" class=\"row_heading level0 row0\" >1</th>\n",
       "      <td id=\"T_f71d1_row0_col0\" class=\"data row0 col0\" >63.80%</td>\n",
       "      <td id=\"T_f71d1_row0_col1\" class=\"data row0 col1\" >80.24%</td>\n",
       "      <td id=\"T_f71d1_row0_col2\" class=\"data row0 col2\" >83.29%</td>\n",
       "      <td id=\"T_f71d1_row0_col3\" class=\"data row0 col3\" >84.03%</td>\n",
       "      <td id=\"T_f71d1_row0_col4\" class=\"data row0 col4\" >84.42%</td>\n",
       "      <td id=\"T_f71d1_row0_col5\" class=\"data row0 col5\" >84.74%</td>\n",
       "      <td id=\"T_f71d1_row0_col6\" class=\"data row0 col6\" >85.86%</td>\n",
       "      <td id=\"T_f71d1_row0_col7\" class=\"data row0 col7\" >85.05%</td>\n",
       "      <td id=\"T_f71d1_row0_col8\" class=\"data row0 col8\" >84.48%</td>\n",
       "      <td id=\"T_f71d1_row0_col9\" class=\"data row0 col9\" >83.47%</td>\n",
       "      <td id=\"T_f71d1_row0_col10\" class=\"data row0 col10\" >76.52%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_f71d1_level0_row1\" class=\"row_heading level0 row1\" >2</th>\n",
       "      <td id=\"T_f71d1_row1_col0\" class=\"data row1 col0\" >64.17%</td>\n",
       "      <td id=\"T_f71d1_row1_col1\" class=\"data row1 col1\" >80.40%</td>\n",
       "      <td id=\"T_f71d1_row1_col2\" class=\"data row1 col2\" >83.04%</td>\n",
       "      <td id=\"T_f71d1_row1_col3\" class=\"data row1 col3\" >84.33%</td>\n",
       "      <td id=\"T_f71d1_row1_col4\" class=\"data row1 col4\" >84.82%</td>\n",
       "      <td id=\"T_f71d1_row1_col5\" class=\"data row1 col5\" >84.86%</td>\n",
       "      <td id=\"T_f71d1_row1_col6\" class=\"data row1 col6\" >85.80%</td>\n",
       "      <td id=\"T_f71d1_row1_col7\" class=\"data row1 col7\" >85.01%</td>\n",
       "      <td id=\"T_f71d1_row1_col8\" class=\"data row1 col8\" >84.59%</td>\n",
       "      <td id=\"T_f71d1_row1_col9\" class=\"data row1 col9\" >84.14%</td>\n",
       "      <td id=\"T_f71d1_row1_col10\" class=\"data row1 col10\" >76.98%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_f71d1_level0_row2\" class=\"row_heading level0 row2\" >3</th>\n",
       "      <td id=\"T_f71d1_row2_col0\" class=\"data row2 col0\" >64.42%</td>\n",
       "      <td id=\"T_f71d1_row2_col1\" class=\"data row2 col1\" >80.47%</td>\n",
       "      <td id=\"T_f71d1_row2_col2\" class=\"data row2 col2\" >83.11%</td>\n",
       "      <td id=\"T_f71d1_row2_col3\" class=\"data row2 col3\" >84.38%</td>\n",
       "      <td id=\"T_f71d1_row2_col4\" class=\"data row2 col4\" >84.53%</td>\n",
       "      <td id=\"T_f71d1_row2_col5\" class=\"data row2 col5\" >84.97%</td>\n",
       "      <td id=\"T_f71d1_row2_col6\" class=\"data row2 col6\" >85.99%</td>\n",
       "      <td id=\"T_f71d1_row2_col7\" class=\"data row2 col7\" >85.18%</td>\n",
       "      <td id=\"T_f71d1_row2_col8\" class=\"data row2 col8\" >84.44%</td>\n",
       "      <td id=\"T_f71d1_row2_col9\" class=\"data row2 col9\" >84.31%</td>\n",
       "      <td id=\"T_f71d1_row2_col10\" class=\"data row2 col10\" >78.18%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_f71d1_level0_row3\" class=\"row_heading level0 row3\" >4</th>\n",
       "      <td id=\"T_f71d1_row3_col0\" class=\"data row3 col0\" >64.21%</td>\n",
       "      <td id=\"T_f71d1_row3_col1\" class=\"data row3 col1\" >80.49%</td>\n",
       "      <td id=\"T_f71d1_row3_col2\" class=\"data row3 col2\" >83.00%</td>\n",
       "      <td id=\"T_f71d1_row3_col3\" class=\"data row3 col3\" >84.46%</td>\n",
       "      <td id=\"T_f71d1_row3_col4\" class=\"data row3 col4\" >84.78%</td>\n",
       "      <td id=\"T_f71d1_row3_col5\" class=\"data row3 col5\" >84.94%</td>\n",
       "      <td id=\"T_f71d1_row3_col6\" class=\"data row3 col6\" >85.46%</td>\n",
       "      <td id=\"T_f71d1_row3_col7\" class=\"data row3 col7\" >85.44%</td>\n",
       "      <td id=\"T_f71d1_row3_col8\" class=\"data row3 col8\" >84.41%</td>\n",
       "      <td id=\"T_f71d1_row3_col9\" class=\"data row3 col9\" >84.26%</td>\n",
       "      <td id=\"T_f71d1_row3_col10\" class=\"data row3 col10\" >74.61%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_f71d1_level0_row4\" class=\"row_heading level0 row4\" >5</th>\n",
       "      <td id=\"T_f71d1_row4_col0\" class=\"data row4 col0\" >64.55%</td>\n",
       "      <td id=\"T_f71d1_row4_col1\" class=\"data row4 col1\" >80.41%</td>\n",
       "      <td id=\"T_f71d1_row4_col2\" class=\"data row4 col2\" >82.87%</td>\n",
       "      <td id=\"T_f71d1_row4_col3\" class=\"data row4 col3\" >84.32%</td>\n",
       "      <td id=\"T_f71d1_row4_col4\" class=\"data row4 col4\" >84.70%</td>\n",
       "      <td id=\"T_f71d1_row4_col5\" class=\"data row4 col5\" >85.02%</td>\n",
       "      <td id=\"T_f71d1_row4_col6\" class=\"data row4 col6\" >85.31%</td>\n",
       "      <td id=\"T_f71d1_row4_col7\" class=\"data row4 col7\" >85.58%</td>\n",
       "      <td id=\"T_f71d1_row4_col8\" class=\"data row4 col8\" >84.65%</td>\n",
       "      <td id=\"T_f71d1_row4_col9\" class=\"data row4 col9\" >84.11%</td>\n",
       "      <td id=\"T_f71d1_row4_col10\" class=\"data row4 col10\" >77.88%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x2249851b448>"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = []\n",
    "for i in range(1, 6):\n",
    "    result = []\n",
    "    for threshold in range(0, 11):\n",
    "        metric = \"{:.2f}\".format(model_results[i - 1][threshold][WEIGHTED][PRECISION] * 100)\n",
    "        result.append(f'{metric}%')\n",
    "\n",
    "    results.append(result)\n",
    "        \n",
    "print(\"Weighted Precision\")\n",
    "results_df = pd.DataFrame(results, columns = [str(_) + '%' for _ in range(0, 101, 10)], index = [i for i in range(1, 6)])\n",
    "results_df.style.highlight_max(color = 'lightgreen', axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "254bc199",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Weighted Recall\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_d978a_row1_col2, #T_d978a_row1_col6, #T_d978a_row2_col8, #T_d978a_row2_col10, #T_d978a_row3_col1, #T_d978a_row3_col4, #T_d978a_row3_col5, #T_d978a_row3_col7, #T_d978a_row4_col0, #T_d978a_row4_col3, #T_d978a_row4_col9 {\n",
       "  background-color: lightgreen;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_d978a_\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th class=\"col_heading level0 col0\" >0%</th>\n",
       "      <th class=\"col_heading level0 col1\" >10%</th>\n",
       "      <th class=\"col_heading level0 col2\" >20%</th>\n",
       "      <th class=\"col_heading level0 col3\" >30%</th>\n",
       "      <th class=\"col_heading level0 col4\" >40%</th>\n",
       "      <th class=\"col_heading level0 col5\" >50%</th>\n",
       "      <th class=\"col_heading level0 col6\" >60%</th>\n",
       "      <th class=\"col_heading level0 col7\" >70%</th>\n",
       "      <th class=\"col_heading level0 col8\" >80%</th>\n",
       "      <th class=\"col_heading level0 col9\" >90%</th>\n",
       "      <th class=\"col_heading level0 col10\" >100%</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_d978a_level0_row0\" class=\"row_heading level0 row0\" >1</th>\n",
       "      <td id=\"T_d978a_row0_col0\" class=\"data row0 col0\" >71.39%</td>\n",
       "      <td id=\"T_d978a_row0_col1\" class=\"data row0 col1\" >76.27%</td>\n",
       "      <td id=\"T_d978a_row0_col2\" class=\"data row0 col2\" >74.05%</td>\n",
       "      <td id=\"T_d978a_row0_col3\" class=\"data row0 col3\" >70.93%</td>\n",
       "      <td id=\"T_d978a_row0_col4\" class=\"data row0 col4\" >67.76%</td>\n",
       "      <td id=\"T_d978a_row0_col5\" class=\"data row0 col5\" >64.11%</td>\n",
       "      <td id=\"T_d978a_row0_col6\" class=\"data row0 col6\" >59.71%</td>\n",
       "      <td id=\"T_d978a_row0_col7\" class=\"data row0 col7\" >54.51%</td>\n",
       "      <td id=\"T_d978a_row0_col8\" class=\"data row0 col8\" >49.14%</td>\n",
       "      <td id=\"T_d978a_row0_col9\" class=\"data row0 col9\" >41.29%</td>\n",
       "      <td id=\"T_d978a_row0_col10\" class=\"data row0 col10\" >27.16%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d978a_level0_row1\" class=\"row_heading level0 row1\" >2</th>\n",
       "      <td id=\"T_d978a_row1_col0\" class=\"data row1 col0\" >71.83%</td>\n",
       "      <td id=\"T_d978a_row1_col1\" class=\"data row1 col1\" >76.54%</td>\n",
       "      <td id=\"T_d978a_row1_col2\" class=\"data row1 col2\" >74.56%</td>\n",
       "      <td id=\"T_d978a_row1_col3\" class=\"data row1 col3\" >71.64%</td>\n",
       "      <td id=\"T_d978a_row1_col4\" class=\"data row1 col4\" >68.20%</td>\n",
       "      <td id=\"T_d978a_row1_col5\" class=\"data row1 col5\" >64.85%</td>\n",
       "      <td id=\"T_d978a_row1_col6\" class=\"data row1 col6\" >60.55%</td>\n",
       "      <td id=\"T_d978a_row1_col7\" class=\"data row1 col7\" >54.95%</td>\n",
       "      <td id=\"T_d978a_row1_col8\" class=\"data row1 col8\" >49.22%</td>\n",
       "      <td id=\"T_d978a_row1_col9\" class=\"data row1 col9\" >41.51%</td>\n",
       "      <td id=\"T_d978a_row1_col10\" class=\"data row1 col10\" >27.37%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d978a_level0_row2\" class=\"row_heading level0 row2\" >3</th>\n",
       "      <td id=\"T_d978a_row2_col0\" class=\"data row2 col0\" >71.85%</td>\n",
       "      <td id=\"T_d978a_row2_col1\" class=\"data row2 col1\" >76.50%</td>\n",
       "      <td id=\"T_d978a_row2_col2\" class=\"data row2 col2\" >74.42%</td>\n",
       "      <td id=\"T_d978a_row2_col3\" class=\"data row2 col3\" >71.66%</td>\n",
       "      <td id=\"T_d978a_row2_col4\" class=\"data row2 col4\" >68.21%</td>\n",
       "      <td id=\"T_d978a_row2_col5\" class=\"data row2 col5\" >64.77%</td>\n",
       "      <td id=\"T_d978a_row2_col6\" class=\"data row2 col6\" >60.45%</td>\n",
       "      <td id=\"T_d978a_row2_col7\" class=\"data row2 col7\" >55.01%</td>\n",
       "      <td id=\"T_d978a_row2_col8\" class=\"data row2 col8\" >49.37%</td>\n",
       "      <td id=\"T_d978a_row2_col9\" class=\"data row2 col9\" >41.93%</td>\n",
       "      <td id=\"T_d978a_row2_col10\" class=\"data row2 col10\" >27.39%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d978a_level0_row3\" class=\"row_heading level0 row3\" >4</th>\n",
       "      <td id=\"T_d978a_row3_col0\" class=\"data row3 col0\" >71.77%</td>\n",
       "      <td id=\"T_d978a_row3_col1\" class=\"data row3 col1\" >76.74%</td>\n",
       "      <td id=\"T_d978a_row3_col2\" class=\"data row3 col2\" >74.47%</td>\n",
       "      <td id=\"T_d978a_row3_col3\" class=\"data row3 col3\" >71.61%</td>\n",
       "      <td id=\"T_d978a_row3_col4\" class=\"data row3 col4\" >68.48%</td>\n",
       "      <td id=\"T_d978a_row3_col5\" class=\"data row3 col5\" >65.08%</td>\n",
       "      <td id=\"T_d978a_row3_col6\" class=\"data row3 col6\" >60.31%</td>\n",
       "      <td id=\"T_d978a_row3_col7\" class=\"data row3 col7\" >55.04%</td>\n",
       "      <td id=\"T_d978a_row3_col8\" class=\"data row3 col8\" >49.25%</td>\n",
       "      <td id=\"T_d978a_row3_col9\" class=\"data row3 col9\" >41.67%</td>\n",
       "      <td id=\"T_d978a_row3_col10\" class=\"data row3 col10\" >26.81%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d978a_level0_row4\" class=\"row_heading level0 row4\" >5</th>\n",
       "      <td id=\"T_d978a_row4_col0\" class=\"data row4 col0\" >72.09%</td>\n",
       "      <td id=\"T_d978a_row4_col1\" class=\"data row4 col1\" >76.58%</td>\n",
       "      <td id=\"T_d978a_row4_col2\" class=\"data row4 col2\" >74.38%</td>\n",
       "      <td id=\"T_d978a_row4_col3\" class=\"data row4 col3\" >71.82%</td>\n",
       "      <td id=\"T_d978a_row4_col4\" class=\"data row4 col4\" >68.37%</td>\n",
       "      <td id=\"T_d978a_row4_col5\" class=\"data row4 col5\" >65.07%</td>\n",
       "      <td id=\"T_d978a_row4_col6\" class=\"data row4 col6\" >60.33%</td>\n",
       "      <td id=\"T_d978a_row4_col7\" class=\"data row4 col7\" >54.87%</td>\n",
       "      <td id=\"T_d978a_row4_col8\" class=\"data row4 col8\" >49.33%</td>\n",
       "      <td id=\"T_d978a_row4_col9\" class=\"data row4 col9\" >41.94%</td>\n",
       "      <td id=\"T_d978a_row4_col10\" class=\"data row4 col10\" >27.13%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x2249806e548>"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = []\n",
    "for i in range(1, 6):\n",
    "    result = []\n",
    "    for threshold in range(0, 11):\n",
    "        metric = \"{:.2f}\".format(model_results[i - 1][threshold][WEIGHTED][RECALL] * 100)\n",
    "        result.append(f'{metric}%')\n",
    "\n",
    "    results.append(result)\n",
    "        \n",
    "print(\"Weighted Recall\")\n",
    "results_df = pd.DataFrame(results, columns = [str(_) + '%' for _ in range(0, 101, 10)], index = [i for i in range(1, 6)])\n",
    "results_df.style.highlight_max(color = 'lightgreen', axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6ab60020",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Weighted F1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_6196e_row1_col2, #T_6196e_row1_col6, #T_6196e_row2_col7, #T_6196e_row2_col8, #T_6196e_row2_col10, #T_6196e_row3_col1, #T_6196e_row3_col4, #T_6196e_row3_col5, #T_6196e_row3_col7, #T_6196e_row4_col0, #T_6196e_row4_col3, #T_6196e_row4_col9 {\n",
       "  background-color: lightgreen;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_6196e_\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th class=\"col_heading level0 col0\" >0%</th>\n",
       "      <th class=\"col_heading level0 col1\" >10%</th>\n",
       "      <th class=\"col_heading level0 col2\" >20%</th>\n",
       "      <th class=\"col_heading level0 col3\" >30%</th>\n",
       "      <th class=\"col_heading level0 col4\" >40%</th>\n",
       "      <th class=\"col_heading level0 col5\" >50%</th>\n",
       "      <th class=\"col_heading level0 col6\" >60%</th>\n",
       "      <th class=\"col_heading level0 col7\" >70%</th>\n",
       "      <th class=\"col_heading level0 col8\" >80%</th>\n",
       "      <th class=\"col_heading level0 col9\" >90%</th>\n",
       "      <th class=\"col_heading level0 col10\" >100%</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_6196e_level0_row0\" class=\"row_heading level0 row0\" >1</th>\n",
       "      <td id=\"T_6196e_row0_col0\" class=\"data row0 col0\" >66.41%</td>\n",
       "      <td id=\"T_6196e_row0_col1\" class=\"data row0 col1\" >76.78%</td>\n",
       "      <td id=\"T_6196e_row0_col2\" class=\"data row0 col2\" >75.58%</td>\n",
       "      <td id=\"T_6196e_row0_col3\" class=\"data row0 col3\" >73.23%</td>\n",
       "      <td id=\"T_6196e_row0_col4\" class=\"data row0 col4\" >70.63%</td>\n",
       "      <td id=\"T_6196e_row0_col5\" class=\"data row0 col5\" >67.47%</td>\n",
       "      <td id=\"T_6196e_row0_col6\" class=\"data row0 col6\" >63.49%</td>\n",
       "      <td id=\"T_6196e_row0_col7\" class=\"data row0 col7\" >58.36%</td>\n",
       "      <td id=\"T_6196e_row0_col8\" class=\"data row0 col8\" >52.56%</td>\n",
       "      <td id=\"T_6196e_row0_col9\" class=\"data row0 col9\" >43.37%</td>\n",
       "      <td id=\"T_6196e_row0_col10\" class=\"data row0 col10\" >24.93%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6196e_level0_row1\" class=\"row_heading level0 row1\" >2</th>\n",
       "      <td id=\"T_6196e_row1_col0\" class=\"data row1 col0\" >66.79%</td>\n",
       "      <td id=\"T_6196e_row1_col1\" class=\"data row1 col1\" >76.97%</td>\n",
       "      <td id=\"T_6196e_row1_col2\" class=\"data row1 col2\" >75.95%</td>\n",
       "      <td id=\"T_6196e_row1_col3\" class=\"data row1 col3\" >73.87%</td>\n",
       "      <td id=\"T_6196e_row1_col4\" class=\"data row1 col4\" >71.04%</td>\n",
       "      <td id=\"T_6196e_row1_col5\" class=\"data row1 col5\" >68.08%</td>\n",
       "      <td id=\"T_6196e_row1_col6\" class=\"data row1 col6\" >64.26%</td>\n",
       "      <td id=\"T_6196e_row1_col7\" class=\"data row1 col7\" >58.76%</td>\n",
       "      <td id=\"T_6196e_row1_col8\" class=\"data row1 col8\" >52.70%</td>\n",
       "      <td id=\"T_6196e_row1_col9\" class=\"data row1 col9\" >43.75%</td>\n",
       "      <td id=\"T_6196e_row1_col10\" class=\"data row1 col10\" >25.13%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6196e_level0_row2\" class=\"row_heading level0 row2\" >3</th>\n",
       "      <td id=\"T_6196e_row2_col0\" class=\"data row2 col0\" >66.82%</td>\n",
       "      <td id=\"T_6196e_row2_col1\" class=\"data row2 col1\" >76.91%</td>\n",
       "      <td id=\"T_6196e_row2_col2\" class=\"data row2 col2\" >75.82%</td>\n",
       "      <td id=\"T_6196e_row2_col3\" class=\"data row2 col3\" >73.84%</td>\n",
       "      <td id=\"T_6196e_row2_col4\" class=\"data row2 col4\" >71.02%</td>\n",
       "      <td id=\"T_6196e_row2_col5\" class=\"data row2 col5\" >68.07%</td>\n",
       "      <td id=\"T_6196e_row2_col6\" class=\"data row2 col6\" >64.17%</td>\n",
       "      <td id=\"T_6196e_row2_col7\" class=\"data row2 col7\" >58.83%</td>\n",
       "      <td id=\"T_6196e_row2_col8\" class=\"data row2 col8\" >52.81%</td>\n",
       "      <td id=\"T_6196e_row2_col9\" class=\"data row2 col9\" >44.20%</td>\n",
       "      <td id=\"T_6196e_row2_col10\" class=\"data row2 col10\" >25.33%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6196e_level0_row3\" class=\"row_heading level0 row3\" >4</th>\n",
       "      <td id=\"T_6196e_row3_col0\" class=\"data row3 col0\" >66.72%</td>\n",
       "      <td id=\"T_6196e_row3_col1\" class=\"data row3 col1\" >77.14%</td>\n",
       "      <td id=\"T_6196e_row3_col2\" class=\"data row3 col2\" >75.81%</td>\n",
       "      <td id=\"T_6196e_row3_col3\" class=\"data row3 col3\" >73.83%</td>\n",
       "      <td id=\"T_6196e_row3_col4\" class=\"data row3 col4\" >71.30%</td>\n",
       "      <td id=\"T_6196e_row3_col5\" class=\"data row3 col5\" >68.34%</td>\n",
       "      <td id=\"T_6196e_row3_col6\" class=\"data row3 col6\" >64.03%</td>\n",
       "      <td id=\"T_6196e_row3_col7\" class=\"data row3 col7\" >58.83%</td>\n",
       "      <td id=\"T_6196e_row3_col8\" class=\"data row3 col8\" >52.67%</td>\n",
       "      <td id=\"T_6196e_row3_col9\" class=\"data row3 col9\" >43.85%</td>\n",
       "      <td id=\"T_6196e_row3_col10\" class=\"data row3 col10\" >24.44%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6196e_level0_row4\" class=\"row_heading level0 row4\" >5</th>\n",
       "      <td id=\"T_6196e_row4_col0\" class=\"data row4 col0\" >67.03%</td>\n",
       "      <td id=\"T_6196e_row4_col1\" class=\"data row4 col1\" >77.04%</td>\n",
       "      <td id=\"T_6196e_row4_col2\" class=\"data row4 col2\" >75.79%</td>\n",
       "      <td id=\"T_6196e_row4_col3\" class=\"data row4 col3\" >73.98%</td>\n",
       "      <td id=\"T_6196e_row4_col4\" class=\"data row4 col4\" >71.17%</td>\n",
       "      <td id=\"T_6196e_row4_col5\" class=\"data row4 col5\" >68.33%</td>\n",
       "      <td id=\"T_6196e_row4_col6\" class=\"data row4 col6\" >64.01%</td>\n",
       "      <td id=\"T_6196e_row4_col7\" class=\"data row4 col7\" >58.68%</td>\n",
       "      <td id=\"T_6196e_row4_col8\" class=\"data row4 col8\" >52.77%</td>\n",
       "      <td id=\"T_6196e_row4_col9\" class=\"data row4 col9\" >44.21%</td>\n",
       "      <td id=\"T_6196e_row4_col10\" class=\"data row4 col10\" >24.80%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x2249852bd08>"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = []\n",
    "for i in range(1, 6):\n",
    "    result = []\n",
    "    for threshold in range(0, 11):\n",
    "        metric = \"{:.2f}\".format(model_results[i - 1][threshold][WEIGHTED][F1] * 100)\n",
    "        result.append(f'{metric}%')\n",
    "\n",
    "    results.append(result)\n",
    "        \n",
    "print(\"Weighted F1\")\n",
    "results_df = pd.DataFrame(results, columns = [str(_) + '%' for _ in range(0, 101, 10)], index = [i for i in range(1, 6)])\n",
    "results_df.style.highlight_max(color = 'lightgreen', axis = 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "972d0435",
   "metadata": {},
   "source": [
    "### Protein Features Only: ProtT5 + Handcrafted (Top n)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "470be846",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_results = []\n",
    "for i in range(1, 6):\n",
    "    with open(f'{constants.TEMP_RESULTS}/prott5_handcrafted_protein_{i}.pickle', 'rb') as f:\n",
    "        model_results.append(pickle.load(f))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "2c744f42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Weighted Precision\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_443ca_row0_col4, #T_443ca_row0_col6, #T_443ca_row0_col9, #T_443ca_row1_col1, #T_443ca_row2_col0, #T_443ca_row2_col2, #T_443ca_row2_col5, #T_443ca_row2_col7, #T_443ca_row2_col10, #T_443ca_row3_col3, #T_443ca_row4_col8 {\n",
       "  background-color: lightgreen;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_443ca_\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th class=\"col_heading level0 col0\" >0%</th>\n",
       "      <th class=\"col_heading level0 col1\" >10%</th>\n",
       "      <th class=\"col_heading level0 col2\" >20%</th>\n",
       "      <th class=\"col_heading level0 col3\" >30%</th>\n",
       "      <th class=\"col_heading level0 col4\" >40%</th>\n",
       "      <th class=\"col_heading level0 col5\" >50%</th>\n",
       "      <th class=\"col_heading level0 col6\" >60%</th>\n",
       "      <th class=\"col_heading level0 col7\" >70%</th>\n",
       "      <th class=\"col_heading level0 col8\" >80%</th>\n",
       "      <th class=\"col_heading level0 col9\" >90%</th>\n",
       "      <th class=\"col_heading level0 col10\" >100%</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_443ca_level0_row0\" class=\"row_heading level0 row0\" >1</th>\n",
       "      <td id=\"T_443ca_row0_col0\" class=\"data row0 col0\" >63.62%</td>\n",
       "      <td id=\"T_443ca_row0_col1\" class=\"data row0 col1\" >80.09%</td>\n",
       "      <td id=\"T_443ca_row0_col2\" class=\"data row0 col2\" >83.08%</td>\n",
       "      <td id=\"T_443ca_row0_col3\" class=\"data row0 col3\" >84.27%</td>\n",
       "      <td id=\"T_443ca_row0_col4\" class=\"data row0 col4\" >84.53%</td>\n",
       "      <td id=\"T_443ca_row0_col5\" class=\"data row0 col5\" >84.62%</td>\n",
       "      <td id=\"T_443ca_row0_col6\" class=\"data row0 col6\" >85.54%</td>\n",
       "      <td id=\"T_443ca_row0_col7\" class=\"data row0 col7\" >84.87%</td>\n",
       "      <td id=\"T_443ca_row0_col8\" class=\"data row0 col8\" >84.24%</td>\n",
       "      <td id=\"T_443ca_row0_col9\" class=\"data row0 col9\" >83.98%</td>\n",
       "      <td id=\"T_443ca_row0_col10\" class=\"data row0 col10\" >78.06%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_443ca_level0_row1\" class=\"row_heading level0 row1\" >2</th>\n",
       "      <td id=\"T_443ca_row1_col0\" class=\"data row1 col0\" >63.57%</td>\n",
       "      <td id=\"T_443ca_row1_col1\" class=\"data row1 col1\" >80.56%</td>\n",
       "      <td id=\"T_443ca_row1_col2\" class=\"data row1 col2\" >83.14%</td>\n",
       "      <td id=\"T_443ca_row1_col3\" class=\"data row1 col3\" >84.02%</td>\n",
       "      <td id=\"T_443ca_row1_col4\" class=\"data row1 col4\" >84.49%</td>\n",
       "      <td id=\"T_443ca_row1_col5\" class=\"data row1 col5\" >84.78%</td>\n",
       "      <td id=\"T_443ca_row1_col6\" class=\"data row1 col6\" >85.53%</td>\n",
       "      <td id=\"T_443ca_row1_col7\" class=\"data row1 col7\" >84.88%</td>\n",
       "      <td id=\"T_443ca_row1_col8\" class=\"data row1 col8\" >84.29%</td>\n",
       "      <td id=\"T_443ca_row1_col9\" class=\"data row1 col9\" >83.24%</td>\n",
       "      <td id=\"T_443ca_row1_col10\" class=\"data row1 col10\" >75.05%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_443ca_level0_row2\" class=\"row_heading level0 row2\" >3</th>\n",
       "      <td id=\"T_443ca_row2_col0\" class=\"data row2 col0\" >63.82%</td>\n",
       "      <td id=\"T_443ca_row2_col1\" class=\"data row2 col1\" >80.40%</td>\n",
       "      <td id=\"T_443ca_row2_col2\" class=\"data row2 col2\" >83.27%</td>\n",
       "      <td id=\"T_443ca_row2_col3\" class=\"data row2 col3\" >84.00%</td>\n",
       "      <td id=\"T_443ca_row2_col4\" class=\"data row2 col4\" >84.30%</td>\n",
       "      <td id=\"T_443ca_row2_col5\" class=\"data row2 col5\" >84.79%</td>\n",
       "      <td id=\"T_443ca_row2_col6\" class=\"data row2 col6\" >85.19%</td>\n",
       "      <td id=\"T_443ca_row2_col7\" class=\"data row2 col7\" >85.32%</td>\n",
       "      <td id=\"T_443ca_row2_col8\" class=\"data row2 col8\" >84.12%</td>\n",
       "      <td id=\"T_443ca_row2_col9\" class=\"data row2 col9\" >83.90%</td>\n",
       "      <td id=\"T_443ca_row2_col10\" class=\"data row2 col10\" >78.62%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_443ca_level0_row3\" class=\"row_heading level0 row3\" >4</th>\n",
       "      <td id=\"T_443ca_row3_col0\" class=\"data row3 col0\" >63.67%</td>\n",
       "      <td id=\"T_443ca_row3_col1\" class=\"data row3 col1\" >80.24%</td>\n",
       "      <td id=\"T_443ca_row3_col2\" class=\"data row3 col2\" >83.21%</td>\n",
       "      <td id=\"T_443ca_row3_col3\" class=\"data row3 col3\" >84.28%</td>\n",
       "      <td id=\"T_443ca_row3_col4\" class=\"data row3 col4\" >84.50%</td>\n",
       "      <td id=\"T_443ca_row3_col5\" class=\"data row3 col5\" >84.67%</td>\n",
       "      <td id=\"T_443ca_row3_col6\" class=\"data row3 col6\" >85.18%</td>\n",
       "      <td id=\"T_443ca_row3_col7\" class=\"data row3 col7\" >84.98%</td>\n",
       "      <td id=\"T_443ca_row3_col8\" class=\"data row3 col8\" >83.99%</td>\n",
       "      <td id=\"T_443ca_row3_col9\" class=\"data row3 col9\" >83.86%</td>\n",
       "      <td id=\"T_443ca_row3_col10\" class=\"data row3 col10\" >77.62%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_443ca_level0_row4\" class=\"row_heading level0 row4\" >5</th>\n",
       "      <td id=\"T_443ca_row4_col0\" class=\"data row4 col0\" >63.68%</td>\n",
       "      <td id=\"T_443ca_row4_col1\" class=\"data row4 col1\" >80.34%</td>\n",
       "      <td id=\"T_443ca_row4_col2\" class=\"data row4 col2\" >83.25%</td>\n",
       "      <td id=\"T_443ca_row4_col3\" class=\"data row4 col3\" >83.93%</td>\n",
       "      <td id=\"T_443ca_row4_col4\" class=\"data row4 col4\" >84.33%</td>\n",
       "      <td id=\"T_443ca_row4_col5\" class=\"data row4 col5\" >84.40%</td>\n",
       "      <td id=\"T_443ca_row4_col6\" class=\"data row4 col6\" >85.42%</td>\n",
       "      <td id=\"T_443ca_row4_col7\" class=\"data row4 col7\" >85.29%</td>\n",
       "      <td id=\"T_443ca_row4_col8\" class=\"data row4 col8\" >84.51%</td>\n",
       "      <td id=\"T_443ca_row4_col9\" class=\"data row4 col9\" >83.83%</td>\n",
       "      <td id=\"T_443ca_row4_col10\" class=\"data row4 col10\" >75.81%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x2249806e908>"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = []\n",
    "for i in range(1, 6):\n",
    "    result = []\n",
    "    for threshold in range(0, 11):\n",
    "        metric = \"{:.2f}\".format(model_results[i - 1][threshold][WEIGHTED][PRECISION] * 100)\n",
    "        result.append(f'{metric}%')\n",
    "\n",
    "    results.append(result)\n",
    "        \n",
    "print(\"Weighted Precision\")\n",
    "results_df = pd.DataFrame(results, columns = [str(_) + '%' for _ in range(0, 101, 10)], index = [i for i in range(1, 6)])\n",
    "results_df.style.highlight_max(color = 'lightgreen', axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "c1480bd5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Weighted Recall\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_67d8f_row0_col5, #T_67d8f_row0_col8, #T_67d8f_row1_col1, #T_67d8f_row1_col6, #T_67d8f_row2_col0, #T_67d8f_row2_col2, #T_67d8f_row3_col3, #T_67d8f_row3_col4, #T_67d8f_row3_col9, #T_67d8f_row3_col10, #T_67d8f_row4_col7 {\n",
       "  background-color: lightgreen;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_67d8f_\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th class=\"col_heading level0 col0\" >0%</th>\n",
       "      <th class=\"col_heading level0 col1\" >10%</th>\n",
       "      <th class=\"col_heading level0 col2\" >20%</th>\n",
       "      <th class=\"col_heading level0 col3\" >30%</th>\n",
       "      <th class=\"col_heading level0 col4\" >40%</th>\n",
       "      <th class=\"col_heading level0 col5\" >50%</th>\n",
       "      <th class=\"col_heading level0 col6\" >60%</th>\n",
       "      <th class=\"col_heading level0 col7\" >70%</th>\n",
       "      <th class=\"col_heading level0 col8\" >80%</th>\n",
       "      <th class=\"col_heading level0 col9\" >90%</th>\n",
       "      <th class=\"col_heading level0 col10\" >100%</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_67d8f_level0_row0\" class=\"row_heading level0 row0\" >1</th>\n",
       "      <td id=\"T_67d8f_row0_col0\" class=\"data row0 col0\" >70.92%</td>\n",
       "      <td id=\"T_67d8f_row0_col1\" class=\"data row0 col1\" >75.73%</td>\n",
       "      <td id=\"T_67d8f_row0_col2\" class=\"data row0 col2\" >73.58%</td>\n",
       "      <td id=\"T_67d8f_row0_col3\" class=\"data row0 col3\" >70.68%</td>\n",
       "      <td id=\"T_67d8f_row0_col4\" class=\"data row0 col4\" >67.20%</td>\n",
       "      <td id=\"T_67d8f_row0_col5\" class=\"data row0 col5\" >63.64%</td>\n",
       "      <td id=\"T_67d8f_row0_col6\" class=\"data row0 col6\" >59.22%</td>\n",
       "      <td id=\"T_67d8f_row0_col7\" class=\"data row0 col7\" >53.67%</td>\n",
       "      <td id=\"T_67d8f_row0_col8\" class=\"data row0 col8\" >48.63%</td>\n",
       "      <td id=\"T_67d8f_row0_col9\" class=\"data row0 col9\" >41.13%</td>\n",
       "      <td id=\"T_67d8f_row0_col10\" class=\"data row0 col10\" >27.23%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_67d8f_level0_row1\" class=\"row_heading level0 row1\" >2</th>\n",
       "      <td id=\"T_67d8f_row1_col0\" class=\"data row1 col0\" >70.85%</td>\n",
       "      <td id=\"T_67d8f_row1_col1\" class=\"data row1 col1\" >76.12%</td>\n",
       "      <td id=\"T_67d8f_row1_col2\" class=\"data row1 col2\" >73.56%</td>\n",
       "      <td id=\"T_67d8f_row1_col3\" class=\"data row1 col3\" >70.39%</td>\n",
       "      <td id=\"T_67d8f_row1_col4\" class=\"data row1 col4\" >67.26%</td>\n",
       "      <td id=\"T_67d8f_row1_col5\" class=\"data row1 col5\" >63.44%</td>\n",
       "      <td id=\"T_67d8f_row1_col6\" class=\"data row1 col6\" >59.23%</td>\n",
       "      <td id=\"T_67d8f_row1_col7\" class=\"data row1 col7\" >53.82%</td>\n",
       "      <td id=\"T_67d8f_row1_col8\" class=\"data row1 col8\" >48.47%</td>\n",
       "      <td id=\"T_67d8f_row1_col9\" class=\"data row1 col9\" >40.91%</td>\n",
       "      <td id=\"T_67d8f_row1_col10\" class=\"data row1 col10\" >27.37%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_67d8f_level0_row2\" class=\"row_heading level0 row2\" >3</th>\n",
       "      <td id=\"T_67d8f_row2_col0\" class=\"data row2 col0\" >70.95%</td>\n",
       "      <td id=\"T_67d8f_row2_col1\" class=\"data row2 col1\" >75.79%</td>\n",
       "      <td id=\"T_67d8f_row2_col2\" class=\"data row2 col2\" >73.63%</td>\n",
       "      <td id=\"T_67d8f_row2_col3\" class=\"data row2 col3\" >70.50%</td>\n",
       "      <td id=\"T_67d8f_row2_col4\" class=\"data row2 col4\" >66.94%</td>\n",
       "      <td id=\"T_67d8f_row2_col5\" class=\"data row2 col5\" >63.60%</td>\n",
       "      <td id=\"T_67d8f_row2_col6\" class=\"data row2 col6\" >59.15%</td>\n",
       "      <td id=\"T_67d8f_row2_col7\" class=\"data row2 col7\" >53.63%</td>\n",
       "      <td id=\"T_67d8f_row2_col8\" class=\"data row2 col8\" >48.10%</td>\n",
       "      <td id=\"T_67d8f_row2_col9\" class=\"data row2 col9\" >41.03%</td>\n",
       "      <td id=\"T_67d8f_row2_col10\" class=\"data row2 col10\" >27.44%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_67d8f_level0_row3\" class=\"row_heading level0 row3\" >4</th>\n",
       "      <td id=\"T_67d8f_row3_col0\" class=\"data row3 col0\" >70.92%</td>\n",
       "      <td id=\"T_67d8f_row3_col1\" class=\"data row3 col1\" >75.79%</td>\n",
       "      <td id=\"T_67d8f_row3_col2\" class=\"data row3 col2\" >73.57%</td>\n",
       "      <td id=\"T_67d8f_row3_col3\" class=\"data row3 col3\" >70.69%</td>\n",
       "      <td id=\"T_67d8f_row3_col4\" class=\"data row3 col4\" >67.41%</td>\n",
       "      <td id=\"T_67d8f_row3_col5\" class=\"data row3 col5\" >63.58%</td>\n",
       "      <td id=\"T_67d8f_row3_col6\" class=\"data row3 col6\" >59.19%</td>\n",
       "      <td id=\"T_67d8f_row3_col7\" class=\"data row3 col7\" >53.84%</td>\n",
       "      <td id=\"T_67d8f_row3_col8\" class=\"data row3 col8\" >48.34%</td>\n",
       "      <td id=\"T_67d8f_row3_col9\" class=\"data row3 col9\" >41.26%</td>\n",
       "      <td id=\"T_67d8f_row3_col10\" class=\"data row3 col10\" >27.53%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_67d8f_level0_row4\" class=\"row_heading level0 row4\" >5</th>\n",
       "      <td id=\"T_67d8f_row4_col0\" class=\"data row4 col0\" >70.81%</td>\n",
       "      <td id=\"T_67d8f_row4_col1\" class=\"data row4 col1\" >75.63%</td>\n",
       "      <td id=\"T_67d8f_row4_col2\" class=\"data row4 col2\" >73.60%</td>\n",
       "      <td id=\"T_67d8f_row4_col3\" class=\"data row4 col3\" >70.50%</td>\n",
       "      <td id=\"T_67d8f_row4_col4\" class=\"data row4 col4\" >67.07%</td>\n",
       "      <td id=\"T_67d8f_row4_col5\" class=\"data row4 col5\" >63.48%</td>\n",
       "      <td id=\"T_67d8f_row4_col6\" class=\"data row4 col6\" >58.99%</td>\n",
       "      <td id=\"T_67d8f_row4_col7\" class=\"data row4 col7\" >54.00%</td>\n",
       "      <td id=\"T_67d8f_row4_col8\" class=\"data row4 col8\" >48.51%</td>\n",
       "      <td id=\"T_67d8f_row4_col9\" class=\"data row4 col9\" >40.89%</td>\n",
       "      <td id=\"T_67d8f_row4_col10\" class=\"data row4 col10\" >27.11%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x224c2ef1cc8>"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = []\n",
    "for i in range(1, 6):\n",
    "    result = []\n",
    "    for threshold in range(0, 11):\n",
    "        metric = \"{:.2f}\".format(model_results[i - 1][threshold][WEIGHTED][RECALL] * 100)\n",
    "        result.append(f'{metric}%')\n",
    "\n",
    "    results.append(result)\n",
    "        \n",
    "print(\"Weighted Recall\")\n",
    "results_df = pd.DataFrame(results, columns = [str(_) + '%' for _ in range(0, 101, 10)], index = [i for i in range(1, 6)])\n",
    "results_df.style.highlight_max(color = 'lightgreen', axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "5f6fe6e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Weighted F1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_77070_row0_col6, #T_77070_row0_col8, #T_77070_row1_col1, #T_77070_row2_col0, #T_77070_row2_col2, #T_77070_row2_col5, #T_77070_row3_col3, #T_77070_row3_col4, #T_77070_row3_col9, #T_77070_row3_col10, #T_77070_row4_col7 {\n",
       "  background-color: lightgreen;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_77070_\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th class=\"col_heading level0 col0\" >0%</th>\n",
       "      <th class=\"col_heading level0 col1\" >10%</th>\n",
       "      <th class=\"col_heading level0 col2\" >20%</th>\n",
       "      <th class=\"col_heading level0 col3\" >30%</th>\n",
       "      <th class=\"col_heading level0 col4\" >40%</th>\n",
       "      <th class=\"col_heading level0 col5\" >50%</th>\n",
       "      <th class=\"col_heading level0 col6\" >60%</th>\n",
       "      <th class=\"col_heading level0 col7\" >70%</th>\n",
       "      <th class=\"col_heading level0 col8\" >80%</th>\n",
       "      <th class=\"col_heading level0 col9\" >90%</th>\n",
       "      <th class=\"col_heading level0 col10\" >100%</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_77070_level0_row0\" class=\"row_heading level0 row0\" >1</th>\n",
       "      <td id=\"T_77070_row0_col0\" class=\"data row0 col0\" >66.10%</td>\n",
       "      <td id=\"T_77070_row0_col1\" class=\"data row0 col1\" >76.32%</td>\n",
       "      <td id=\"T_77070_row0_col2\" class=\"data row0 col2\" >75.15%</td>\n",
       "      <td id=\"T_77070_row0_col3\" class=\"data row0 col3\" >73.03%</td>\n",
       "      <td id=\"T_77070_row0_col4\" class=\"data row0 col4\" >70.18%</td>\n",
       "      <td id=\"T_77070_row0_col5\" class=\"data row0 col5\" >66.96%</td>\n",
       "      <td id=\"T_77070_row0_col6\" class=\"data row0 col6\" >63.00%</td>\n",
       "      <td id=\"T_77070_row0_col7\" class=\"data row0 col7\" >57.49%</td>\n",
       "      <td id=\"T_77070_row0_col8\" class=\"data row0 col8\" >52.00%</td>\n",
       "      <td id=\"T_77070_row0_col9\" class=\"data row0 col9\" >43.13%</td>\n",
       "      <td id=\"T_77070_row0_col10\" class=\"data row0 col10\" >25.07%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_77070_level0_row1\" class=\"row_heading level0 row1\" >2</th>\n",
       "      <td id=\"T_77070_row1_col0\" class=\"data row1 col0\" >66.07%</td>\n",
       "      <td id=\"T_77070_row1_col1\" class=\"data row1 col1\" >76.69%</td>\n",
       "      <td id=\"T_77070_row1_col2\" class=\"data row1 col2\" >75.16%</td>\n",
       "      <td id=\"T_77070_row1_col3\" class=\"data row1 col3\" >72.78%</td>\n",
       "      <td id=\"T_77070_row1_col4\" class=\"data row1 col4\" >70.24%</td>\n",
       "      <td id=\"T_77070_row1_col5\" class=\"data row1 col5\" >66.85%</td>\n",
       "      <td id=\"T_77070_row1_col6\" class=\"data row1 col6\" >62.96%</td>\n",
       "      <td id=\"T_77070_row1_col7\" class=\"data row1 col7\" >57.57%</td>\n",
       "      <td id=\"T_77070_row1_col8\" class=\"data row1 col8\" >51.79%</td>\n",
       "      <td id=\"T_77070_row1_col9\" class=\"data row1 col9\" >42.87%</td>\n",
       "      <td id=\"T_77070_row1_col10\" class=\"data row1 col10\" >25.17%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_77070_level0_row2\" class=\"row_heading level0 row2\" >3</th>\n",
       "      <td id=\"T_77070_row2_col0\" class=\"data row2 col0\" >66.18%</td>\n",
       "      <td id=\"T_77070_row2_col1\" class=\"data row2 col1\" >76.44%</td>\n",
       "      <td id=\"T_77070_row2_col2\" class=\"data row2 col2\" >75.23%</td>\n",
       "      <td id=\"T_77070_row2_col3\" class=\"data row2 col3\" >72.86%</td>\n",
       "      <td id=\"T_77070_row2_col4\" class=\"data row2 col4\" >69.93%</td>\n",
       "      <td id=\"T_77070_row2_col5\" class=\"data row2 col5\" >67.02%</td>\n",
       "      <td id=\"T_77070_row2_col6\" class=\"data row2 col6\" >62.95%</td>\n",
       "      <td id=\"T_77070_row2_col7\" class=\"data row2 col7\" >57.47%</td>\n",
       "      <td id=\"T_77070_row2_col8\" class=\"data row2 col8\" >51.38%</td>\n",
       "      <td id=\"T_77070_row2_col9\" class=\"data row2 col9\" >43.09%</td>\n",
       "      <td id=\"T_77070_row2_col10\" class=\"data row2 col10\" >25.19%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_77070_level0_row3\" class=\"row_heading level0 row3\" >4</th>\n",
       "      <td id=\"T_77070_row3_col0\" class=\"data row3 col0\" >66.11%</td>\n",
       "      <td id=\"T_77070_row3_col1\" class=\"data row3 col1\" >76.40%</td>\n",
       "      <td id=\"T_77070_row3_col2\" class=\"data row3 col2\" >75.18%</td>\n",
       "      <td id=\"T_77070_row3_col3\" class=\"data row3 col3\" >73.06%</td>\n",
       "      <td id=\"T_77070_row3_col4\" class=\"data row3 col4\" >70.36%</td>\n",
       "      <td id=\"T_77070_row3_col5\" class=\"data row3 col5\" >66.95%</td>\n",
       "      <td id=\"T_77070_row3_col6\" class=\"data row3 col6\" >62.92%</td>\n",
       "      <td id=\"T_77070_row3_col7\" class=\"data row3 col7\" >57.63%</td>\n",
       "      <td id=\"T_77070_row3_col8\" class=\"data row3 col8\" >51.67%</td>\n",
       "      <td id=\"T_77070_row3_col9\" class=\"data row3 col9\" >43.36%</td>\n",
       "      <td id=\"T_77070_row3_col10\" class=\"data row3 col10\" >25.34%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_77070_level0_row4\" class=\"row_heading level0 row4\" >5</th>\n",
       "      <td id=\"T_77070_row4_col0\" class=\"data row4 col0\" >66.05%</td>\n",
       "      <td id=\"T_77070_row4_col1\" class=\"data row4 col1\" >76.29%</td>\n",
       "      <td id=\"T_77070_row4_col2\" class=\"data row4 col2\" >75.16%</td>\n",
       "      <td id=\"T_77070_row4_col3\" class=\"data row4 col3\" >72.83%</td>\n",
       "      <td id=\"T_77070_row4_col4\" class=\"data row4 col4\" >70.01%</td>\n",
       "      <td id=\"T_77070_row4_col5\" class=\"data row4 col5\" >66.86%</td>\n",
       "      <td id=\"T_77070_row4_col6\" class=\"data row4 col6\" >62.79%</td>\n",
       "      <td id=\"T_77070_row4_col7\" class=\"data row4 col7\" >57.82%</td>\n",
       "      <td id=\"T_77070_row4_col8\" class=\"data row4 col8\" >51.83%</td>\n",
       "      <td id=\"T_77070_row4_col9\" class=\"data row4 col9\" >42.86%</td>\n",
       "      <td id=\"T_77070_row4_col10\" class=\"data row4 col10\" >24.80%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x224c2edecc8>"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = []\n",
    "for i in range(1, 6):\n",
    "    result = []\n",
    "    for threshold in range(0, 11):\n",
    "        metric = \"{:.2f}\".format(model_results[i - 1][threshold][WEIGHTED][F1] * 100)\n",
    "        result.append(f'{metric}%')\n",
    "\n",
    "    results.append(result)\n",
    "        \n",
    "print(\"Weighted F1\")\n",
    "results_df = pd.DataFrame(results, columns = [str(_) + '%' for _ in range(0, 101, 10)], index = [i for i in range(1, 6)])\n",
    "results_df.style.highlight_max(color = 'lightgreen', axis = 0)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
